#MOVICS_RShiny是在root环境下运行的,查看相应设置时请用root用户登录查看
#包在服务器上的安装路径为:"/usr/local/lib/R/site-library"
library(shiny)
library(shinyBS)
library(shinyjs)
library(shinydashboard)
library(rhandsontable)
library(DT)
library(ggplot2)
library(jsonlite)
library(pwr)
library(TCGAbiolinks)
library(data.table)
library(ChAMPdata)
library(tidyverse)
library(magrittr)
library(readxl)
library(stringr)
library(forcats)
library(iClusterPlus)
library(data.table)
library(RColorBrewer)
library(ComplexHeatmap)
library(circlize)
library(survival)
library(survminer)
library(ggpp)
library(MOVICS)
library(patchwork)
library(ggalluvial)
library(cowplot)
library(GSVA)
library(HelpersMG)
library(CMScaller)
library(pamr)
library(clusterRepro)
library(limma)
#library(filesstrings) #安装的话使用github安装install_github('rorynolan/filesstrings')
source("getClustNums.R")
source("getSilhouettes.R")
source("getMoHeatmaps.R")
source("compSurvs.R")
source("quiet.R")
source("compClinvars.R")
source("compMuts.R")
source("compTMBs.R")
source("read_maf.R")
source("compFGAs.R")
source("compDrugsens.R")
source("pRRophetic.R")
source("compAgrees.R")
source("runDEAs.R")
source("runMarkers.R")
source("runGSEAs.R")
source("runGSVAs.R")
source("runNTPs.R")
source("runPAMs.R")
source("runKappas.R")
source("customized_functions.R")
options(timeout = 360000,shiny.maxRequestSize=30000*1024^2) #设置文件上传大小限制为30GB且响应时间为100h
# data.table::setDTthreads(threads = 1) #设置程序运行线程为1

shinyServer(function(input, output, session) {
    
    #1.cancerType填写表格
    output$table.cancerType<-renderRHandsontable({
        
        DF=data.frame(matrix(NA,ncol=input$cancerType_number,nrow=1))
        colnames(DF)=unlist(lapply(1:input$cancerType_number,function(i) paste("Type",i,sep=" ")))
        rownames(DF)="Cancer Type"
        
        default <- c()
        for(d in 1:input$cancerType_number){
            default <- c(default,"")
        }
        DF[1,]=default
        
        #设置table的参数
        rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
            hot_cols(colWidths=75,halign="htMiddle") %>%
            hot_rows(rowHeights=30)%>%
            hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=100)
    })
    
    #2.cancerType的临床数据集下载链接填写表格
    output$table.cliSurUrl<-renderRHandsontable({
        
        cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
        DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
        rownames(DF)=cancerType[1,]
        colnames(DF)="Url"
        
        default <- c()
        for(d in 1:input$cancerType_number){
            default <- c(default,"")
        }
        DF[,1]=default
        
        #设置table的参数
        rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
            hot_cols(colWidths=350,halign="htMiddle") %>%
            hot_rows(rowHeights=30)%>%
            hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #3.cancerType的临床数据集(验证)下载链接填写表格
    output$table.validation_cliSurUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #4.cancerType的RNA数据集下载链接填写表格
    output$table.RNAUrl<-renderRHandsontable({
        
        cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
        DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
        rownames(DF)=cancerType[1,]
        colnames(DF)="Url"
        
        default <- c()
        for(d in 1:input$cancerType_number){
            default <- c(default,"")
        }
        DF[,1]=default
        
        #设置table的参数
        rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
            hot_cols(colWidths=350,halign="htMiddle") %>%
            hot_rows(rowHeights=30)%>%
            hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #5.cancerType的RNA数据集(验证)下载链接填写表格
    output$table.validation_RNAUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #6.cancerType的DNA methylation数据集下载链接填写表格
    output$table.DNAMethylationUrl<-renderRHandsontable({
        
        cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
        DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
        rownames(DF)=cancerType[1,]
        colnames(DF)="Url"
        
        default <- c()
        for(d in 1:input$cancerType_number){
            default <- c(default,"")
        }
        DF[,1]=default
        
        #设置table的参数
        rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
            hot_cols(colWidths=350,halign="htMiddle") %>%
            hot_rows(rowHeights=30)%>%
            hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #7.cancerType的DNA methylation数据集(验证)下载链接填写表格
    output$table.validation_DNAMethylationUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #8.cancerType的copy number alterations数据集下载链接填写表格
    output$table.copyNumberAlterationsUrl<-renderRHandsontable({
        
        cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
        DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
        rownames(DF)=cancerType[1,]
        colnames(DF)="Url"
        
        default <- c()
        for(d in 1:input$cancerType_number){
            default <- c(default,"")
        }
        DF[,1]=default
        
        #设置table的参数
        rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
            hot_cols(colWidths=350,halign="htMiddle") %>%
            hot_rows(rowHeights=30)%>%
            hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #9.cancerType的copy number alterations数据集(验证)下载链接填写表格
    output$table.validation_copyNumberAlterationsUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #10.cancerType的binary somatic mutation数据集下载链接填写表格
    output$table.binarySomaticMutationUrl<-renderRHandsontable({
        
        cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
        DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
        rownames(DF)=cancerType[1,]
        colnames(DF)="Url"
        
        default <- c()
        for(d in 1:input$cancerType_number){
            default <- c(default,"")
        }
        DF[,1]=default
        
        #设置table的参数
        rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
            hot_cols(colWidths=350,halign="htMiddle") %>%
            hot_rows(rowHeights=30)%>%
            hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #11.cancerType的binary somatic mutation数据集(验证)下载链接填写表格
    output$table.validation_binarySomaticMutationUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #12.cancerType的radiomics数据集下载链接填写表格(Internal)
    output$table.internalRadiomicsUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #13.cancerType的radiomics数据集下载链接填写表格(External)
    output$table.externalRadiomicsUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #14.cancerType的radiomics数据集(验证)下载链接填写表格(External)
    output$table.validation_externalRadiomicsUrl<-renderRHandsontable({
      
      cancerType <- data.frame(rbind(hot_to_r(input$table.cancerType)))
      DF=data.frame(matrix(NA,nrow=input$cancerType_number,ncol=1))
      rownames(DF)=cancerType[1,]
      colnames(DF)="Url"
      
      default <- c()
      for(d in 1:input$cancerType_number){
        default <- c(default,"")
      }
      DF[,1]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=350,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=50)
    })
    
    #15.用户创建按钮
    observeEvent(input$basic_create1,{
        
      #Basic Settings
      ##设置工作目录
      # setwd("C:\\Users\\LEGION\\Desktop\\MOVICS_RShiny\\workingDirectory") #本地测试目录
      # setwd("/shinyapps/MOVICS_RShiny/workingDirectory") #服务器测试目录
      setwd("/srv/shiny-server/workingDirectory") #服务器测试目录
      workdir <- getwd()
      workdir_origin <- workdir
      system(paste0("chmod -R 777 ",workdir)) #级联更新总工作目录下所有文件的权限为“777”
      #判断用户是否已经发送邮件申请Username和User Account
      username <- trimws(input$user_name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      sam <- trimws(input$user_account,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      usersInfo <- read.table("usersInfo.txt",sep = "\t",check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
      if((username %in% usersInfo$user_name) & (sam %in% usersInfo$user_account)){
        
        if(input$log_in_first_time==1){
          
          dir.create(file.path(workdir,sam,"TCGA_data_preparing"),recursive = TRUE)
          #Sys.chmod(workdir_origin, "777", use_umask = FALSE)
          system(paste0("chmod -R 777 ",workdir)) #级联更新总工作目录下所有文件的权限为“777”
          setwd(file.path(workdir,sam,"TCGA_data_preparing"))
          workdir <- getwd()
          
          #将三个公共数据集移动到工作目录
          workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
          source_dir <- file.path(workdir_origin,"data_resources") #获取存放内置数据的路径
          file.copy(file.path(source_dir,"pancancerSurvivalData_XLu.txt"), file.path(workdir,"pancancerSurvivalData_XLu.txt"),overwrite=TRUE)
          file.copy(file.path(source_dir,"overlapTable27.txt"), file.path(workdir,"overlapTable27.txt"),overwrite=TRUE)
          file.copy(file.path(source_dir,"Mutation Flags 100.txt"), file.path(workdir,"Mutation Flags 100.txt"),overwrite=TRUE)
          
        }else{
          setwd(file.path(workdir,sam,"TCGA_data_preparing"))
          workdir <- getwd()
        }
      }
      
      #用户创建好后给予用户的反馈提示
      output$user_create<-renderUI({
        
        if((username %in% usersInfo$user_name) & (sam %in% usersInfo$user_account)){
          if(input$log_in_first_time==1){
            list(br(),column(12,h3(strong(paste0("Welcome to MOVICS RShiny! Your user account is ",sam,". Now let's start preaparing tcga datasets first."))),style="color:darkblack"))
          }else{
            list(br(),column(12,h3(strong(paste0("Welcome back MOVICS RShiny! Your user account is ",sam,". Now you need to specify 'The Number of Cancer Types', 'Input Cancer Types From TCGA', 'Project Name', 'Multi-omics Types' in TCGA datasets preparation, 'Multi-omics Types' in Validation datasets preparation, 'Clustering algorithms' and 'The number of clusters' in Consensus Clustering first. The values of these parameters need to be the same as previously set. After that, you can keep on analysis."))),style="color:darkblack"))
          }
        }else{
          list(br(),column(12,h3(strong("The system detected that the 'Username' or 'User Account' you entered was not correct, please check and enter again. If you do not have a 'Username' or a 'User Account', please write an application mail to 'xlu.cpu@foxmail.com' (Name, Affiliation, purpose, etc.).")),style="color:darkblack"))
        }
      })
      
      # rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    observeEvent(input$basic_create2,{
      
      #Basic Settings
      ##设置工作目录
      # setwd("C:\\Users\\LEGION\\Desktop\\MOVICS_RShiny\\workingDirectory") #本地测试目录
      # setwd("/shinyapps/MOVICS_RShiny/workingDirectory") #服务器测试目录
      setwd("/srv/shiny-server/workingDirectory") #服务器测试目录
      workdir <- getwd()
      workdir_origin <- workdir
      system(paste0("chmod -R 777 ",workdir)) #级联更新总工作目录下所有文件的权限为“777”
      #判断用户是否已经发送邮件申请Username和User Account
      username <- trimws(input$user_name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      sam <- trimws(input$user_account,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      usersInfo <- read.table("usersInfo.txt",sep = "\t",check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
      if((username %in% usersInfo$user_name) & (sam %in% usersInfo$user_account)){
        
        if(input$log_in_first_time==1){
          
          dir.create(file.path(workdir,sam,"TCGA_data_preparing"),recursive = TRUE)
          #Sys.chmod(workdir_origin, "777", use_umask = FALSE)
          system(paste0("chmod -R 777 ",workdir)) #级联更新总工作目录下所有文件的权限为“777”
          setwd(file.path(workdir,sam,"TCGA_data_preparing"))
          workdir <- getwd()
          
          #将三个公共数据集移动到工作目录
          workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
          source_dir <- file.path(workdir_origin,"data_resources") #获取存放内置数据的路径
          file.copy(file.path(source_dir,"pancancerSurvivalData_XLu.txt"), file.path(workdir,"pancancerSurvivalData_XLu.txt"),overwrite=TRUE)
          file.copy(file.path(source_dir,"overlapTable27.txt"), file.path(workdir,"overlapTable27.txt"),overwrite=TRUE)
          file.copy(file.path(source_dir,"Mutation Flags 100.txt"), file.path(workdir,"Mutation Flags 100.txt"),overwrite=TRUE)
          
        }else{
          setwd(file.path(workdir,sam,"TCGA_data_preparing"))
          workdir <- getwd()
        }
      }
      
      #用户创建好后给予用户的反馈提示
      output$user_create<-renderUI({
        
        if((username %in% usersInfo$user_name) & (sam %in% usersInfo$user_account)){
          if(input$log_in_first_time==1){
            list(br(),column(12,h3(strong(paste0("Welcome to MOVICS RShiny! Your user account is ",sam,". Now let's start preaparing tcga datasets first."))),style="color:darkblack"))
          }else{
            list(br(),column(12,h3(strong(paste0("Welcome back MOVICS RShiny! Your user account is ",sam,". Now you need to specify 'The Number of Cancer Types', 'Input Cancer Types From TCGA', 'Project Name', 'Multi-omics Types' in TCGA datasets preparation, 'Multi-omics Types' in Validation datasets preparation, 'Clustering algorithms' and 'The number of clusters' in Consensus Clustering first. The values of these parameters need to be the same as previously set. After that, you can keep on analysis."))),style="color:darkblack"))
          }
        }else{
          list(br(),column(12,h3(strong("The system detected that the 'Username' or 'User Account' you entered was not correct, please check and enter again. If you do not have a 'Username' or a 'User Account', please write an application mail to 'xlu.cpu@foxmail.com' (Name, Affiliation, purpose, etc.).")),style="color:darkblack"))
        }
      })
      
      # rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    observeEvent(input$basic_create3,{
      
      #Basic Settings
      ##设置工作目录
      # setwd("C:\\Users\\LEGION\\Desktop\\MOVICS_RShiny\\workingDirectory") #本地测试目录
      # setwd("/shinyapps/MOVICS_RShiny/workingDirectory") #服务器测试目录
      setwd("/srv/shiny-server/workingDirectory") #服务器测试目录
      workdir <- getwd()
      workdir_origin <- workdir
      system(paste0("chmod -R 777 ",workdir)) #级联更新总工作目录下所有文件的权限为“777”
      #判断用户是否已经发送邮件申请Username和User Account
      username <- trimws(input$user_name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      sam <- trimws(input$user_account,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      usersInfo <- read.table("usersInfo.txt",sep = "\t",check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
      if((username %in% usersInfo$user_name) & (sam %in% usersInfo$user_account)){
        
        if(input$log_in_first_time==1){
          
          dir.create(file.path(workdir,sam,"TCGA_data_preparing"),recursive = TRUE)
          #Sys.chmod(workdir_origin, "777", use_umask = FALSE)
          system(paste0("chmod -R 777 ",workdir)) #级联更新总工作目录下所有文件的权限为“777”
          setwd(file.path(workdir,sam,"TCGA_data_preparing"))
          workdir <- getwd()
          
          #将三个公共数据集移动到工作目录
          workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
          source_dir <- file.path(workdir_origin,"data_resources") #获取存放内置数据的路径
          file.copy(file.path(source_dir,"pancancerSurvivalData_XLu.txt"), file.path(workdir,"pancancerSurvivalData_XLu.txt"),overwrite=TRUE)
          file.copy(file.path(source_dir,"overlapTable27.txt"), file.path(workdir,"overlapTable27.txt"),overwrite=TRUE)
          file.copy(file.path(source_dir,"Mutation Flags 100.txt"), file.path(workdir,"Mutation Flags 100.txt"),overwrite=TRUE)
          
        }else{
          setwd(file.path(workdir,sam,"TCGA_data_preparing"))
          workdir <- getwd()
        }
      }
      
      #用户创建好后给予用户的反馈提示
      output$user_create<-renderUI({
        
        if((username %in% usersInfo$user_name) & (sam %in% usersInfo$user_account)){
          if(input$log_in_first_time==1){
            list(br(),column(12,h3(strong(paste0("Welcome to MOVICS RShiny! Your user account is ",sam,". Now let's start preaparing tcga datasets first."))),style="color:darkblack"))
          }else{
            list(br(),column(12,h3(strong(paste0("Welcome back MOVICS RShiny! Your user account is ",sam,". Now you need to specify 'The Number of Cancer Types', 'Input Cancer Types From TCGA', 'Project Name', 'Multi-omics Types' in TCGA datasets preparation, 'Multi-omics Types' in Validation datasets preparation, 'Clustering algorithms' and 'The number of clusters' in Consensus Clustering first. The values of these parameters need to be the same as previously set. After that, you can keep on analysis."))),style="color:darkblack"))
          }
        }else{
          list(br(),column(12,h3(strong("The system detected that the 'Username' or 'User Account' you entered was not correct, please check and enter again. If you do not have a 'Username' or a 'User Account', please write an application mail to 'xlu.cpu@foxmail.com' (Name, Affiliation, purpose, etc.).")),style="color:darkblack"))
        }
      })
      
      # rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #16.使用已整合数据处理按钮(TCGA)
    observeEvent(input$use_integrated_datasets_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$use_integrated_datasets_approaches==1,"Download integrated tcga datasets (.RData) using specified url","Upload integrated tcga datasets (.RData)"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 2)
      if(input$use_integrated_datasets_approaches==1){
        
        #从用户指定路径下载整合好的TCGA数据
        integrated_datasets_downloadUrl <- input$integrated_datasets_downloadUrl
        integrated_datasets_downloadUrl <- trimws(integrated_datasets_downloadUrl,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(integrated_datasets_downloadUrl,method="auto",destfile = file.path(workdir,paste0(tolower(project),".tcga.RData")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
      }else{
        
        #已上传整合好的TCGA数据转移
        integrated_datasets_uploadPath <- input$integrated_datasets_fileUpload$datapath #整合好的TCGA数据上传路径
        # file.move(integrated_datasets_uploadPath, file.path(workdir,paste0(tolower(project),".tcga.RData"))) #将上传的文件移动到工作目录下
        file.copy(integrated_datasets_uploadPath, file.path(workdir,paste0(tolower(project),".tcga.RData")),overwrite=TRUE) #将上传的文件复制到工作目录下
      }
      
      progress$set(value = 3)
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      progress$set(value = 4)
      #已整合TCGA数据处理好后给予用户的反馈提示
      output$use_integrated_datasets_finish<-renderUI({
        if(input$use_integrated_datasets_approaches==1){
          if(input$external_validation==1){
            list(br(),column(12,h3(strong("The tcga integrated dataset (.RData) has been downloaded in the corresponding folder. Now let's start to prepare validation datasets.")),style="color:darkblack"))
          }else{
            list(br(),column(12,h3(strong("The tcga integrated dataset (.RData) has been downloaded in the corresponding folder. Now let's start the first step of MOVICS--GET Module!")),style="color:darkblack"))
          }
        }else{
          if(input$external_validation==1){
            list(br(),column(12,h3(strong("The tcga integrated dataset (.RData) has been uploaded in the corresponding folder. Now let's start to prepare validation datasets.")),style="color:darkblack"))
          }else{
            list(br(),column(12,h3(strong("The tcga integrated dataset (.RData) has been uploaded in the corresponding folder. Now let's start the first step of MOVICS--GET Module!")),style="color:darkblack"))
          }
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #17.验证数据目录创建按钮
    observeEvent(input$validation_basic_create,{
      
      ##设置验证数据目录
      workdir <- getwd() #获取当前工作路径
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      dir.create(validation.path,recursive = TRUE)
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      #Sys.chmod(workdir_origin, "777", use_umask = FALSE)
      setwd(validation.path)
      validation_workdir <- getwd()
      
      #将三个公共数据集移动到工作目录
      source_dir <- file.path(workdir_origin,"data_resources") #获取存放内置数据的路径
      file.copy(file.path(source_dir,"pancancerSurvivalData_XLu.txt"), file.path(validation_workdir,"pancancerSurvivalData_XLu.txt"),overwrite=TRUE)
      file.copy(file.path(source_dir,"overlapTable27.txt"), file.path(validation_workdir,"overlapTable27.txt"),overwrite=TRUE)
      file.copy(file.path(source_dir,"Mutation Flags 100.txt"), file.path(validation_workdir,"Mutation Flags 100.txt"),overwrite=TRUE)
      
      #创建好验证数据目录后给予用户的反馈提示
      output$validation_create<-renderUI({
        
        list(br(),column(12,h3(strong(paste0("Now a folder storing validation datasets has been created and then let's start preaparing vadlidation datasets."))),style="color:darkblack"))
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      # rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #18.使用已整合数据处理按钮(验证数据)
    observeEvent(input$validation_use_integrated_datasets_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$validation_use_integrated_datasets_approaches==1,"Download integrated validation datasets (.RData) using specified url","Upload integrated validation datasets (.RData)"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 2)
      if(input$validation_use_integrated_datasets_approaches==1){
        
        #从用户指定路径下载整合好的验证数据
        validation_integrated_datasets_downloadUrl <- input$validation_integrated_datasets_downloadUrl
        validation_integrated_datasets_downloadUrl <- trimws(validation_integrated_datasets_downloadUrl,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(validation_integrated_datasets_downloadUrl,method="auto",destfile = file.path(validation_workdir,paste0(tolower(project),".validation.RData")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
      }else{
        
        #已上传整合好的验证数据转移
        validation_integrated_datasets_uploadPath <- input$validation_integrated_datasets_fileUpload$datapath #整合好的验证数据上传路径
        # file.move(validation_integrated_datasets_uploadPath, file.path(validation_workdir,paste0(tolower(project),".validation.RData"))) #将上传的文件移动到工作目录下
        file.copy(validation_integrated_datasets_uploadPath, file.path(validation_workdir,paste0(tolower(project),".validation.RData")),overwrite=TRUE) #将上传的文件复制到工作目录下
      }
      
      progress$set(value = 3)
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      progress$set(value = 4)
      #已整合验证数据处理好后给予用户的反馈提示
      output$validation_use_integrated_datasets_finish<-renderUI({
        if(input$validation_use_integrated_datasets_approaches==1){
          list(br(),column(12,h3(strong("The validation integrated dataset (.RData) has been downloaded in the corresponding folder. Now let's start the first step of MOVICS--GET Module!")),style="color:darkblack"))
        }else{
          list(br(),column(12,h3(strong("The validation integrated dataset (.RData) has been uploaded in the corresponding folder. Now let's start the first step of MOVICS--GET Module!")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #19.内置数据处理按钮
    observeEvent(input$basic_process,{
      
      progress <- Progress$new(session, min=1, max=3) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Extract and process datasets from internal resources',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd() #获取工作目录
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      source_dir <- file.path(workdir_origin,"data_resources") #获取存放内置数据的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      data_list <- list() #数据集
      #根据选定的组学类别依次对各cancerType的文件进行拷贝和预处理
      progress$set(value = 2)
      #1.Clinical & Survival
      clin.info <- data.frame()
      for(ca in 1:length(cancers)){
        
        #从存放内置数据的目录拷贝相应数据到工作目录
        file.copy(file.path(source_dir,cancers[ca],paste0(cancers[ca],"_GDCdata")),file.path(workdir,paste0(cancers[ca],"_GDCdata")),overwrite=TRUE)
        file.copy(file.path(source_dir,cancers[ca],paste0("TCGA_",cancers[ca],"_Clinical.txt")),file.path(workdir,paste0("TCGA_",cancers[ca],"_Clinical.txt")),overwrite=TRUE)
        
        #导入并预处理各cancerType的临床数据集
        clin_info <- read.delim(paste0("TCGA_",cancers[ca],"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        #clin_info <- clin_info[which(clin_info$SampleType=="PrimaryTumor"),]
        #rownames(clin_info) <- paste0(clin_info$bcr_patient_barcode,"-",sapply(strsplit(clin_info$sampleID_XSU,"-",fixed = T),"[",3))
        colnames(clin_info)[1] <- "Tumor_Sample_Barcode"
        clin_info$Tumor_Sample_Barcode <- paste0(clin_info$Tumor_Sample_Barcode,"-01A")
        clin_info <- clin_info[!duplicated(clin_info$Tumor_Sample_Barcode),]
        rownames(clin_info) <- clin_info[,1]
        write.table(clin_info,paste0("TCGA_",cancers[ca],"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        
        #Combine clinical files from different cancers
        if(ca==1){
          clin.info <- clin_info
        }else{
          intersect_clin <- intersect(colnames(clin.info),colnames(clin_info))
          clin.info <- rbind(clin.info[,intersect_clin],clin_info[,intersect_clin])
        }
        
        gc() #释放内存
        
      }
      write.table(clin.info,paste0("TCGA_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
      data_list$"clin_info" <- clin_info
      
      #准备生存数据
      surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
      #循环合并各类型癌症的生存数据
      surv.infos <- data.frame()
      for(ca in 1:length(cancers)){
        surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
        surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
        rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
        write.table(surv.info,paste0("TCGA_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
        
        #Combine survival files from different cancers
        if(ca==1){
          surv.infos <- surv.info
        }else{
          intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
          surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
        }
        
        gc() #释放内存
        
      }
      
      # Final save
      write.table(surv.infos,paste0("TCGA_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      data_list$"surv.infos" <- surv.infos
      gc() #释放内存
      
      progress$set(value = 3)      
      #2.RNA
      if((1 %in% input$multiOmics) | (2 %in% input$multiOmics)){
        
        expMatrixs <- data.frame()
        for(ca in 1:length(cancers)){
          
          #从存放内置数据的目录拷贝相应数据到工作目录
          file.copy(file.path(source_dir,cancers[ca],paste0("TCGA_",cancers[ca],"_expMatrix.txt")),file.path(workdir,paste0("TCGA_",cancers[ca],"_expMatrix.txt")),overwrite=TRUE)
          
          #Load gene annotation
          Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          rownames(Ginfo) <- Ginfo[,1]
          
          #导入并预处理各cancerType的RNA数据集
          expMatrix <- read.table(paste0("TCGA_",cancers[ca],"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          
          #Combine expression matrix from different cancers
          if(ca==1){
            expMatrixs <- expMatrix
          }else{
            intersect_expMatrix <- intersect(rownames(expMatrixs),rownames(expMatrix))
            expMatrixs <- cbind(expMatrixs[intersect_expMatrix,],expMatrix[intersect_expMatrix,])
          }
          
          colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
          normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
          tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
          
          #Tumor sample
          expMatrix <- expMatrix[,tumorsamples]
          expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
          #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
          comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
          count <- as.data.frame(expMatrix)[comgene,]
          Ginfo <- Ginfo[comgene,]
          
          #Extract mRNA and lncRNA
          Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          
          #count data
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            count <- count[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            count <- count[rownames(Mid),]
          }else{
            count <- count[rownames(Lid),]
          }
          
          countsTable <- count
          countsTable$Gene <- Ginfo[rownames(count),"genename"]
          countsTable <- countsTable[!duplicated(countsTable$Gene),]
          Ginfo_origin <- Ginfo
          Ginfo <- Ginfo[rownames(countsTable),]
          Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          Lids_gene <- Lids$genename
          Mids_gene <- Mids$genename
          rownames(countsTable) <- countsTable$Gene
          countsTable <- countsTable[,-ncol(countsTable)]
          write.table(countsTable,paste0("TCGA_",cancers[ca],"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            
            write.table(fpkm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            write.table(fpkm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            write.table(fpkm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(tpm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ##Normal sample
          ###count
          tmp <- read.table(paste0("TCGA_",cancers[ca],"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
          tmp <- tmp[,normsamples]
          tmp <- as.data.frame(tmp[rownames(count),])
          
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            tmp <- tmp[rownames(Mid),]
          }else{
            tmp <- tmp[rownames(Lid),]
          }
          
          tmpTable <- tmp
          tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
          tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
          rownames(tmpTable) <- tmpTable$Gene
          tmpTable <- tmpTable[,-ncol(tmpTable)]
          write.table(tmpTable,paste0("TCGA_",cancers[ca],"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            
            write.table(fpkm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            write.table(fpkm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            write.table(fpkm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            write.table(tpm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else{
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          }
          
          gc() #释放内存
          
        }
        
        #对合并后的RNA数据集进行处理
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        rownames(Ginfo) <- Ginfo[,1]
        expMatrix <- expMatrixs
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("TCGA_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }
        
        ##Normal sample
        ###count
        tmp <- expMatrixs
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("TCGA_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }
        
        gc() #释放内存
      }
      
      progress$set(value = 4)      
      #3.DNA methylation
      if(3 %in% input$multiOmics){
        
        meths <- data.frame()
        for(ca in 1:length(cancers)){
          
          #从存放内置数据的目录拷贝相应数据到工作目录
          file.copy(file.path(source_dir,cancers[ca],paste0("TCGA_",cancers[ca],"_Methylation450.gz")),file.path(workdir,paste0("TCGA_",cancers[ca],"_Methylation450.gz")),overwrite=TRUE)
          
          #导入并预处理各cancerType的DNA methylation数据集
          meth <- fread(paste0("TCGA_",cancers[ca],"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
          meth <- as.data.frame(meth)
          rownames(meth) <- meth[,1]
          meth <- as.data.frame(na.omit(meth[,-1]))
          #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
          meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
          colnames(meth) <- paste0(colnames(meth),"A")
          
          # Get promoter
          data("probe.features")
          promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
          meth <- round(meth[promoter,],3)
          
          # Map to gene name to take median value if multiple mapping
          meth$gene <- as.character(probe.features[promoter,"gene"])
          meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
          meth <- as.data.frame(round(meth,3))
          
          write.table(meth,paste0("TCGA_",cancers[ca],"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine DNA methylation files from different cancers
          if(ca==1){
            meths <- meth
          }else{
            intersect_meth <- intersect(rownames(meths),rownames(meth))
            meths <- cbind(meths[intersect_meth,],meth[intersect_meth,])
          }
          
          gc() #释放内存
          
        }
        
        ### Final save
        write.table(meths,paste0("TCGA_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
        gc() #释放内存
      }
      
      progress$set(value = 5)      
      #4.copy number alterations
      if(4 %in% input$multiOmics){
        
        segments <- data.frame()
        for(ca in 1:length(cancers)){
          
          #从存放内置数据的目录拷贝相应数据到工作目录
          file.copy(file.path(source_dir,cancers[ca],paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),overwrite=TRUE)
          
          #导入并预处理各cancerType的copy number alterations数据集
          untar(file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压.tar.gz文件
          untar_list <- untar(file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          segment_file <- file.path(workdir,untar_list[grep("seg.seg",untar_list)])
          segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
          segment$sample <- substr(segment$sample, start = 1,stop = 16)
          segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
          
          write.table(segment, paste0("TCGA_",cancers[ca],"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
          cna <- CNregions(segment,
                           epsilon = 0,
                           adaptive = FALSE,
                           rmCNV = FALSE,
                           cnv = FALSE,
                           frac.overlap = 0.5, 
                           rmSmallseg = TRUE, 
                           nProbes = 15)
          cna <- as.data.frame(t(cna))
          
          write.table(cna,paste0("TCGA_",cancers[ca],"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine segment files from different cancers
          if(ca==1){
            segments <- segment
          }else{
            intersect_segment <- intersect(colnames(segments),colnames(segment))
            segments <- rbind(segments[,intersect_segment],segment[,intersect_segment])
          }
          
          gc() #释放内存
        }
        
        #获取合并cna
        cnas <- CNregions(segments,
                          epsilon = 0,
                          adaptive = FALSE,
                          rmCNV = FALSE,
                          cnv = FALSE,
                          frac.overlap = 0.5, 
                          rmSmallseg = TRUE, 
                          nProbes = 15)
        cnas <- as.data.frame(t(cnas))
        
        ### Final save
        write.table(segments, paste0("TCGA_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")  #segment
        data_list$"segment" <- segments
        write.table(cnas,paste0("TCGA_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #cna
        data_list$"cna" <- cnas
        
        gc() #释放内存
      }
      
      progress$set(value = 6)      
      #5.binary somatic mutation
      if(5 %in% input$multiOmics){
        
        mafs <- data.frame()
        for(ca in 1:length(cancers)){
          
          #从存放内置数据的目录拷贝相应数据到工作目录
          file.copy(file.path(source_dir,cancers[ca],paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")),file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")),overwrite=TRUE)
          
          #导入并预处理各cancerType的binary somatic mutation数据集
          untar(file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")))  #解压.tar.gz文件
          untar_list <- untar(file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          maf_file <- file.path(workdir,untar_list[grep("data_mutations",untar_list)])
          
          maf <- read_tsv(maf_file, comment = "#")
          flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
          label <- c("Tumor_Sample_Barcode",
                     "Hugo_Symbol",
                     "Chromosome",
                     "Start_Position",
                     "End_Position",
                     "Variant_Classification",
                     "Variant_Type",
                     "Reference_Allele",
                     "Tumor_Seq_Allele1",
                     "Tumor_Seq_Allele2")
          maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
          maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
          maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
          maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
          write.table(maf, paste0("TCGA_",cancers[ca],"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # Prepare binary somatic mutation
          # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
          # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
          # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          # mut <- as.data.frame(na.omit(mut))
          # rownames(mut) <- mut$sample
          # mut <- mut[,-1]
          # mut <- mut[rowSums(mut) > 0,]
          # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
          # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
          
          # option 2: convert from MAF file without silent mutation (may not be the same with above file)
          # binary mutation
          # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
          
          # for (i in colnames(mut.binary)) {
          #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          #   if(is.element("Silent",tmp$Variant_Classification)) {
          #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
          #   }
          #   for (j in tmp$Hugo_Symbol)
          #     mut.binary[j,i] <- 1
          # }
          for (i in colnames(mut.binary)) {
            tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
            tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
            for (j in tmp$Hugo_Symbol){
              mut.binary[j,i] <- 1
            }
          }
          mut.binary <- as.data.frame(mut.binary)
          #rownames(mut.binary) <- toupper(rownames(mut.binary))
          mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
          #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
          #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
          
          write.table(mut.binary,paste0("TCGA_",cancers[ca],"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine maf files from different cancers
          if(ca==1){
            mafs <- maf
          }else{
            intersect_maf <- intersect(colnames(mafs),colnames(maf))
            mafs <- rbind(mafs[,intersect_maf],maf[,intersect_maf])
          }
          
          gc() #释放内存
          
        }
        
        ##Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # mafs.ns <- mafs[which(mafs$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        muts.binary <- matrix(0,nrow = length(unique(mafs$Hugo_Symbol)),ncol = length(unique(mafs$Tumor_Sample_Barcode)),dimnames = list(unique(mafs$Hugo_Symbol),unique(mafs$Tumor_Sample_Barcode)))
        
        # for (i in colnames(muts.binary)) {
        #   tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     muts.binary[j,i] <- 1
        # }
        for (i in colnames(muts.binary)) {
          tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            muts.binary[j,i] <- 1
          }
        }
        muts.binary <- as.data.frame(muts.binary)
        #rownames(muts.binary) <- toupper(rownames(muts.binary))
        muts.binary <- muts.binary[rowSums(muts.binary) > 0,]
        #muts.binary <- muts.binary[,which(substr(colnames(muts.binary), 14, 15) == "01")]
        #colnames(muts.binary) <- paste0(colnames(muts.binary),"A")
        
        ### Final save
        write.table(mafs, paste0("TCGA_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- mafs
        write.table(muts.binary,paste0("TCGA_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- muts.binary
        
        gc() #释放内存
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$internal_datalist_download<-downloadHandler(
        filename = "data_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$internal_process_finish<-renderUI({
        
        if(!(6 %in% input$multiOmics)){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort."))),style="color:darkblack",br(),
               column(12,downloadButton("internal_datalist_download",strong("Download RData file of internal datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("Internal datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare radiomics datasets if you have specified."))),style="color:darkblack",br(),
               column(12,downloadButton("internal_datalist_download",strong("Download RData file of internal datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
    })
    
    #20.临床生存数据集自动下载处理按钮
    observeEvent(input$cli_sur_process1,{
        
        progress <- Progress$new(session, min=1, max=4) #设置进度条
        on.exit(progress$close())
        progress$set(message = 'Download Clinical & Survival datasets automatically',
                     detail = 'This may take a while...')
        progress$set(value = 1)
        
        #每点击一次按钮就重新进行数据的下载
        workdir <- getwd()
        workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
        cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
        cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        
        progress$set(value = 2)
        data_list <- list() #数据集
        #临床数据集下载
        clinical <- GDCquery(project = paste0("TCGA-",cancers), 
                             data.category = "Clinical", 
                             file.type = "xml")
        GDCdownload(clinical,directory = "GDCdata")
        cliquery <- GDCprepare_clinic(clinical, clinical.info = "patient",directory = "GDCdata")
        gc() #释放内存
        
        progress$set(value = 3)
        colnames(cliquery)[1] <- "Tumor_Sample_Barcode"
        write.table(cliquery,paste0("TCGA_",project,"_Clinical.txt"), quote=F, row.names=F,sep = "\t") #上传此文件
        
        ##导入并预处理下载好的临床数据集
        clin.info <- read.delim(paste0("TCGA_",project,"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        #clin.info <- clin.info[which(clin.info$SampleType=="PrimaryTumor"),]
        #rownames(clin.info) <- paste0(clin.info$bcr_patient_barcode,"-",sapply(strsplit(clin.info$sampleID_XSU,"-",fixed = T),"[",3))
        clin.info$Tumor_Sample_Barcode <- paste0(clin.info$Tumor_Sample_Barcode,"-01A")
        clin.info <- clin.info[!duplicated(clin.info$Tumor_Sample_Barcode),]
        rownames(clin.info) <- clin.info[,1]
        write.table(clin.info,paste0("TCGA_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        data_list$"clin.info" <- clin.info

        progress$set(value = 4)
        #准备生存数据
        surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        ##循环合并各类型癌症的生存数据
        surv.infos <- data.frame()
        for(ca in 1:length(cancers)){
            surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
            surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
            rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
            write.table(surv.info,paste0("TCGA_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
            
            #Combine survival files from different cancers
            if(ca==1){
              surv.infos <- surv.info
            }else{
              intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
              surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
            }
            
            gc() #释放内存
            
        }

        ## Final save
        write.table(surv.infos,paste0("TCGA_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"surv.infos" <- surv.infos

        system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
        
        ##RData结果下载按钮响应事件
        output$external_clisur_download1<-downloadHandler(
          filename = "clisur_preparation(TCGA).RData",
          content = function(theFile) {
            save(data_list,file=theFile)
          }
        )
        
        #数据下载好后给予用户的反馈提示
        output$clisur_download_finish<-renderUI({
            
            list(br(),column(12,h3(strong(paste0("Clinical & Survival datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare ",ifelse((('1' %in% input$multiOmics) & ('2' %in% input$multiOmics)),"mRNA & lncRNA",ifelse('1' %in% input$multiOmics.includes,"mRNA","lncRNA"))," datasets if you have specified."))),style="color:darkblack"),br(),
                 column(12,downloadButton("external_clisur_download1",strong("Download RData file of Clinical & Survival datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        })
        
        output$clisur_finish<-renderUI({
          
          if((!(1 %in% input$multiOmics)) & (!(2 %in% input$multiOmics)) & (!(3 %in% input$multiOmics)) & (!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
            list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
          }
        })
        
        #rm(list=ls()) #清空变量
        gc() #释放内存
    })
    
    #21.临床生存数据集链接下载处理按钮
    observeEvent(input$cli_sur_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download Clinical & Survival datasets using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.cliSurUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #临床和生存数据集下载及预处理
      data_list <- list() #数据集
      if(input$cli_sur_url_sources==1){
          
          progress$set(value = 2)
          #合并临床数据下载
          cli_sur_url <- input$cli_sur_downloadUrl
          cli_sur_url <- trimws(cli_sur_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          download.file(cli_sur_url,method="auto",destfile = file.path(workdir,paste0("TCGA_",project,"_Clinical.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          gc() #释放内存
          
          progress$set(value = 3)
          #导入并预处理下载好的临床数据集
          clin.info <- read.delim(paste0("TCGA_",project,"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          #clin.info <- clin.info[which(clin.info$SampleType=="PrimaryTumor"),]
          #rownames(clin.info) <- paste0(clin.info$bcr_patient_barcode,"-",sapply(strsplit(clin.info$sampleID_XSU,"-",fixed = T),"[",3))
          colnames(clin.info)[1] <- "Tumor_Sample_Barcode"
          clin.info$Tumor_Sample_Barcode <- paste0(clin.info$Tumor_Sample_Barcode,"-01A")
          clin.info <- clin.info[!duplicated(clin.info$Tumor_Sample_Barcode),]
          rownames(clin.info) <- clin.info[,1]
          write.table(clin.info,paste0("TCGA_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
          data_list$"clin.info" <- clin.info
          
          progress$set(value = 4)
          #准备生存数据
          surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
          ##循环合并各类型癌症的生存数据
          surv.infos <- data.frame()
          for(ca in 1:length(cancers)){
            surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
            surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
            rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
            write.table(surv.info,paste0("TCGA_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
            
            #Combine survival files from different cancers
            if(ca==1){
              surv.infos <- surv.info
            }else{
              intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
              surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
            }
            
            gc() #释放内存
            
          }
          
          ## Final save
          write.table(surv.infos,paste0("TCGA_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          data_list$"surv.infos" <- surv.infos
          
      }else{
        
          progress$set(value = 2)
          #循环下载各cancerType的临床数据
          clin.info <- data.frame()
          for(url in 1:length(download_urls)){
            download.file(download_urls[url],method="auto",destfile = file.path(workdir,paste0("TCGA_",cancers[url],"_Clinical.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
            
            #导入并预处理下载好的各cancerType的临床数据集
            clin_info <- read.delim(paste0("TCGA_",cancers[url],"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
            #clin_info <- clin_info[which(clin_info$SampleType=="PrimaryTumor"),]
            #rownames(clin_info) <- paste0(clin_info$bcr_patient_barcode,"-",sapply(strsplit(clin_info$sampleID_XSU,"-",fixed = T),"[",3))
            colnames(clin_info)[1] <- "Tumor_Sample_Barcode"
            clin_info$Tumor_Sample_Barcode <- paste0(clin_info$Tumor_Sample_Barcode,"-01A")
            clin_info <- clin_info[!duplicated(clin_info$Tumor_Sample_Barcode),]
            rownames(clin_info) <- clin_info[,1]
            write.table(clin_info,paste0("TCGA_",cancers[url],"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
            
            #Combine clinical files from different cancers
            if(url==1){
              clin.info <- clin_info
            }else{
              intersect_clin <- intersect(colnames(clin.info),colnames(clin_info))
              clin.info <- rbind(clin.info[,intersect_clin],clin_info[,intersect_clin])
            }
            
            gc() #释放内存
            
          }
          write.table(clin.info,paste0("TCGA_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
          data_list$"clin.info" <- clin.info
          
          progress$set(value = 3)
          #准备生存数据
          surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
          ##循环合并各类型癌症的生存数据
          surv.infos <- data.frame()
          for(ca in 1:length(cancers)){
            surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
            surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
            rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
            write.table(surv.info,paste0("TCGA_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
            
            #Combine survival files from different cancers
            if(ca==1){
              surv.infos <- surv.info
            }else{
              intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
              surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
            }
            
            gc() #释放内存
            
          }
          
          progress$set(value = 4)
          ## Final save
          write.table(surv.infos,paste0("TCGA_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          data_list$"surv.infos" <- surv.infos
      }

      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_clisur_download2<-downloadHandler(
        filename = "clisur_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$clisur_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0("Clinical & Survival datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse((1 %in% input$multiOmics),"mRNA","lncRNA"))," datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("external_clisur_download2",strong("Download RData file of Clinical & Survival datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$clisur_finish<-renderUI({
        
        if((!(1 %in% input$multiOmics)) & (!(2 %in% input$multiOmics)) & (!(3 %in% input$multiOmics)) & (!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    # #文件上传控件测试代码
    # output$files <- renderTable(input$cli_sur_fileUpload2)
    # output$files_show<-renderUI({
    #   
    #   list(br(),column(12,h3(strong(input$cli_sur_fileUpload2$datapath[1])),style="color:darkblack"))
    # })
    
    
    #22.临床生存数据集链接下载处理按钮(验证数据集)
    observeEvent(input$validation_cli_sur_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download Clinical & Survival(Validation) datasets using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.validation_cliSurUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #临床和生存数据集下载及预处理
      data_list <- list() #数据集
      if(input$validation_cli_sur_url_sources==1){
        
        progress$set(value = 2)
        #合并临床数据下载
        cli_sur_url <- input$validation_cli_sur_downloadUrl
        cli_sur_url <- trimws(cli_sur_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(cli_sur_url,method="auto",destfile = file.path(validation_workdir,paste0("Validation_",project,"_Clinical.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        #导入并预处理下载好的临床数据集
        clin.info <- read.delim(paste0("Validation_",project,"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        #clin.info <- clin.info[which(clin.info$SampleType=="PrimaryTumor"),]
        #rownames(clin.info) <- paste0(clin.info$bcr_patient_barcode,"-",sapply(strsplit(clin.info$sampleID_XSU,"-",fixed = T),"[",3))
        colnames(clin.info)[1] <- "Tumor_Sample_Barcode"
        clin.info$Tumor_Sample_Barcode <- paste0(clin.info$Tumor_Sample_Barcode,"-01A")
        clin.info <- clin.info[!duplicated(clin.info$Tumor_Sample_Barcode),]
        rownames(clin.info) <- clin.info[,1]
        write.table(clin.info,paste0("Validation_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        data_list$"clin.info" <- clin.info
        
        progress$set(value = 4)
        #准备生存数据
        surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        ##循环合并各类型癌症的生存数据
        surv.infos <- data.frame()
        for(ca in 1:length(cancers)){
          surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
          surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
          rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
          write.table(surv.info,paste0("Validation_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
          
          #Combine survival files from different cancers
          if(ca==1){
            surv.infos <- surv.info
          }else{
            intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
            surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
          }
          
          gc() #释放内存
          
        }
        
        ## Final save
        write.table(surv.infos,paste0("Validation_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"surv.infos" <- surv.infos
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的临床数据
        clin.info <- data.frame()
        for(url in 1:length(download_urls)){
          download.file(download_urls[url],method="auto",destfile = file.path(validation_workdir,paste0("Validation_",cancers[url],"_Clinical.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          
          #导入并预处理下载好的各cancerType的临床数据集
          clin_info <- read.delim(paste0("Validation_",cancers[url],"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          #clin_info <- clin_info[which(clin_info$SampleType=="PrimaryTumor"),]
          #rownames(clin_info) <- paste0(clin_info$bcr_patient_barcode,"-",sapply(strsplit(clin_info$sampleID_XSU,"-",fixed = T),"[",3))
          colnames(clin_info)[1] <- "Tumor_Sample_Barcode"
          clin_info$Tumor_Sample_Barcode <- paste0(clin_info$Tumor_Sample_Barcode,"-01A")
          clin_info <- clin_info[!duplicated(clin_info$Tumor_Sample_Barcode),]
          rownames(clin_info) <- clin_info[,1]
          write.table(clin_info,paste0("Validation_",cancers[url],"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
          
          #Combine clinical files from different cancers
          if(url==1){
            clin.info <- clin_info
          }else{
            intersect_clin <- intersect(colnames(clin.info),colnames(clin_info))
            clin.info <- rbind(clin.info[,intersect_clin],clin_info[,intersect_clin])
          }
          
          gc() #释放内存
          
        }
        write.table(clin.info,paste0("Validation_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        data_list$"clin.info" <- clin.info
        
        progress$set(value = 3)
        #准备生存数据
        surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        ##循环合并各类型癌症的生存数据
        surv.infos <- data.frame()
        for(ca in 1:length(cancers)){
          surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
          surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
          rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
          write.table(surv.info,paste0("Validation_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
          
          #Combine survival files from different cancers
          if(ca==1){
            surv.infos <- surv.info
          }else{
            intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
            surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ## Final save
        write.table(surv.infos,paste0("Validation_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"surv.infos" <- surv.infos
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_clisur_download1<-downloadHandler(
        filename = "clisur_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_clisur_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0("Clinical & Survival(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA(Validation)",ifelse((1 %in% input$multiOmics),"mRNA(Validation)","lncRNA(Validation)"))," datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("validation_clisur_download1",strong("Download RData file of Clinical & Survival datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_clisur_finish<-renderUI({
        
        if((!(1 %in% input$validation_multiOmics)) & (!(2 %in% input$validation_multiOmics)) & (!(3 %in% input$validation_multiOmics)) & (!(4 %in% input$validation_multiOmics)) & (!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #23.临床生存数据集上传处理按钮
    observeEvent(input$cli_sur_process3,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded Clinical & Survival datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$cli_sur_fileUpload1$datapath #合并临床数据集上传路径
      upload_paths2 <- input$cli_sur_fileUpload2$datapath #各cancerType临床数据集上传路径
      
      # 临床和生存数据集转移及预处理
      data_list <- list() #数据集
      if(input$cli_sur_upload_sources==1){

        progress$set(value = 2)
        #合并临床数据集
        # file.move(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_Clinical.txt"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_Clinical.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下

        progress$set(value = 3)
        #导入并预处理上传的合并临床数据集
        clin.info <- read.delim(paste0("TCGA_",project,"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        #clin.info <- clin.info[which(clin.info$SampleType=="PrimaryTumor"),]
        #rownames(clin.info) <- paste0(clin.info$bcr_patient_barcode,"-",sapply(strsplit(clin.info$sampleID_XSU,"-",fixed = T),"[",3))
        colnames(clin.info)[1] <- "Tumor_Sample_Barcode"
        clin.info$Tumor_Sample_Barcode <- paste0(clin.info$Tumor_Sample_Barcode,"-01A")
        clin.info <- clin.info[!duplicated(clin.info$Tumor_Sample_Barcode),]
        rownames(clin.info) <- clin.info[,1]
        write.table(clin.info,paste0("TCGA_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        data_list$"clin.info" <- clin.info

        progress$set(value = 4)
        #准备生存数据
        surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        ##循环合并各类型癌症的生存数据
        surv.infos <- data.frame()
        for(ca in 1:length(cancers)){
          surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
          surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
          rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
          write.table(surv.info,paste0("TCGA_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据

          #Combine survival files from different cancers
          if(ca==1){
            surv.infos <- surv.info
          }else{
            intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
            surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
          }

          gc() #释放内存

        }

        ## Final save
        write.table(surv.infos,paste0("TCGA_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"surv.infos" <- surv.infos

      }else{
        
          progress$set(value = 2)
          #循环拷贝并预处理各cancerType的临床数据
          clin.info <- data.frame()
          for(ca in 1:length(cancers)){
            
            # file.move(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_Clinical.txt"))) #将上传的文件移动到工作目录下
            file.copy(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_Clinical.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
            
            #导入并预处理上传的临床数据集
            clin_info <- read.delim(paste0("TCGA_",cancers[ca],"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
            #clin_info <- clin_info[which(clin_info$SampleType=="PrimaryTumor"),]
            #rownames(clin_info) <- paste0(clin_info$bcr_patient_barcode,"-",sapply(strsplit(clin_info$sampleID_XSU,"-",fixed = T),"[",3))
            colnames(clin_info)[1] <- "Tumor_Sample_Barcode"
            clin_info$Tumor_Sample_Barcode <- paste0(clin_info$Tumor_Sample_Barcode,"-01A")
            clin_info <- clin_info[!duplicated(clin_info$Tumor_Sample_Barcode),]
            rownames(clin_info) <- clin_info[,1]
            write.table(clin_info,paste0("TCGA_",cancers[ca],"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
            
            #Combine clinical files from different cancers
            if(ca==1){
              clin.info <- clin_info
            }else{
              intersect_clin <- intersect(colnames(clin.info),colnames(clin_info))
              clin.info <- rbind(clin.info[,intersect_clin],clin_info[,intersect_clin])
            }
            
            gc() #释放内存
          
        }
        write.table(clin.info,paste0("TCGA_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        data_list$"clin.info" <- clin.info
        
        progress$set(value = 3)
        #准备生存数据
        surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        ##循环合并各类型癌症的生存数据
        surv.infos <- data.frame()
        for(ca in 1:length(cancers)){
          surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
          surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
          rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
          write.table(surv.info,paste0("TCGA_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
          
          #Combine survival files from different cancers
          if(ca==1){
            surv.infos <- surv.info
          }else{
            intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
            surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ## Final save
        write.table(surv.infos,paste0("TCGA_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"surv.infos" <- surv.infos
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_clisur_download3<-downloadHandler(
        filename = "clisur_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$clisur_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0("Clinical & Survival datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse((1 %in% input$multiOmics),"mRNA","lncRNA"))," datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("external_clisur_download3",strong("Download RData file of Clinical & Survival datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$clisur_finish<-renderUI({
        
        if((!(1 %in% input$multiOmics)) & (!(2 %in% input$multiOmics)) & (!(3 %in% input$multiOmics)) & (!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #24.临床生存数据集上传处理按钮(验证数据集)
    observeEvent(input$validation_cli_sur_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded Clinical & Survival(Validation) datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$validation_cli_sur_fileUpload1$datapath #合并临床数据集上传路径
      upload_paths2 <- input$validation_cli_sur_fileUpload2$datapath #各cancerType临床数据集上传路径
      
      # 临床和生存数据集转移及预处理
      data_list <- list() #数据集
      if(input$validation_cli_sur_upload_sources==1){
        
        progress$set(value = 2)
        #合并临床数据集
        # file.move(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_Clinical.txt"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_Clinical.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理上传的合并临床数据集
        clin.info <- read.delim(paste0("Validation_",project,"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        #clin.info <- clin.info[which(clin.info$SampleType=="PrimaryTumor"),]
        #rownames(clin.info) <- paste0(clin.info$bcr_patient_barcode,"-",sapply(strsplit(clin.info$sampleID_XSU,"-",fixed = T),"[",3))
        colnames(clin.info)[1] <- "Tumor_Sample_Barcode"
        clin.info$Tumor_Sample_Barcode <- paste0(clin.info$Tumor_Sample_Barcode,"-01A")
        clin.info <- clin.info[!duplicated(clin.info$Tumor_Sample_Barcode),]
        rownames(clin.info) <- clin.info[,1]
        write.table(clin.info,paste0("Validation_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        data_list$"clin.info" <- clin.info
        
        progress$set(value = 4)
        #准备生存数据
        surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        ##循环合并各类型癌症的生存数据
        surv.infos <- data.frame()
        for(ca in 1:length(cancers)){
          surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
          surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
          rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
          write.table(surv.info,paste0("Validation_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
          
          #Combine survival files from different cancers
          if(ca==1){
            surv.infos <- surv.info
          }else{
            intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
            surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
          }
          
          gc() #释放内存
          
        }
        
        ## Final save
        write.table(surv.infos,paste0("Validation_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"surv.infos" <- surv.infos
        
      }else{
        
        progress$set(value = 2)
        #循环拷贝并预处理各cancerType的临床数据
        clin.info <- data.frame()
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_Clinical.txt"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_Clinical.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理上传的临床数据集
          clin_info <- read.delim(paste0("Validation_",cancers[ca],"_Clinical.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          #clin_info <- clin_info[which(clin_info$SampleType=="PrimaryTumor"),]
          #rownames(clin_info) <- paste0(clin_info$bcr_patient_barcode,"-",sapply(strsplit(clin_info$sampleID_XSU,"-",fixed = T),"[",3))
          colnames(clin_info)[1] <- "Tumor_Sample_Barcode"
          clin_info$Tumor_Sample_Barcode <- paste0(clin_info$Tumor_Sample_Barcode,"-01A")
          clin_info <- clin_info[!duplicated(clin_info$Tumor_Sample_Barcode),]
          rownames(clin_info) <- clin_info[,1]
          write.table(clin_info,paste0("Validation_",cancers[ca],"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
          
          #Combine clinical files from different cancers
          if(ca==1){
            clin.info <- clin_info
          }else{
            intersect_clin <- intersect(colnames(clin.info),colnames(clin_info))
            clin.info <- rbind(clin.info[,intersect_clin],clin_info[,intersect_clin])
          }
          
          gc() #释放内存
          
        }
        write.table(clin.info,paste0("Validation_",project,"_Clinical_preprocess.txt"), quote=F, row.names=F,sep = "\t")
        data_list$"clin.info" <- clin.info
        
        progress$set(value = 3)
        #准备生存数据
        surv.info_origin <- read.table("pancancerSurvivalData_XLu.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        ##循环合并各类型癌症的生存数据
        surv.infos <- data.frame()
        for(ca in 1:length(cancers)){
          surv.info <- as.data.frame(surv.info_origin[which(surv.info_origin$type == cancers[ca]),c(3,6,25:36)])
          surv.info <- surv.info[which(surv.info$OS.time > 0),] # get survival time greater than 0
          rownames(surv.info) <- paste0(rownames(surv.info),"-01A")
          write.table(surv.info,paste0("Validation_",cancers[ca],"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #保存数据
          
          #Combine survival files from different cancers
          if(ca==1){
            surv.infos <- surv.info
          }else{
            intersect_surv <- intersect(colnames(surv.infos),colnames(surv.info))
            surv.infos <- rbind(surv.infos[,intersect_surv],surv.info[,intersect_surv])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ## Final save
        write.table(surv.infos,paste0("Validation_",project,"_Survival.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"surv.infos" <- surv.infos
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_clisur_download2<-downloadHandler(
        filename = "clisur_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_clisur_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0("Clinical & Survival(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA(Validation)",ifelse((1 %in% input$multiOmics),"mRNA(Validation)","lncRNA(Validation)"))," datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("validation_clisur_download2",strong("Download RData file of Clinical & Survival datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_clisur_finish<-renderUI({
        
        if((!(1 %in% input$validation_multiOmics)) & (!(2 %in% input$validation_multiOmics)) & (!(3 %in% input$validation_multiOmics)) & (!(4 %in% input$validation_multiOmics)) & (!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #25.RNA数据集自动下载处理按钮
    observeEvent(input$RNA_process1,{
        
        progress <- Progress$new(session, min=1, max=4) #设置进度条
        on.exit(progress$close())
        progress$set(message = paste0("Download ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse(1 %in% input$multiOmics,"mRNA","lncRNA"))," datasets automatically"),
                     detail = 'This may take a while...')
        progress$set(value = 1)
        
        #每点击一次按钮就重新进行数据的下载
        workdir <- getwd()
        workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
        cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
        cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        
        #RNA数据集下载及预处理
        data_list <- list() #数据集
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        
        progress$set(value = 2)
        expquery <- GDCquery(project = paste0("TCGA-",cancers),
                             data.category = "Transcriptome Profiling",
                             data.type = "Gene Expression Quantification",
                             workflow.type = "STAR - Counts"
        )
        GDCdownload(expquery,directory = "GDCdata")
        expquery2 <- GDCprepare(expquery,directory = "GDCdata",summarizedExperiment = T)
        save(expquery2,file = paste0(project,".gdc.RData"))
        gc() #释放内存
        
        progress$set(value = 3)
        expMatrix <- TCGAanalyze_Preprocessing(expquery2) #上传此文件
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("TCGA_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            data_list$"tpm.mrna" <- tpm.mrna
            write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            data_list$"tpm.lncrna" <- tpm.lncrna
            
        }else if(1 %in% input$multiOmics){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            data_list$"tpm.mrna" <- tpm.mrna
            
        }else{
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- TCGAanalyze_Preprocessing(expquery2)
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("TCGA_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            data_list$"tpm.mrna.norm" <- tpm.mrna.norm
            write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
            
        }else if(1 %in% input$multiOmics){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            data_list$"tpm.mrna.norm" <- tpm.mrna.norm
            
        }else{
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
        
        system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
        
        ##RData结果下载按钮响应事件
        output$external_RNA_download1<-downloadHandler(
          filename = "RNA_preparation(TCGA).RData",
          content = function(theFile) {
            save(data_list,file=theFile)
          }
        )
        
        #数据下载好后给予用户的反馈提示
        output$RNA_download_finish<-renderUI({
            
            list(br(),column(12,h3(strong(paste0(ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse(1 %in% input$multiOmics,"mRNA","lncRNA"))," datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare DNA methylation datasets if you have specified."))),style="color:darkblack"),br(),
                 column(12,downloadButton("external_RNA_download1",strong("Download RData file of RNA datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        })
        
        output$RNA_finish<-renderUI({
          
          if((!(3 %in% input$multiOmics)) & (!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
            list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
          }
        })
        
        #rm(list=ls()) #清空变量
        gc() #释放内存
    })
    
    #26.RNA数据集链接下载处理按钮
    observeEvent(input$RNA_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = paste0("Download ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse(1 %in% input$multiOmics,"mRNA","lncRNA"))," datasets using specified urls"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.RNAUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #RNA数据集下载及预处理
      data_list <- list() #数据集
      if(input$RNA_url_sources==1){
        
        progress$set(value = 2)
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        
        #合并RNA数据下载
        RNA_url <- input$RNA_downloadUrl
        RNA_url <- trimws(RNA_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(RNA_url,method="auto",destfile = file.path(workdir,paste0("TCGA_",project,"_expMatrix.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        #导入并预处理下载好的RNA数据集
        expMatrix <- read.table(paste0("TCGA_",project,"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("TCGA_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable 
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- read.table(file.path(workdir,paste0("TCGA_",project,"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("TCGA_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的RNA数据
        expMatrixs <- data.frame()
        for(url in 1:length(download_urls)){
          
          #Load gene annotation
          Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
          rownames(Ginfo) <- Ginfo[,1]
          
          download.file(download_urls[url],method="auto",destfile = file.path(workdir,paste0("TCGA_",cancers[url],"_expMatrix.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          
          #导入并预处理下载好的RNA数据集
          expMatrix <- read.table(paste0("TCGA_",cancers[url],"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
          
          #Combine expression matrix from different cancers
          if(url==1){
            expMatrixs <- expMatrix
          }else{
            intersect_expMatrix <- intersect(rownames(expMatrixs),rownames(expMatrix))
            expMatrixs <- cbind(expMatrixs[intersect_expMatrix,],expMatrix[intersect_expMatrix,])
          }
          
          colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
          normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
          tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
          
          #Tumor sample
          expMatrix <- expMatrix[,tumorsamples]
          expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
          #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
          comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
          count <- as.data.frame(expMatrix)[comgene,]
          Ginfo <- Ginfo[comgene,]
          
          #Extract mRNA and lncRNA
          Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          
          #count data
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            count <- count[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            count <- count[rownames(Mid),]
          }else{
            count <- count[rownames(Lid),]
          }
          
          countsTable <- count
          countsTable$Gene <- Ginfo[rownames(count),"genename"]
          countsTable <- countsTable[!duplicated(countsTable$Gene),]
          Ginfo_origin <- Ginfo
          Ginfo <- Ginfo[rownames(countsTable),]
          Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          Lids_gene <- Lids$genename
          Mids_gene <- Mids$genename
          rownames(countsTable) <- countsTable$Gene
          countsTable <- countsTable[,-ncol(countsTable)]
          write.table(countsTable,paste0("TCGA_",cancers[url],"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            
            write.table(fpkm.mrna,paste0("TCGA_",cancers[url],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna,paste0("TCGA_",cancers[url],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            write.table(fpkm.mrna,paste0("TCGA_",cancers[url],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            write.table(fpkm.lncrna,paste0("TCGA_",cancers[url],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("TCGA_",cancers[url],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(tpm.lncrna,paste0("TCGA_",cancers[url],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("TCGA_",cancers[url],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.lncrna,paste0("TCGA_",cancers[url],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ##Normal sample
          ###count
          tmp <- read.table(file.path(workdir,paste0("TCGA_",cancers[url],"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
          colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
          tmp <- tmp[,normsamples]
          tmp <- as.data.frame(tmp[rownames(count),])
          
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            tmp <- tmp[rownames(Mid),]
          }else{
            tmp <- tmp[rownames(Lid),]
          }
          
          tmpTable <- tmp
          tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
          tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
          rownames(tmpTable) <- tmpTable$Gene
          tmpTable <- tmpTable[,-ncol(tmpTable)]
          write.table(tmpTable,paste0("TCGA_",cancers[url],"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            
            write.table(fpkm.mrna.norm,paste0("TCGA_",cancers[url],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna.norm,paste0("TCGA_",cancers[url],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            write.table(fpkm.mrna.norm,paste0("TCGA_",cancers[url],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            write.table(fpkm.lncrna.norm,paste0("TCGA_",cancers[url],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("TCGA_",cancers[url],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            write.table(tpm.lncrna.norm,paste0("TCGA_",cancers[url],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("TCGA_",cancers[url],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else{
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.lncrna.norm,paste0("TCGA_",cancers[url],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        #对合并后的RNA数据集进行处理
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        expMatrix <- expMatrixs
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("TCGA_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- expMatrixs
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("TCGA_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
      }   
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_RNA_download2<-downloadHandler(
        filename = "RNA_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$RNA_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0(ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse(1 %in% input$multiOmics,"mRNA","lncRNA"))," datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare DNA methylation datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("external_RNA_download2",strong("Download RData file of RNA datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$RNA_finish<-renderUI({
        
        if((!(3 %in% input$multiOmics)) & (!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #27.RNA数据集链接下载处理按钮(验证数据集)
    observeEvent(input$validation_RNA_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = paste0("Download ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA(Validation)",ifelse(1 %in% input$multiOmics,"mRNA(Validation)","lncRNA(Validation)"))," datasets using specified urls"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.validation_RNAUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #RNA数据集下载及预处理
      data_list <- list() #数据集
      if(input$validation_RNA_url_sources==1){
        
        progress$set(value = 2)
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        
        #合并RNA数据下载
        RNA_url <- input$validation_RNA_downloadUrl
        RNA_url <- trimws(RNA_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(RNA_url,method="auto",destfile = file.path(validation_workdir,paste0("Validation_",project,"_expMatrix.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        #导入并预处理下载好的RNA数据集
        expMatrix <- read.table(paste0("Validation_",project,"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("Validation_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- read.table(file.path(workdir,paste0("Validation_",project,"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("Validation_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的RNA数据
        expMatrixs <- data.frame()
        for(url in 1:length(download_urls)){
          
          #Load gene annotation
          Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
          rownames(Ginfo) <- Ginfo[,1]
          
          download.file(download_urls[url],method="auto",destfile = file.path(validation_workdir,paste0("Validation_",cancers[url],"_expMatrix.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          
          #导入并预处理下载好的RNA数据集
          expMatrix <- read.table(paste0("Validation_",cancers[url],"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
          
          #Combine expression matrix from different cancers
          if(url==1){
            expMatrixs <- expMatrix
          }else{
            intersect_expMatrix <- intersect(rownames(expMatrixs),rownames(expMatrix))
            expMatrixs <- cbind(expMatrixs[intersect_expMatrix,],expMatrix[intersect_expMatrix,])
          }
          
          colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
          normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
          tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
          
          #Tumor sample
          expMatrix <- expMatrix[,tumorsamples]
          expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
          #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
          comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
          count <- as.data.frame(expMatrix)[comgene,]
          Ginfo <- Ginfo[comgene,]
          
          #Extract mRNA and lncRNA
          Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          
          #count data
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            count <- count[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            count <- count[rownames(Mid),]
          }else{
            count <- count[rownames(Lid),]
          }
          
          countsTable <- count
          countsTable$Gene <- Ginfo[rownames(count),"genename"]
          countsTable <- countsTable[!duplicated(countsTable$Gene),]
          Ginfo_origin <- Ginfo
          Ginfo <- Ginfo[rownames(countsTable),]
          Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          Lids_gene <- Lids$genename
          Mids_gene <- Mids$genename
          rownames(countsTable) <- countsTable$Gene
          countsTable <- countsTable[,-ncol(countsTable)]
          write.table(countsTable,paste0("Validation_",cancers[url],"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            
            write.table(fpkm.mrna,paste0("Validation_",cancers[url],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna,paste0("Validation_",cancers[url],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            write.table(fpkm.mrna,paste0("Validation_",cancers[url],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            write.table(fpkm.lncrna,paste0("Validation_",cancers[url],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("Validation_",cancers[url],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(tpm.lncrna,paste0("Validation_",cancers[url],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("Validation_",cancers[url],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.lncrna,paste0("Validation_",cancers[url],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ##Normal sample
          ###count
          tmp <- read.table(file.path(workdir,paste0("Validation_",cancers[url],"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入下载好的RNA数据集
          colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
          tmp <- tmp[,normsamples]
          tmp <- as.data.frame(tmp[rownames(count),])
          
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            tmp <- tmp[rownames(Mid),]
          }else{
            tmp <- tmp[rownames(Lid),]
          }
          
          tmpTable <- tmp
          tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
          tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
          rownames(tmpTable) <- tmpTable$Gene
          tmpTable <- tmpTable[,-ncol(tmpTable)]
          write.table(tmpTable,paste0("Validation_",cancers[url],"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            
            write.table(fpkm.mrna.norm,paste0("Validation_",cancers[url],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna.norm,paste0("Validation_",cancers[url],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            write.table(fpkm.mrna.norm,paste0("Validation_",cancers[url],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            write.table(fpkm.lncrna.norm,paste0("Validation_",cancers[url],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("Validation_",cancers[url],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            write.table(tpm.lncrna.norm,paste0("Validation_",cancers[url],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("Validation_",cancers[url],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else{
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.lncrna.norm,paste0("Validation_",cancers[url],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        #对合并后的RNA数据集进行处理
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        expMatrix <- expMatrixs
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("Validation_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- expMatrixs
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("Validation_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
      }   
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_RNA_download1<-downloadHandler(
        filename = "RNA_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_RNA_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0(ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA(Validation)",ifelse(1 %in% input$multiOmics,"mRNA(Validation)","lncRNA(Validation)"))," datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare DNA methylation(Validation) datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("validation_RNA_download1",strong("Download RData file of RNA datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_RNA_finish<-renderUI({
        
        if((!(3 %in% input$validation_multiOmics)) & (!(4 %in% input$validation_multiOmics)) & (!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #28.RNA数据集文件上传处理按钮
    observeEvent(input$RNA_process3,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = paste0("Upload downloaded ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse(1 %in% input$multiOmics,"mRNA","lncRNA"))," datasets from local paths"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$RNA_fileUpload1$datapath #合并RNA数据集上传路径
      upload_paths2 <- input$RNA_fileUpload2$datapath #各cancerType的RNA数据集上传路径
      
      # RNA数据集转移及预处理
      data_list <- list() #数据集
      if(input$RNA_upload_sources==1){
        
        #合并RNA数据集
        progress$set(value = 2)
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
       
        # file.move(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_expMatrix.txt"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_expMatrix.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并RNA数据集
        expMatrix <- read.table(paste0("TCGA_",project,"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的合并RNA数据集
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("TCGA_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- read.table(file.path(workdir,paste0("TCGA_",project,"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的RNA数据集
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("TCGA_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
        
      }else{
        
          progress$set(value = 2)
          #循环拷贝并预处理各cancerType的RNA数据
          for(ca in 1:length(cancers)){
            
            #Load gene annotation
            Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
            rownames(Ginfo) <- Ginfo[,1]
            
            # file.move(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_expMatrix.txt"))) #将上传的文件移动到工作目录下
            file.copy(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_expMatrix.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
            
            #导入并预处理已上传的RNA数据集
            expMatrix <- read.table(paste0("TCGA_",cancers[ca],"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的RNA数据集
            
            #Combine expression matrix from different cancers
            if(ca==1){
              expMatrixs <- expMatrix
            }else{
              intersect_expMatrix <- intersect(rownames(expMatrixs),rownames(expMatrix))
              expMatrixs <- cbind(expMatrixs[intersect_expMatrix,],expMatrix[intersect_expMatrix,])
            }
            
            colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
            normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
            tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
            
            #Tumor sample
            expMatrix <- expMatrix[,tumorsamples]
            expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
            #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
            comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
            count <- as.data.frame(expMatrix)[comgene,]
            Ginfo <- Ginfo[comgene,]
            
            #Extract mRNA and lncRNA
            Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
            Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
            
            #count data
            if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
              count <- count[c(rownames(Mid),rownames(Lid)),]
            }else if(1 %in% input$multiOmics){
              count <- count[rownames(Mid),]
            }else{
              count <- count[rownames(Lid),]
            }
            
            countsTable <- count
            countsTable$Gene <- Ginfo[rownames(count),"genename"]
            countsTable <- countsTable[!duplicated(countsTable$Gene),]
            Ginfo_origin <- Ginfo
            Ginfo <- Ginfo[rownames(countsTable),]
            Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
            Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
            Lids_gene <- Lids$genename
            Mids_gene <- Mids$genename
            rownames(countsTable) <- countsTable$Gene
            countsTable <- countsTable[,-ncol(countsTable)]
            write.table(countsTable,paste0("TCGA_",cancers[ca],"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
            ###FPKM
            if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
              fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
              rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
              fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
              rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
              
              write.table(fpkm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              write.table(fpkm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              
            }else if(1 %in% input$multiOmics){
              fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
              rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
              write.table(fpkm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              
            }else{
              fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
              rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
              write.table(fpkm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            }
            
            ###TPM
            if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
              tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
              tpm.mrna$Gene <- rownames(tpm.mrna)
              tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
              
              tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
              tpm.lncrna$Gene <- rownames(tpm.lncrna)
              tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
              
              write.table(tpm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              write.table(tpm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              
            }else if(1 %in% input$multiOmics){
              tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
              tpm.mrna$Gene <- rownames(tpm.mrna)
              tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
              
              write.table(tpm.mrna,paste0("TCGA_",cancers[ca],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              
            }else{
              tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
              tpm.lncrna$Gene <- rownames(tpm.lncrna)
              tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
              
              write.table(tpm.lncrna,paste0("TCGA_",cancers[ca],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            }
            
            ##Normal sample
            ###count
            tmp <- read.table(file.path(workdir,paste0("TCGA_",cancers[ca],"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的RNA数据集
            colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
            tmp <- tmp[,normsamples]
            tmp <- as.data.frame(tmp[rownames(count),])
            
            if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
              tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
            }else if(1 %in% input$multiOmics){
              tmp <- tmp[rownames(Mid),]
            }else{
              tmp <- tmp[rownames(Lid),]
            }
            
            tmpTable <- tmp
            tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
            tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
            rownames(tmpTable) <- tmpTable$Gene
            tmpTable <- tmpTable[,-ncol(tmpTable)]
            write.table(tmpTable,paste0("TCGA_",cancers[ca],"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
            ###FPKM
            if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
              fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
              rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
              fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
              rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
              
              write.table(fpkm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              write.table(fpkm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              
            }else if(1 %in% input$multiOmics){
              fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
              rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
              write.table(fpkm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
              
            }else{
              fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
              rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
              write.table(fpkm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            }
            
            ###TPM
            if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
              tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
              tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
              tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
              
              tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
              tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
              tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
              
              write.table(tpm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
              write.table(tpm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
              
            }else if(1 %in% input$multiOmics){
              tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
              tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
              tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
              
              write.table(tpm.mrna.norm,paste0("TCGA_",cancers[ca],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
              
            }else{
              tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
              tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
              tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
              
              write.table(tpm.lncrna.norm,paste0("TCGA_",cancers[ca],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            }
            
            gc() #释放内存
          }
          
        progress$set(value = 3)
        #对合并后的RNA数据集进行处理
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        expMatrix <- expMatrixs
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("TCGA_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("TCGA_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("TCGA_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("TCGA_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("TCGA_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- expMatrixs
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("TCGA_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("TCGA_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("TCGA_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
          
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_RNA_download3<-downloadHandler(
        filename = "RNA_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$RNA_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0(ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA",ifelse(1 %in% input$multiOmics,"mRNA","lncRNA"))," datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare DNA methylation datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("external_RNA_download3",strong("Download RData file of RNA datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$RNA_finish<-renderUI({
        
        if((!(3 %in% input$multiOmics)) & (!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #29.RNA数据集文件上传处理按钮(验证数据集)
    observeEvent(input$validation_RNA_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = paste0("Upload downloaded ",ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA(Validation)",ifelse(1 %in% input$multiOmics,"mRNA(Validation)","lncRNA(Validation)"))," datasets from local paths"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$validation_RNA_fileUpload1$datapath #合并RNA数据集上传路径
      upload_paths2 <- input$validation_RNA_fileUpload2$datapath #各cancerType的RNA数据集上传路径
      
      # RNA数据集转移及预处理
      data_list <- list() #数据集
      if(input$validation_RNA_upload_sources==1){
        
        #合并RNA数据集
        progress$set(value = 2)
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        
        # file.move(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_expMatrix.txt"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_expMatrix.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并RNA数据集
        expMatrix <- read.table(paste0("Validation_",project,"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的合并RNA数据集
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("Validation_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- read.table(file.path(workdir,paste0("Validation_",project,"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的RNA数据集
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("Validation_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
        
      }else{
        
        progress$set(value = 2)
        #循环拷贝并预处理各cancerType的RNA数据
        for(ca in 1:length(cancers)){
          
          #Load gene annotation
          Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
          rownames(Ginfo) <- Ginfo[,1]
          
          # file.move(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_expMatrix.txt"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_expMatrix.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的RNA数据集
          expMatrix <- read.table(paste0("Validation_",cancers[ca],"_expMatrix.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的RNA数据集
          
          #Combine expression matrix from different cancers
          if(ca==1){
            expMatrixs <- expMatrix
          }else{
            intersect_expMatrix <- intersect(rownames(expMatrixs),rownames(expMatrix))
            expMatrixs <- cbind(expMatrixs[intersect_expMatrix,],expMatrix[intersect_expMatrix,])
          }
          
          colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
          normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
          tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
          
          #Tumor sample
          expMatrix <- expMatrix[,tumorsamples]
          expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
          #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
          comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
          count <- as.data.frame(expMatrix)[comgene,]
          Ginfo <- Ginfo[comgene,]
          
          #Extract mRNA and lncRNA
          Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          
          #count data
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            count <- count[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            count <- count[rownames(Mid),]
          }else{
            count <- count[rownames(Lid),]
          }
          
          countsTable <- count
          countsTable$Gene <- Ginfo[rownames(count),"genename"]
          countsTable <- countsTable[!duplicated(countsTable$Gene),]
          Ginfo_origin <- Ginfo
          Ginfo <- Ginfo[rownames(countsTable),]
          Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
          Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
          Lids_gene <- Lids$genename
          Mids_gene <- Mids$genename
          rownames(countsTable) <- countsTable$Gene
          countsTable <- countsTable[,-ncol(countsTable)]
          write.table(countsTable,paste0("Validation_",cancers[ca],"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            
            write.table(fpkm.mrna,paste0("Validation_",cancers[ca],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna,paste0("Validation_",cancers[ca],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
            write.table(fpkm.mrna,paste0("Validation_",cancers[ca],"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
            write.table(fpkm.lncrna,paste0("Validation_",cancers[ca],"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("Validation_",cancers[ca],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(tpm.lncrna,paste0("Validation_",cancers[ca],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
            tpm.mrna$Gene <- rownames(tpm.mrna)
            tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.mrna,paste0("Validation_",cancers[ca],"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
            tpm.lncrna$Gene <- rownames(tpm.lncrna)
            tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
            
            write.table(tpm.lncrna,paste0("Validation_",cancers[ca],"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ##Normal sample
          ###count
          tmp <- read.table(file.path(workdir,paste0("Validation_",cancers[ca],"_expMatrix.txt")),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #导入已上传的RNA数据集
          colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
          tmp <- tmp[,normsamples]
          tmp <- as.data.frame(tmp[rownames(count),])
          
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
          }else if(1 %in% input$multiOmics){
            tmp <- tmp[rownames(Mid),]
          }else{
            tmp <- tmp[rownames(Lid),]
          }
          
          tmpTable <- tmp
          tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
          tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
          rownames(tmpTable) <- tmpTable$Gene
          tmpTable <- tmpTable[,-ncol(tmpTable)]
          write.table(tmpTable,paste0("Validation_",cancers[ca],"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          
          ###FPKM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            
            write.table(fpkm.mrna.norm,paste0("Validation_",cancers[ca],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            write.table(fpkm.lncrna.norm,paste0("Validation_",cancers[ca],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else if(1 %in% input$multiOmics){
            fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
            rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
            write.table(fpkm.mrna.norm,paste0("Validation_",cancers[ca],"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
            
          }else{
            fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
            rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
            write.table(fpkm.lncrna.norm,paste0("Validation_",cancers[ca],"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          }
          
          ###TPM
          if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("Validation_",cancers[ca],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            write.table(tpm.lncrna.norm,paste0("Validation_",cancers[ca],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else if(1 %in% input$multiOmics){
            tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
            tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
            tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.mrna.norm,paste0("Validation_",cancers[ca],"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
            
          }else{
            tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
            tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
            tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
            
            write.table(tpm.lncrna.norm,paste0("Validation_",cancers[ca],"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        #对合并后的RNA数据集进行处理
        #Load gene annotation
        Ginfo <- read.table("overlapTable27.txt",sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T) #此文件放置于公共目录下
        rownames(Ginfo) <- Ginfo[,1]
        expMatrix <- expMatrixs
        colnames(expMatrix) <- substr(colnames(expMatrix),start = 1,stop = 16)
        normsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="11")] #get normal samples
        tumorsamples <- colnames(expMatrix)[which(substr(colnames(expMatrix),14,15)=="01")] #get tumor samples
        
        #Tumor sample
        expMatrix <- expMatrix[,tumorsamples]
        expMatrix <- as.data.frame(expMatrix[rowSums(expMatrix)>0,])
        #expMatrix <- expMatrix[apply(expMatrix,1,function(x) sum(x>0)>0.9*length(tumorsamples)),]
        comgene <- intersect(rownames(expMatrix),rownames(Ginfo))
        count <- as.data.frame(expMatrix)[comgene,]
        Ginfo <- Ginfo[comgene,]
        
        #Extract mRNA and lncRNA
        Lid <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mid <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        
        #count data
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          count <- count[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          count <- count[rownames(Mid),]
        }else{
          count <- count[rownames(Lid),]
        }
        
        countsTable <- count
        countsTable$Gene <- Ginfo[rownames(count),"genename"]
        countsTable <- countsTable[!duplicated(countsTable$Gene),]
        Ginfo_origin <- Ginfo
        Ginfo <- Ginfo[rownames(countsTable),]
        Lids <- Ginfo[Ginfo$genetype %in% c("non_coding","3prime_overlapping_ncRNA","antisense_RNA","lincRNA","sense_intronic","sense_overlapping","macro_lncRNA","bidirectional_promoter_lncRNA"),] #lncRNA
        Mids <- Ginfo[Ginfo$genetype == "protein_coding",] #mRNA
        Lids_gene <- Lids$genename
        Mids_gene <- Mids$genename
        rownames(countsTable) <- countsTable$Gene
        countsTable <- countsTable[,-ncol(countsTable)]
        write.table(countsTable,paste0("Validation_",project,"_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"tumor_counts" <- countsTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Mids),]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna) <- rownames(countsTable[Mids_gene,])
          write.table(fpkm.mrna,paste0("Validation_",project,"_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna" <- fpkm.mrna
          
        }else{
          fpkm.lncrna <- as.data.frame(apply(as.matrix(countsTable[rownames(Lids),]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna) <- rownames(countsTable[Lids_gene,])
          write.table(fpkm.lncrna,paste0("Validation_",project,"_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna" <- fpkm.lncrna
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna <- as.data.frame(round(apply(fpkm.mrna,2,fpkmToTpm),2))
          tpm.mrna$Gene <- rownames(tpm.mrna)
          tpm.mrna <- apply(tpm.mrna[,setdiff(colnames(tpm.mrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.mrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.mrna,paste0("Validation_",project,"_mRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.mrna" <- tpm.mrna
          
        }else{
          tpm.lncrna <- as.data.frame(round(apply(fpkm.lncrna,2,fpkmToTpm),2))
          tpm.lncrna$Gene <- rownames(tpm.lncrna)
          tpm.lncrna <- apply(tpm.lncrna[,setdiff(colnames(tpm.lncrna),"Gene")],2,function(x) tapply(x,INDEX = factor(tpm.lncrna$Gene),FUN = median,na.rm=TRUE))
          
          write.table(tpm.lncrna,paste0("Validation_",project,"_lncRNA_TPM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"tpm.lncrna" <- tpm.lncrna
        }
        
        progress$set(value = 4)
        ##Normal sample
        ###count
        tmp <- expMatrixs
        colnames(tmp) <- substr(colnames(tmp), start = 1,stop = 16)
        tmp <- tmp[,normsamples]
        tmp <- as.data.frame(tmp[rownames(count),])
        
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tmp <- tmp[c(rownames(Mid),rownames(Lid)),]
        }else if(1 %in% input$multiOmics){
          tmp <- tmp[rownames(Mid),]
        }else{
          tmp <- tmp[rownames(Lid),]
        }
        
        tmpTable <- tmp
        tmpTable$Gene <- Ginfo_origin[rownames(tmp),"genename"]
        tmpTable <- tmpTable[!duplicated(tmpTable$Gene),]
        rownames(tmpTable) <- tmpTable$Gene
        tmpTable <- tmpTable[,-ncol(tmpTable)]
        write.table(tmpTable,paste0("Validation_",project,"_norm_Count.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        data_list$"normal_counts" <- tmpTable
        
        ###FPKM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          fpkm.mrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Mids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Mids),"unqlen"]))
          rownames(fpkm.mrna.norm) <- rownames(tmpTable[Mids_gene,])
          write.table(fpkm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.mrna.norm" <- fpkm.mrna.norm
          
        }else{
          fpkm.lncrna.norm <- as.data.frame(apply(as.matrix(tmpTable[Lids_gene,]), 2, countToFpkm, effLen = Ginfo[rownames(Lids),"unqlen"]))
          rownames(fpkm.lncrna.norm) <- rownames(tmpTable[Lids_gene,])
          write.table(fpkm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_FPKM.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
          data_list$"fpkm.lncrna.norm" <- fpkm.lncrna.norm
        }
        
        ###TPM
        if((1 %in% input$multiOmics) & (2 %in% input$multiOmics)){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
          
        }else if(1 %in% input$multiOmics){
          tpm.mrna.norm <- as.data.frame(round(apply(fpkm.mrna.norm,2,fpkmToTpm),2))
          tpm.mrna.norm$Gene <- rownames(tpm.mrna.norm)
          tpm.mrna.norm <- apply(tpm.mrna.norm[,setdiff(colnames(tpm.mrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.mrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.mrna.norm,paste0("Validation_",project,"_norm_mRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.mrna.norm" <- tpm.mrna.norm
          
        }else{
          tpm.lncrna.norm <- as.data.frame(round(apply(fpkm.lncrna.norm,2,fpkmToTpm),2))
          tpm.lncrna.norm$Gene <- rownames(tpm.lncrna.norm)
          tpm.lncrna.norm <- apply(tpm.lncrna.norm[,setdiff(colnames(tpm.lncrna.norm), "Gene")], 2, function(x) tapply(x, INDEX=factor(tpm.lncrna.norm$Gene), FUN=median, na.rm=TRUE))
          
          write.table(tpm.lncrna.norm,paste0("Validation_",project,"_norm_lncRNA_TPM.txt"), quote=F, row.names=T,col.names = NA,sep = "\t")
          data_list$"tpm.lncrna.norm" <- tpm.lncrna.norm
        }
        
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_RNA_download2<-downloadHandler(
        filename = "RNA_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_RNA_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong(paste0(ifelse((1 %in% input$multiOmics) & (2 %in% input$multiOmics),"mRNA & lncRNA(Validation)",ifelse(1 %in% input$multiOmics,"mRNA(Validation)","lncRNA(Validation)"))," datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare DNA methylation(Validation) datasets if you have specified."))),style="color:darkblack"),br(),
             column(12,downloadButton("validation_RNA_download2",strong("Download RData file of RNA datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_RNA_finish<-renderUI({
        
        if((!(3 %in% input$validation_multiOmics)) & (!(4 %in% input$validation_multiOmics)) & (!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #30.DNA methylation数据集自动下载处理按钮
    observeEvent(input$DNA_methylation_process1,{
        
        progress <- Progress$new(session, min=1, max=3) #设置进度条
        on.exit(progress$close())
        progress$set(message = 'Download DNA methylation dataset automatically',
                     detail = 'This may take a while...')
        progress$set(value = 1)
        
        #每点击一次按钮就重新进行数据的下载
        workdir <- getwd()
        workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
        cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
        cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        
        progress$set(value = 2)
        #DNA methylation数据集下载
        data_list <- list() #数据集
        meths <- data.frame()
        for(ca in 1:length(cancers)){
            methylation_download_link <- paste0("https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.",cancers[ca],".sampleMap%2FHumanMethylation450.gz")
            download.file(methylation_download_link,method="auto",destfile = file.path(workdir,paste0("TCGA_",cancers[ca],"_Methylation450.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
            meth <- fread(paste0("TCGA_",cancers[ca],"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
            meth <- as.data.frame(meth)
            rownames(meth) <- meth[,1]
            meth <- as.data.frame(na.omit(meth[,-1]))
            #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
            meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
            colnames(meth) <- paste0(colnames(meth),"A")
            
            # Get promoter
            data("probe.features")
            promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
            meth <- round(meth[promoter,],3)
            
            # Map to gene name to take median value if multiple mapping
            meth$gene <- as.character(probe.features[promoter,"gene"])
            meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
            meth <- as.data.frame(round(meth,3))
            
            # Save
            write.table(meth,paste0("TCGA_",cancers[ca],"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
            
            # Combine DNA methylation files from different cancers
            if(ca==1){
                meths <- meth
            }else{
                intersect_meth <- intersect(rownames(meths),rownames(meth))
                meths <- cbind(meths[intersect_meth,],meth[intersect_meth,])
            }
            
            gc() #释放内存
        }
        
        progress$set(value = 3)
        ### Final save
        write.table(meths,paste0("TCGA_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
        
        system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
        
        ##RData结果下载按钮响应事件
        output$external_meth_download1<-downloadHandler(
          filename = "DNA_methylation_preparation(TCGA).RData",
          content = function(theFile) {
            save(data_list,file=theFile)
          }
        )
        
        #数据下载好后给予用户的反馈提示
        output$DNA_methylation_download_finish<-renderUI({
          
          list(br(),column(12,h3(strong("DNA methylation datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare copy number alterations datasets if you have specified.")),style="color:darkblack"),br(),
               column(12,downloadButton("external_meth_download1",strong("Download RData file of DNA methylation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        })
        
        output$DNA_methylation_finish<-renderUI({
          
          if((!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
            list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
          }
        })
        
        #rm(list=ls()) #清空变量
        gc() #释放内存
    })
    
    #31.DNA methylation数据集链接下载处理按钮
    observeEvent(input$DNA_methylation_process2,{
      
        progress <- Progress$new(session, min=1, max=4) #设置进度条
        on.exit(progress$close())
        progress$set(message = 'Download DNA methylation dataset using specified urls',
                     detail = 'This may take a while...')
        progress$set(value = 1)
        
        #每点击一次按钮就重新进行数据的下载
        workdir <- getwd()
        workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
        cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
        cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download_urls <- data.frame(rbind(hot_to_r(input$table.DNAMethylationUrl)))[,1]
        download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        
        #DNA methylation数据集下载及预处理
        data_list <- list() #数据集
        if(input$DNA_methylation_url_sources==1){
          
          progress$set(value = 2)
          meths <- data.frame()
          #合并DNA methylation数据下载
          DNA_methylation_url <- input$DNA_methylation_downloadUrl
          DNA_methylation_url <- trimws(DNA_methylation_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          download.file(DNA_methylation_url,method="auto",destfile = file.path(workdir,paste0("TCGA_",project,"_Methylation450.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          gc() #释放内存
          
          progress$set(value = 3)
          meth <- fread(paste0("TCGA_",project,"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
          meth <- as.data.frame(meth)
          rownames(meth) <- meth[,1]
          meth <- as.data.frame(na.omit(meth[,-1]))
          #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
          meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
          colnames(meth) <- paste0(colnames(meth),"A")
          
          # Get promoter
          data("probe.features")
          promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
          meth <- round(meth[promoter,],3)
          
          # Map to gene name to take median value if multiple mapping
          meth$gene <- as.character(probe.features[promoter,"gene"])
          meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
          meth <- as.data.frame(round(meth,3))
          
          progress$set(value = 4)
          ### Final save
          meths <- meth
          write.table(meths,paste0("TCGA_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          data_list$"meth" <- meths
          
        }else{
          
          progress$set(value = 2)
          #循环下载各cancerType的DNA methylation数据
          meths <- data.frame()
          for(url in 1:length(download_urls)){
            download.file(download_urls[url],method="auto",destfile = file.path(workdir,paste0("TCGA_",cancers[url],"_Methylation450.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
            gc() #释放内存
            
            progress$set(value = 3)
            #导入并预处理下载好的各cancerType的DNA methylation数据集
            meth <- fread(paste0("TCGA_",cancers[url],"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
            meth <- as.data.frame(meth)
            rownames(meth) <- meth[,1]
            meth <- as.data.frame(na.omit(meth[,-1]))
            #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
            meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
            colnames(meth) <- paste0(colnames(meth),"A")
            
            # Get promoter
            data("probe.features")
            promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
            meth <- round(meth[promoter,],3)
            
            # Map to gene name to take median value if multiple mapping
            meth$gene <- as.character(probe.features[promoter,"gene"])
            meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
            meth <- as.data.frame(round(meth,3))
            
            write.table(meth,paste0("TCGA_",cancers[url],"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
            
            # Combine DNA methylation files from different cancers
            if(url==1){
              meths <- meth
            }else{
              intersect_meth <- intersect(rownames(meths),rownames(meth))
              meths <- cbind(meths[intersect_meth,],meth[intersect_meth,])
            }
            
            gc() #释放内存
            
          }
          
          progress$set(value = 4)
          ### Final save
          write.table(meths,paste0("TCGA_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          data_list$"meth" <- meths
        }
        
        system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
        
        ##RData结果下载按钮响应事件
        output$external_meth_download2<-downloadHandler(
          filename = "DNA_methylation_preparation(TCGA).RData",
          content = function(theFile) {
            save(data_list,file=theFile)
          }
        )
        
        #数据下载好后给予用户的反馈提示
        output$DNA_methylation_download_finish<-renderUI({
          
          list(br(),column(12,h3(strong("DNA methylation datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare copy number alterations datasets if you have specified.")),style="color:darkblack"),br(),
               column(12,downloadButton("external_meth_download2",strong("Download RData file of DNA methylation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        })
        
        output$DNA_methylation_finish<-renderUI({
          
          if((!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
            list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
          }
        })
        
        #rm(list=ls()) #清空变量
        gc() #释放内存
    
    })
    
    #32.DNA methylation数据集链接下载处理按钮(验证数据集)
    observeEvent(input$validation_DNA_methylation_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download DNA methylation(Validation) dataset using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.validation_DNAMethylationUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #DNA methylation数据集下载及预处理
      data_list <- list() #数据集
      if(input$validation_DNA_methylation_url_sources==1){
        
        progress$set(value = 2)
        meths <- data.frame()
        #合并DNA methylation数据下载
        DNA_methylation_url <- input$validation_DNA_methylation_downloadUrl
        DNA_methylation_url <- trimws(DNA_methylation_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(DNA_methylation_url,method="auto",destfile = file.path(validation_workdir,paste0("Validation_",project,"_Methylation450.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        meth <- fread(paste0("Validation_",project,"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
        meth <- as.data.frame(meth)
        rownames(meth) <- meth[,1]
        meth <- as.data.frame(na.omit(meth[,-1]))
        #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
        meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
        colnames(meth) <- paste0(colnames(meth),"A")
        
        # Get promoter
        data("probe.features")
        promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
        meth <- round(meth[promoter,],3)
        
        # Map to gene name to take median value if multiple mapping
        meth$gene <- as.character(probe.features[promoter,"gene"])
        meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
        meth <- as.data.frame(round(meth,3))
        
        progress$set(value = 4)
        ### Final save
        meths <- meth
        write.table(meths,paste0("Validation_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的DNA methylation数据
        meths <- data.frame()
        for(url in 1:length(download_urls)){
          download.file(download_urls[url],method="auto",destfile = file.path(validation_workdir,paste0("Validation_",cancers[url],"_Methylation450.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          gc() #释放内存
          
          progress$set(value = 3)
          #导入并预处理下载好的各cancerType的DNA methylation数据集
          meth <- fread(paste0("Validation_",cancers[url],"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
          meth <- as.data.frame(meth)
          rownames(meth) <- meth[,1]
          meth <- as.data.frame(na.omit(meth[,-1]))
          #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
          meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
          colnames(meth) <- paste0(colnames(meth),"A")
          
          # Get promoter
          data("probe.features")
          promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
          meth <- round(meth[promoter,],3)
          
          # Map to gene name to take median value if multiple mapping
          meth$gene <- as.character(probe.features[promoter,"gene"])
          meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
          meth <- as.data.frame(round(meth,3))
          
          write.table(meth,paste0("Validation_",cancers[url],"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine DNA methylation files from different cancers
          if(url==1){
            meths <- meth
          }else{
            intersect_meth <- intersect(rownames(meths),rownames(meth))
            meths <- cbind(meths[intersect_meth,],meth[intersect_meth,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(meths,paste0("Validation_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_meth_download1<-downloadHandler(
        filename = "DNA_methylation_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_DNA_methylation_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("DNA methylation(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare copy number alterations(Validation) datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_meth_download1",strong("Download RData file of DNA methylation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_DNA_methylation_finish<-renderUI({
        
        if((!(4 %in% input$validation_multiOmics)) & (!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #33.DNA methylation数据集文件上传处理按钮
    observeEvent(input$DNA_methylation_process3,{
      
      # data.table::setDTthreads(threads = 1) #设置程序运行线程为1
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded DNA methylation datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$DNA_methylation_fileUpload1$datapath #合并DNA methylation数据集上传路径
      upload_paths2 <- input$DNA_methylation_fileUpload2$datapath #各cancerType的DNA methylation数据集上传路径
      
      # DNA methylation数据集转移及预处理
      data_list <- list() #数据集
      if(input$DNA_methylation_upload_sources==1){
        
        progress$set(value = 2)
        #合并DNA methylation数据集
        # file.move(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_Methylation450.gz"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_Methylation450.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
        meths <- data.frame()
        
        progress$set(value = 3)
        #导入并预处理已上传的合并DNA methylation数据集
        meth <- fread(paste0("TCGA_",project,"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
        meth <- as.data.frame(meth)
        rownames(meth) <- meth[,1]
        meth <- as.data.frame(na.omit(meth[,-1]))
        #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
        meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
        colnames(meth) <- paste0(colnames(meth),"A")
        
        # Get promoter
        data("probe.features")
        promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
        meth <- round(meth[promoter,],3)
        
        # Map to gene name to take median value if multiple mapping
        meth$gene <- as.character(probe.features[promoter,"gene"])
        meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
        meth <- as.data.frame(round(meth,3))
        
        progress$set(value = 4)
        ### Final save
        meths <- meth
        write.table(meths,paste0("TCGA_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
        
      }else{
        
        progress$set(value = 2)
        #循环转移并处理各cancerType的DNA methylation数据
        meths <- data.frame()
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_Methylation450.gz"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_Methylation450.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的DNA methylation数据集
          meth <- fread(paste0("TCGA_",cancers[ca],"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
          meth <- as.data.frame(meth)
          rownames(meth) <- meth[,1]
          meth <- as.data.frame(na.omit(meth[,-1]))
          #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
          meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
          colnames(meth) <- paste0(colnames(meth),"A")
          
          # Get promoter
          data("probe.features")
          promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
          meth <- round(meth[promoter,],3)
          
          # Map to gene name to take median value if multiple mapping
          meth$gene <- as.character(probe.features[promoter,"gene"])
          meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
          meth <- as.data.frame(round(meth,3))
          
          write.table(meth,paste0("TCGA_",cancers[ca],"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine DNA methylation files from different cancers
          if(ca==1){
            meths <- meth
          }else{
            intersect_meth <- intersect(rownames(meths),rownames(meth))
            meths <- cbind(meths[intersect_meth,],meth[intersect_meth,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(meths,paste0("TCGA_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_meth_download3<-downloadHandler(
        filename = "DNA_methylation_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$DNA_methylation_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("DNA methylation datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare copy number alterations datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("external_meth_download3",strong("Download RData file of DNA methylation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$DNA_methylation_finish<-renderUI({
        
        if((!(4 %in% input$multiOmics)) & (!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #34.DNA methylation数据集文件上传处理按钮(验证数据集)
    observeEvent(input$validation_DNA_methylation_process2,{
      
      # data.table::setDTthreads(threads = 1) #设置程序运行线程为1
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded DNA methylation(Validation) datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$validation_DNA_methylation_fileUpload1$datapath #合并DNA methylation数据集上传路径
      upload_paths2 <- input$validation_DNA_methylation_fileUpload2$datapath #各cancerType的DNA methylation数据集上传路径
      
      # DNA methylation数据集转移及预处理
      data_list <- list() #数据集
      if(input$validation_DNA_methylation_upload_sources==1){
        
        progress$set(value = 2)
        #合并DNA methylation数据集
        # file.move(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_Methylation450.gz"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_Methylation450.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
        meths <- data.frame()
        
        progress$set(value = 3)
        #导入并预处理已上传的合并DNA methylation数据集
        meth <- fread(paste0("Validation_",project,"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
        meth <- as.data.frame(meth)
        rownames(meth) <- meth[,1]
        meth <- as.data.frame(na.omit(meth[,-1]))
        #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
        meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
        colnames(meth) <- paste0(colnames(meth),"A")
        
        # Get promoter
        data("probe.features")
        promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
        meth <- round(meth[promoter,],3)
        
        # Map to gene name to take median value if multiple mapping
        meth$gene <- as.character(probe.features[promoter,"gene"])
        meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
        meth <- as.data.frame(round(meth,3))
        
        progress$set(value = 4)
        ### Final save
        meths <- meth
        write.table(meths,paste0("Validation_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
        
      }else{
        
        progress$set(value = 2)
        #循环转移并处理各cancerType的DNA methylation数据
        meths <- data.frame()
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_Methylation450.gz"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_Methylation450.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的DNA methylation数据集
          meth <- fread(paste0("Validation_",cancers[ca],"_Methylation450.gz"),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
          meth <- as.data.frame(meth)
          rownames(meth) <- meth[,1]
          meth <- as.data.frame(na.omit(meth[,-1]))
          #colnames(meth) <- substr(colnames(meth), start = 1,stop = 16)
          meth <- meth[,which(substr(colnames(meth), 14, 15) == "01")]
          colnames(meth) <- paste0(colnames(meth),"A")
          
          # Get promoter
          data("probe.features")
          promoter <- intersect(rownames(probe.features[which(probe.features$cgi == "island" & probe.features$feature %in% c("TSS1500","TSS200")),]),rownames(meth))
          meth <- round(meth[promoter,],3)
          
          # Map to gene name to take median value if multiple mapping
          meth$gene <- as.character(probe.features[promoter,"gene"])
          meth <- apply(meth[,setdiff(colnames(meth), "gene")], 2, function(x) tapply(x, INDEX=factor(meth$gene), FUN=median, na.rm=TRUE)) # be patient because this will take a while
          meth <- as.data.frame(round(meth,3))
          
          write.table(meth,paste0("Validation_",cancers[ca],"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine DNA methylation files from different cancers
          if(ca==1){
            meths <- meth
          }else{
            intersect_meth <- intersect(rownames(meths),rownames(meth))
            meths <- cbind(meths[intersect_meth,],meth[intersect_meth,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(meths,paste0("Validation_",project,"_Methpromoter.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"meth" <- meths
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_meth_download2<-downloadHandler(
        filename = "DNA_methylation_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_DNA_methylation_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("DNA methylation(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare copy number alterations(Validation) datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_meth_download2",strong("Download RData file of DNA methylation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_DNA_methylation_finish<-renderUI({
        
        if((!(4 %in% input$validation_multiOmics)) & (!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #35.copy number alterations数据集自动下载处理按钮
    observeEvent(input$copy_number_alterations_process1,{
        
        progress <- Progress$new(session, min=1, max=3) #设置进度条
        on.exit(progress$close())
        progress$set(message = 'Download copy number alterations dataset automatically',
                     detail = 'This may take a while...')
        progress$set(value = 1)
        
        #每点击一次按钮就重新进行数据的下载
        workdir <- getwd()
        workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
        cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
        cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        project <- toupper(trimws(input$projectName,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        
        progress$set(value = 2)
        #copy number alterations数据集下载
        data_list <- list() #数据集
        segments <- data.frame()
        for(ca in 1:length(cancers)){
            segment_download_link <- paste0("https://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/",cancers[ca],"/20160128/gdac.broadinstitute.org_",cancers[ca],".Merge_snp__genome_wide_snp_6__broad_mit_edu__Level_3__segmented_scna_minus_germline_cnv_hg19__seg.Level_3.2016012800.0.0.tar.gz")
            download.file(segment_download_link,method = "auto",destfile = file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
            untar(file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
            untar_list <- untar(file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
            segment_file <- file.path(workdir,untar_list[grep("seg.seg",untar_list)])
            segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
            colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
            segment$sample <- substr(segment$sample, start = 1,stop = 16)
            segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
            
            write.table(segment, paste0("TCGA_",cancers[ca],"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
            
            # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
            cna <- CNregions(segment,
                             epsilon = 0,
                             adaptive = FALSE,
                             rmCNV = FALSE,
                             cnv = FALSE,
                             frac.overlap = 0.5, 
                             rmSmallseg = TRUE, 
                             nProbes = 15)
            cna <- as.data.frame(t(cna))
            
            write.table(cna,paste0("TCGA_",cancers[ca],"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
            
            #Combine segment files from different cancers
            if(ca==1){
              segments <- segment
            }else{
              intersect_segment <- intersect(colnames(segments),colnames(segment))
              segments <- rbind(segments[,intersect_segment],segment[,intersect_segment])
            }
            
            gc() #释放内存
        }
        
        progress$set(value = 3)
        ###cna
        cnas <- CNregions(segments,
                          epsilon = 0,
                          adaptive = FALSE,
                          rmCNV = FALSE,
                          cnv = FALSE,
                          frac.overlap = 0.5, 
                          rmSmallseg = TRUE, 
                          nProbes = 15)
        cnas <- as.data.frame(t(cnas))
        
        ### Final save
        write.table(segments, paste0("TCGA_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")  #segment
        data_list$"segment" <- segments
        write.table(cnas,paste0("TCGA_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #cna
        data_list$"cna" <- cnas
        
        system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
        
        ##RData结果下载按钮响应事件
        output$external_cna_download1<-downloadHandler(
          filename = "copy_number_alterations_preparation(TCGA).RData",
          content = function(theFile) {
            save(data_list,file=theFile)
          }
        )
        
        #数据下载好后给予用户的反馈提示
        output$copy_number_alterations_download_finish<-renderUI({
          
          list(br(),column(12,h3(strong("Copy number alterations datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare binary somatic mutation datasets if you have specified.")),style="color:darkblack"),br(),
               column(12,downloadButton("external_cna_download1",strong("Download RData file of copy number alterations datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        })
        
        output$copy_number_alterations_finish<-renderUI({
          
          if((!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
            list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
          }
        })
        
        #rm(list=ls()) #清空变量
        gc() #释放内存
    })
    
    #36.copy number alterations数据集链接下载处理按钮
    observeEvent(input$copy_number_alterations_process2,{
      
        progress <- Progress$new(session, min=1, max=4) #设置进度条
        on.exit(progress$close())
        progress$set(message = 'Download copy number alterations dataset using specified urls',
                     detail = 'This may take a while...')
        progress$set(value = 1)
        
        #每点击一次按钮就重新进行数据的下载
        workdir <- getwd()
        workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
        cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
        cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        project <- toupper(trimws(input$projectName,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        download_urls <- data.frame(rbind(hot_to_r(input$table.copyNumberAlterationsUrl)))[,1]
        download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        
        #copy number alterations数据集下载及预处理
        data_list <- list() #数据集
        if(input$copy_number_alterations_url_sources==1){
          
          progress$set(value = 2)
          #合并copy number alterations数据下载
          copy_number_alterations_url <- input$copy_number_alterations_downloadUrl
          copy_number_alterations_url <- trimws(copy_number_alterations_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          download.file(copy_number_alterations_url,method="auto",destfile = file.path(workdir,paste0("TCGA_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          gc() #释放内存
          
          progress$set(value = 3)
          #copy number alterations数据预处理
          untar(file.path(workdir,paste0("TCGA_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(workdir,paste0("TCGA_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          segment_file <- file.path(workdir,untar_list[grep("seg.seg",untar_list)])
          segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
          segment$sample <- substr(segment$sample, start = 1,stop = 16)
          segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]

          write.table(segment, paste0("TCGA_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          data_list$"segment" <- segment
          
          progress$set(value = 4)
          # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
          cna <- CNregions(segment,
                           epsilon = 0,
                           adaptive = FALSE,
                           rmCNV = FALSE,
                           cnv = FALSE,
                           frac.overlap = 0.5, 
                           rmSmallseg = TRUE, 
                           nProbes = 15)
          cna <- as.data.frame(t(cna))
          
          write.table(cna,paste0("TCGA_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          data_list$"cna" <- cna
          gc() #释放内存
          
        }else{
          
          progress$set(value = 2)
          #循环下载各cancerType的copy number alterations数据
          segments <- data.frame()
          for(url in 1:length(download_urls)){
            
            download.file(download_urls[url],method = "auto",destfile = file.path(workdir,paste0("TCGA_",cancers[url],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #CentOS服务器代码自动下载
            
            #下载数据预处理
            untar(file.path(workdir,paste0("TCGA_",cancers[url],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
            untar_list <- untar(file.path(workdir,paste0("TCGA_",cancers[url],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
            segment_file <- file.path(workdir,untar_list[grep("seg.seg",untar_list)])
            segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
            colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
            segment$sample <- substr(segment$sample, start = 1,stop = 16)
            segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
            
            write.table(segment, paste0("TCGA_",cancers[url],"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
            
            # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
            cna <- CNregions(segment,
                             epsilon = 0,
                             adaptive = FALSE,
                             rmCNV = FALSE,
                             cnv = FALSE,
                             frac.overlap = 0.5, 
                             rmSmallseg = TRUE, 
                             nProbes = 15)
            cna <- as.data.frame(t(cna))
            
            write.table(cna,paste0("TCGA_",cancers[url],"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
            
            #Combine segment files from different cancers
            if(url==1){
              segments <- segment
            }else{
              intersect_segment <- intersect(colnames(segments),colnames(segment))
              segments <- rbind(segments[,intersect_segment],segment[,intersect_segment])
            }
            
            gc() #释放内存
          }
          
          progress$set(value = 3)
          ###cna
          cnas <- CNregions(segments,
                            epsilon = 0,
                            adaptive = FALSE,
                            rmCNV = FALSE,
                            cnv = FALSE,
                            frac.overlap = 0.5, 
                            rmSmallseg = TRUE, 
                            nProbes = 15)
          cnas <- as.data.frame(t(cnas))
          
          progress$set(value = 4)
          ### Final save
          write.table(segments, paste0("TCGA_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")  #segment
          data_list$"segment" <- segments
          write.table(cnas,paste0("TCGA_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #cna
          data_list$"cna" <- cnas
        }
        
        system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
        
        ##RData结果下载按钮响应事件
        output$external_cna_download2<-downloadHandler(
          filename = "copy_number_alterations_preparation(TCGA).RData",
          content = function(theFile) {
            save(data_list,file=theFile)
          }
        )
        
        #数据下载好后给予用户的反馈提示
        output$copy_number_alterations_download_finish<-renderUI({
          
          list(br(),column(12,h3(strong("Copy number alterations datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare binary somatic mutation datasets if you have specified.")),style="color:darkblack"),br(),
               column(12,downloadButton("external_cna_download2",strong("Download RData file of copy number alterations datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        })
        
        output$copy_number_alterations_finish<-renderUI({
          
          if((!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
            list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
          }
        })
        
        #rm(list=ls()) #清空变量
        gc() #释放内存
        
    })
    
    #37.copy number alterations数据集链接下载处理按钮(验证数据集)
    observeEvent(input$validation_copy_number_alterations_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download copy number alterations(Validation) dataset using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- toupper(trimws(input$projectName,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      download_urls <- data.frame(rbind(hot_to_r(input$table.validation_copyNumberAlterationsUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #copy number alterations数据集下载及预处理
      data_list <- list() #数据集
      if(input$validation_copy_number_alterations_url_sources==1){
        
        progress$set(value = 2)
        #合并copy number alterations数据下载
        copy_number_alterations_url <- input$validation_copy_number_alterations_downloadUrl
        copy_number_alterations_url <- trimws(copy_number_alterations_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(copy_number_alterations_url,method="auto",destfile = file.path(validation_workdir,paste0("Validation_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        #copy number alterations数据预处理
        untar(file.path(validation_workdir,paste0("Validation_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
        untar_list <- untar(file.path(validation_workdir,paste0("Validation_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
        segment_file <- file.path(validation_workdir,untar_list[grep("seg.seg",untar_list)])
        segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
        segment$sample <- substr(segment$sample, start = 1,stop = 16)
        segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
        
        write.table(segment, paste0("Validation_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"segment" <- segment
        
        progress$set(value = 4)
        # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
        cna <- CNregions(segment,
                         epsilon = 0,
                         adaptive = FALSE,
                         rmCNV = FALSE,
                         cnv = FALSE,
                         frac.overlap = 0.5, 
                         rmSmallseg = TRUE, 
                         nProbes = 15)
        cna <- as.data.frame(t(cna))
        
        write.table(cna,paste0("Validation_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"cna" <- cna
        gc() #释放内存
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的copy number alterations数据
        segments <- data.frame()
        for(url in 1:length(download_urls)){
          
          download.file(download_urls[url],method = "auto",destfile = file.path(validation_workdir,paste0("Validation_",cancers[url],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          
          #下载数据预处理
          untar(file.path(validation_workdir,paste0("Validation_",cancers[url],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(validation_workdir,paste0("Validation_",cancers[url],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          segment_file <- file.path(validation_workdir,untar_list[grep("seg.seg",untar_list)])
          segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
          segment$sample <- substr(segment$sample, start = 1,stop = 16)
          segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
          
          write.table(segment, paste0("Validation_",cancers[url],"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
          cna <- CNregions(segment,
                           epsilon = 0,
                           adaptive = FALSE,
                           rmCNV = FALSE,
                           cnv = FALSE,
                           frac.overlap = 0.5, 
                           rmSmallseg = TRUE, 
                           nProbes = 15)
          cna <- as.data.frame(t(cna))
          
          write.table(cna,paste0("Validation_",cancers[url],"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine segment files from different cancers
          if(url==1){
            segments <- segment
          }else{
            intersect_segment <- intersect(colnames(segments),colnames(segment))
            segments <- rbind(segments[,intersect_segment],segment[,intersect_segment])
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        ###cna
        cnas <- CNregions(segments,
                          epsilon = 0,
                          adaptive = FALSE,
                          rmCNV = FALSE,
                          cnv = FALSE,
                          frac.overlap = 0.5, 
                          rmSmallseg = TRUE, 
                          nProbes = 15)
        cnas <- as.data.frame(t(cnas))
        
        progress$set(value = 4)
        ### Final save
        write.table(segments, paste0("Validation_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")  #segment
        data_list$"segment" <- segments
        write.table(cnas,paste0("Validation_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #cna
        data_list$"cna" <- cnas
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_cna_download1<-downloadHandler(
        filename = "copy_number_alterations_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_copy_number_alterations_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Copy number alterations(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare binary somatic mutation(Validation) datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_cna_download1",strong("Download RData file of copy number alterations datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_copy_number_alterations_finish<-renderUI({
        
        if((!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #38.copy number alterations数据集文件上传处理按钮
    observeEvent(input$copy_number_alterations_process3,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded copy number alterations datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$copy_number_alterations_fileUpload1$datapath #合并copy number alterations数据集上传路径
      upload_paths2 <- input$copy_number_alterations_fileUpload2$datapath #各cancerType的copy number alterations数据集上传路径
      
      # copy number alterations数据集转移及预处理
      data_list <- list() #数据集
      if(input$copy_number_alterations_upload_sources==1){
        
        progress$set(value = 2)
        #合并copy number alterations数据集
        # file.move(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并copy number alterations数据集
        untar(file.path(workdir,paste0("TCGA_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
        untar_list <- untar(file.path(workdir,paste0("TCGA_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
        segment_file <- file.path(workdir,untar_list[grep("seg.seg",untar_list)])
        segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
        segment$sample <- substr(segment$sample, start = 1,stop = 16)
        segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
        
        write.table(segment, paste0("TCGA_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"segment" <- segment
        
        progress$set(value = 4)
        # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
        cna <- CNregions(segment,
                         epsilon = 0,
                         adaptive = FALSE,
                         rmCNV = FALSE,
                         cnv = FALSE,
                         frac.overlap = 0.5, 
                         rmSmallseg = TRUE, 
                         nProbes = 15)
        cna <- as.data.frame(t(cna))
        
        write.table(cna,paste0("TCGA_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"cna" <- cna
        gc() #释放内存
        
      }else{
        
        progress$set(value = 2)
        #循环移动并预处理各cancerType的copy number alterations数据
        segments <- data.frame()
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的copy number alterations数据集
          untar(file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(workdir,paste0("TCGA_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          segment_file <- file.path(workdir,untar_list[grep("seg.seg",untar_list)])
          segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
          segment$sample <- substr(segment$sample, start = 1,stop = 16)
          segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
          
          write.table(segment, paste0("TCGA_",cancers[ca],"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
          cna <- CNregions(segment,
                           epsilon = 0,
                           adaptive = FALSE,
                           rmCNV = FALSE,
                           cnv = FALSE,
                           frac.overlap = 0.5, 
                           rmSmallseg = TRUE, 
                           nProbes = 15)
          cna <- as.data.frame(t(cna))
          
          write.table(cna,paste0("TCGA_",cancers[ca],"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine segment files from different cancers
          if(ca==1){
            segments <- segment
          }else{
            intersect_segment <- intersect(colnames(segments),colnames(segment))
            segments <- rbind(segments[,intersect_segment],segment[,intersect_segment])
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        ###cna
        cnas <- CNregions(segments,
                          epsilon = 0,
                          adaptive = FALSE,
                          rmCNV = FALSE,
                          cnv = FALSE,
                          frac.overlap = 0.5, 
                          rmSmallseg = TRUE, 
                          nProbes = 15)
        cnas <- as.data.frame(t(cnas))
        
        progress$set(value = 4)
        ### Final save
        write.table(segments, paste0("TCGA_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")  #segment
        data_list$"segment" <- segments
        write.table(cnas,paste0("TCGA_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #cna
        data_list$"cna" <- cnas
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_cna_download3<-downloadHandler(
        filename = "copy_number_alterations_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$copy_number_alterations_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Copy number alterations datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare binary somatic mutation datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("external_cna_download3",strong("Download RData file of copy number alterations datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$copy_number_alterations_finish<-renderUI({
        
        if((!(5 %in% input$multiOmics)) & (!(6 %in% input$multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #39.copy number alterations数据集文件上传处理按钮(验证数据集)
    observeEvent(input$validation_copy_number_alterations_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded copy number alterations(Validation) datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$validation_copy_number_alterations_fileUpload1$datapath #合并copy number alterations数据集上传路径
      upload_paths2 <- input$validation_copy_number_alterations_fileUpload2$datapath #各cancerType的copy number alterations数据集上传路径
      
      # copy number alterations数据集转移及预处理
      data_list <- list() #数据集
      if(input$validation_copy_number_alterations_upload_sources==1){
        
        progress$set(value = 2)
        #合并copy number alterations数据集
        # file.move(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并copy number alterations数据集
        untar(file.path(validation_workdir,paste0("Validation_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
        untar_list <- untar(file.path(validation_workdir,paste0("Validation_",project,"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
        segment_file <- file.path(validation_workdir,untar_list[grep("seg.seg",untar_list)])
        segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
        segment$sample <- substr(segment$sample, start = 1,stop = 16)
        segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
        
        write.table(segment, paste0("Validation_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"segment" <- segment
        
        progress$set(value = 4)
        # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
        cna <- CNregions(segment,
                         epsilon = 0,
                         adaptive = FALSE,
                         rmCNV = FALSE,
                         cnv = FALSE,
                         frac.overlap = 0.5, 
                         rmSmallseg = TRUE, 
                         nProbes = 15)
        cna <- as.data.frame(t(cna))
        
        write.table(cna,paste0("Validation_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"cna" <- cna
        gc() #释放内存
        
      }else{
        
        progress$set(value = 2)
        #循环移动并预处理各cancerType的copy number alterations数据
        segments <- data.frame()
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的copy number alterations数据集
          untar(file.path(validation_workdir,paste0("Validation_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(validation_workdir,paste0("Validation_",cancers[ca],"_genome_wide_snp_6-segmented_scna_minus_germline_cnv_hg19.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          segment_file <- file.path(validation_workdir,untar_list[grep("seg.seg",untar_list)])
          segment <- read.table(segment_file,sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          colnames(segment) <- c("sample","chromosome","start","end","num.mark","seg.mean")
          segment$sample <- substr(segment$sample, start = 1,stop = 16)
          segment <- segment[which(substr(segment$sample, 14, 15) == "01"),]
          
          write.table(segment, paste0("Validation_",cancers[ca],"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # cnv <- read.delim("TCGA.LUAD.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz",sep = "\t",header = T,check.names = F,stringsAsFactors = F)
          cna <- CNregions(segment,
                           epsilon = 0,
                           adaptive = FALSE,
                           rmCNV = FALSE,
                           cnv = FALSE,
                           frac.overlap = 0.5, 
                           rmSmallseg = TRUE, 
                           nProbes = 15)
          cna <- as.data.frame(t(cna))
          
          write.table(cna,paste0("Validation_",cancers[ca],"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine segment files from different cancers
          if(ca==1){
            segments <- segment
          }else{
            intersect_segment <- intersect(colnames(segments),colnames(segment))
            segments <- rbind(segments[,intersect_segment],segment[,intersect_segment])
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        ###cna
        cnas <- CNregions(segments,
                          epsilon = 0,
                          adaptive = FALSE,
                          rmCNV = FALSE,
                          cnv = FALSE,
                          frac.overlap = 0.5, 
                          rmSmallseg = TRUE, 
                          nProbes = 15)
        cnas <- as.data.frame(t(cnas))
        
        progress$set(value = 4)
        ### Final save
        write.table(segments, paste0("Validation_",project,"_segment.txt"), quote=F, row.names=F,col.names = T,sep = "\t")  #segment
        data_list$"segment" <- segments
        write.table(cnas,paste0("Validation_",project,"_cna.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")  #cna
        data_list$"cna" <- cnas
        
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_cna_download2<-downloadHandler(
        filename = "copy_number_alterations_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_copy_number_alterations_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Copy number alterations(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare binary somatic mutation(Validation) datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_cna_download2",strong("Download RData file of copy number alterations datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_copy_number_alterations_finish<-renderUI({
        
        if((!(5 %in% input$validation_multiOmics)) & (!(6 %in% input$validation_multiOmics))){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #40.binary somatic mutation数据集自动下载按钮
    observeEvent(input$binary_somatic_mutation_process1,{
      
        progress <- Progress$new(session, min=1, max=3) #设置进度条
        on.exit(progress$close())
        progress$set(message = 'Download binary somatic mutation dataset automatically',
                     detail = 'This may take a while...')
        progress$set(value = 1)
        
        #每点击一次按钮就重新进行数据的下载
        workdir <- getwd()
        workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
        cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
        cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        project <- toupper(trimws(input$projectName,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
        
        progress$set(value = 2)
        #binary somatic mutation数据集下载
        data_list <- list() #数据集
        mafs <- data.frame()
        for(ca in 1:length(cancers)){
          maf_download_link <- paste0("https://cbioportal-datahub.s3.amazonaws.com/",tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")
          download.file(maf_download_link,method = "auto",destfile = file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          untar(file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          maf_file <- file.path(workdir,untar_list[grep("data_mutations",untar_list)])
          
          maf <- read_tsv(maf_file, comment = "#")
          flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
          label <- c("Tumor_Sample_Barcode",
                     "Hugo_Symbol",
                     "Chromosome",
                     "Start_Position",
                     "End_Position",
                     "Variant_Classification",
                     "Variant_Type",
                     "Reference_Allele",
                     "Tumor_Seq_Allele1",
                     "Tumor_Seq_Allele2")
          maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
          maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
          maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
          maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
          write.table(maf, paste0("TCGA_",cancers[ca],"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # Prepare binary somatic mutation
          # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
          # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
          # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          # mut <- as.data.frame(na.omit(mut))
          # rownames(mut) <- mut$sample
          # mut <- mut[,-1]
          # mut <- mut[rowSums(mut) > 0,]
          # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
          # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
          
          # option 2: convert from MAF file without silent mutation (may not be the same with above file)
          # binary mutation
          # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
          
          # for (i in colnames(mut.binary)) {
          #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          #   if(is.element("Silent",tmp$Variant_Classification)) {
          #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
          #   }
          #   for (j in tmp$Hugo_Symbol)
          #     mut.binary[j,i] <- 1
          # }
          for (i in colnames(mut.binary)) {
            tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
            tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
            for (j in tmp$Hugo_Symbol){
              mut.binary[j,i] <- 1
            }
          }
          mut.binary <- as.data.frame(mut.binary)
          #rownames(mut.binary) <- toupper(rownames(mut.binary))
          mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
          #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
          #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
          
          write.table(mut.binary,paste0("TCGA_",cancers[ca],"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine maf files from different cancers
          if(ca==1){
            mafs <- maf
          }else{
            intersect_maf <- intersect(colnames(mafs),colnames(maf))
            mafs <- rbind(mafs[,intersect_maf],maf[,intersect_maf])
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        ##Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # mafs.ns <- mafs[which(mafs$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        muts.binary <- matrix(0,nrow = length(unique(mafs$Hugo_Symbol)),ncol = length(unique(mafs$Tumor_Sample_Barcode)),dimnames = list(unique(mafs$Hugo_Symbol),unique(mafs$Tumor_Sample_Barcode)))
        
        # for (i in colnames(muts.binary)) {
        #   tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     muts.binary[j,i] <- 1
        # }
        for (i in colnames(muts.binary)) {
          tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            muts.binary[j,i] <- 1
          }
        }
        muts.binary <- as.data.frame(muts.binary)
        #rownames(muts.binary) <- toupper(rownames(muts.binary))
        muts.binary <- muts.binary[rowSums(muts.binary) > 0,]
        #muts.binary <- muts.binary[,which(substr(colnames(muts.binary), 14, 15) == "01")]
        #colnames(muts.binary) <- paste0(colnames(muts.binary),"A")
        
        ### Final save
        write.table(mafs, paste0("TCGA_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- mafs
        write.table(muts.binary,paste0("TCGA_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- muts.binary
        
        system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
        
        ##RData结果下载按钮响应事件
        output$external_mut_download1<-downloadHandler(
          filename = "binary_somatic_mutation_preparation(TCGA).RData",
          content = function(theFile) {
            save(data_list,file=theFile)
          }
        )
        
        #数据下载好后给予用户的反馈提示
        output$binary_somatic_mutation_download_finish<-renderUI({
          
          list(br(),column(12,h3(strong("Binary somatic mutation datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare radiomics datasets if you have specified.")),style="color:darkblack"),br(),
               column(12,downloadButton("external_mut_download1",strong("Download RData file of binary somatic mutation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
        })
        
        output$binary_somatic_mutation_finish<-renderUI({
          
          if(!(6 %in% input$multiOmics)){
            list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
          }
        })
        
        #rm(list=ls()) #清空变量
        gc() #释放内存
    })
    
    #41.binary somatic mutation数据集链接下载按钮
    observeEvent(input$binary_somatic_mutation_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download binary somatic mutation dataset using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- toupper(trimws(input$projectName,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      download_urls <- data.frame(rbind(hot_to_r(input$table.binarySomaticMutationUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #binary somatic mutation数据集下载及预处理
      data_list <- list() #数据集
      if(input$binary_somatic_mutation_url_sources==1){
        
        progress$set(value = 2)
        #合并binary somatic mutation数据下载
        binary_somatic_mutation_url <- input$binary_somatic_mutation_downloadUrl
        binary_somatic_mutation_url <- trimws(binary_somatic_mutation_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(binary_somatic_mutation_url,method="auto",destfile = file.path(workdir,paste0(tolower(project),"_tcga_pan_can_atlas_2018.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        #binary somatic mutation数据预处理
        untar(file.path(workdir,paste0(tolower(project),"_tcga_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
        untar_list <- untar(file.path(workdir,paste0(tolower(project),"_tcga_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
        maf_file <- file.path(workdir,untar_list[grep("data_mutations",untar_list)])
        maf <- read_tsv(maf_file, comment = "#")
        flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
        label <- c("Tumor_Sample_Barcode",
                   "Hugo_Symbol",
                   "Chromosome",
                   "Start_Position",
                   "End_Position",
                   "Variant_Classification",
                   "Variant_Type",
                   "Reference_Allele",
                   "Tumor_Seq_Allele1",
                   "Tumor_Seq_Allele2")
        maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
        maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
        maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
        maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
        write.table(maf, paste0("TCGA_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- maf
        
        # Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
        
        # for (i in colnames(mut.binary)) {
        #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     mut.binary[j,i] <- 1
        # }
        for (i in colnames(mut.binary)) {
          tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            mut.binary[j,i] <- 1
          }
        }
        mut.binary <- as.data.frame(mut.binary)
        #rownames(mut.binary) <- toupper(rownames(mut.binary))
        mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
        #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
        #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
        
        write.table(mut.binary,paste0("TCGA_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- mut.binary
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的binary somatic mutation数据
        mafs <- data.frame()
        for(url in 1:length(download_urls)){
          download.file(download_urls[url],method = "auto",destfile = file.path(workdir,paste0(tolower(cancers[url]),"_tcga_pan_can_atlas_2018.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          
          #下载数据预处理
          untar(file.path(workdir,paste0(tolower(cancers[url]),"_tcga_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(workdir,paste0(tolower(cancers[url]),"_tcga_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          maf_file <- file.path(workdir,untar_list[grep("data_mutations",untar_list)])
          
          maf <- read_tsv(maf_file, comment = "#")
          flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
          label <- c("Tumor_Sample_Barcode",
                     "Hugo_Symbol",
                     "Chromosome",
                     "Start_Position",
                     "End_Position",
                     "Variant_Classification",
                     "Variant_Type",
                     "Reference_Allele",
                     "Tumor_Seq_Allele1",
                     "Tumor_Seq_Allele2")
          maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
          maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
          maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
          maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
          write.table(maf, paste0("TCGA_",cancers[url],"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # Prepare binary somatic mutation
          # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
          # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
          # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          # mut <- as.data.frame(na.omit(mut))
          # rownames(mut) <- mut$sample
          # mut <- mut[,-1]
          # mut <- mut[rowSums(mut) > 0,]
          # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
          # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
          
          # option 2: convert from MAF file without silent mutation (may not be the same with above file)
          # binary mutation
          # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
          
          # for (i in colnames(mut.binary)) {
          #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          #   if(is.element("Silent",tmp$Variant_Classification)) {
          #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
          #   }
          #   for (j in tmp$Hugo_Symbol)
          #     mut.binary[j,i] <- 1
          # }
          for (i in colnames(mut.binary)) {
            tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
            tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
            for (j in tmp$Hugo_Symbol){
              mut.binary[j,i] <- 1
            }
          }
          mut.binary <- as.data.frame(mut.binary)
          #rownames(mut.binary) <- toupper(rownames(mut.binary))
          mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
          #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
          #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
          
          write.table(mut.binary,paste0("TCGA_",cancers[url],"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine maf files from different cancers
          if(url==1){
            mafs <- maf
          }else{
            intersect_maf <- intersect(colnames(mafs),colnames(maf))
            mafs <- rbind(mafs[,intersect_maf],maf[,intersect_maf])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 3)
        ##Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # mafs.ns <- mafs[which(mafs$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        muts.binary <- matrix(0,nrow = length(unique(mafs$Hugo_Symbol)),ncol = length(unique(mafs$Tumor_Sample_Barcode)),dimnames = list(unique(mafs$Hugo_Symbol),unique(mafs$Tumor_Sample_Barcode)))
        
        # for (i in colnames(muts.binary)) {
        #   tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     muts.binary[j,i] <- 1
        # }
        for (i in colnames(muts.binary)) {
          tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            muts.binary[j,i] <- 1
          }
        }
        muts.binary <- as.data.frame(muts.binary)
        #rownames(muts.binary) <- toupper(rownames(muts.binary))
        muts.binary <- muts.binary[rowSums(muts.binary) > 0,]
        #muts.binary <- muts.binary[,which(substr(colnames(muts.binary), 14, 15) == "01")]
        #colnames(muts.binary) <- paste0(colnames(muts.binary),"A")
        
        progress$set(value = 4)
        ### Final save
        write.table(mafs, paste0("TCGA_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- mafs
        write.table(muts.binary,paste0("TCGA_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- muts.binary
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_mut_download2<-downloadHandler(
        filename = "binary_somatic_mutation_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$binary_somatic_mutation_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Binary somatic mutation datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare radiomics datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("external_mut_download2",strong("Download RData file of binary somatic mutation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$binary_somatic_mutation_finish<-renderUI({
        
        if(!(6 %in% input$multiOmics)){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })

    #42.binary somatic mutation数据集链接下载按钮(验证数据集)
    observeEvent(input$validation_binary_somatic_mutation_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download binary somatic mutation(Validation) dataset using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- toupper(trimws(input$projectName,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      download_urls <- data.frame(rbind(hot_to_r(input$table.validation_binarySomaticMutationUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #binary somatic mutation数据集下载及预处理
      data_list <- list() #数据集
      if(input$validation_binary_somatic_mutation_url_sources==1){
        
        progress$set(value = 2)
        #合并binary somatic mutation数据下载
        binary_somatic_mutation_url <- input$validation_binary_somatic_mutation_downloadUrl
        binary_somatic_mutation_url <- trimws(binary_somatic_mutation_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(binary_somatic_mutation_url,method="auto",destfile = file.path(validation_workdir,paste0(tolower(project),"_validation_pan_can_atlas_2018.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        #binary somatic mutation数据预处理
        untar(file.path(validation_workdir,paste0(tolower(project),"_validation_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
        untar_list <- untar(file.path(validation_workdir,paste0(tolower(project),"_validation_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
        maf_file <- file.path(validation_workdir,untar_list[grep("data_mutations",untar_list)])
        maf <- read_tsv(maf_file, comment = "#")
        flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
        label <- c("Tumor_Sample_Barcode",
                   "Hugo_Symbol",
                   "Chromosome",
                   "Start_Position",
                   "End_Position",
                   "Variant_Classification",
                   "Variant_Type",
                   "Reference_Allele",
                   "Tumor_Seq_Allele1",
                   "Tumor_Seq_Allele2")
        maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
        maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
        maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
        maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
        write.table(maf, paste0("Validation_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- maf
        
        # Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
        
        # for (i in colnames(mut.binary)) {
        #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     mut.binary[j,i] <- 1
        # }
        for (i in colnames(mut.binary)) {
          tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            mut.binary[j,i] <- 1
          }
        }
        mut.binary <- as.data.frame(mut.binary)
        #rownames(mut.binary) <- toupper(rownames(mut.binary))
        mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
        #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
        #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
        
        write.table(mut.binary,paste0("Validation_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- mut.binary
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的binary somatic mutation数据
        mafs <- data.frame()
        for(url in 1:length(download_urls)){
          download.file(download_urls[url],method = "auto",destfile = file.path(validation_workdir,paste0(tolower(cancers[url]),"_validation_pan_can_atlas_2018.tar.gz")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          
          #下载数据预处理
          untar(file.path(validation_workdir,paste0(tolower(cancers[url]),"_validation_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(validation_workdir,paste0(tolower(cancers[url]),"_validation_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          maf_file <- file.path(validation_workdir,untar_list[grep("data_mutations",untar_list)])
          
          maf <- read_tsv(maf_file, comment = "#")
          flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
          label <- c("Tumor_Sample_Barcode",
                     "Hugo_Symbol",
                     "Chromosome",
                     "Start_Position",
                     "End_Position",
                     "Variant_Classification",
                     "Variant_Type",
                     "Reference_Allele",
                     "Tumor_Seq_Allele1",
                     "Tumor_Seq_Allele2")
          maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
          maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
          maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
          maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
          write.table(maf, paste0("Validation_",cancers[url],"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # Prepare binary somatic mutation
          # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
          # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
          # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          # mut <- as.data.frame(na.omit(mut))
          # rownames(mut) <- mut$sample
          # mut <- mut[,-1]
          # mut <- mut[rowSums(mut) > 0,]
          # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
          # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
          
          # option 2: convert from MAF file without silent mutation (may not be the same with above file)
          # binary mutation
          # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
          
          # for (i in colnames(mut.binary)) {
          #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          #   if(is.element("Silent",tmp$Variant_Classification)) {
          #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
          #   }
          #   for (j in tmp$Hugo_Symbol)
          #     mut.binary[j,i] <- 1
          # }
          for (i in colnames(mut.binary)) {
            tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
            tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
            for (j in tmp$Hugo_Symbol){
              mut.binary[j,i] <- 1
            }
          }
          mut.binary <- as.data.frame(mut.binary)
          #rownames(mut.binary) <- toupper(rownames(mut.binary))
          mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
          #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
          #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
          
          write.table(mut.binary,paste0("Validation_",cancers[url],"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine maf files from different cancers
          if(url==1){
            mafs <- maf
          }else{
            intersect_maf <- intersect(colnames(mafs),colnames(maf))
            mafs <- rbind(mafs[,intersect_maf],maf[,intersect_maf])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 3)
        ##Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # mafs.ns <- mafs[which(mafs$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        muts.binary <- matrix(0,nrow = length(unique(mafs$Hugo_Symbol)),ncol = length(unique(mafs$Tumor_Sample_Barcode)),dimnames = list(unique(mafs$Hugo_Symbol),unique(mafs$Tumor_Sample_Barcode)))
        
        # for (i in colnames(muts.binary)) {
        #   tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     muts.binary[j,i] <- 1
        # }
        for (i in colnames(muts.binary)) {
          tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            muts.binary[j,i] <- 1
          }
        }
        muts.binary <- as.data.frame(muts.binary)
        #rownames(muts.binary) <- toupper(rownames(muts.binary))
        muts.binary <- muts.binary[rowSums(muts.binary) > 0,]
        #muts.binary <- muts.binary[,which(substr(colnames(muts.binary), 14, 15) == "01")]
        #colnames(muts.binary) <- paste0(colnames(muts.binary),"A")
        
        progress$set(value = 4)
        ### Final save
        write.table(mafs, paste0("Validation_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- mafs
        write.table(muts.binary,paste0("Validation_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- muts.binary
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_mut_download1<-downloadHandler(
        filename = "binary_somatic_mutation_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_binary_somatic_mutation_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Binary somatic mutation(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare radiomics(Validation) datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_mut_download1",strong("Download RData file of binary somatic mutation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_binary_somatic_mutation_finish<-renderUI({
        
        if(!(6 %in% input$validation_multiOmics)){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #43.binary somatic mutation数据集文件上传处理按钮
    observeEvent(input$binary_somatic_mutation_process3,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded binary somatic mutation datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$binary_somatic_mutation_fileUpload1$datapath #合并binary somatic mutation数据集上传路径
      upload_paths2 <- input$binary_somatic_mutation_fileUpload2$datapath #各cancerType的binary somatic mutation数据集上传路径
      
      # binary somatic mutation数据集转移及预处理
      data_list <- list() #数据集
      if(input$binary_somatic_mutation_upload_sources==1){
        
        progress$set(value = 2)
        #合并binary somatic mutation数据集
        # file.move(upload_paths1, file.path(workdir,paste0(tolower(project),"_tcga_pan_can_atlas_2018.tar.gz"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(workdir,paste0(tolower(project),"_tcga_pan_can_atlas_2018.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并binary somatic mutation数据集
        untar(file.path(workdir,paste0(tolower(project),"_tcga_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
        untar_list <- untar(file.path(workdir,paste0(tolower(project),"_tcga_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
        maf_file <- file.path(workdir,untar_list[grep("data_mutations",untar_list)])
        maf <- read_tsv(maf_file, comment = "#")
        flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
        label <- c("Tumor_Sample_Barcode",
                   "Hugo_Symbol",
                   "Chromosome",
                   "Start_Position",
                   "End_Position",
                   "Variant_Classification",
                   "Variant_Type",
                   "Reference_Allele",
                   "Tumor_Seq_Allele1",
                   "Tumor_Seq_Allele2")
        maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
        maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
        maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
        maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
        write.table(maf, paste0("TCGA_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- maf
        
        # Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
        
        # for (i in colnames(mut.binary)) {
        #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     mut.binary[j,i] <- 1
        # }
        for (i in colnames(mut.binary)) {
          tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            mut.binary[j,i] <- 1
          }
        }
        mut.binary <- as.data.frame(mut.binary)
        #rownames(mut.binary) <- toupper(rownames(mut.binary))
        mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
        #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
        #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
        
        write.table(mut.binary,paste0("TCGA_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- mut.binary
        
      }else{
        
        progress$set(value = 2)
        #循环转移并处理各cancerType的binary somatic mutation数据
        mafs <- data.frame()
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的binary somatic mutation数据集
          untar(file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(workdir,paste0(tolower(cancers[ca]),"_tcga_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          maf_file <- file.path(workdir,untar_list[grep("data_mutations",untar_list)])
          
          maf <- read_tsv(maf_file, comment = "#")
          flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
          label <- c("Tumor_Sample_Barcode",
                     "Hugo_Symbol",
                     "Chromosome",
                     "Start_Position",
                     "End_Position",
                     "Variant_Classification",
                     "Variant_Type",
                     "Reference_Allele",
                     "Tumor_Seq_Allele1",
                     "Tumor_Seq_Allele2")
          maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
          maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
          maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
          maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
          write.table(maf, paste0("TCGA_",cancers[ca],"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # Prepare binary somatic mutation
          # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
          # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
          # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          # mut <- as.data.frame(na.omit(mut))
          # rownames(mut) <- mut$sample
          # mut <- mut[,-1]
          # mut <- mut[rowSums(mut) > 0,]
          # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
          # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
          
          # option 2: convert from MAF file without silent mutation (may not be the same with above file)
          # binary mutation
          # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
          
          # for (i in colnames(mut.binary)) {
          #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          #   if(is.element("Silent",tmp$Variant_Classification)) {
          #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
          #   }
          #   for (j in tmp$Hugo_Symbol)
          #     mut.binary[j,i] <- 1
          # }
          for (i in colnames(mut.binary)) {
            tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
            tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
            for (j in tmp$Hugo_Symbol){
              mut.binary[j,i] <- 1
            }
          }
          mut.binary <- as.data.frame(mut.binary)
          #rownames(mut.binary) <- toupper(rownames(mut.binary))
          mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
          #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
          #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
          
          write.table(mut.binary,paste0("TCGA_",cancers[ca],"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine maf files from different cancers
          if(ca==1){
            mafs <- maf
          }else{
            intersect_maf <- intersect(colnames(mafs),colnames(maf))
            mafs <- rbind(mafs[,intersect_maf],maf[,intersect_maf])
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        ##Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # mafs.ns <- mafs[which(mafs$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        muts.binary <- matrix(0,nrow = length(unique(mafs$Hugo_Symbol)),ncol = length(unique(mafs$Tumor_Sample_Barcode)),dimnames = list(unique(mafs$Hugo_Symbol),unique(mafs$Tumor_Sample_Barcode)))
        
        # for (i in colnames(muts.binary)) {
        #   tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     muts.binary[j,i] <- 1
        # }
        for (i in colnames(muts.binary)) {
          tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            muts.binary[j,i] <- 1
          }
        }
        muts.binary <- as.data.frame(muts.binary)
        #rownames(muts.binary) <- toupper(rownames(muts.binary))
        muts.binary <- muts.binary[rowSums(muts.binary) > 0,]
        #muts.binary <- muts.binary[,which(substr(colnames(muts.binary), 14, 15) == "01")]
        #colnames(muts.binary) <- paste0(colnames(muts.binary),"A")
        
        progress$set(value = 4)
        ### Final save
        write.table(mafs, paste0("TCGA_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- mafs
        write.table(muts.binary,paste0("TCGA_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- muts.binary
        
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_mut_download3<-downloadHandler(
        filename = "binary_somatic_mutation_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$binary_somatic_mutation_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Binary somatic mutation datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare radiomics datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("external_mut_download3",strong("Download RData file of binary somatic mutation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$binary_somatic_mutation_finish<-renderUI({
        
        if(!(6 %in% input$multiOmics)){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
    })
    
    #44.binary somatic mutation数据集文件上传处理按钮(验证数据集)
    observeEvent(input$validation_binary_somatic_mutation_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload downloaded binary somatic mutation(Validation) datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$validation_binary_somatic_mutation_fileUpload1$datapath #合并binary somatic mutation数据集上传路径
      upload_paths2 <- input$validation_binary_somatic_mutation_fileUpload2$datapath #各cancerType的binary somatic mutation数据集上传路径
      
      # binary somatic mutation数据集转移及预处理
      data_list <- list() #数据集
      if(input$validation_binary_somatic_mutation_upload_sources==1){
        
        progress$set(value = 2)
        #合并binary somatic mutation数据集
        # file.move(upload_paths1, file.path(validation_workdir,paste0(tolower(project),"_validation_pan_can_atlas_2018.tar.gz"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(validation_workdir,paste0(tolower(project),"_validation_pan_can_atlas_2018.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并binary somatic mutation数据集
        untar(file.path(validation_workdir,paste0(tolower(project),"_validation_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
        untar_list <- untar(file.path(validation_workdir,paste0(tolower(project),"_validation_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
        maf_file <- file.path(validation_workdir,untar_list[grep("data_mutations",untar_list)])
        maf <- read_tsv(maf_file, comment = "#")
        flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
        label <- c("Tumor_Sample_Barcode",
                   "Hugo_Symbol",
                   "Chromosome",
                   "Start_Position",
                   "End_Position",
                   "Variant_Classification",
                   "Variant_Type",
                   "Reference_Allele",
                   "Tumor_Seq_Allele1",
                   "Tumor_Seq_Allele2")
        maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
        maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
        maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
        maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
        write.table(maf, paste0("Validation_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- maf
        
        # Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
        
        # for (i in colnames(mut.binary)) {
        #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     mut.binary[j,i] <- 1
        # }
        for (i in colnames(mut.binary)) {
          tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            mut.binary[j,i] <- 1
          }
        }
        mut.binary <- as.data.frame(mut.binary)
        #rownames(mut.binary) <- toupper(rownames(mut.binary))
        mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
        #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
        #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
        
        write.table(mut.binary,paste0("Validation_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- mut.binary
        
      }else{
        
        progress$set(value = 2)
        #循环转移并处理各cancerType的binary somatic mutation数据
        mafs <- data.frame()
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(validation_workdir,paste0(tolower(cancers[ca]),"_validation_pan_can_atlas_2018.tar.gz"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(validation_workdir,paste0(tolower(cancers[ca]),"_validation_pan_can_atlas_2018.tar.gz")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的binary somatic mutation数据集
          untar(file.path(validation_workdir,paste0(tolower(cancers[ca]),"_validation_pan_can_atlas_2018.tar.gz")))  #解压下载好的.tar.gz文件
          untar_list <- untar(file.path(validation_workdir,paste0(tolower(cancers[ca]),"_validation_pan_can_atlas_2018.tar.gz")),list = TRUE)  #列出解压后文件夹中的文件路径
          maf_file <- file.path(validation_workdir,untar_list[grep("data_mutations",untar_list)])
          
          maf <- read_tsv(maf_file, comment = "#")
          flag <- read.table("Mutation Flags 100.txt",sep = "\t")$V1
          label <- c("Tumor_Sample_Barcode",
                     "Hugo_Symbol",
                     "Chromosome",
                     "Start_Position",
                     "End_Position",
                     "Variant_Classification",
                     "Variant_Type",
                     "Reference_Allele",
                     "Tumor_Seq_Allele1",
                     "Tumor_Seq_Allele2")
          maf <- maf[-which(maf$Hugo_Symbol %in% flag),label]
          maf$Hugo_Symbol <- toupper(maf$Hugo_Symbol) # transfer gene name to capital. (eg, C1orf198 to C1ORF198)
          maf <- maf[which(substr(maf$Tumor_Sample_Barcode, 14, 15) == "01"),]
          maf$Tumor_Sample_Barcode <- paste0(maf$Tumor_Sample_Barcode,"A")
          write.table(maf, paste0("Validation_",cancers[ca],"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
          
          # Prepare binary somatic mutation
          # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
          # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
          # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
          # mut <- as.data.frame(na.omit(mut))
          # rownames(mut) <- mut$sample
          # mut <- mut[,-1]
          # mut <- mut[rowSums(mut) > 0,]
          # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
          # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
          
          # option 2: convert from MAF file without silent mutation (may not be the same with above file)
          # binary mutation
          # maf.ns <- maf[which(maf$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          mut.binary <- matrix(0,nrow = length(unique(maf$Hugo_Symbol)),ncol = length(unique(maf$Tumor_Sample_Barcode)),dimnames = list(unique(maf$Hugo_Symbol),unique(maf$Tumor_Sample_Barcode)))
          
          # for (i in colnames(mut.binary)) {
          #   tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
          #   if(is.element("Silent",tmp$Variant_Classification)) {
          #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
          #   }
          #   for (j in tmp$Hugo_Symbol)
          #     mut.binary[j,i] <- 1
          # }
          for (i in colnames(mut.binary)) {
            tmp <- maf[which(maf$Tumor_Sample_Barcode == i),]
            tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
            for (j in tmp$Hugo_Symbol){
              mut.binary[j,i] <- 1
            }
          }
          mut.binary <- as.data.frame(mut.binary)
          #rownames(mut.binary) <- toupper(rownames(mut.binary))
          mut.binary <- mut.binary[rowSums(mut.binary) > 0,]
          #mut.binary <- mut.binary[,which(substr(colnames(mut.binary), 14, 15) == "01")]
          #colnames(mut.binary) <- paste0(colnames(mut.binary),"A")
          
          write.table(mut.binary,paste0("Validation_",cancers[ca],"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #Combine maf files from different cancers
          if(ca==1){
            mafs <- maf
          }else{
            intersect_maf <- intersect(colnames(mafs),colnames(maf))
            mafs <- rbind(mafs[,intersect_maf],maf[,intersect_maf])
          }
          
          gc() #释放内存
        }
        
        progress$set(value = 3)
        ##Prepare binary somatic mutation
        # option 1: please download from XENA (TCGA Non-Small Cell Lung Cancer (NSCLC))
        # https://xenabrowser.net/datapages/?dataset=mc3_gene_level%2FLUNG_mc3_gene_level.txt&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443
        # mut <- read.delim("LUNG_mc3_gene_level.txt",sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        # mut <- as.data.frame(na.omit(mut))
        # rownames(mut) <- mut$sample
        # mut <- mut[,-1]
        # mut <- mut[rowSums(mut) > 0,]
        # mut <- mut[,which(substr(colnames(mut), 14, 15) == "01")]
        # write.table(mut, "TCGA_LUNG_mut1.txt", quote=F, row.names=T,col.names = NA,sep = "\t")
        
        # option 2: convert from MAF file without silent mutation (may not be the same with above file)
        # binary mutation
        # mafs.ns <- mafs[which(mafs$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
        muts.binary <- matrix(0,nrow = length(unique(mafs$Hugo_Symbol)),ncol = length(unique(mafs$Tumor_Sample_Barcode)),dimnames = list(unique(mafs$Hugo_Symbol),unique(mafs$Tumor_Sample_Barcode)))
        
        # for (i in colnames(muts.binary)) {
        #   tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
        #   if(is.element("Silent",tmp$Variant_Classification)) {
        #     tmp <- tmp[-which(tmp$Variant_Classification %in% "Silent"),]
        #   }
        #   for (j in tmp$Hugo_Symbol)
        #     muts.binary[j,i] <- 1
        # }
        for (i in colnames(muts.binary)) {
          tmp <- mafs[which(mafs$Tumor_Sample_Barcode == i),]
          tmp <- tmp[which(tmp$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation")),]
          for (j in tmp$Hugo_Symbol){
            muts.binary[j,i] <- 1
          }
        }
        muts.binary <- as.data.frame(muts.binary)
        #rownames(muts.binary) <- toupper(rownames(muts.binary))
        muts.binary <- muts.binary[rowSums(muts.binary) > 0,]
        #muts.binary <- muts.binary[,which(substr(colnames(muts.binary), 14, 15) == "01")]
        #colnames(muts.binary) <- paste0(colnames(muts.binary),"A")
        
        progress$set(value = 4)
        ### Final save
        write.table(mafs, paste0("Validation_",project,"_MAF.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
        data_list$"maf" <- mafs
        write.table(muts.binary,paste0("Validation_",project,"_mut2.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"muts.binary" <- muts.binary
        
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_mut_download2<-downloadHandler(
        filename = "binary_somatic_mutation_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_binary_somatic_mutation_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Binary somatic mutation(Validation) datasets have been well prepared, and you can download what you want below. Next, we will turn to prepare radiomics(Validation) datasets if you have specified.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_mut_download2",strong("Download RData file of binary somatic mutation datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_binary_somatic_mutation_finish<-renderUI({
        
        if(!(6 %in% input$validation_multiOmics)){
          list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
        }
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
    })
    
    #45.radiomics数据集链接下载处理按钮(Internal)
    observeEvent(input$internal_radiomics_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download radiomics dataset using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.internalRadiomicsUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #radiomics数据集下载及预处理
      data_list <- list() #数据集
      if(input$internal_radiomics_url_sources==1){
        
        progress$set(value = 2)
        #合并radiomics数据下载
        radiomics_url <- input$internal_radiomics_downloadUrl
        radiomics_url <- trimws(radiomics_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(radiomics_url,method="auto",destfile = file.path(workdir,paste0("TCGA_",project,"_radiomics_origin.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        radiomics <- read.table(paste0("TCGA_",project,"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        radiomics <- radiomics[,which(substr(colnames(radiomics), 14, 15) == "01")]
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的radiomics数据
        radiomics <- data.frame()
        progress$set(value = 3)
        for(url in 1:length(download_urls)){
          download.file(download_urls[url],method="auto",destfile = file.path(workdir,paste0("TCGA_",cancers[url],"_radiomics_origin.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          gc() #释放内存
          
          #导入并预处理下载好的各cancerType的radiomics数据集
          ra <- read.table(paste0("TCGA_",cancers[url],"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          ra <- as.data.frame(ra)
          ra <- ra[,which(substr(colnames(ra), 14, 15) == "01")]
          
          write.table(ra,paste0("TCGA_",cancers[url],"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine radiomics files from different cancers
          if(url==1){
            radiomics <- ra
          }else{
            intersect_radiomics <- intersect(rownames(radiomics),rownames(ra))
            radiomics <- cbind(radiomics[intersect_radiomics,],ra[intersect_radiomics,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$internal_radiomics_download1<-downloadHandler(
        filename = "radiomics_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$radiomics_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Radiomics datasets have been well prepared, and you can download what you want below.")),style="color:darkblack"),br(),
             column(12,downloadButton("internal_radiomics_download1",strong("Download RData file of radiomics datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$radiomics_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
    })
    
    #46.radiomics数据集链接下载处理按钮(External)
    observeEvent(input$external_radiomics_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download radiomics dataset using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.externalRadiomicsUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #radiomics数据集下载及预处理
      data_list <- list() #数据集
      if(input$external_radiomics_url_sources==1){
        
        progress$set(value = 2)
        #合并radiomics数据下载
        radiomics_url <- input$external_radiomics_downloadUrl
        radiomics_url <- trimws(radiomics_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(radiomics_url,method="auto",destfile = file.path(workdir,paste0("TCGA_",project,"_radiomics_origin.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        radiomics <- read.table(paste0("TCGA_",project,"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        radiomics <- radiomics[,which(substr(colnames(radiomics), 14, 15) == "01")]
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的radiomics数据
        radiomics <- data.frame()
        progress$set(value = 3)
        for(url in 1:length(download_urls)){
          download.file(download_urls[url],method="auto",destfile = file.path(workdir,paste0("TCGA_",cancers[url],"_radiomics_origin.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          gc() #释放内存
          
          #导入并预处理下载好的各cancerType的radiomics数据集
          ra <- read.table(paste0("TCGA_",cancers[url],"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          ra <- as.data.frame(ra)
          ra <- ra[,which(substr(colnames(ra), 14, 15) == "01")]
          
          write.table(ra,paste0("TCGA_",cancers[url],"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine radiomics files from different cancers
          if(url==1){
            radiomics <- ra
          }else{
            intersect_radiomics <- intersect(rownames(radiomics),rownames(ra))
            radiomics <- cbind(radiomics[intersect_radiomics,],ra[intersect_radiomics,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_radiomics_download1<-downloadHandler(
        filename = "radiomics_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$radiomics_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Radiomics datasets have been well prepared, and you can download what you want below.")),style="color:darkblack"),br(),
             column(12,downloadButton("external_radiomics_download1",strong("Download RData file of radiomics datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$radiomics_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
    })
    
    #47.radiomics数据集链接下载处理按钮(External)(验证数据集)
    observeEvent(input$validation_external_radiomics_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download radiomics dataset using specified urls',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      download_urls <- data.frame(rbind(hot_to_r(input$table.validation_externalRadiomicsUrl)))[,1]
      download_urls <- trimws(download_urls,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #radiomics数据集下载及预处理
      data_list <- list() #数据集
      if(input$validation_external_radiomics_url_sources==1){
        
        progress$set(value = 2)
        #合并radiomics数据下载
        radiomics_url <- input$validation_external_radiomics_downloadUrl
        radiomics_url <- trimws(radiomics_url,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        download.file(radiomics_url,method="auto",destfile = file.path(validation_workdir,paste0("Validation_",project,"_radiomics_origin.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
        gc() #释放内存
        
        progress$set(value = 3)
        radiomics <- read.table(paste0("Validation_",project,"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        radiomics <- radiomics[,which(substr(colnames(radiomics), 14, 15) == "01")]
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("Validation_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
        
      }else{
        
        progress$set(value = 2)
        #循环下载各cancerType的radiomics数据
        radiomics <- data.frame()
        progress$set(value = 3)
        for(url in 1:length(download_urls)){
          download.file(download_urls[url],method="auto",destfile = file.path(validation_workdir,paste0("Validation_",cancers[url],"_radiomics_origin.txt")))  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
          gc() #释放内存
          
          #导入并预处理下载好的各cancerType的radiomics数据集
          ra <- read.table(paste0("Validation_",cancers[url],"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          ra <- as.data.frame(ra)
          ra <- ra[,which(substr(colnames(ra), 14, 15) == "01")]
          
          write.table(ra,paste0("Validation_",cancers[url],"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine radiomics files from different cancers
          if(url==1){
            radiomics <- ra
          }else{
            intersect_radiomics <- intersect(rownames(radiomics),rownames(ra))
            radiomics <- cbind(radiomics[intersect_radiomics,],ra[intersect_radiomics,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("Validation_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_radiomics_download1<-downloadHandler(
        filename = "radiomics_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_radiomics_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Radiomics(Validation) datasets have been well prepared, and you can download what you want below.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_radiomics_download1",strong("Download RData file of radiomics datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_radiomics_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
    })
    
    #48.radiomics数据集文件上传处理按钮(Internal)
    observeEvent(input$internal_radiomics_process2,{
      
      # data.table::setDTthreads(threads = 1) #设置程序运行线程为1
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload radiomics datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$internal_radiomics_fileUpload1$datapath #合并radiomics数据集上传路径
      upload_paths2 <- input$internal_radiomics_fileUpload2$datapath #各cancerType的radiomics数据集上传路径
      
      #radiomics数据集转移及预处理
      data_list <- list() #数据集
      if(input$internal_radiomics_upload_sources==1){
        
        progress$set(value = 2)
        #合并radiomics数据集
        # file.move(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_radiomics_origin.txt"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_radiomics_origin.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并radiomics数据集
        radiomics <- read.table(paste0("TCGA_",project,"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        radiomics <- radiomics[,which(substr(colnames(radiomics), 14, 15) == "01")]
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
        
      }else{
        
        progress$set(value = 2)
        #循环转移并处理各cancerType的radiomics数据
        radiomics <- data.frame()
        progress$set(value = 3)
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_radiomics_origin.txt"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_radiomics_origin.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的radiomics数据集
          ra <- read.table(paste0("TCGA_",cancers[ca],"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          ra <- as.data.frame(ra)
          ra <- ra[,which(substr(colnames(ra), 14, 15) == "01")]
          
          write.table(ra,paste0("TCGA_",cancers[ca],"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine radiomics files from different cancers
          if(ca==1){
            radiomics <- ra
          }else{
            intersect_radiomics <- intersect(rownames(radiomics),rownames(ra))
            radiomics <- cbind(radiomics[intersect_radiomics,],ra[intersect_radiomics,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$internal_radiomics_download2<-downloadHandler(
        filename = "radiomics_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$radiomics_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Radiomics datasets have been well prepared, and you can download what you want below.")),style="color:darkblack"),br(),
             column(12,downloadButton("internal_radiomics_download2",strong("Download RData file of radiomics datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$radiomics_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #49.radiomics数据集文件上传处理按钮(External)
    observeEvent(input$external_radiomics_process2,{
      
      # data.table::setDTthreads(threads = 1) #设置程序运行线程为1
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload radiomics datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$external_radiomics_fileUpload1$datapath #合并radiomics数据集上传路径
      upload_paths2 <- input$external_radiomics_fileUpload2$datapath #各cancerType的radiomics数据集上传路径
      
      #radiomics数据集转移及预处理
      data_list <- list() #数据集
      if(input$external_radiomics_upload_sources==1){
        
        progress$set(value = 2)
        #合并radiomics数据集
        # file.move(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_radiomics_origin.txt"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(workdir,paste0("TCGA_",project,"_radiomics_origin.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并radiomics数据集
        radiomics <- read.table(paste0("TCGA_",project,"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        radiomics <- radiomics[,which(substr(colnames(radiomics), 14, 15) == "01")]
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
        
      }else{
        
        progress$set(value = 2)
        #循环转移并处理各cancerType的radiomics数据
        radiomics <- data.frame()
        progress$set(value = 3)
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_radiomics_origin.txt"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(workdir,paste0("TCGA_",cancers[ca],"_radiomics_origin.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的radiomics数据集
          ra <- read.table(paste0("TCGA_",cancers[ca],"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          ra <- as.data.frame(ra)
          ra <- ra[,which(substr(colnames(ra), 14, 15) == "01")]
          
          write.table(ra,paste0("TCGA_",cancers[ca],"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine radiomics files from different cancers
          if(ca==1){
            radiomics <- ra
          }else{
            intersect_radiomics <- intersect(rownames(radiomics),rownames(ra))
            radiomics <- cbind(radiomics[intersect_radiomics,],ra[intersect_radiomics,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("TCGA_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$external_radiomics_download2<-downloadHandler(
        filename = "radiomics_preparation(TCGA).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$radiomics_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Radiomics datasets have been well prepared, and you can download what you want below.")),style="color:darkblack"),br(),
             column(12,downloadButton("external_radiomics_download2",strong("Download RData file of radiomics datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$radiomics_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Now all omics datasets you specified for tcga cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for tcga cohort.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #50.radiomics数据集文件上传处理按钮(External)(验证数据集)
    observeEvent(input$validation_external_radiomics_process2,{
      
      # data.table::setDTthreads(threads = 1) #设置程序运行线程为1
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload radiomics datasets from local paths',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      upload_paths1 <- input$validation_external_radiomics_fileUpload1$datapath #合并radiomics数据集上传路径
      upload_paths2 <- input$validation_external_radiomics_fileUpload2$datapath #各cancerType的radiomics数据集上传路径
      
      #radiomics数据集转移及预处理
      data_list <- list() #数据集
      if(input$validation_external_radiomics_upload_sources==1){
        
        progress$set(value = 2)
        #合并radiomics数据集
        # file.move(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_radiomics_origin.txt"))) #将上传的文件移动到工作目录下
        file.copy(upload_paths1, file.path(validation_workdir,paste0("Validation_",project,"_radiomics_origin.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
        
        progress$set(value = 3)
        #导入并预处理已上传的合并radiomics数据集
        radiomics <- read.table(paste0("Validation_",project,"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        radiomics <- radiomics[,which(substr(colnames(radiomics), 14, 15) == "01")]
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("Validation_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
        
      }else{
        
        progress$set(value = 2)
        #循环转移并处理各cancerType的radiomics数据
        radiomics <- data.frame()
        progress$set(value = 3)
        for(ca in 1:length(cancers)){
          
          # file.move(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_radiomics_origin.txt"))) #将上传的文件移动到工作目录下
          file.copy(upload_paths2[ca], file.path(validation_workdir,paste0("Validation_",cancers[ca],"_radiomics_origin.txt")),overwrite=TRUE) #将上传的文件复制到工作目录下
          
          #导入并预处理已上传的各cancerType的radiomics数据集
          ra <- read.table(paste0("Validation_",cancers[ca],"_radiomics_origin.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
          ra <- as.data.frame(ra)
          ra <- ra[,which(substr(colnames(ra), 14, 15) == "01")]
          
          write.table(ra,paste0("Validation_",cancers[ca],"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          # Combine radiomics files from different cancers
          if(ca==1){
            radiomics <- ra
          }else{
            intersect_radiomics <- intersect(rownames(radiomics),rownames(ra))
            radiomics <- cbind(radiomics[intersect_radiomics,],ra[intersect_radiomics,])
          }
          
          gc() #释放内存
          
        }
        
        progress$set(value = 4)
        ### Final save
        write.table(radiomics,paste0("Validation_",project,"_radiomics.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
        data_list$"radiomics" <- radiomics
      }
      
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      ##RData结果下载按钮响应事件
      output$validation_radiomics_download2<-downloadHandler(
        filename = "radiomics_preparation(Validation).RData",
        content = function(theFile) {
          save(data_list,file=theFile)
        }
      )
      
      #数据下载好后给予用户的反馈提示
      output$validation_radiomics_download_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Radiomics(Validation) datasets have been well prepared, and you can download what you want below.")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_radiomics_download2",strong("Download RData file of radiomics datasets before integration"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      output$validation_radiomics_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Now all omics datasets you specified for validation cohort have been well prepared, and you can download what you want below. Then we will turn to integrate all omics datasets for validation cohort.")),style="color:darkblack"))
      })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #51.TCGA数据集整合处理按钮
    observeEvent(input$datasets_integrate,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Integrate specified omics datasets and extra datasets (TCGA)',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##1.获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #2.导入整理好的生存数据
      surv.infos <- read.table(paste0("TCGA_",project,"_Survival.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
      comsam <- rownames(surv.infos)
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      #3.导入整理好的临床数据
      clin.info <- read.delim(paste0("TCGA_",project,"_Clinical_preprocess.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
      rownames(clin.info) <- clin.info[,1]
      comsam <- intersect(comsam,rownames(clin.info))
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      progress$set(value = 2)
      #4.导入整理好的mRNA数据(如果有的话)
      if(1 %in% input$multiOmics){
        
        countsTable <- read.table(paste0("TCGA_",project,"_Count.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        fpkm.mrna <- read.table(paste0("TCGA_",project,"_mRNA_FPKM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        tpm.mrna <- read.table(paste0("TCGA_",project,"_mRNA_TPM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(tpm.mrna))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #5.导入整理好的lncRNA数据(如果有的话)
      if(2 %in% input$multiOmics){
        
        countsTable <- read.table(paste0("TCGA_",project,"_Count.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        fpkm.lncrna <- read.table(paste0("TCGA_",project,"_lncRNA_FPKM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        tpm.lncrna <- read.table(paste0("TCGA_",project,"_lncRNA_TPM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(tpm.lncrna))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #6.导入整理好的DNA methylation数据(如果有的话)
      if(3 %in% input$multiOmics){
        
        meths <- read.table(paste0("TCGA_",project,"_Methpromoter.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(meths))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      progress$set(value = 3)
      #7.导入整理好的copy number alterations数据(如果有的话)
      if(4 %in% input$multiOmics){
        
        segments <- read.table(paste0("TCGA_",project,"_segment.txt"),sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        cnas <- read.table(paste0("TCGA_",project,"_cna.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(cnas))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #8.导入整理好的binary somatic mutation数据(如果有的话)
      if(5 %in% input$multiOmics){
        
        mafs <- read_tsv(paste0("TCGA_",project,"_MAF.txt"), comment = "#")
        muts.binary <- read.table(paste0("TCGA_",project,"_mut2.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(muts.binary))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #9.导入整理好的radiomics数据(如果有的话)
      if(6 %in% input$multiOmics){
        
        radiomics <- read.table(paste0("TCGA_",project,"_radiomics.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        comsam <- intersect(comsam,colnames(radiomics))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      progress$set(value = 4)
      #10.构造一个存放整合数据的列表
      integrated_data <- list()
      
      #11.整合数据
      ##Five types of multi-omics data
      if(1 %in% input$multiOmics){
        mRNA.expr <- tpm.mrna[,comsam]
        mRNA.expr <- as.data.frame(mRNA.expr)
        integrated_data$"mRNA.expr" <- mRNA.expr
        write.table(mRNA.expr,paste0("TCGA_",project,"_mRNA_TPM_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(2 %in% input$multiOmics){
        lncRNA.expr <- tpm.lncrna[,comsam]
        lncRNA.expr <- as.data.frame(lncRNA.expr)
        integrated_data$"lncRNA.expr" <- lncRNA.expr
        write.table(lncRNA.expr,paste0("TCGA_",project,"_lncRNA_TPM_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(3 %in% input$multiOmics){
        meth.beta <- meths[,comsam]
        meth.beta <- as.data.frame(meth.beta)
        integrated_data$"meth.beta" <- meth.beta
        write.table(meth.beta,paste0("TCGA_",project,"_Methpromoter_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(4 %in% input$multiOmics){
        cna.beta <- cnas[,comsam]
        integrated_data$"cna" <- cna.beta
        write.table(cna.beta,paste0("TCGA_",project,"_cna_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(5 %in% input$multiOmics){
        mut.status <- muts.binary[,comsam]
        integrated_data$"mut.status" <- mut.status
        write.table(mut.status,paste0("TCGA_",project,"_mut2_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(6 %in% input$multiOmics){
        radiomics <- radiomics[,comsam]
        integrated_data$"radiomics" <- radiomics
        write.table(radiomics,paste0("TCGA_",project,"_radiomics_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      ##Extra data for downstream analyses
      if(1 %in% input$multiOmics){
        
        counts <- countsTable[,comsam]
        integrated_data$"count" <- counts
        write.table(counts,paste0("TCGA_",project,"_Count_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        fpkm_mrna <- fpkm.mrna[,comsam]
        integrated_data$"fpkm_mrna" <- fpkm_mrna
        write.table(fpkm_mrna,paste0("TCGA_",project,"_mRNA_FPKM_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
      }
      
      if(2 %in% input$multiOmics){
        counts <- countsTable[,comsam]
        integrated_data$"count" <- counts
        write.table(counts,paste0("TCGA_",project,"_Count_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        fpkm_lncrna <- fpkm.lncrna[,comsam]
        integrated_data$"fpkm_lncrna" <- fpkm_lncrna
        write.table(fpkm_lncrna,paste0("TCGA_",project,"_lncRNA_FPKM_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
      }
      
      if(4 %in% input$multiOmics){
        segment_comsam <- segments[which(segments$sample %in% comsam),]
        integrated_data$"segment" <- segment_comsam
        write.table(segment_comsam,paste0("TCGA_",project,"_segment_comsam.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
      }
      
      if(5 %in% input$multiOmics){
        maf_comsam <- mafs[which(mafs$Tumor_Sample_Barcode %in% comsam),]
        integrated_data$"maf" <- maf_comsam
        write.table(maf_comsam,paste0("TCGA_",project,"_MAF_comsam.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
      }
      
      clin.infos <- cbind.data.frame(surv.infos[comsam,],clin.info[comsam,])
      integrated_data$"clin.info" <- clin.infos
      write.table(clin.infos,paste0("TCGA_",project,"_surv_clin_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      #12.保存整合好的数据(.RData)
      save(integrated_data, file = paste0(tolower(project),".tcga.RData"))
      save.image(paste0("Data_Preparation_for_tcga_",project,"_MOVICS_Pipeline.RData"))
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      ##RData结果下载按钮响应事件
      output$datasets_integration_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.RData"),
        content = function(theFile) {
          save(integrated_data,file=theFile)
        }
      )
      
      #数据整理好后给予用户的反馈提示
      output$datasets_integration_finish<-renderUI({
        
        if(input$external_validation==1){
          ##有外部验证数据集
          list(br(),column(12,h3(strong("All specified omics datasets and extra datasets from TCGA cohort have been integrated into a list for the following MOVICS pipeline, and you can download and check the obtained files you want below. Now let's start to prepare validation datasets.")),style="color:darkblack"),br(),
               column(12,downloadButton("datasets_integration_download",strong("Download RData file of integrated datasets"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          ##无外部验证数据集
          list(br(),column(12,h3(strong("All specified omics datasets and extra datasets from TCGA cohort have been integrated into a list for the following MOVICS pipeline, and you can download and check the obtained files you want below. Now let's start the first step of MOVICS--GET Module!")),style="color:darkblack"),br(),
               column(12,downloadButton("datasets_integration_download",strong("Download RData file of integrated datasets"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      # ##整合好的mRNA(tpm)表格标题
      # output$table.integrated_mRNA_tpm_TCGA_Title<-renderUI({
      #   
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: mRNA (tpm) from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的mRNA(tpm)表格
      # output$table.integrated_mRNA_tpm_TCGA<-DT::renderDataTable({
      #   
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"mRNA.expr"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"mRNA.expr")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的mRNA(tpm)表格脚注
      # output$table.integrated_mRNA_tpm_TCGA_Note<-renderUI({
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的lncRNA(tpm)表格标题
      # output$table.integrated_lncRNA_tpm_TCGA_Title<-renderUI({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: lncRNA (tpm) from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的lncRNA(tpm)表格
      # output$table.integrated_lncRNA_tpm_TCGA<-DT::renderDataTable({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"lncRNA.expr"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"lncRNA.expr")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的lncRNA(tpm)表格脚注
      # output$table.integrated_lncRNA_tpm_TCGA_Note<-renderUI({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的DNA methylation表格标题
      # output$table.integrated_DNA_methylation_TCGA_Title<-renderUI({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: DNA methylation from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的DNA methylation表格
      # output$table.integrated_DNA_methylation_TCGA<-DT::renderDataTable({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"meth.beta"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"meth.beta")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "DNA_methylation_from_integrated_dataset(TCGA)",
      #                            title = "Table: DNA methylation from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "DNA_methylation_from_integrated_dataset(TCGA)",
      #                            title = "Table: DNA methylation from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "DNA_methylation_from_integrated_dataset(TCGA)",
      #                            title = "Table: DNA methylation from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "DNA_methylation_from_integrated_dataset(TCGA)",
      #                            title = "Table: DNA methylation from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的DNA methylation表格脚注
      # output$table.integrated_DNA_methylation_TCGA_Note<-renderUI({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的copy number alterations表格标题
      # output$table.integrated_copy_number_alterations_TCGA_Title<-renderUI({
      #   
      #   if("cna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: copy number alterations from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的copy number alterations表格
      # output$table.integrated_copy_number_alterations_TCGA<-DT::renderDataTable({
      #   
      #   if("cna" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"cna"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"cna")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "copy_number_alterations_from_integrated_dataset(TCGA)",
      #                            title = "Table: copy number alterations from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "copy_number_alterations_from_integrated_dataset(TCGA)",
      #                            title = "Table: copy number alterations from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "copy_number_alterations_from_integrated_dataset(TCGA)",
      #                            title = "Table: copy number alterations from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "copy_number_alterations_from_integrated_dataset(TCGA)",
      #                            title = "Table: copy number alterations from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的copy number alterations表格脚注
      # output$table.integrated_copy_number_alterations_TCGA_Note<-renderUI({
      #  
      #   if("cna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的binary somatic mutation表格标题
      # output$table.integrated_binary_somatic_mutation_TCGA_Title<-renderUI({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: binary somatic mutation from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的binary somatic mutation表格
      # output$table.integrated_binary_somatic_mutation_TCGA<-DT::renderDataTable({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"mut.status"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"mut.status")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(TCGA)",
      #                            title = "Table: binary somatic mutation from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(TCGA)",
      #                            title = "Table: binary somatic mutation from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(TCGA)",
      #                            title = "Table: binary somatic mutation from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(TCGA)",
      #                            title = "Table: binary somatic mutation from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的binary somatic mutation表格脚注
      # output$table.integrated_binary_somatic_mutation_TCGA_Note<-renderUI({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的radiomics表格标题
      # output$table.integrated_radiomics_TCGA_Title<-renderUI({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: radiomics from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的radiomics表格
      # output$table.integrated_radiomics_TCGA<-DT::renderDataTable({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"radiomics"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"radiomics")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "radiomics_from_integrated_dataset(TCGA)",
      #                            title = "Table: radiomics from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "radiomics_from_integrated_dataset(TCGA)",
      #                            title = "Table: radiomics from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "radiomics_from_integrated_dataset(TCGA)",
      #                            title = "Table: radiomics from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "radiomics_from_integrated_dataset(TCGA)",
      #                            title = "Table: radiomics from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的radiomics表格脚注
      # output$table.integrated_radiomics_TCGA_Note<-renderUI({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的count表格标题
      # output$table.integrated_count_TCGA_Title<-renderUI({
      #   
      #   if("count" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: count from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的count表格
      # output$table.integrated_count_TCGA<-DT::renderDataTable({
      #   
      #   if("count" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"count"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"count")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "count_from_integrated_dataset(TCGA)",
      #                            title = "Table: count from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "count_from_integrated_dataset(TCGA)",
      #                            title = "Table: count from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "count_from_integrated_dataset(TCGA)",
      #                            title = "Table: count from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "count_from_integrated_dataset(TCGA)",
      #                            title = "Table: count from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的count表格脚注
      # output$table.integrated_count_TCGA_Note<-renderUI({
      #   
      #   if("count" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的mRNA(fpkm)表格标题
      # output$table.integrated_mRNA_fpkm_TCGA_Title<-renderUI({
      #   
      #   if("fpkm_mrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: mRNA (fpkm) from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的mRNA(fpkm)表格
      # output$table.integrated_mRNA_fpkm_TCGA<-DT::renderDataTable({
      #   
      #   if("fpkm_mrna" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"fpkm_mrna"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"fpkm_mrna")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的mRNA(fpkm)表格脚注
      # output$table.integrated_mRNA_fpkm_TCGA_Note<-renderUI({
      #   if("fpkm_mrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的lncRNA(fpkm)表格标题
      # output$table.integrated_lncRNA_fpkm_TCGA_Title<-renderUI({
      #   
      #   if("fpkm_lncrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: lncRNA (fpkm) from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的lncRNA(fpkm)表格
      # output$table.integrated_lncRNA_fpkm_TCGA<-DT::renderDataTable({
      #   
      #   if("fpkm_lncrna" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"fpkm_lncrna"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"fpkm_lncrna")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(TCGA)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的lncRNA(fpkm)表格脚注
      # output$table.integrated_lncRNA_fpkm_TCGA_Note<-renderUI({
      #   
      #   if("fpkm_lncrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # ##整合好的segment表格标题
      # output$table.integrated_segment_TCGA_Title<-renderUI({
      #   
      #   if("segment" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: segment from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的segment表格
      # output$table.integrated_segment_TCGA<-DT::renderDataTable({
      #   
      #   if("segment" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"segment"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"segment")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "segment_from_integrated_dataset(TCGA)",
      #                            title = "Table: segment from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "segment_from_integrated_dataset(TCGA)",
      #                            title = "Table: segment from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "segment_from_integrated_dataset(TCGA)",
      #                            title = "Table: segment from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "segment_from_integrated_dataset(TCGA)",
      #                            title = "Table: segment from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   ),rownames = FALSE
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的segment表格脚注
      # output$table.integrated_segment_TCGA_Note<-renderUI({
      #   
      #   if("segment" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # ##整合好的maf表格标题
      # output$table.integrated_maf_TCGA_Title<-renderUI({
      #   
      #   if("maf" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: maf from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的maf表格
      # output$table.integrated_maf_TCGA<-DT::renderDataTable({
      #   
      #   if("maf" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"maf"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"maf")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "maf_from_integrated_dataset(TCGA)",
      #                            title = "Table: maf from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "maf_from_integrated_dataset(TCGA)",
      #                            title = "Table: maf from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "maf_from_integrated_dataset(TCGA)",
      #                            title = "Table: maf from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "maf_from_integrated_dataset(TCGA)",
      #                            title = "Table: maf from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   ),rownames = FALSE
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的maf表格脚注
      # output$table.integrated_maf_TCGA_Note<-renderUI({
      #   
      #   if("maf" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # ##整合好的clinSurv表格标题
      # output$table.integrated_clinSurv_TCGA_Title<-renderUI({
      #   
      #   if("clin.info" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: clinical and survival information from integrated dataset (TCGA)" )
      #   }
      # })
      # 
      # ##整合好的clinSurv表格
      # output$table.integrated_clinSurv_TCGA<-DT::renderDataTable({
      #   
      #   if("clin.info" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"clin.info"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"clin.info")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "clinSurv_from_integrated_dataset(TCGA)",
      #                            title = "Table: clinical and survival information from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "clinSurv_from_integrated_dataset(TCGA)",
      #                            title = "Table: clinical and survival information from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "clinSurv_from_integrated_dataset(TCGA)",
      #                            title = "Table: clinical and survival information from integrated dataset (TCGA)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "clinSurv_from_integrated_dataset(TCGA)",
      #                            title = "Table: clinical and survival information from integrated dataset (TCGA)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的clinSurv表格脚注
      # output$table.integrated_clinSurv_TCGA_Note<-renderUI({
      #   
      #   if("clin.info" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #52.数据集整合处理按钮(验证数据集)
    observeEvent(input$validation_datasets_integrate,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Integrate specified omics datasets and extra datasets (Validation)',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##1.获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      validation.path <- file.path(user_workdir,"Validation_data_preparing")
      setwd(validation.path)
      validation_workdir <- getwd()
      
      #2.导入整理好的生存数据
      surv.infos <- read.table(paste0("Validation_",project,"_Survival.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
      comsam <- rownames(surv.infos)
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      #3.导入整理好的临床数据
      clin.info <- read.delim(paste0("Validation_",project,"_Clinical_preprocess.txt"),sep="\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
      rownames(clin.info) <- clin.info[,1]
      comsam <- intersect(comsam,rownames(clin.info))
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      progress$set(value = 2)
      #4.导入整理好的mRNA数据(如果有的话)
      if(1 %in% input$validation_multiOmics){
        
        countsTable <- read.table(paste0("Validation_",project,"_Count.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        fpkm.mrna <- read.table(paste0("Validation_",project,"_mRNA_FPKM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        tpm.mrna <- read.table(paste0("Validation_",project,"_mRNA_TPM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(tpm.mrna))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #5.导入整理好的lncRNA数据(如果有的话)
      if(2 %in% input$validation_multiOmics){
        
        countsTable <- read.table(paste0("Validation_",project,"_Count.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        fpkm.lncrna <- read.table(paste0("Validation_",project,"_lncRNA_FPKM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        tpm.lncrna <- read.table(paste0("Validation_",project,"_lncRNA_TPM.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(tpm.lncrna))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #6.导入整理好的DNA methylation数据(如果有的话)
      if(3 %in% input$validation_multiOmics){
        
        meths <- read.table(paste0("Validation_",project,"_Methpromoter.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(meths))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      progress$set(value = 3)
      #7.导入整理好的copy number alterations数据(如果有的话)
      if(4 %in% input$validation_multiOmics){
        
        segments <- read.table(paste0("Validation_",project,"_segment.txt"),sep = "\t",row.names = NULL,check.names = F,stringsAsFactors = F,header = T)
        cnas <- read.table(paste0("Validation_",project,"_cna.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(cnas))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #8.导入整理好的binary somatic mutation数据(如果有的话)
      if(5 %in% input$validation_multiOmics){
        
        mafs <- read_tsv(paste0("Validation_",project,"_MAF.txt"), comment = "#")
        muts.binary <- read.table(paste0("Validation_",project,"_mut2.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        comsam <- intersect(comsam,colnames(muts.binary))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      #9.导入整理好的radiomics数据(如果有的话)
      if(6 %in% input$validation_multiOmics){
        
        radiomics <- read.table(paste0("Validation_",project,"_radiomics.txt"),sep = "\t",row.names = 1,check.names = F,stringsAsFactors = F,header = T)
        radiomics <- as.data.frame(radiomics)
        comsam <- intersect(comsam,colnames(radiomics))
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
      
      progress$set(value = 4)
      #10.构造一个存放整合数据的列表
      integrated_data <- list()
      
      #11.整合数据
      ##Five types of multi-omics data
      if(1 %in% input$validation_multiOmics){
        mRNA.expr <- tpm.mrna[,comsam]
        mRNA.expr <- as.data.frame(mRNA.expr)
        integrated_data$"mRNA.expr" <- mRNA.expr
        write.table(mRNA.expr,paste0("Validation_",project,"_mRNA_TPM_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(2 %in% input$validation_multiOmics){
        lncRNA.expr <- tpm.lncrna[,comsam]
        lncRNA.expr <- as.data.frame(lncRNA.expr)
        integrated_data$"lncRNA.expr" <- lncRNA.expr
        write.table(lncRNA.expr,paste0("Validation_",project,"_lncRNA_TPM_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(3 %in% input$validation_multiOmics){
        meth.beta <- meths[,comsam]
        meth.beta <- as.data.frame(meth.beta)
        integrated_data$"meth.beta" <- meth.beta
        write.table(meth.beta,paste0("Validation_",project,"_Methpromoter_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(4 %in% input$validation_multiOmics){
        cna.beta <- cnas[,comsam]
        integrated_data$"cna" <- cna.beta
        write.table(cna.beta,paste0("Validation_",project,"_cna_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(5 %in% input$validation_multiOmics){
        mut.status <- muts.binary[,comsam]
        integrated_data$"mut.status" <- mut.status
        write.table(mut.status,paste0("Validation_",project,"_mut2_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      if(6 %in% input$validation_multiOmics){
        radiomics <- radiomics[,comsam]
        integrated_data$"radiomics" <- radiomics
        write.table(radiomics,paste0("Validation_",project,"_radiomics_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      }
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      ##Extra data for downstream analyses
      if(1 %in% input$validation_multiOmics){
        
        counts <- countsTable[,comsam]
        integrated_data$"count" <- counts
        write.table(counts,paste0("Validation_",project,"_Count_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        fpkm_mrna <- fpkm.mrna[,comsam]
        integrated_data$"fpkm_mrna" <- fpkm_mrna
        write.table(fpkm_mrna,paste0("Validation_",project,"_mRNA_FPKM_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
      }
      
      if(2 %in% input$validation_multiOmics){
        counts <- countsTable[,comsam]
        integrated_data$"count" <- counts
        write.table(counts,paste0("Validation_",project,"_Count_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
        fpkm_lncrna <- fpkm.lncrna[,comsam]
        integrated_data$"fpkm_lncrna" <- fpkm_lncrna
        write.table(fpkm_lncrna,paste0("Validation_",project,"_lncRNA_FPKM_comsam.txt"),quote=F,row.names=T,col.names=NA,sep="\t")
      }
      
      if(4 %in% input$validation_multiOmics){
        segment_comsam <- segments[which(segments$sample %in% comsam),]
        integrated_data$"segment" <- segment_comsam
        write.table(segment_comsam,paste0("Validation_",project,"_segment_comsam.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
      }
      
      if(5 %in% input$validation_multiOmics){
        maf_comsam <- mafs[which(mafs$Tumor_Sample_Barcode %in% comsam),]
        integrated_data$"maf" <- maf_comsam
        write.table(maf_comsam,paste0("Validation_",project,"_MAF_comsam.txt"), quote=F, row.names=F,col.names = T,sep = "\t")
      }
      
      clin.infos <- cbind.data.frame(surv.infos[comsam,],clin.info[comsam,])
      integrated_data$"clin.info" <- clin.infos
      write.table(clin.infos,paste0("Validation_",project,"_surv_clin_comsam.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      #12.保存整合好的数据(.RData)
      save(integrated_data, file = paste0(tolower(project),".validation.RData"))
      save.image(paste0("Data_Preparation_for_validation_",project,"_MOVICS_Pipeline.RData"))
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      ##RData结果下载按钮响应事件
      output$validation_datasets_integration_download<-downloadHandler(
        filename = paste0(tolower(project),".validation.RData"),
        content = function(theFile) {
          save(integrated_data,file=theFile)
        }
      )
      
      #数据整理好后给予用户的反馈提示
      output$validation_datasets_integration_finish<-renderUI({
        
        list(br(),column(12,h3(strong("All specified omics datasets and extra datasets from validation cohort have been integrated into a list for the following MOVICS pipeline, and you can download and check the obtained files you want below. Now let's start the first step of MOVICS--GET Module!")),style="color:darkblack"),br(),
             column(12,downloadButton("validation_datasets_integration_download",strong("Download RData file of integrated datasets"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      # ##整合好的mRNA(tpm)表格标题
      # output$table.integrated_mRNA_tpm_Validation_Title<-renderUI({
      #   
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: mRNA (tpm) from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的mRNA(tpm)表格
      # output$table.integrated_mRNA_tpm_Validation<-DT::renderDataTable({
      #   
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"mRNA.expr"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"mRNA.expr")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "mRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (tpm) from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的mRNA(tpm)表格脚注
      # output$table.integrated_mRNA_tpm_Validation_Note<-renderUI({
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的lncRNA(tpm)表格标题
      # output$table.integrated_lncRNA_tpm_Validation_Title<-renderUI({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: lncRNA (tpm) from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的lncRNA(tpm)表格
      # output$table.integrated_lncRNA_tpm_Validation<-DT::renderDataTable({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"lncRNA.expr"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"lncRNA.expr")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "lncRNA(tpm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (tpm) from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的lncRNA(tpm)表格脚注
      # output$table.integrated_lncRNA_tpm_Validation_Note<-renderUI({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的DNA methylation表格标题
      # output$table.integrated_DNA_methylation_Validation_Title<-renderUI({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: DNA methylation from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的DNA methylation表格
      # output$table.integrated_DNA_methylation_Validation<-DT::renderDataTable({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"meth.beta"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"meth.beta")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "DNA_methylation_from_integrated_dataset(Validation)",
      #                            title = "Table: DNA methylation from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "DNA_methylation_from_integrated_dataset(Validation)",
      #                            title = "Table: DNA methylation from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "DNA_methylation_from_integrated_dataset(Validation)",
      #                            title = "Table: DNA methylation from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "DNA_methylation_from_integrated_dataset(Validation)",
      #                            title = "Table: DNA methylation from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的DNA methylation表格脚注
      # output$table.integrated_DNA_methylation_Validation_Note<-renderUI({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的copy number alterations表格标题
      # output$table.integrated_copy_number_alterations_Validation_Title<-renderUI({
      #   
      #   if("cna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: copy number alterations from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的copy number alterations表格
      # output$table.integrated_copy_number_alterations_Validation<-DT::renderDataTable({
      #   
      #   if("cna" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"cna"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"cna")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "copy_number_alterations_from_integrated_dataset(Validation)",
      #                            title = "Table: copy number alterations from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "copy_number_alterations_from_integrated_dataset(Validation)",
      #                            title = "Table: copy number alterations from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "copy_number_alterations_from_integrated_dataset(Validation)",
      #                            title = "Table: copy number alterations from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "copy_number_alterations_from_integrated_dataset(Validation)",
      #                            title = "Table: copy number alterations from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的copy number alterations表格脚注
      # output$table.integrated_copy_number_alterations_Validation_Note<-renderUI({
      #   
      #   if("cna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的binary somatic mutation表格标题
      # output$table.integrated_binary_somatic_mutation_Validation_Title<-renderUI({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: binary somatic mutation from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的binary somatic mutation表格
      # output$table.integrated_binary_somatic_mutation_Validation<-DT::renderDataTable({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"mut.status"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"mut.status")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(Validation)",
      #                            title = "Table: binary somatic mutation from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(Validation)",
      #                            title = "Table: binary somatic mutation from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(Validation)",
      #                            title = "Table: binary somatic mutation from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "binary_somatic_mutation_from_integrated_dataset(Validation)",
      #                            title = "Table: binary somatic mutation from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的binary somatic mutation表格脚注
      # output$table.integrated_binary_somatic_mutation_Validation_Note<-renderUI({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的radiomics表格标题
      # output$table.integrated_radiomics_Validation_Title<-renderUI({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: radiomics from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的radiomics表格
      # output$table.integrated_radiomics_Validation<-DT::renderDataTable({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"radiomics"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"radiomics")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "radiomics_from_integrated_dataset(Validation)",
      #                            title = "Table: radiomics from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "radiomics_from_integrated_dataset(Validation)",
      #                            title = "Table: radiomics from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "radiomics_from_integrated_dataset(Validation)",
      #                            title = "Table: radiomics from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "radiomics_from_integrated_dataset(Validation)",
      #                            title = "Table: radiomics from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的radiomics表格脚注
      # output$table.integrated_radiomics_Validation_Note<-renderUI({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的count表格标题
      # output$table.integrated_count_Validation_Title<-renderUI({
      #   
      #   if("count" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: count from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的count表格
      # output$table.integrated_count_Validation<-DT::renderDataTable({
      #   
      #   if("count" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"count"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"count")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "count_from_integrated_dataset(Validation)",
      #                            title = "Table: count from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "count_from_integrated_dataset(Validation)",
      #                            title = "Table: count from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "count_from_integrated_dataset(Validation)",
      #                            title = "Table: count from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "count_from_integrated_dataset(Validation)",
      #                            title = "Table: count from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的count表格脚注
      # output$table.integrated_count_Validation_Note<-renderUI({
      #   
      #   if("count" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的mRNA(fpkm)表格标题
      # output$table.integrated_mRNA_fpkm_Validation_Title<-renderUI({
      #   
      #   if("fpkm_mrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: mRNA (fpkm) from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的mRNA(fpkm)表格
      # output$table.integrated_mRNA_fpkm_Validation<-DT::renderDataTable({
      #   
      #   if("fpkm_mrna" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"fpkm_mrna"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"fpkm_mrna")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "mRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: mRNA (fpkm) from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的mRNA(fpkm)表格脚注
      # output$table.integrated_mRNA_fpkm_Validation_Note<-renderUI({
      #   if("fpkm_mrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##整合好的lncRNA(fpkm)表格标题
      # output$table.integrated_lncRNA_fpkm_Validation_Title<-renderUI({
      #   
      #   if("fpkm_lncrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: lncRNA (fpkm) from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的lncRNA(fpkm)表格
      # output$table.integrated_lncRNA_fpkm_Validation<-DT::renderDataTable({
      #   
      #   if("fpkm_lncrna" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"fpkm_lncrna"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"fpkm_lncrna")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "lncRNA(fpkm)_from_integrated_dataset(Validation)",
      #                            title = "Table: lncRNA (fpkm) from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的lncRNA(fpkm)表格脚注
      # output$table.integrated_lncRNA_fpkm_Validation_Note<-renderUI({
      #   
      #   if("fpkm_lncrna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # ##整合好的segment表格标题
      # output$table.integrated_segment_Validation_Title<-renderUI({
      #   
      #   if("segment" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: segment from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的segment表格
      # output$table.integrated_segment_Validation<-DT::renderDataTable({
      #   
      #   if("segment" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"segment"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"segment")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "segment_from_integrated_dataset(Validation)",
      #                            title = "Table: segment from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "segment_from_integrated_dataset(Validation)",
      #                            title = "Table: segment from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "segment_from_integrated_dataset(Validation)",
      #                            title = "Table: segment from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "segment_from_integrated_dataset(Validation)",
      #                            title = "Table: segment from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   ),rownames = FALSE
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的segment表格脚注
      # output$table.integrated_segment_Validation_Note<-renderUI({
      #   
      #   if("segment" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # ##整合好的maf表格标题
      # output$table.integrated_maf_Validation_Title<-renderUI({
      #   
      #   if("maf" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: maf from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的maf表格
      # output$table.integrated_maf_Validation<-DT::renderDataTable({
      #   
      #   if("maf" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"maf"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"maf")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "maf_from_integrated_dataset(Validation)",
      #                            title = "Table: maf from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "maf_from_integrated_dataset(Validation)",
      #                            title = "Table: maf from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "maf_from_integrated_dataset(Validation)",
      #                            title = "Table: maf from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "maf_from_integrated_dataset(Validation)",
      #                            title = "Table: maf from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   ),rownames = FALSE
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的maf表格脚注
      # output$table.integrated_maf_Validation_Note<-renderUI({
      #   
      #   if("maf" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # ##整合好的clinSurv表格标题
      # output$table.integrated_clinSurv_Validation_Title<-renderUI({
      #   
      #   if("clin.info" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Table: clinical and survival information from integrated dataset (Validation)" )
      #   }
      # })
      # 
      # ##整合好的clinSurv表格
      # output$table.integrated_clinSurv_Validation<-DT::renderDataTable({
      #   
      #   if("clin.info" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"clin.info"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"clin.info")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = "clinSurv_from_integrated_dataset(Validation)",
      #                            title = "Table: clinical and survival information from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='csv',
      #                            filename = "clinSurv_from_integrated_dataset(Validation)",
      #                            title = "Table: clinical and survival information from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='excel',
      #                            filename = "clinSurv_from_integrated_dataset(Validation)",
      #                            title = "Table: clinical and survival information from integrated dataset (Validation)"
      #                       ),
      #                       list(extend='print',
      #                            filename = "clinSurv_from_integrated_dataset(Validation)",
      #                            title = "Table: clinical and survival information from integrated dataset (Validation)"
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##整合好的clinSurv表格脚注
      # output$table.integrated_clinSurv_Validation_Note<-renderUI({
      #   
      #   if("clin.info" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      
      setwd(workdir) #将工作目录修改回TCGA的数据存储目录
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #53.Get Elites事件响应按钮
    observeEvent(input$Get_Elites_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$tcga_vali_getElites==1,'Get Elites on tcga datasets for specified omics datasets','Get Elites on validation datasets for specified omics datasets'),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Data Preparation步骤整理好的数据
      ##获取项目名称及相关文件路径
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      
      if(input$tcga_vali_getElites==1){
        #TCGA
        ##导入数据
        integrated_data <- get(load(paste0(tolower(project),".tcga.RData")))
        
        #1.Get Elites for mRNA dataset
        if((1 %in% input$multiOmics) && (input$mRNA_getElites==1)){
          
          #生存信息
          if(input$mRNA_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$mRNA_getElites_survival==1,input$mRNA_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$mRNA_getElites_na==1){
            na.action = "rm"
          }else if(input$mRNA_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$mRNA_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$mRNA_getElites_lowpct==1){
            lowpct <- input$mRNA_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$mRNA_getElites_survival==2){
            if(input$mRNA_getElites_method==1){
              method <- "mad"
              if(input$mRNA_getElites_filtration1==1){
                elite.num <- input$mRNA_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$mRNA_getElites_elite_pct1
              }
            }else if(input$mRNA_getElites_method==2){
              method <- "sd"
              if(input$mRNA_getElites_filtration2==1){
                elite.num <- input$mRNA_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$mRNA_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$mRNA_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$mRNA_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$mRNA_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          mRNA.expr <- getElites(dat = integrated_data$"mRNA.expr",
                                 surv.info = surv.info,
                                 method = method,
                                 na.action = na.action,
                                 doLog2 = doLog2,
                                 lowpct = lowpct,
                                 p.cutoff = p.cutoff,
                                 elite.num = elite.num,
                                 elite.pct = elite.pct,
                                 pca.ratio  = pca.ratio,
                                 scaleFlag = scaleFlag,
                                 centerFlag = centerFlag)
          
          # deal with missing value
          mRNA.expr$elite.dat <- as.data.frame(na.omit(mRNA.expr$elite.dat))
          
          integrated_data$"mRNA.expr" <- mRNA.expr$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的mRNA数据
          write.table(mRNA.expr$elite.dat,paste0("TCGA_",project,"_mRNA_TPM_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        #2.Get Elites for lncRNA dataset
        if((2 %in% input$multiOmics) && (input$lncRNA_getElites==1)){
          
          #生存信息
          if(input$lncRNA_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$lncRNA_getElites_survival==1,input$lncRNA_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$lncRNA_getElites_na==1){
            na.action = "rm"
          }else if(input$lncRNA_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$lncRNA_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$lncRNA_getElites_lowpct==1){
            lowpct <- input$lncRNA_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$lncRNA_getElites_survival==2){
            if(input$lncRNA_getElites_method==1){
              method <- "mad"
              if(input$lncRNA_getElites_filtration1==1){
                elite.num <- input$lncRNA_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$lncRNA_getElites_elite_pct1
              }
            }else if(input$lncRNA_getElites_method==2){
              method <- "sd"
              if(input$lncRNA_getElites_filtration2==1){
                elite.num <- input$lncRNA_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$lncRNA_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$lncRNA_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$lncRNA_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$lncRNA_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          lncRNA.expr <- getElites(dat = integrated_data$"lncRNA.expr",
                                   surv.info = surv.info,
                                   method = method,
                                   na.action = na.action,
                                   doLog2 = doLog2,
                                   lowpct = lowpct,
                                   p.cutoff = p.cutoff,
                                   elite.num = elite.num,
                                   elite.pct = elite.pct,
                                   pca.ratio  = pca.ratio,
                                   scaleFlag = scaleFlag,
                                   centerFlag = centerFlag)
          
          # deal with missing value
          lncRNA.expr$elite.dat <- as.data.frame(na.omit(lncRNA.expr$elite.dat))
          
          integrated_data$"lncRNA.expr" <- lncRNA.expr$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的lncRNA数据
          write.table(lncRNA.expr$elite.dat,paste0("TCGA_",project,"_lncRNA_TPM_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        progress$set(value = 2)
        #3.Get Elites for DNA methylation dataset
        if((3 %in% input$multiOmics) && (input$DNA_methylation_getElites==1)){
          
          #生存信息
          if(input$DNA_methylation_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$DNA_methylation_getElites_survival==1,input$DNA_methylation_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$DNA_methylation_getElites_na==1){
            na.action = "rm"
          }else if(input$DNA_methylation_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$DNA_methylation_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$DNA_methylation_getElites_lowpct==1){
            lowpct <- input$DNA_methylation_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$DNA_methylation_getElites_survival==2){
            if(input$DNA_methylation_getElites_method==1){
              method <- "mad"
              if(input$DNA_methylation_getElites_filtration1==1){
                elite.num <- input$DNA_methylation_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$DNA_methylation_getElites_elite_pct1
              }
            }else if(input$DNA_methylation_getElites_method==2){
              method <- "sd"
              if(input$DNA_methylation_getElites_filtration2==1){
                elite.num <- input$DNA_methylation_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$DNA_methylation_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$DNA_methylation_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$DNA_methylation_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$DNA_methylation_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          meth.beta <- getElites(dat = integrated_data$"meth.beta",
                                 surv.info = surv.info,
                                 method = method,
                                 na.action = na.action,
                                 doLog2 = doLog2,
                                 lowpct = lowpct,
                                 p.cutoff = p.cutoff,
                                 elite.num = elite.num,
                                 elite.pct = elite.pct,
                                 pca.ratio  = pca.ratio,
                                 scaleFlag = scaleFlag,
                                 centerFlag = centerFlag)
          
          # deal with missing value
          meth.beta$elite.dat <- as.data.frame(na.omit(meth.beta$elite.dat))
          
          integrated_data$"meth.beta" <- meth.beta$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的DNA methylation数据
          write.table(meth.beta$elite.dat,paste0("TCGA_",project,"_Methpromoter_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        #4.Get Elites for copy number alterations dataset
        if((4 %in% input$multiOmics) && (input$copy_number_alterations_getElites==1)){
          
          #生存信息
          if(input$copy_number_alterations_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$copy_number_alterations_getElites_survival==1,input$copy_number_alterations_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$copy_number_alterations_getElites_na==1){
            na.action = "rm"
          }else if(input$copy_number_alterations_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$copy_number_alterations_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$copy_number_alterations_getElites_lowpct==1){
            lowpct <- input$copy_number_alterations_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$copy_number_alterations_getElites_survival==2){
            if(input$copy_number_alterations_getElites_method==1){
              method <- "mad"
              if(input$copy_number_alterations_getElites_filtration1==1){
                elite.num <- input$copy_number_alterations_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$copy_number_alterations_getElites_elite_pct1
              }
            }else if(input$copy_number_alterations_getElites_method==2){
              method <- "sd"
              if(input$copy_number_alterations_getElites_filtration2==1){
                elite.num <- input$copy_number_alterations_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$copy_number_alterations_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$copy_number_alterations_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$copy_number_alterations_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$copy_number_alterations_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          cna <- getElites(dat = integrated_data$"cna",
                           surv.info = surv.info,
                           method = method,
                           na.action = na.action,
                           doLog2 = doLog2,
                           lowpct = lowpct,
                           p.cutoff = p.cutoff,
                           elite.num = elite.num,
                           elite.pct = elite.pct,
                           pca.ratio  = pca.ratio,
                           scaleFlag = scaleFlag,
                           centerFlag = centerFlag)
          
          # deal with missing value
          cna$elite.dat <- as.data.frame(na.omit(cna$elite.dat))
          
          integrated_data$"cna" <- cna$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的copy number alterations数据
          write.table(cna$elite.dat,paste0("TCGA_",project,"_cna_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        progress$set(value = 3)
        #5.Get Elites for binary somatic mutation dataset
        if((5 %in% input$multiOmics) && (input$binary_somatic_mutation_getElites==1)){
          
          #生存信息
          if(input$binary_somatic_mutation_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$binary_somatic_mutation_getElites_survival==1,input$binary_somatic_mutation_getElites_p_cutoff,0.05)
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$binary_somatic_mutation_getElites_na==1){
            na.action = "rm"
          }else if(input$binary_somatic_mutation_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$binary_somatic_mutation_getElites_log2==1,TRUE,FALSE)
          
          #Get Elites Method
          if(input$binary_somatic_mutation_getElites_survival==2){
            method <- "freq"
            if(input$binary_somatic_mutation_getElites_filtration==1){
              elite.num <- input$binary_somatic_mutation_getElites_elite_num
              elite.pct <- NULL
            }else{
              elite.num <- NULL
              elite.pct <- input$binary_somatic_mutation_getElites_elite_pct
            }
          }else{
            elite.num <- NULL
            elite.pct <- NULL
          }
          
          #scale & center
          scaleFlag <- ifelse(input$binary_somatic_mutation_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$binary_somatic_mutation_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          mut.status <- getElites(dat = integrated_data$"mut.status",
                                  surv.info = surv.info,
                                  method = method,
                                  na.action = na.action,
                                  doLog2 = doLog2,
                                  lowpct = lowpct,
                                  p.cutoff = p.cutoff,
                                  elite.num = elite.num,
                                  elite.pct = elite.pct,
                                  pca.ratio  = pca.ratio,
                                  scaleFlag = scaleFlag,
                                  centerFlag = centerFlag)
          
          # deal with missing value
          mut.status$elite.dat <- as.data.frame(na.omit(mut.status$elite.dat))
          
          integrated_data$"mut.status" <- mut.status$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的binary somatic mutation数据
          write.table(mut.status$elite.dat,paste0("TCGA_",project,"_mut2_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        #6.Get Elites for radiomics dataset
        if((6 %in% input$multiOmics) && (input$radiomics_getElites==1)){
          
          #生存信息
          if(input$radiomics_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$radiomics_getElites_survival==1,input$radiomics_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$radiomics_getElites_na==1){
            na.action = "rm"
          }else if(input$radiomics_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$radiomics_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$radiomics_getElites_lowpct==1){
            lowpct <- input$radiomics_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$radiomics_getElites_survival==2){
            if(input$radiomics_getElites_method==1){
              method <- "mad"
              if(input$radiomics_getElites_filtration1==1){
                elite.num <- input$radiomics_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$radiomics_getElites_elite_pct1
              }
            }else if(input$radiomics_getElites_method==2){
              method <- "sd"
              if(input$radiomics_getElites_filtration2==1){
                elite.num <- input$radiomics_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$radiomics_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$radiomics_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$radiomics_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$radiomics_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          radiomics <- getElites(dat = integrated_data$"radiomics",
                                 surv.info = surv.info,
                                 method = method,
                                 na.action = na.action,
                                 doLog2 = doLog2,
                                 lowpct = lowpct,
                                 p.cutoff = p.cutoff,
                                 elite.num = elite.num,
                                 elite.pct = elite.pct,
                                 pca.ratio  = pca.ratio,
                                 scaleFlag = scaleFlag,
                                 centerFlag = centerFlag)
          
          # deal with missing value
          radiomics$elite.dat <- as.data.frame(na.omit(radiomics$elite.dat))
          
          integrated_data$"radiomics" <- radiomics$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的radiomics数据
          write.table(radiomics$elite.dat,paste0("TCGA_",project,"_radiomics_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        progress$set(value = 4)
        #7.保存Get Elites后的整合数据(.RData)
        ##RData
        workdir <- getwd() #获取当前工作路径
        user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
        RData.path <- file.path(user_workdir,"RData")
        if(!file.exists(RData.path)) {dir.create(RData.path,recursive = TRUE)}
        
        save(integrated_data, file = file.path(RData.path,paste0(tolower(project),".tcga.getElites.RData")))
        #rm(list=ls()) #清空变量
        gc() #释放内存
          
      }else{
        #Validation
        validation.path <- file.path(user_workdir,"Validation_data_preparing")
        setwd(validation.path)
        validation_workdir <- getwd()
        
        ##导入数据
        integrated_data <- get(load(paste0(tolower(project),".validation.RData")))
        
        #1.Get Elites for mRNA dataset
        if((1 %in% input$validation_multiOmics) && (input$validation_mRNA_getElites==1)){
          
          #生存信息
          if(input$validation_mRNA_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$validation_mRNA_getElites_survival==1,input$validation_mRNA_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$validation_mRNA_getElites_na==1){
            na.action = "rm"
          }else if(input$validation_mRNA_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$validation_mRNA_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$validation_mRNA_getElites_lowpct==1){
            lowpct <- input$validation_mRNA_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$validation_mRNA_getElites_survival==2){
            if(input$validation_mRNA_getElites_method==1){
              method <- "mad"
              if(input$validation_mRNA_getElites_filtration1==1){
                elite.num <- input$validation_mRNA_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_mRNA_getElites_elite_pct1
              }
            }else if(input$validation_mRNA_getElites_method==2){
              method <- "sd"
              if(input$validation_mRNA_getElites_filtration2==1){
                elite.num <- input$validation_mRNA_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_mRNA_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$validation_mRNA_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$validation_mRNA_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$validation_mRNA_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          mRNA.expr <- getElites(dat = integrated_data$"mRNA.expr",
                                 surv.info = surv.info,
                                 method = method,
                                 na.action = na.action,
                                 doLog2 = doLog2,
                                 lowpct = lowpct,
                                 p.cutoff = p.cutoff,
                                 elite.num = elite.num,
                                 elite.pct = elite.pct,
                                 pca.ratio  = pca.ratio,
                                 scaleFlag = scaleFlag,
                                 centerFlag = centerFlag)
          
          # deal with missing value
          mRNA.expr$elite.dat <- as.data.frame(na.omit(mRNA.expr$elite.dat))
          
          integrated_data$"mRNA.expr" <- mRNA.expr$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的mRNA数据
          write.table(mRNA.expr$elite.dat,paste0("Validation_",project,"_mRNA_TPM_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        #2.Get Elites for lncRNA dataset
        if((2 %in% input$validation_multiOmics) && (input$validation_lncRNA_getElites==1)){
          
          #生存信息
          if(input$validation_lncRNA_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$validation_lncRNA_getElites_survival==1,input$validation_lncRNA_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$validation_lncRNA_getElites_na==1){
            na.action = "rm"
          }else if(input$validation_lncRNA_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$validation_lncRNA_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$validation_lncRNA_getElites_lowpct==1){
            lowpct <- input$validation_lncRNA_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$validation_lncRNA_getElites_survival==2){
            if(input$validation_lncRNA_getElites_method==1){
              method <- "mad"
              if(input$validation_lncRNA_getElites_filtration1==1){
                elite.num <- input$validation_lncRNA_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_lncRNA_getElites_elite_pct1
              }
            }else if(input$validation_lncRNA_getElites_method==2){
              method <- "sd"
              if(input$validation_lncRNA_getElites_filtration2==1){
                elite.num <- input$validation_lncRNA_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_lncRNA_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$validation_lncRNA_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$validation_lncRNA_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$validation_lncRNA_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          lncRNA.expr <- getElites(dat = integrated_data$"lncRNA.expr",
                                   surv.info = surv.info,
                                   method = method,
                                   na.action = na.action,
                                   doLog2 = doLog2,
                                   lowpct = lowpct,
                                   p.cutoff = p.cutoff,
                                   elite.num = elite.num,
                                   elite.pct = elite.pct,
                                   pca.ratio  = pca.ratio,
                                   scaleFlag = scaleFlag,
                                   centerFlag = centerFlag)
          
          # deal with missing value
          lncRNA.expr$elite.dat <- as.data.frame(na.omit(lncRNA.expr$elite.dat))
          
          integrated_data$"lncRNA.expr" <- lncRNA.expr$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的lncRNA数据
          write.table(lncRNA.expr$elite.dat,paste0("Validation_",project,"_lncRNA_TPM_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        progress$set(value = 2)
        #3.Get Elites for DNA methylation dataset
        if((3 %in% input$validation_multiOmics) && (input$validation_DNA_methylation_getElites==1)){
          
          #生存信息
          if(input$validation_DNA_methylation_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$validation_DNA_methylation_getElites_survival==1,input$validation_DNA_methylation_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$validation_DNA_methylation_getElites_na==1){
            na.action = "rm"
          }else if(input$validation_DNA_methylation_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$validation_DNA_methylation_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$validation_DNA_methylation_getElites_lowpct==1){
            lowpct <- input$validation_DNA_methylation_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$validation_DNA_methylation_getElites_survival==2){
            if(input$validation_DNA_methylation_getElites_method==1){
              method <- "mad"
              if(input$validation_DNA_methylation_getElites_filtration1==1){
                elite.num <- input$validation_DNA_methylation_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_DNA_methylation_getElites_elite_pct1
              }
            }else if(input$validation_DNA_methylation_getElites_method==2){
              method <- "sd"
              if(input$validation_DNA_methylation_getElites_filtration2==1){
                elite.num <- input$validation_DNA_methylation_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_DNA_methylation_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$validation_DNA_methylation_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$validation_DNA_methylation_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$validation_DNA_methylation_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          meth.beta <- getElites(dat = integrated_data$"meth.beta",
                                 surv.info = surv.info,
                                 method = method,
                                 na.action = na.action,
                                 doLog2 = doLog2,
                                 lowpct = lowpct,
                                 p.cutoff = p.cutoff,
                                 elite.num = elite.num,
                                 elite.pct = elite.pct,
                                 pca.ratio  = pca.ratio,
                                 scaleFlag = scaleFlag,
                                 centerFlag = centerFlag)
          
          # deal with missing value
          meth.beta$elite.dat <- as.data.frame(na.omit(meth.beta$elite.dat))
          
          integrated_data$"meth.beta" <- meth.beta$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的DNA methylation数据
          write.table(meth.beta$elite.dat,paste0("Validation_",project,"_Methpromoter_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        #4.Get Elites for copy number alterations dataset
        if((4 %in% input$validation_multiOmics) && (input$validation_copy_number_alterations_getElites==1)){
          
          #生存信息
          if(input$validation_copy_number_alterations_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$validation_copy_number_alterations_getElites_survival==1,input$validation_copy_number_alterations_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$validation_copy_number_alterations_getElites_na==1){
            na.action = "rm"
          }else if(input$validation_copy_number_alterations_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$validation_copy_number_alterations_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$validation_copy_number_alterations_getElites_lowpct==1){
            lowpct <- input$validation_copy_number_alterations_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$validation_copy_number_alterations_getElites_survival==2){
            if(input$validation_copy_number_alterations_getElites_method==1){
              method <- "mad"
              if(input$validation_copy_number_alterations_getElites_filtration1==1){
                elite.num <- input$validation_copy_number_alterations_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_copy_number_alterations_getElites_elite_pct1
              }
            }else if(input$validation_copy_number_alterations_getElites_method==2){
              method <- "sd"
              if(input$validation_copy_number_alterations_getElites_filtration2==1){
                elite.num <- input$validation_copy_number_alterations_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_copy_number_alterations_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$validation_copy_number_alterations_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$validation_copy_number_alterations_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$validation_copy_number_alterations_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          cna <- getElites(dat = integrated_data$"cna",
                           surv.info = surv.info,
                           method = method,
                           na.action = na.action,
                           doLog2 = doLog2,
                           lowpct = lowpct,
                           p.cutoff = p.cutoff,
                           elite.num = elite.num,
                           elite.pct = elite.pct,
                           pca.ratio  = pca.ratio,
                           scaleFlag = scaleFlag,
                           centerFlag = centerFlag)
          
          # deal with missing value
          cna$elite.dat <- as.data.frame(na.omit(cna$elite.dat))
          
          integrated_data$"cna" <- cna$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的copy number alterations数据
          write.table(cna$elite.dat,paste0("Validation_",project,"_cna_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        progress$set(value = 3)
        #5.Get Elites for binary somatic mutation dataset
        if((5 %in% input$validation_multiOmics) && (input$validation_binary_somatic_mutation_getElites==1)){
          
          #生存信息
          if(input$validation_binary_somatic_mutation_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$validation_binary_somatic_mutation_getElites_survival==1,input$validation_binary_somatic_mutation_getElites_p_cutoff,0.05)
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$validation_binary_somatic_mutation_getElites_na==1){
            na.action = "rm"
          }else if(input$validation_binary_somatic_mutation_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$validation_binary_somatic_mutation_getElites_log2==1,TRUE,FALSE)
          
          #Get Elites Method
          if(input$validation_binary_somatic_mutation_getElites_survival==2){
            method <- "freq"
            if(input$validation_binary_somatic_mutation_getElites_filtration==1){
              elite.num <- input$validation_binary_somatic_mutation_getElites_elite_num
              elite.pct <- NULL
            }else{
              elite.num <- NULL
              elite.pct <- input$validation_binary_somatic_mutation_getElites_elite_pct
            }
          }else{
            elite.num <- NULL
            elite.pct <- NULL
          }
          
          #scale & center
          scaleFlag <- ifelse(input$validation_binary_somatic_mutation_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$validation_binary_somatic_mutation_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          mut.status <- getElites(dat = integrated_data$"mut.status",
                                  surv.info = surv.info,
                                  method = method,
                                  na.action = na.action,
                                  doLog2 = doLog2,
                                  lowpct = lowpct,
                                  p.cutoff = p.cutoff,
                                  elite.num = elite.num,
                                  elite.pct = elite.pct,
                                  pca.ratio  = pca.ratio,
                                  scaleFlag = scaleFlag,
                                  centerFlag = centerFlag)
          
          # deal with missing value
          mut.status$elite.dat <- as.data.frame(na.omit(mut.status$elite.dat))
          
          integrated_data$"mut.status" <- mut.status$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的binary somatic mutation数据
          write.table(mut.status$elite.dat,paste0("Validation_",project,"_mut2_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        #6.Get Elites for radiomics dataset
        if((6 %in% input$validation_multiOmics) && (input$validation_radiomics_getElites==1)){
          
          #生存信息
          if(input$validation_radiomics_getElites_survival==1){
            surv.info <- integrated_data$clin.info
            if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
            } #数据预处理(列名替换)
            if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
              colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
            } #数据预处理(列名替换)
          }else{
            surv.info <- NULL
          }
          method <- "cox"
          p.cutoff <- ifelse(input$validation_radiomics_getElites_survival==1,input$validation_radiomics_getElites_p_cutoff,0.05)
          
          #mad/sd
          elite.num <- NULL
          elite.pct <- NULL
          
          #pca
          pca.ratio <- 0.9
          
          #NA值处理
          if(input$validation_radiomics_getElites_na==1){
            na.action = "rm"
          }else if(input$validation_radiomics_getElites_na==2){
            na.action = "impute"
          }else{
            na.action = NULL
          }
          
          #log2操作
          doLog2 <- ifelse(input$validation_radiomics_getElites_log2==1,TRUE,FALSE)
          
          #lowpct处理
          if(input$validation_radiomics_getElites_lowpct==1){
            lowpct <- input$validation_radiomics_getElites_lowpct_value
          }else{
            lowpct <- NULL
          }
          
          #Get Elites Method
          if(input$validation_radiomics_getElites_survival==2){
            if(input$validation_radiomics_getElites_method==1){
              method <- "mad"
              if(input$validation_radiomics_getElites_filtration1==1){
                elite.num <- input$validation_radiomics_getElites_elite_num1
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_radiomics_getElites_elite_pct1
              }
            }else if(input$validation_radiomics_getElites_method==2){
              method <- "sd"
              if(input$validation_radiomics_getElites_filtration2==1){
                elite.num <- input$validation_radiomics_getElites_elite_num2
                elite.pct <- NULL
              }else{
                elite.num <- NULL
                elite.pct <- input$validation_radiomics_getElites_elite_pct2
              }
            }else{
              method <- "pca"
              pca.ratio <- input$validation_radiomics_getElites_pca_ratio
            }
          }
          
          #scale & center
          scaleFlag <- ifelse(input$validation_radiomics_getElites_scaleFlag==1,TRUE,FALSE)
          centerFlag <- ifelse(input$validation_radiomics_getElites_centerFlag==1,TRUE,FALSE)
          
          # Get Elites
          radiomics <- getElites(dat = integrated_data$"radiomics",
                                 surv.info = surv.info,
                                 method = method,
                                 na.action = na.action,
                                 doLog2 = doLog2,
                                 lowpct = lowpct,
                                 p.cutoff = p.cutoff,
                                 elite.num = elite.num,
                                 elite.pct = elite.pct,
                                 pca.ratio  = pca.ratio,
                                 scaleFlag = scaleFlag,
                                 centerFlag = centerFlag)
          
          # deal with missing value
          radiomics$elite.dat <- as.data.frame(na.omit(radiomics$elite.dat))
          
          integrated_data$"radiomics" <- radiomics$elite.dat #重新整合回原来的数据集
          
          #保存Get Elites后的radiomics数据
          write.table(radiomics$elite.dat,paste0("Validation_",project,"_radiomics_comsam_getElites.txt"),quote = F,row.names = T,col.names = NA,sep = "\t")
          
          #rm(list=ls()) #清空变量
          gc() #释放内存
        }
        
        progress$set(value = 4)
        #7.保存Get Elites后的整合数据(.RData)
        ##RData
        user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
        RData.path <- file.path(user_workdir,"RData")
        if(!file.exists(RData.path)) {dir.create(RData.path,recursive = TRUE)}
        
        save(integrated_data, file = file.path(RData.path,paste0(tolower(project),".validation.getElites.RData")))
        setwd(workdir) #将工作目录修改回TCGA的数据存储目录
        #rm(list=ls()) #清空变量
        gc() #释放内存
      }
     
      ##RData结果下载按钮响应事件
      output$getElites_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_getElites==1,paste0(tolower(project),".tcga.getElites.RData"),paste0(tolower(project),".validation.getElites.RData")),
        content = function(theFile) {
          save(integrated_data,file=theFile)
        }
      )
      
      #Get Elites后给予用户的反馈提示
      output$getElites_finish<-renderUI({
        
        if(input$tcga_vali_getElites==1){
          list(br(),column(12,h3(strong("All omics datasets you specified have been processed 'Get Elites' on tcga datasets to extract important features, and you can download and check the obtained files you want below. Now let's start to get the clustering number using elites from tcga datasets.")),style="color:darkblack"),br(),
               column(12,downloadButton("getElites_download",strong("Download RData file of the integrated data after 'Get Elites'"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("All omics datasets you specified have been processed 'Get Elites' on validation datasets to extract important features, and you can download and check the obtained files you want below. Now let's start to get the clustering number using elites from tcga datasets.")),style="color:darkblack"),br(),
               column(12,downloadButton("getElites_download",strong("Download RData file of the integrated data after 'Get Elites'"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      # ##getElites后的mRNA(tpm)表格标题
      # output$table.getElites_mRNA_tpm_Title<-renderUI({
      #   
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     if(input$tcga_vali_getElites==1){
      #       tags$strong("Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)" )
      #     }else{
      #       tags$strong("Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)" )
      #     }
      #   }
      # })
      # 
      # ##getElites后的mRNA(tpm)表格
      # output$table.getElites_mRNA_tpm<-DT::renderDataTable({
      #   
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"mRNA.expr"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"mRNA.expr")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='csv',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='excel',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='print',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","mRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: mRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##getElites后的mRNA(tpm)表格脚注
      # output$table.getElites_mRNA_tpm_Note<-renderUI({
      #   if("mRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##getElites后lncRNA(tpm)表格标题
      # output$table.getElites_lncRNA_tpm_Title<-renderUI({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     if(input$tcga_vali_getElites==1){
      #       tags$strong("Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)" )
      #     }else{
      #       tags$strong("Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)" )
      #     }
      #   }
      # })
      # 
      # ##getElites后的lncRNA(tpm)表格
      # output$table.getElites_lncRNA_tpm<-DT::renderDataTable({
      #   
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"lncRNA.expr"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"lncRNA.expr")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='csv',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='excel',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='print',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","lncRNA(tpm)_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (TCGA)","Table: lncRNA (tpm) from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##getElites后的lncRNA(tpm)表格脚注
      # output$table.getElites_lncRNA_tpm_Note<-renderUI({
      #   if("lncRNA.expr" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##getElites后DNA methylation表格标题
      # output$table.getElites_DNA_methylation_Title<-renderUI({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     if(input$tcga_vali_getElites==1){
      #       tags$strong("Table: DNA methylation from integrated dataset after 'Get Elites' procedure (TCGA)" )
      #     }else{
      #       tags$strong("Table: DNA methylation from integrated dataset after 'Get Elites' procedure (Validation)" )
      #     }
      #   }
      # })
      # 
      # ##getElites后的DNA methylation表格
      # output$table.getElites_DNA_methylation<-DT::renderDataTable({
      #   
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"meth.beta"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"meth.beta")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: DNA methylation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: DNA methylation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='csv',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: DNA methylation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: DNA methylation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='excel',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: DNA methylation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: DNA methylation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='print',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","DNA_methylation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: DNA methylation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: DNA methylation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##getElites后的DNA methylation表格脚注
      # output$table.getElites_DNA_methylation_Note<-renderUI({
      #   if("meth.beta" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##getElites后copy number alterations表格标题
      # output$table.getElites_copy_number_alterations_Title<-renderUI({
      #   
      #   if("cna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     if(input$tcga_vali_getElites==1){
      #       tags$strong("Table: copy number alterations from integrated dataset after 'Get Elites' procedure (TCGA)" )
      #     }else{
      #       tags$strong("Table: copy number alterations from integrated dataset after 'Get Elites' procedure (Validation)" )
      #     }
      #   }
      # })
      # 
      # ##getElites后的copy number alterations表格
      # output$table.getElites_copy_number_alterations<-DT::renderDataTable({
      #   
      #   if("cna" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"cna"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"cna")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: copy number alterations from integrated dataset after 'Get Elites' procedure (TCGA)","Table: copy number alterations from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='csv',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: copy number alterations from integrated dataset after 'Get Elites' procedure (TCGA)","Table: copy number alterations from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='excel',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: copy number alterations from integrated dataset after 'Get Elites' procedure (TCGA)","Table: copy number alterations from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='print',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","copy_number_alterations_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: copy number alterations from integrated dataset after 'Get Elites' procedure (TCGA)","Table: copy number alterations from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##getElites后的copy number alterations表格脚注
      # output$table.getElites_copy_number_alterations_Note<-renderUI({
      #   if("cna" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##getElites后binary somatic mutation表格标题
      # output$table.getElites_binary_somatic_mutation_Title<-renderUI({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     if(input$tcga_vali_getElites==1){
      #       tags$strong("Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (TCGA)" )
      #     }else{
      #       tags$strong("Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (Validation)" )
      #     }
      #   }
      # })
      # 
      # ##getElites后的binary somatic mutation表格
      # output$table.getElites_binary_somatic_mutation<-DT::renderDataTable({
      #   
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"mut.status"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"mut.status")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='csv',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='excel',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='print',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","binary_somatic_mutation_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (TCGA)","Table: binary somatic mutation from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##getElites后的binary somatic mutation表格脚注
      # output$table.getElites_binary_somatic_mutation_Note<-renderUI({
      #   if("mut.status" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      # 
      # 
      # ##getElites后radiomics表格标题
      # output$table.getElites_radiomics_Title<-renderUI({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     if(input$tcga_vali_getElites==1){
      #       tags$strong("Table: radiomics from integrated dataset after 'Get Elites' procedure (TCGA)" )
      #     }else{
      #       tags$strong("Table: radiomics from integrated dataset after 'Get Elites' procedure (Validation)" )
      #     }
      #   }
      # })
      # 
      # ##getElites后的radiomics表格
      # output$table.getElites_radiomics<-DT::renderDataTable({
      #   
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate(validInputTS())
      #     DT::datatable(isolate(integrated_data$"radiomics"),
      #                   extensions = 'Buttons', options =list(
      #                     columnDefs = list(list(className = 'dt-center', targets = 1:(dim(integrated_data$"radiomics")[2]-1))),
      #                     colReorder = TRUE,
      #                     scrollX = TRUE,
      #                     rowReorder = TRUE,
      #                     dom = 'Blfrtip',
      #                     # dom = 'Bfrtip',
      #                     # searching= FALSE,
      #                     # bInfo=FALSE, #not to shown 1/N message
      #                     # display=NULL,
      #                     # bPaginate= FALSE, #suppress page
      #                     pageLength =10,
      #                     
      #                     buttons=list(
      #                       list(extend='copy',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"radiomics_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","radiomics_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: radiomics from integrated dataset after 'Get Elites' procedure (TCGA)","Table: radiomics from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='csv',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"radiomics_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","radiomics_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: radiomics from integrated dataset after 'Get Elites' procedure (TCGA)","Table: radiomics from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='excel',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"radiomics_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","radiomics_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: radiomics from integrated dataset after 'Get Elites' procedure (TCGA)","Table: radiomics from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       ),
      #                       list(extend='print',
      #                            filename = ifelse(input$tcga_vali_getElites==1,"radiomics_from_integrated_dataset_after_Get_Elites_procedure(TCGA)","radiomics_from_integrated_dataset_after_Get_Elites_procedure(Validation)"),
      #                            title = ifelse(input$tcga_vali_getElites==1,"Table: radiomics from integrated dataset after 'Get Elites' procedure (TCGA)","Table: radiomics from integrated dataset after 'Get Elites' procedure (Validation)")
      #                       )
      #                     )
      #                   )
      #     )
      #   }
      # }, server=FALSE)
      # 
      # ##getElites后的radiomics表格脚注
      # output$table.getElites_radiomics_Note<-renderUI({
      #   if("radiomics" %in% names(integrated_data)){
      #     # isolate (validInputTS.caption())
      #     tags$strong("Note: ...")
      #   }
      # })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #54.Get Clustering Number事件响应按钮
    observeEvent(input$getClusteringNumber_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Get Clustering Number',
                   detail = 'This may take a while...')
      progress$set(value = 1)

      #读取Get Elites步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getElites.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]

      progress$set(value = 2)
      #判断是否包含binary数据
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          is.binary <- c(rep(FALSE, length(data)-2),TRUE,FALSE)
        }else{
          is.binary <- rep(FALSE, length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          is.binary <- c(rep(FALSE, length(data)-1),TRUE)
        }else{
          is.binary <- rep(FALSE, length(data))
        }
      }
      
      #获取聚类数范围
      getClusteringNumber_min <- input$getClusteringNumber_min
      getClusteringNumber_max <- input$getClusteringNumber_max
      try.N.clust <- getClusteringNumber_min:getClusteringNumber_max
      
      #Center&Scale
      if(input$getClusteringNumber_center==1){
        center = TRUE
      }else{
        center = FALSE
      }
      if(input$getClusteringNumber_scale==1){
        scale = TRUE
      }else{
        scale = FALSE
      }
      
      #Result Save
      ##Figures
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      if(!file.exists(fig.path)) {dir.create(fig.path,recursive = TRUE)}
      fig.name <- input$getClusteringNumber_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "optimal_number_cluster"
      }
      
      progress$set(value = 3)
      #getClustNum
      # optClustNum <- getClustNum(data = data,
      #                            is.binary = is.binary,
      #                            try.N.clust = try.N.clust,
      #                            center = center,
      #                            scale = scale,
      #                            fig.path = fig.path,
      #                            fig.name = fig.name)
      
      optClustNum <- getClustNums(data = data,
                                  is.binary = is.binary,
                                  try.N.clust = try.N.clust,
                                  center = center,
                                  scale = scale,
                                  fig.path = fig.path,
                                  fig.name = fig.name)
      
      #保存getClustNum后的整合数据(.RData)
      integrated_data$"optClustNum" <- optClustNum
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(integrated_data, file = file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData")))
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      progress$set(value = 4)
      N.clust <- optClustNum$N.clust
      CPI <- optClustNum$CPI
      Gapk <- optClustNum$Gapk
      
      #绘图展示
      output$getClusteringNumber_figure_unit<-renderUI({
        list(column(12,downloadButton("getClusteringNumber_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("getClusteringNumber_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$getClusteringNumber_figure_download<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          pdf(file = theFile,width = 5.5, height = 5)
          par(bty="o", mgp = c(1.9,.33,0), mar=c(3.1,3.1,2.1,3.1)+.1, las=1, tcl=-.25)
          plot(NULL, NULL,
               xlim = c(min(try.N.clust),max(try.N.clust)),
               ylim = c(0,1),
               xlab = "Number of Multi-Omics Clusters",ylab = "")
          rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col = "#EAE9E9",border = FALSE)
          grid(col = "white", lty = 1, lwd = 1.5)
          points(try.N.clust, apply(CPI, 1, mean), pch = 19, col = ggplot2::alpha("#224A8D"), cex = 1.5)
          lines(try.N.clust, apply(CPI, 1, mean), col = "#224A8D", lwd = 2, lty = 4)
          mtext("Cluster Prediction Index", side = 2, line = 2, cex = 1.5, col = "#224A8D", las = 3)
          
          par(new = TRUE, xpd = FALSE)
          plot(NULL,NULL,
               xlim = c(min(try.N.clust),max(try.N.clust)),
               ylim = c(0,1),
               xlab = "",ylab = "",xaxt = "n",yaxt = "n")
          points(try.N.clust, Gapk$gap, pch = 19, col = ggplot2::alpha("#E51718",0.8), cex = 1.5)
          lines(try.N.clust, Gapk$gap, col = "#E51718", lwd = 2, lty = 4)
          axis(side = 4, at = seq(0,1,0.2), labels = c("0.0","0.2","0.4","0.6","0.8","1.0"))
          mtext("Gap-statistics", side = 4, line = 2,las = 3, cex = 1.5, col = "#E51718")
          dev.off()
        }
      )
      
      ##页面展示事件
      output$getClusteringNumber_figure<-renderPlot({
        
        par(bty="o", mgp = c(1.9,.33,0), mar=c(3.1,3.1,2.1,3.1)+.1, las=1, tcl=-.25)
        plot(NULL, NULL,
             xlim = c(min(try.N.clust),max(try.N.clust)),
             ylim = c(0,1),
             xlab = "Number of Multi-Omics Clusters",ylab = "")
        rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col = "#EAE9E9",border = FALSE)
        grid(col = "white", lty = 1, lwd = 1.5)
        points(try.N.clust, apply(CPI, 1, mean), pch = 19, col = ggplot2::alpha("#224A8D"), cex = 1.5)
        lines(try.N.clust, apply(CPI, 1, mean), col = "#224A8D", lwd = 2, lty = 4)
        mtext("Cluster Prediction Index", side = 2, line = 2, cex = 1.5, col = "#224A8D", las = 3)
        
        par(new = TRUE, xpd = FALSE)
        plot(NULL,NULL,
             xlim = c(min(try.N.clust),max(try.N.clust)),
             ylim = c(0,1),
             xlab = "",ylab = "",xaxt = "n",yaxt = "n")
        points(try.N.clust, Gapk$gap, pch = 19, col = ggplot2::alpha("#E51718",0.8), cex = 1.5)
        lines(try.N.clust, Gapk$gap, col = "#E51718", lwd = 2, lty = 4)
        axis(side = 4, at = seq(0,1,0.2), labels = c("0.0","0.2","0.4","0.6","0.8","1.0"))
        mtext("Gap-statistics", side = 4, line = 2,las = 3, cex = 1.5, col = "#E51718")
      })
      
      ##RData结果下载按钮响应事件
      output$getClustNums_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.getClustNum.RData"),
        content = function(theFile) {
          save(integrated_data,file=theFile)
        }
      )
      
      #getClustNums后给予用户的反馈提示
      output$getClusteringNumber_finish<-renderUI({
        
        if(N.clust > 1){
          list(br(),column(12,h4(paste0("The imputed optimal cluster number is ", N.clust, " arbitrarily, but it would be better referring to other priori knowledge."))),br(),column(12,h3(strong("Now we have determined the optimal number of clustering, you can download and check the figure as well as the '.RData' file. Now let's start clustering based on all omics datasets.")),style="color:darkblack"),br(),
               column(12,downloadButton("getClustNums_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("There is something wrong with the determined optimal cluster number, please check!")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #55.The number of neighbors for each omic(NEMO)填写表格
    output$NEMO_numNeighbors2<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="The number of neighbors"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=160)
    })
    
    #56.The absolute number of non-zero coefficients for the variable loading vectors(MoCluster)填写表格
    output$table.MoCluster_k1<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Absolute number"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #57.The proportion of non-zero coefficients for the variable loading vectors(MoCluster)填写表格
    output$table.MoCluster_k2<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Proportion"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=80)
    })
    
    #58.Heatmap mapping colors填写表格
    output$table.consensus_clustering_mapcolor<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=5,nrow=1))
      colnames(DF)=1:5
      rownames(DF)="Heatmap mapping colors"
      
      default <- c()
      for(d in 1:5){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=160)
    })
    
    #59.Colors for clustering subtypes填写表格
    output$table.consensus_clustering_clustcolor<-renderRHandsontable({
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      ##获取聚类数
      if(input$Consensus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$Consensus_N_clusts #用户指定聚类数
      }
        
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #60.iClusterBayes_process事件响应按钮
    observeEvent(input$iClusterBayes_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using iClusterBayes algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$iClusterBayes_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$iClusterBayes_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #The number of MCMC burnin
      n.burnin <- input$iClusterBayes_burnin
      
      #The number of MCMC draw
      n.draw <- input$iClusterBayes_draw
      
      #The standard deviation of random walk proposal for the latent variable
      sdev <- input$iClusterBayes_sdev
      
      #A numerical value to thin the MCMC chain in order to reduce autocorrelation
      thin <- input$iClusterBayes_thin
      
      #The prior probability for the indicator variable gamma of each subdataset
      prior.gamma <- c()
      ##mRNA
      if(1 %in% input$multiOmics){
        prior.gamma <- c(prior.gamma,input$iClusterBayes_mRNA_prior_gamma)
      }
      ##lncRNA
      if(2 %in% input$multiOmics){
        prior.gamma <- c(prior.gamma,input$iClusterBayes_lncRNA_prior_gamma)
      }
      ##DNA methylation
      if(3 %in% input$multiOmics){
        prior.gamma <- c(prior.gamma,input$iClusterBayes_DNA_methylation_prior_gamma)
      }
      ##copy number alterations
      if(4 %in% input$multiOmics){
        prior.gamma <- c(prior.gamma,input$iClusterBayes_copy_number_alterations_prior_gamma)
      }
      ##binary somatic mutation
      if(5 %in% input$multiOmics){
        prior.gamma <- c(prior.gamma,input$iClusterBayes_binary_somatic_mutation_prior_gamma)
      }
      ##radiomics
      if(6 %in% input$multiOmics){
        prior.gamma <- c(prior.gamma,input$iClusterBayes_radiomics_prior_gamma)
      }
      
      progress$set(value = 3)
      iClusterBayes <- getiClusterBayes(data = data,
                                        N.clust = N.clust,
                                        type = type,
                                        n.burnin = n.burnin,
                                        n.draw = n.draw,
                                        prior.gamma = prior.gamma,
                                        sdev = sdev,
                                        thin = thin)
      
      progress$set(value = 4)
      #保存iClusterBayes结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(iClusterBayes, file = file.path(RData.path,paste0(tolower(project),".tcga.iClusterBayes.RData")))
      
      ##RData结果下载按钮响应事件
      output$getiClusterBayes_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.iClusterBayes.RData"),
        content = function(theFile) {
          save(iClusterBayes,file=theFile)
        }
      )
      
      #iClusterBayes后给予用户的反馈提示
      output$iClusterBayes_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'iClusterBayes' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getiClusterBayes_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #61.SNF_process事件响应按钮
    observeEvent(input$SNF_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using SNF algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$SNF_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$SNF_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #The number of neighbors in K-nearest neighbors part of the algorithm
      K <- input$SNF_K
      
      #The number of interations for the diffusion process
      t <- input$SNF_t
      
      #The variance for local model
      sigma <- input$SNF_sigma
      
      progress$set(value = 3)
      SNF <- getSNF(data = data,
                    N.clust = N.clust,
                    type = type,
                    K = K,
                    t = t,
                    sigma = sigma)
      
      progress$set(value = 4)
      #保存SNF结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(SNF, file = file.path(RData.path,paste0(tolower(project),".tcga.SNF.RData")))
      
      ##RData结果下载按钮响应事件
      output$getSNF_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.SNF.RData"),
        content = function(theFile) {
          save(SNF,file=theFile)
        }
      )
      
      #SNF后给予用户的反馈提示
      output$SNF_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'SNF' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getSNF_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #62.PINSPlus_process事件响应按钮
    observeEvent(input$PINSPlus_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using PINSPlus algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$PINSPlus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$PINSPlus_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #The normalization method for consensus clustering
      if(input$PINSPlus_norMethod==1){
        norMethod <- "median-centered"
      }else if(input$PINSPlus_norMethod==2){
        norMethod <- "mean-centered"
      }else if(input$PINSPlus_norMethod==3){
        norMethod <- "z-score"
      }else{
        norMethod <- "none"
      }
      
      #Built-in clustering algorithm that PerturbationClustering will use
      if(input$PINSPlus_clusteringMethod==1){
        clusteringMethod <- "kmeans"
      }else if(input$PINSPlus_clusteringMethod==2){
        clusteringMethod <- "pam"
      }else{
        clusteringMethod <- "hclust"
      }
      
      #The minimum number of iterations
      iterMin <- input$PINSPlus_iterMin
      
      #The maximum number of iterations
      iterMax <- input$PINSPlus_iterMax
      
      progress$set(value = 3)
      PINSPlus <- getPINSPlus(data = data,
                              N.clust = N.clust,
                              type = type,
                              norMethod = norMethod,
                              clusteringMethod = clusteringMethod,
                              iterMin = iterMin,
                              iterMax = iterMax)
      
      progress$set(value = 4)
      #保存PINSPlus结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(PINSPlus, file = file.path(RData.path,paste0(tolower(project),".tcga.PINSPlus.RData")))
      
      ##RData结果下载按钮响应事件
      output$getPINSPlus_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.PINSPlus.RData"),
        content = function(theFile) {
          save(PINSPlus,file=theFile)
        }
      )
      
      #PINSPlus后给予用户的反馈提示
      output$PINSPlus_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'PINSPlus' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getPINSPlus_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #63.NEMO_process事件响应按钮
    observeEvent(input$NEMO_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using NEMO algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$NEMO_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$NEMO_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #The number of neighbors to use for each omic
      if(input$NEMO_numNeighbors==1){
        num.neighbors <- input$EMO_numNeighbors1
      }else if(input$NEMO_numNeighbors==2){
        num.neighbors <- data.frame(rbind(hot_to_r(input$NEMO_numNeighbors2)))[1,]
        num.neighbors <- trimws(num.neighbors,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        num.neighbors <- as.numeric(num.neighbors) #字符串向量转换成数值型向量
      }else{
        num.neighbors <- NA
      }
      
      progress$set(value = 3)
      NEMO <- getNEMO(data = data,
                      N.clust = N.clust,
                      type = type,
                      num.neighbors = num.neighbors)
      
      progress$set(value = 4)
      #保存NEMO结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(NEMO, file = file.path(RData.path,paste0(tolower(project),".tcga.NEMO.RData")))
      
      ##RData结果下载按钮响应事件
      output$getNEMO_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.NEMO.RData"),
        content = function(theFile) {
          save(NEMO,file=theFile)
        }
      )
      
      #NEMO后给予用户的反馈提示
      output$NEMO_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'NEMO' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getNEMO_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #64.COCA_process事件响应按钮
    observeEvent(input$COCA_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using COCA algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$COCA_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$COCA_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #Clustering methods to be used to cluster the observations in each subdataset
      methods <- input$COCA_methods
      methods <- trimws(methods,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Distances to be used in the clustering step for each subdataset
      distances <- input$COCA_distances
      distances <- trimws(distances,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 3)
      COCA <- getCOCA(data = data,
                      N.clust = N.clust,
                      type = type,
                      methods = methods,
                      distances = distances)
      
      progress$set(value = 4)
      #保存COCA结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(COCA, file = file.path(RData.path,paste0(tolower(project),".tcga.COCA.RData")))
      
      ##RData结果下载按钮响应事件
      output$getCOCA_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.COCA.RData"),
        content = function(theFile) {
          save(COCA,file=theFile)
        }
      )
      
      #COCA后给予用户的反馈提示
      output$COCA_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'COCA' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getCOCA_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #65.LRAcluster_process事件响应按钮
    observeEvent(input$LRAcluster_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using LRAcluster algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$LRAcluster_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$LRAcluster_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #The cluster algorithm for similarity matrix
      clusterAlg <- input$LRAcluster_clusterAlg
      clusterAlg <- trimws(clusterAlg,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 3)
      LRAcluster <- getLRAcluster(data = data,
                                  N.clust = N.clust,
                                  type = type,
                                  clusterAlg = clusterAlg)
      
      progress$set(value = 4)
      #保存LRAcluster结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(LRAcluster, file = file.path(RData.path,paste0(tolower(project),".tcga.LRAcluster.RData")))
      
      ##RData结果下载按钮响应事件
      output$getLRAcluster_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.LRAcluster.RData"),
        content = function(theFile) {
          save(LRAcluster,file=theFile)
        }
      )
      
      #LRAcluster后给予用户的反馈提示
      output$LRAcluster_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'LRAcluster' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getLRAcluster_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #66.ConsensusClustering_process事件响应按钮
    observeEvent(input$ConsensusClustering_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using ConsensusClustering algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$ConsensusClustering_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$ConsensusClustering_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #The normalization method for consensus clustering
      if(input$ConsensusClustering_norMethod==1){
        norMethod <- "median-centered"
      }else if(input$ConsensusClustering_norMethod==2){
        norMethod <- "mean-centered"
      }else if(input$ConsensusClustering_norMethod==3){
        norMethod <- "z-score"
      }else{
        norMethod <- "none"
      }
      
      #The number of subsamples
      reps <- input$ConsensusClustering_reps
      
      #The proportion of items to sample
      pItem <- input$ConsensusClustering_pItem
      
      #The proportion of features to sample
      pFeature <- input$ConsensusClustering_pFeature
      
      #The cluster algorithm
      clusterAlg <- input$ConsensusClustering_clusterAlg
      clusterAlg <- trimws(clusterAlg,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The heirachical linakge method for subsampling
      innerLinkage <- input$ConsensusClustering_innerLinkage
      innerLinkage <- trimws(innerLinkage,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The heirarchical method for consensus matrix
      finalLinkage <- input$ConsensusClustering_finalLinkage
      finalLinkage <- trimws(finalLinkage,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The distance function
      distance <- input$ConsensusClustering_distance
      distance <- trimws(distance,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The output format for heatmap
      if(input$ConsensusClustering_plot==1){
        plot <= "pdf"
      }else if(input$ConsensusClustering_plot==2){
        plot <- "png"
      }else if(input$ConsensusClustering_plot==3){
        plot <- "pngBMP"
      }else{
        plot <- NULL
      }
      
      #Write output and log to csv or not
      if(input$ConsensusClustering_writeTable==1){
        writeTable <- TRUE
      }else{
        writeTable <- FALSE
      }
      
      #The name of output directory
      if((input$ConsensusClustering_plot != 4) | (input$ConsensusClustering_writeTable == 1)){
        title <- input$ConsensusClustering_title
        title <- trimws(title,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        title <- file.path(RData.path,title)
      }else{
        title <- "consensuscluster"
        title <- file.path(RData.path,title)
      }
      
      #A random seed for reproducible results
      seed <- input$ConsensusClustering_seed
      
      #Don't print messages to the screen to indicate progress
      verbose <- FALSE
      
      progress$set(value = 3)
      ConsensusClustering <- getConsensusClustering(data = data,
                                                    N.clust = N.clust,
                                                    type = type,
                                                    norMethod = norMethod,
                                                    reps = reps,
                                                    pItem = pItem,
                                                    pFeature = pFeature,
                                                    clusterAlg = clusterAlg,
                                                    innerLinkage = innerLinkage,
                                                    finalLinkage = finalLinkage,
                                                    distance = distance,
                                                    plot = plot,
                                                    writeTable = writeTable,
                                                    title = title,
                                                    seed = seed,
                                                    verbose = verbose)
      
      progress$set(value = 4)
      #保存ConsensusClustering结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(ConsensusClustering, file = file.path(RData.path,paste0(tolower(project),".tcga.ConsensusClustering.RData")))
      
      ##RData结果下载按钮响应事件
      output$getConsensusClustering_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.ConsensusClustering.RData"),
        content = function(theFile) {
          save(ConsensusClustering,file=theFile)
        }
      )
      
      #ConsensusClustering后给予用户的反馈提示
      output$ConsensusClustering_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'ConsensusClustering' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getConsensusClustering_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #67.IntNMF_process事件响应按钮
    observeEvent(input$IntNMF_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using IntNMF algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$IntNMF_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$IntNMF_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      progress$set(value = 3)
      IntNMF <- getIntNMF(data = data,
                          N.clust = N.clust,
                          type = type)
      
      progress$set(value = 4)
      #保存IntNMF结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(IntNMF, file = file.path(RData.path,paste0(tolower(project),".tcga.IntNMF.RData")))
      
      ##RData结果下载按钮响应事件
      output$getIntNMF_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.IntNMF.RData"),
        content = function(theFile) {
          save(IntNMF,file=theFile)
        }
      )
      
      #IntNMF后给予用户的反馈提示
      output$IntNMF_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'IntNMF' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getIntNMF_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #68.CIMLR_process事件响应按钮
    observeEvent(input$CIMLR_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using CIMLR algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$CIMLR_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$CIMLR_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #Ratio of the number of cores to be used when computing the multi-kernel
      cores.ratio <- input$CIMLR_cores_ratio
      
      #Suppress progression output
      verbose <- TRUE
      
      progress$set(value = 3)
      CIMLR <- getCIMLR(data = data,
                        N.clust = N.clust,
                        type = type,
                        cores.ratio = cores.ratio,
                        verbose = verbose)
      
      progress$set(value = 4)
      #保存CIMLR结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(CIMLR, file = file.path(RData.path,paste0(tolower(project),".tcga.CIMLR.RData")))
      
      ##RData结果下载按钮响应事件
      output$getCIMLR_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.CIMLR.RData"),
        content = function(theFile) {
          save(CIMLR,file=theFile)
        }
      )
      
      #CIMLR后给予用户的反馈提示
      output$CIMLR_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'CIMLR' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getCIMLR_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #69.MoCluster_process事件响应按钮
    observeEvent(input$MoCluster_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using MoCluster algorithm',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$MoCluster_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$MoCluster_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #The number of components to calculate
      ncomp <- input$MoCluster_ncomp
      
      #Method
      if(input$MoCluster_method==1){
        method <- "CPCA"
      }else if(input$MoCluster_method==2){
        method <- "GCCA"
      }else{
        method <- "MCIA"
      }
      
      #Normalization of different matrices
      if(input$MoCluster_option==1){
        option <- "lambda1"
      }else if(input$MoCluster_option==2){
        option <- "inertia"
      }else{
        option <- "uniform"
      }
      
      #A numeric value to indicate the absolute number (if k >= 1) or the proportion (if 0 < k < 1) of non-zero coefficients for the variable loading vectors
      if(input$MoCluster_k_format==1){
        if(input$MoCluster_k_setting1==1){
          k <- input$MoCluster_k1
        }else{
          k <- data.frame(rbind(hot_to_r(input$table.MoCluster_k1)))[1,]
          k <- trimws(k,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          k <- as.numeric(k) #字符串向量转换成数值型向量
        }
      }else{
        if(input$MoCluster_k_setting2==1){
          k <- input$MoCluster_k2
        }else{
          k <- data.frame(rbind(hot_to_r(input$table.MoCluster_k2)))[1,]
          k <- trimws(k,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          k <- as.numeric(k) #字符串向量转换成数值型向量
        }
      }
      
      #The variables should be centered or not
      if(input$MoCluster_center==1){
        center = TRUE
      }else{
        center = FALSE
      }
      
      #The variables should be scaled or not
      if(input$MoCluster_scale==1){
        scale = TRUE
      }else{
        scale = FALSE
      }
      
      #The cluster algorithm for distance
      clusterAlg <- input$MoCluster_clusterAlg
      clusterAlg <- trimws(clusterAlg,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 3)
      MoCluster <- getMoCluster(data = data,
                                N.clust = N.clust,
                                type = type,
                                ncomp = ncomp,
                                method = method,
                                option = option,
                                k = k,
                                center = center,
                                scale = scale,
                                clusterAlg = clusterAlg)
      
      progress$set(value = 4)
      #保存MoCluster结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(MoCluster, file = file.path(RData.path,paste0(tolower(project),".tcga.MoCluster.RData")))
      
      ##RData结果下载按钮响应事件
      output$getMoCluster_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.MoCluster.RData"),
        content = function(theFile) {
          save(MoCluster,file=theFile)
        }
      )
      
      #MoCluster后给予用户的反馈提示
      output$MoCluster_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using 'MoCluster' algorithm has been finished, you can download and check '.RData' file below. Now let's turn to 'Multi-omics Heatmaps' step directly.")),style="color:darkblack"),br(),
             column(12,downloadButton("getMoCluster_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #70.consensus_clustering_process1事件响应按钮
    observeEvent(input$consensus_clustering_process1,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Clustering using multiple specified algorithms',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取Get Clustering Number步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      data <- integrated_data[1:length(input$multiOmics)]
      
      progress$set(value = 2)
      ##获取聚类数
      if(input$Consensus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$Consensus_N_clusts #用户指定聚类数
      }
      
      #设定各类型数据的假设分布
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-2),"binomial","gaussian")
        }else{
          type <- rep("gaussian", length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          type <- c(rep("gaussian", length(data)-1),"binomial")
        }else{
          type <- rep("gaussian", length(data))
        }
      }
      
      #方法列表
      methodslist <- list()
      for(index in 1:length(input$multiClusteringAlgorithms)){
        methodslist[[index]] <- input$multiClusteringAlgorithms[index]
      }
      
      progress$set(value = 3)
      moic.res.list <- getMOIC(data = data,
                               methodslist = methodslist,
                               N.clust = N.clust,
                               type = type)
      
      progress$set(value = 4)
      #保存moic.res.list结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(moic.res.list, file = file.path(RData.path,paste0(tolower(project),".tcga.moic.res.list.RData")))
      
      ##RData结果下载按钮响应事件
      output$getMOIC_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.moic.res.list.RData"),
        content = function(theFile) {
          save(moic.res.list,file=theFile)
        }
      )
      
      #getMOIC后给予用户的反馈提示
      output$MOIC_finish<-renderUI({
        list(br(),column(12,h3(strong("The process of clustering using multiple specified algorithms has been finished, you can download and check '.RData' file below. Now let's keep on the step of 'Consensus Heatmap'.")),style="color:darkblack"),br(),
             column(12,downloadButton("getMOIC_download",strong("Download RData file of the results"),style="background-color:#47AEE9")),br(),br()) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #71.consensus_clustering_process2事件响应按钮
    observeEvent(input$consensus_clustering_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Get consensus from different algorithms and plot consensus heatmap',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取consensus_clustering_process1步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      moic.res.list <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.moic.res.list.RData"))))
      
      progress$set(value = 2)
      #Distance measurement for hierarchical clustering
      distance <- input$consensus_clustering_distance
      distance <- trimws(distance,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Clustering method for hierarchical clustering
      linkage <- input$consensus_clustering_linkage
      linkage <- trimws(linkage,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Heatmap mapping color
      if(input$consensus_clustering_mapcolor==1){
        mapcolor <- c("#000004FF", "#56106EFF", "#BB3754FF", "#F98C0AFF", "#FCFFA4FF")
      }else{
        mapcolor <- data.frame(rbind(hot_to_r(input$table.consensus_clustering_mapcolor)))[1,]
        mapcolor <- trimws(mapcolor,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Colors for annotating each cluster at the top of heatmap
      if(input$consensus_clustering_clustcolor==1){
        clust.col <- c("#2EC4B6", "#E71D36", "#FF9F1C", "#BDD5EA", "#FFA5AB", "#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.consensus_clustering_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Show sample ID or not
      if(input$consensus_clustering_showID==1){
        showID <- TRUE
      }else{
        showID <- FALSE
      }
      
      #The output path for storing the consensus heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the consensus heatmap
      fig.name <- input$consensus_clustering_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "consensusheatmap"
      }
      
      #The width of output figure
      width <- input$consensus_clustering_FigureWidth
      
      #The height of output figure
      height <- input$consensus_clustering_FigureHeight
      
      progress$set(value = 3)
      cmoic <- getConsensusMOIC(moic.res.list = moic.res.list,
                                distance = distance,
                                linkage = linkage,
                                mapcolor = mapcolor,
                                clust.col = clust.col,
                                showID = showID,
                                fig.path = fig.path,
                                fig.name = fig.name,
                                width = width,
                                height = height)
      
      #保存cmoic结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(cmoic, file = file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData")))
      #rm(list=ls()) #清空变量
      gc() #释放内存
      
      progress$set(value = 4)
      #绘图展示
      output$consensus_heatmap_figure_unit<-renderUI({
        list(br(),column(12,downloadButton("consensus_heatmap_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("consensus_heatmap_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$consensus_heatmap_figure_download<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          pdf(file = theFile,width = width, height = height)
          draw(cmoic$consensus.hm)
          dev.off()
        }
      )
      
      ##页面展示事件
      output$consensus_heatmap_figure<-renderPlot({
        
        cmoic$consensus.hm
      })
      
      ##RData结果下载按钮响应事件
      output$cmoic_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.cmoic.RData"),
        content = function(theFile) {
          save(cmoic,file=theFile)
        }
      )
      
      #getConsensusMOIC后给予用户的反馈提示
      output$cmoic_finish<-renderUI({
        
        list(br(),column(12,h3(strong("The process of getting consensus from different algorithms and plotting consensus heatmap as well as the process of silhouette have been finished, you can download and check the figure as well as the '.RData' file. Now let's turn to the step of 'Silhouette'.")),style="color:darkblack"),br(),
             column(12,downloadButton("cmoic_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
   })
  
    #72.Colors for annotating each cluster填写表格
    output$table.silhouette_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      ##获取聚类数
      if(input$Consensus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        N.clust <- input$Consensus_N_clusts #用户指定聚类数
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })  
    
    #73.silhouette_process事件响应按钮
    observeEvent(input$silhouette_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Visualize silhouette information from consensus clustering',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #读取consensus_clustering_process2步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
      sil <- cmoic$sil
      
      progress$set(value = 2)
      #Colors for annotating each cluster
      if(input$silhouette_clustcolor==1){
        clust.col <- c("#2EC4B6", "#E71D36", "#FF9F1C", "#BDD5EA", "#FFA5AB", "#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.silhouette_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #The output path for storing the silhouette plot
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the silhouette plot
      fig.name <- input$silhouette_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "silhouette"
      }
      
      #The width of output figure
      width <- input$silhouette_FigureWidth
      
      #The height of output figure
      height <- input$silhouette_FigureHeight
      
      progress$set(value = 3)
      getSilhouettes(sil = sil,
                     clust.col = clust.col,
                     fig.path = fig.path,
                     fig.name = fig.name,
                     width = width,
                     height = height)
      
      progress$set(value = 4)
      #绘图展示
      output$silhouette_figure_unit<-renderUI({
        list(column(12,downloadButton("silhouette_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("silhouette_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$silhouette_figure_download<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          pdf(file = theFile,width = width, height = height)
          N.clust <- length(unique(sil[,1]))
          colvec <- clust.col[1:N.clust]
          par(bty="o", mgp = c(2.5,0.33,0), mar=c(5.1,2.1,3.1,2.1)+.1, las=1, tcl=-.25)
          plot(sil,
               border = NA,
               col = colvec)
          dev.off()
        }
      )
      
      ##页面展示事件
      output$silhouette_figure<-renderPlot({
        
        N.clust <- length(unique(sil[,1]))
        colvec <- clust.col[1:N.clust]
        par(bty="o", mgp = c(2.5,0.33,0), mar=c(5.1,2.1,3.1,2.1)+.1, las=1, tcl=-.25)
        plot(sil,
             border = NA,
             col = colvec)
      })
      
      #getSilhouettes后给予用户的反馈提示
      output$silhouette_finish<-renderUI({
        
        list(br(),column(12,h3(strong("The process of visualizing silhouette information from consensus clustering has been finished, you can download and check the figure. Now let's turn to the step of 'Multi-omics Heatmaps'.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #74.Indicate if each omic data should be centered填写表格
    output$table.centerFlag_heatmap<-renderRHandsontable({
      
      if(5 %in% input$multiOmics){
        ncol <- length(input$multiOmics)-1
      }else{
        ncol <- length(input$multiOmics)
      }
      DF=data.frame(matrix(NA,ncol=ncol,nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Center or not"
      
      default <- c()
      for(d in 1:ncol){
        default <- c(default,TRUE)
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=160)
    })
    
    #75.Indicate if each omic data should be scaled填写表格
    output$table.scaleFlag_heatmap<-renderRHandsontable({
      
      if(5 %in% input$multiOmics){
        ncol <- length(input$multiOmics)-1
      }else{
        ncol <- length(input$multiOmics)
      }
      DF=data.frame(matrix(NA,ncol=ncol,nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Scale or not"
      
      default <- c()
      for(d in 1:ncol){
        default <- c(default,TRUE)
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=160)
    })
    
    #76.Assign truncating values for extreme values in continuous normalized multi-omics data填写表格
    output$table.halfwidth_heatmap<-renderRHandsontable({
      
      if(5 %in% input$multiOmics){
        ncol <- length(input$multiOmics)-1
      }else{
        ncol <- length(input$multiOmics)
      }
      DF=data.frame(matrix(NA,ncol=ncol,nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Truncating value"
      
      default <- c()
      for(d in 1:ncol){
        default <- c(default,2)
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=140)
    })
    
    #77.Row title for each omic data填写表格
    output$table.row_title_heatmap<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Row title"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #78.Legend title for each omic data填写表格
    output$table.legend_title_heatmap<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Legend title"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #79.Show the feature names for rows of each omic data填写表格
    output$table.show_rownames_heatmap<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Show rownames"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,FALSE)
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #80.Distance method for clustering each omic data at feature dimension填写表格
    output$table.clust_dist_row_heatmap<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Distance method"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,"pearson")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #81.Clustering method for each omic data at feature dimension填写表格
    output$table.clust_method_row_heatmap<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Clustering method"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,"ward.D")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #82.Show dendrogram for rows of each omic data填写表格
    output$table.show_row_dend_heatmap<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=length(input$multiOmics),nrow=1))
      ##将数字转换成具体的数据类型名称
      DataType = c()
      if(1 %in% input$multiOmics){
        DataType <- c(DataType,"mRNA")
      }
      if(2 %in% input$multiOmics){
        DataType <- c(DataType,"lncRNA")
      }
      if(3 %in% input$multiOmics){
        DataType <- c(DataType,"DNA methylation")
      }
      if(4 %in% input$multiOmics){
        DataType <- c(DataType,"copy number alterations")
      }
      if(5 %in% input$multiOmics){
        DataType <- c(DataType,"binary somatic mutation")
      }
      if(6 %in% input$multiOmics){
        DataType <- c(DataType,"radiomics")
      }
      colnames(DF)=DataType
      rownames(DF)="Show row dendrogram"
      
      default <- c()
      for(d in 1:length(input$multiOmics)){
        default <- c(default,TRUE)
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=160)
    })
    
    #83.Colors for annotating each cluster at the top of heatmap填写表格
    output$table.heatmap_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      ##获取聚类数
      if(input$Consensus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        if(length(input$multiClusteringAlgorithms)==1){
          #用户指定聚类数
          N.clust <- switch(input$multiClusteringAlgorithms,
                            'iClusterBayes' = input$iClusterBayes_N_clusts,
                            'SNF' = input$SNF_N_clusts,
                            'PINSPlus' = input$PINSPlus_N_clusts,
                            'NEMO' = input$NEMO_N_clusts,
                            'COCA' = input$COCA_N_clusts,
                            'LRAcluster' = input$LRAcluster_N_clusts,
                            'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                            'IntNMF' = input$IntNMF_N_clusts,
                            'CIMLR' = input$CIMLR_N_clusts,
                            'MoCluster' = input$MoCluster_N_clusts)
        }
        if(length(input$multiClusteringAlgorithms)>1){
          N.clust <- input$Consensus_N_clusts #用户指定聚类数
        }
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #84.Colors for mRNA dataset heatmap填写表格
    output$table.mRNA_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$mRNA_heatmap_color_number,nrow=1))
      colnames(DF)=1:input$mRNA_heatmap_color_number
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:input$mRNA_heatmap_color_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #85.Colors for lncRNA dataset heatmap填写表格
    output$table.lncRNA_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$lncRNA_heatmap_color_number,nrow=1))
      colnames(DF)=1:input$lncRNA_heatmap_color_number
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:input$lncRNA_heatmap_color_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #86.Colors for DNA methylation dataset heatmap填写表格
    output$table.DNA_methylation_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$DNA_methylation_heatmap_color_number,nrow=1))
      colnames(DF)=1:input$DNA_methylation_heatmap_color_number
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:input$DNA_methylation_heatmap_color_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #87.Colors for copy number alterations dataset heatmap填写表格
    output$table.copy_number_alterations_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$copy_number_alterations_heatmap_color_number,nrow=1))
      colnames(DF)=1:input$copy_number_alterations_heatmap_color_number
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:input$copy_number_alterations_heatmap_color_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #88.Colors for binary somatic mutation dataset heatmap填写表格
    output$table.binary_somatic_mutation_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=2,nrow=1))
      colnames(DF)=1:2
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:2){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #89.Colors for radiomics dataset heatmap填写表格
    output$table.radiomics_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$radiomics_heatmap_color_number,nrow=1))
      colnames(DF)=1:input$radiomics_heatmap_color_number
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:input$radiomics_heatmap_color_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #90.User defined features to be annotated for mRNA dataset heatmap填写表格
    output$table.mRNA_heatmap_annRow<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$mRNA_heatmap_annRow_number,nrow=1))
      colnames(DF)=1:input$mRNA_heatmap_annRow_number
      rownames(DF)="Annotated features"
      
      default <- c()
      for(d in 1:input$mRNA_heatmap_annRow_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=130)
    })
    
    #91.User defined features to be annotated for lncRNA dataset heatmap填写表格
    output$table.lncRNA_heatmap_annRow<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$lncRNA_heatmap_annRow_number,nrow=1))
      colnames(DF)=1:input$lncRNA_heatmap_annRow_number
      rownames(DF)="Annotated features"
      
      default <- c()
      for(d in 1:input$lncRNA_heatmap_annRow_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=130)
    })
    
    #92.User defined features to be annotated for DNA methylation dataset heatmap填写表格
    output$table.DNA_methylation_heatmap_annRow<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$DNA_methylation_heatmap_annRow_number,nrow=1))
      colnames(DF)=1:input$DNA_methylation_heatmap_annRow_number
      rownames(DF)="Annotated features"
      
      default <- c()
      for(d in 1:input$DNA_methylation_heatmap_annRow_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=130)
    })
    
    #93.User defined features to be annotated for copy number alterations dataset heatmap填写表格
    output$table.copy_number_alterations_heatmap_annRow<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$copy_number_alterations_heatmap_annRow_number,nrow=1))
      colnames(DF)=1:input$copy_number_alterations_heatmap_annRow_number
      rownames(DF)="Annotated features"
      
      default <- c()
      for(d in 1:input$copy_number_alterations_heatmap_annRow_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=130)
    })
    
    #94.User defined features to be annotated for binary somatic mutation dataset heatmap填写表格
    output$table.binary_somatic_mutation_heatmap_annRow<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$binary_somatic_mutation_heatmap_annRow_number,nrow=1))
      colnames(DF)=1:input$binary_somatic_mutation_heatmap_annRow_number
      rownames(DF)="Annotated features"
      
      default <- c()
      for(d in 1:input$binary_somatic_mutation_heatmap_annRow_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=130)
    })
    
    #95.User defined features to be annotated for radiomics dataset heatmap填写表格
    output$table.radiomics_heatmap_annRow<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$radiomics_heatmap_annRow_number,nrow=1))
      colnames(DF)=1:input$radiomics_heatmap_annRow_number
      rownames(DF)="Annotated features"
      
      default <- c()
      for(d in 1:input$radiomics_heatmap_annRow_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=130)
    })
    
    #96.Sample annotations from survival information for heatmap填写表格
    output$table.heatmap_annCol<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$heatmap_annCol_number,nrow=3))
      colnames(DF)=1:input$heatmap_annCol_number
      rownames(DF)=c("Sample annotation","Continuous or Categorical","Color settings")
      
      default <- c()
      for(d in 1:input$heatmap_annCol_number){
        default <- c(default,"")
      }
      DF[1,]=default
      DF[2,]=default
      DF[3,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=180)
    })
    
    #97.heatmap_process事件响应按钮
    observeEvent(input$heatmap_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Get multi-omics comprehensive heatmap',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      
      ###聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        if("iClusterBayes" %in% input$multiClusteringAlgorithms){
          iClusterBayes <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.iClusterBayes.RData"))))
          clust.result <- iClusterBayes
        }
        if("SNF" %in% input$multiClusteringAlgorithms){
          SNF <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.SNF.RData"))))
          clust.result <- SNF
        }
        if("PINSPlus" %in% input$multiClusteringAlgorithms){
          PINSPlus <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.PINSPlus.RData"))))
          clust.result <- PINSPlus
        }
        if("NEMO" %in% input$multiClusteringAlgorithms){
          NEMO <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.NEMO.RData"))))
          clust.result <- NEMO
        }
        if("COCA" %in% input$multiClusteringAlgorithms){
          COCA <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.COCA.RData"))))
          clust.result <- COCA
        }
        if("LRAcluster" %in% input$multiClusteringAlgorithms){
          LRAcluster <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.LRAcluster.RData"))))
          clust.result <- LRAcluster
        }
        if("ConsensusClustering" %in% input$multiClusteringAlgorithms){
          ConsensusClustering <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.ConsensusClustering.RData"))))
          clust.result <- ConsensusClustering
        }
        if("IntNMF" %in% input$multiClusteringAlgorithms){
          IntNMF <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.IntNMF.RData"))))
          clust.result <- IntNMF
        }
        if("CIMLR" %in% input$multiClusteringAlgorithms){
          CIMLR <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.CIMLR.RData"))))
          clust.result <- CIMLR
        }
        if("MoCluster" %in% input$multiClusteringAlgorithms){
          MoCluster <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.MoCluster.RData"))))
          clust.result <- MoCluster
        }
      }
      
      ###聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        
        #Multiple clustering algorithms
        if(input$multiple_algorithms_heatmap==1){
          
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          clust.result <- cmoic
        }else{
          specified_algorithm_heatmap <- trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")
          moic.res.list <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.moic.res.list.RData"))))
          clust.result <- moic.res.list[[specified_algorithm_heatmap]]
        }
      }
      
      ###多组学数据导入
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      indata <- integrated_data[1:length(input$multiOmics)]
      
      ##多组学数据预处理
      indata$meth.beta <- log2(indata$meth.beta / (1 - indata$meth.beta)) #甲基化数据转换
      
      ##处理meth.beta可能存在的Inf值(样本中位数填补连续型变量)
      for(i in 1:nrow(indata$meth.beta)){
        for(j in 1:ncol(indata$meth.beta)){
          if(is.infinite(indata$meth.beta[i,j])){
            indata$meth.beta[i,j] <- median(indata$meth.beta[,j])
          }
        }
      }
      
      ###Assign truncating values for extreme values in continuous normalized multi-omics data
      halfwidth <- data.frame(rbind(hot_to_r(input$table.halfwidth_heatmap)))[1,]
      halfwidth <- trimws(halfwidth,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      halfwidth <- as.numeric(halfwidth)
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          halfwidth <- c(halfwidth[1:(length(halfwidth)-1)],NA,halfwidth[length(halfwidth)])
        }
      }else{
        if(5 %in% input$multiOmics){
          halfwidth <- c(halfwidth,NA)
        }
      }
      
      ###Indicate if each omic data should be centered
      centerFlag <- unlist(data.frame(rbind(hot_to_r(input$table.centerFlag_heatmap)))[1,])
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          centerFlag <- c(centerFlag[1:(length(centerFlag)-1)],FALSE,centerFlag[length(centerFlag)])
        }
      }else{
        if(5 %in% input$multiOmics){
          centerFlag <- c(centerFlag,FALSE)
        }
      }
      
      ###Indicate if each omic data should be scaled
      scaleFlag <- unlist(data.frame(rbind(hot_to_r(input$table.scaleFlag_heatmap)))[1,])
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          scaleFlag <- c(scaleFlag[1:(length(scaleFlag)-1)],FALSE,scaleFlag[length(scaleFlag)])
        }
      }else{
        if(5 %in% input$multiOmics){
          scaleFlag <- c(scaleFlag,FALSE)
        }
      }
      
      ###Get standardized multi-omics data
      plotdata <- getStdiz(data = indata,
                           halfwidth = halfwidth,
                           centerFlag = centerFlag,
                           scaleFlag = scaleFlag)
      data <- plotdata
      
      progress$set(value = 2)
      ##Indicate if the omic data is binary matrix of 0 and 1 such as mutation
      if(6 %in% input$multiOmics){
        if(5 %in% input$multiOmics){
          is.binary <- c(rep(FALSE, length(data)-2),TRUE,FALSE)
        }else{
          is.binary <- rep(FALSE, length(data))
        }
      }else{
        if(5 %in% input$multiOmics){
          is.binary <- c(rep(FALSE, length(data)-1),TRUE)
        }else{
          is.binary <- rep(FALSE, length(data))
        }
      }
      
      ##Row title for each omic data
      if(input$row_title_heatmap==1){
        row.title = c("mRNA","lncRNA","Methylation","CNAs","Mutation","Radiomics")
      }else{
        row.title <- data.frame(rbind(hot_to_r(input$table.row_title_heatmap)))[1,]
        row.title <- trimws(row.title,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      ##Legend title for each omic data
      if(input$legend_title_heatmap==1){
        legend.name = c("mRNA","lncRNA","DNA methylation","copy number alterations","binary somatic mutation","radiomics")
      }else{
        legend.name <- data.frame(rbind(hot_to_r(input$table.legend_title_heatmap)))[1,]
        legend.name <- trimws(legend.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      clust.res <- clust.result$clust.res
      
      ##The dendrogram for columns at the top of heatmap
      if((length(input$multiClusteringAlgorithms)==1 && ('COCA' %in% input$multiClusteringAlgorithms | 'LRAcluster' %in% input$multiClusteringAlgorithms | 'ConsensusClustering' %in% input$multiClusteringAlgorithms | 'MoCluster' %in% input$multiClusteringAlgorithms)) | (length(input$multiClusteringAlgorithms)>1 && (input$multiple_algorithms_heatmap==1 | (input$multiple_algorithms_heatmap==2 && (trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")=='COCA' | trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")=='LRAcluster' | trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")=='ConsensusClustering' | trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")=='MoCluster'))))){
        clust.dend <- clust.result$clust.dend
      }else{
        clust.dend <- NULL
      }
      
      ##Show the dendrogram for columns at the top of heatmap or not
      if(is.null(clust.dend)){
        show.col.dend <- TRUE
      }else{
        if(input$show_col_dend_heatmap==1){
          show.col.dend <- TRUE
        }else{
          show.col.dend <- FALSE
        }
      }
      
      ##Show the sample names for columns at the bottom of heatmap
      if(input$show_colnames_heatmap==1){
        show.colnames <- TRUE
      }else{
        show.colnames <- FALSE
      }
      
      ##Show the feature names for rows of each omic data
      show.rownames <- unlist(data.frame(rbind(hot_to_r(input$table.show_rownames_heatmap)))[1,])
      
      ##Distance method for clustering each omic data at feature dimension
      clust.dist.row <- data.frame(rbind(hot_to_r(input$table.clust_dist_row_heatmap)))[1,]
      clust.dist.row <- trimws(clust.dist.row,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##Clustering method for each omic data at feature dimension
      clust.method.row <- data.frame(rbind(hot_to_r(input$table.clust_method_row_heatmap)))[1,]
      clust.method.row <- trimws(clust.method.row,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##Show dendrogram for rows of each omic data
      show.row.dend <- unlist(data.frame(rbind(hot_to_r(input$table.show_row_dend_heatmap)))[1,])
      
      ##Colors for annotating each cluster at the top of heatmap
      if(input$heatmap_clustcolor==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.heatmap_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      ##Colors for heatmap of each omic data
      if(input$heatmap_color==1){
        color   <- list()
        # color <- rep(list(c("#00FF00", "#000000", "#FF0000")),length(data))
        
        if(1 %in% input$multiOmics){
          mRNA.col   <- c("#00FF00", "#008000", "#000000", "#800000", "#FF0000")
          color$"mRNA.col" <- mRNA.col
        }
        if(2 %in% input$multiOmics){
          lncRNA.col <- c("#6699CC", "white", "#FF3C38")
          color$"lncRNA.col" <- lncRNA.col
        }
        if(3 %in% input$multiOmics){
          meth.col <- c("#0074FE", "#96EBF9", "#FEE900", "#F00003")
          color$"meth.col" <- meth.col
        }
        if(4 %in% input$multiOmics){
          cna.col <- c("#00FFFF", "#00CED1", "#87CEFA", "#EE82EE", "#FF00FF")
          color$"cna.col" <- cna.col
        }
        if(5 %in% input$multiOmics){
          mut.col <- c("grey90", "black")
          color$"mut.col" <- mut.col
        }
        if(6 %in% input$multiOmics){
          radi.col <- c("#A52A2A", "#B22222", "#FFDAB9", "#FFA500", "#FF8C00")
          color$"radi.col" <- radi.col
        }
        
      }else{
        color <- list()
        if(1 %in% input$multiOmics){
          mRNA_heatmap_color <- data.frame(rbind(hot_to_r(input$table.mRNA_heatmap_color)))[1,]
          mRNA_heatmap_color <- trimws(mRNA_heatmap_color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          color$"mRNA.col" <- mRNA_heatmap_color
        }
        if(2 %in% input$multiOmics){
          lncRNA_heatmap_color <- data.frame(rbind(hot_to_r(input$table.lncRNA_heatmap_color)))[1,]
          lncRNA_heatmap_color <- trimws(lncRNA_heatmap_color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          color$"lncRNA.col" <- lncRNA_heatmap_color
        }
        if(3 %in% input$multiOmics){
          DNA_methylation_heatmap_color <- data.frame(rbind(hot_to_r(input$table.DNA_methylation_heatmap_color)))[1,]
          DNA_methylation_heatmap_color <- trimws(DNA_methylation_heatmap_color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          color$"meth.col" <- DNA_methylation_heatmap_color
        }
        if(4 %in% input$multiOmics){
          copy_number_alterations_heatmap_color <- data.frame(rbind(hot_to_r(input$table.copy_number_alterations_heatmap_color)))[1,]
          copy_number_alterations_heatmap_color <- trimws(copy_number_alterations_heatmap_color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          color$"cna.col" <- copy_number_alterations_heatmap_color
        }
        if(5 %in% input$multiOmics){
          binary_somatic_mutation_heatmap_color <- data.frame(rbind(hot_to_r(input$table.binary_somatic_mutation_heatmap_color)))[1,]
          binary_somatic_mutation_heatmap_color <- trimws(binary_somatic_mutation_heatmap_color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          color$"mut.col" <- binary_somatic_mutation_heatmap_color
        }
        if(6 %in% input$multiOmics){
          radiomics_heatmap_color <- data.frame(rbind(hot_to_r(input$table.radiomics_heatmap_color)))[1,]
          radiomics_heatmap_color <- trimws(radiomics_heatmap_color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          color$"radi.col" <- radiomics_heatmap_color
        }
      }
      
      ##Features should be annotated specifically in each omic data
      if((length(input$multiClusteringAlgorithms)==1 && ('iClusterBayes' %in% input$multiClusteringAlgorithms | 'CIMLR' %in% input$multiClusteringAlgorithms | 'MoCluster' %in% input$multiClusteringAlgorithms)) | (length(input$multiClusteringAlgorithms)>1 && input$multiple_algorithms_heatmap==2 && (trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")=='iClusterBayes' | trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")=='CIMLR' | trimws(input$specified_algorithm_heatmap,whitespace = "[ \t\r\n]")=='MoCluster'))){
        
        feat <- clust.result$feat.res
        annRow <- list()
        if(1 %in% input$multiOmics){
          if(input$mRNA_heatmap_annRow==1){
            mRNA_heatmap_annRow_number <- input$mRNA_heatmap_annRow_number
            if(input$mRNA_heatmap_annRow_setting==1){
              annRow$"mRNA.annRow" <- feat[which(feat$dataset == "mRNA.expr"),][1:mRNA_heatmap_annRow_number,"feature"]
            }else{
              mRNA_heatmap_annRow <- data.frame(rbind(hot_to_r(input$table.mRNA_heatmap_annRow)))[1,]
              mRNA_heatmap_annRow <- trimws(mRNA_heatmap_annRow,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
              annRow$"mRNA.annRow" <- feat[which(feat$dataset == "mRNA.expr"),][which(feat$feature %in% mRNA_heatmap_annRow),"feature"]
            }
          }else{
            annRow$"mRNA.annRow" <- NULL
          }
        }
        if(2 %in% input$multiOmics){
          if(input$lncRNA_heatmap_annRow==1){
            lncRNA_heatmap_annRow_number <- input$lncRNA_heatmap_annRow_number
            if(input$lncRNA_heatmap_annRow_setting==1){
              annRow$"lncRNA.annRow" <- feat[which(feat$dataset == "lncRNA.expr"),][1:lncRNA_heatmap_annRow_number,"feature"]
            }else{
              lncRNA_heatmap_annRow <- data.frame(rbind(hot_to_r(input$table.lncRNA_heatmap_annRow)))[1,]
              lncRNA_heatmap_annRow <- trimws(lncRNA_heatmap_annRow,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
              annRow$"lncRNA.annRow" <- feat[which(feat$dataset == "lncRNA.expr"),][which(feat$feature %in% lncRNA_heatmap_annRow),"feature"]
            }
          }else{
            annRow$"lncRNA.annRow" <- NULL
          }
        }
        if(3 %in% input$multiOmics){
          if(input$DNA_methylation_heatmap_annRow==1){
            DNA_methylation_heatmap_annRow_number <- input$DNA_methylation_heatmap_annRow_number
            if(input$DNA_methylation_heatmap_annRow_setting==1){
              annRow$"meth.annRow" <- feat[which(feat$dataset == "meth.beta"),][1:DNA_methylation_heatmap_annRow_number,"feature"]
            }else{
              DNA_methylation_heatmap_annRow <- data.frame(rbind(hot_to_r(input$table.DNA_methylation_heatmap_annRow)))[1,]
              DNA_methylation_heatmap_annRow <- trimws(DNA_methylation_heatmap_annRow,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
              annRow$"meth.annRow" <- feat[which(feat$dataset == "meth.beta"),][which(feat$feature %in% DNA_methylation_heatmap_annRow),"feature"]
            }
          }else{
            annRow$"meth.annRow" <- NULL
          }
        }
        if(4 %in% input$multiOmics){
          if(input$copy_number_alterations_heatmap_annRow==1){
            copy_number_alterations_heatmap_annRow_number <- input$copy_number_alterations_heatmap_annRow_number
            if(input$copy_number_alterations_heatmap_annRow_setting==1){
              annRow$"cna.annRow" <- feat[which(feat$dataset == "cna"),][1:copy_number_alterations_heatmap_annRow_number,"feature"]
            }else{
              copy_number_alterations_heatmap_annRow <- data.frame(rbind(hot_to_r(input$table.copy_number_alterations_heatmap_annRow)))[1,]
              copy_number_alterations_heatmap_annRow <- trimws(copy_number_alterations_heatmap_annRow,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
              annRow$"cna.annRow" <- feat[which(feat$dataset == "cna"),][which(feat$feature %in% copy_number_alterations_heatmap_annRow),"feature"]
            }
          }else{
            annRow$"cna.annRow" <- NULL
          }
        }
        if(5 %in% input$multiOmics){
          if(input$binary_somatic_mutation_heatmap_annRow==1){
            binary_somatic_mutation_heatmap_annRow_number <- input$binary_somatic_mutation_heatmap_annRow_number
            if(input$binary_somatic_mutation_heatmap_annRow_setting==1){
              annRow$"mut.annRow" <- feat[which(feat$dataset == "mut.status"),][1:binary_somatic_mutation_heatmap_annRow_number,"feature"]
            }else{
              binary_somatic_mutation_heatmap_annRow <- data.frame(rbind(hot_to_r(input$table.binary_somatic_mutation_heatmap_annRow)))[1,]
              binary_somatic_mutation_heatmap_annRow <- trimws(binary_somatic_mutation_heatmap_annRow,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
              annRow$"mut.annRow" <- feat[which(feat$dataset == "mut.status"),][which(feat$feature %in% binary_somatic_mutation_heatmap_annRow),"feature"]
            }
          }else{
            annRow$"mut.annRow" <- NULL
          }
        }
        if(6 %in% input$multiOmics){
          if(input$radiomics_heatmap_annRow==1){
            radiomics_heatmap_annRow_number <- input$radiomics_heatmap_annRow_number
            if(input$radiomics_heatmap_annRow_setting==1){
              annRow$"radiomics.annRow" <- feat[which(feat$dataset == "radiomics"),][1:radiomics_heatmap_annRow_number,"feature"]
            }else{
              radiomics_heatmap_annRow <- data.frame(rbind(hot_to_r(input$table.radiomics_heatmap_annRow)))[1,]
              radiomics_heatmap_annRow <- trimws(radiomics_heatmap_annRow,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
              annRow$"radiomics.annRow" <- feat[which(feat$dataset == "radiomics"),][which(feat$feature %in% radiomics_heatmap_annRow),"feature"]
            }
          }else{
            annRow$"radiomics.annRow" <- NULL
          }
        }
      }else{
        annRow <- NULL
      }
      
      ##Sample annotations from survival information for heatmap
      if(input$heatmap_annCol==1){
        surv_clin.info <- integrated_data$clin.info #获取生存和临床数据
        # if(!('fustat' %in% colnames(surv_clin.info)) & ('OS' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS')] <- "fustat"
        # } #数据预处理(列名替换)
        # if(!('futime' %in% colnames(surv_clin.info)) & ('OS.time' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS.time')] <- "futime"
        # } #数据预处理(列名替换)
        # if('age_at_initial_pathologic_diagnosis' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='age_at_initial_pathologic_diagnosis')] <- "age"
        # } #数据预处理(列名替换)
        # if('ajcc_pathologic_tumor_stage' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        # } #数据预处理(列名替换)
        # if("pstage" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Discrepancy]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Not Available]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     } 
        #   } #缺失值填补(众数填补多分类变量)
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage I" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IB"){
        #       surv_clin.info$pStage[i] <- "Stage I"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage II" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIB"){
        #       surv_clin.info$pStage[i] <- "Stage II"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage III" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIB"){
        #       surv_clin.info$pStage[i] <- "Stage III"
        #     }
        #   } #类别合并
        # }
        # if("age" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(is.na(surv_clin.info$age[i])==TRUE){
        #       surv_clin.info$age[i] <- median(surv_clin.info$age,na.rm = TRUE)
        #     }
        #   } #缺失值填补(中位数填补连续型变量)
        # }
        ###annCol
        sample_annotation <- data.frame(rbind(hot_to_r(input$table.heatmap_annCol)))[1,]
        sample_annotation <- trimws(sample_annotation,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        annCol <- surv_clin.info[,sample_annotation, drop = FALSE]
        ###annColors
        continuous_or_categorical <- data.frame(rbind(hot_to_r(input$table.heatmap_annCol)))[2,]
        continuous_or_categorical <- trimws(continuous_or_categorical,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        color_settings <- data.frame(rbind(hot_to_r(input$table.heatmap_annCol)))[3,]
        color_settings <- trimws(color_settings,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        ###Process the obtained data
        if(!('fustat' %in% colnames(annCol)) & ('OS' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS')] <- "fustat"
        } #数据预处理(列名替换)
        if(!('futime' %in% colnames(annCol)) & ('OS.time' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS.time')] <- "futime"
        } #数据预处理(列名替换)
        if('age_at_initial_pathologic_diagnosis' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='age_at_initial_pathologic_diagnosis')] <- "age"
        } #数据预处理(列名替换)
        if('ajcc_pathologic_tumor_stage' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        } #数据预处理(列名替换)
        for (j in 1:ncol(annCol)){
          for(i in 1:nrow(annCol)){
            ##NA判定##
            if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))){
              next
            }
            if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Discrepancy]" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Not Available]"){
              annCol[i,j] <- NA
            }
            ##类别合并##
            if(tolower(colnames(annCol)[j])=="pstage"){
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage I" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IB"){
                annCol[i,j] <- "Stage I"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage II" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIB"){
                annCol[i,j] <- "Stage II"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage III" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIB"){
                annCol[i,j] <- "Stage III"
              }
            }
            ##缺失值填补##
            if(continuous_or_categorical[j] == "Categorical"){
              ##众数填补多分类变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- names(which(table(annCol[,j])==max(table(annCol[,j]))))
              }
            }else{
              ##中位数填补连续型变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- median(annCol[,j],na.rm = TRUE)
              }
            }
          }
        }
        annColors <- list()
        for(i in 1:length(sample_annotation)){
          if(continuous_or_categorical[i]=="Continuous"){
            breaks <- c(min(annCol[,sample_annotation[i]]),median(annCol[,sample_annotation[i]]),max(annCol[,sample_annotation[i]]))
            colors <- unlist(strsplit(color_settings[i],";"))
            annColors[[i]] <- circlize::colorRamp2(breaks = breaks, colors = colors)
          }else{
            colors <- unlist(strsplit(color_settings[i],";"))
            names(colors) <- unique(annCol[,sample_annotation[i]])
            annColors[[i]] <- colors
          }
        }
        names(annColors) <- sample_annotation
      }else{
        annCol = NULL
        annColors = NULL
      }
      
      ##The width for each heatmap
      width <- input$width_heatmap
      
      ##The height for each heatmap
      height <- input$height_heatmap
      
      #The output path for storing the heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of multi-omics heatmaps
      fig.name <- input$heatmap_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "moheatmap"
      }
      
      progress$set(value = 3)
      heatmap_list <- getMoHeatmaps(data = data,
                                    is.binary = is.binary,
                                    row.title = row.title,
                                    legend.name = legend.name,
                                    clust.res = clust.res,
                                    clust.dend = clust.dend,
                                    show.col.dend = show.col.dend,
                                    show.colnames = show.colnames,
                                    show.row.dend = show.row.dend,
                                    show.rownames = show.rownames,
                                    clust.dist.row = clust.dist.row,
                                    clust.method.row = clust.method.row,
                                    clust.col = clust.col,
                                    color = color,
                                    annCol = annCol,
                                    annColors = annColors,
                                    annRow = annRow,
                                    width = width,
                                    height = height,
                                    fig.path = fig.path,
                                    fig.name = fig.name)
      
      #保存heatmap绘图结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(heatmap_list, file = file.path(RData.path,paste0(tolower(project),".tcga.heatmap_list.RData")))
      
      progress$set(value = 4)
      #绘图展示
      output$heatmap_figure_unit<-renderUI({
        list(column(12,downloadButton("heatmap_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("heatmap_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$heatmap_figure_download<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          if(is.null(annCol)){
            pdf(file = theFile, width = width, height = height*heatmap_list$n_dat/2)
          }else{
            pdf(file = theFile, width = width, height = height*heatmap_list$n_dat/1.5)
          }
          draw(heatmap_list$ht_list, merge_legend = TRUE, heatmap_legend_side = "right") # output to pdf
          dev.off()
        }
      )
      
      ##页面展示事件
      output$heatmap_figure<-renderPlot({
        
        draw(heatmap_list$ht_list, merge_legend = TRUE, heatmap_legend_side = "right") # output to screen
      })
      
      ##RData结果下载按钮响应事件
      output$getMoHeatmaps_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.heatmap_list.RData"),
        content = function(theFile) {
          save(heatmap_list,file=theFile)
        }
      )
      
      #getMoHeatmaps后给予用户的反馈提示
      output$heatmap_finish<-renderUI({
        
        list(br(),column(12,h3(strong("The process of getting multi-omics comprehensive heatmap has been finished, you can download and check the figure as well as the '.RData' file. Now all steps in 'GET Module' have been finished, let's start the second step of MOVICS--COMP Module!")),style="color:darkblack"),br(),
             column(12,downloadButton("getMoHeatmaps_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #98.Estimating x-year survival填写表格
    output$table.xyrs_est_compSurv<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$xyrs_est_number_compSurv,nrow=1))
      colnames(DF)=1:input$xyrs_est_number_compSurv
      rownames(DF)="Year"
      
      default <- c()
      for(d in 1:input$xyrs_est_number_compSurv){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=60)
    })
    
    #99.Colors for each subtype填写表格
    output$table.compSurv_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ##获取聚类数
      if(input$tcga_vali_compSurv==1){
        #TCGA
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        
        if(input$Consensus_N_clust==1){
          N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
        }else{
          if(length(input$multiClusteringAlgorithms)==1){
            #用户指定聚类数
            N.clust <- switch(input$multiClusteringAlgorithms,
                              'iClusterBayes' = input$iClusterBayes_N_clusts,
                              'SNF' = input$SNF_N_clusts,
                              'PINSPlus' = input$PINSPlus_N_clusts,
                              'NEMO' = input$NEMO_N_clusts,
                              'COCA' = input$COCA_N_clusts,
                              'LRAcluster' = input$LRAcluster_N_clusts,
                              'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                              'IntNMF' = input$IntNMF_N_clusts,
                              'CIMLR' = input$CIMLR_N_clusts,
                              'MoCluster' = input$MoCluster_N_clusts)
          }
          if(length(input$multiClusteringAlgorithms)>1){
            N.clust <- input$Consensus_N_clusts #用户指定聚类数
          }
        }
      }else{
        #Validation
        if(input$validation_method_compSurv==1){
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        N.clust <- length(unique(integrated_data$clust.res$clust))
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #100.compSurv_process事件响应按钮
    observeEvent(input$compSurv_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$tcga_vali_compSurv==1,'Compare the prognosis of different subtypes on tcga datasets','Compare the prognosis of different subtypes on validation datasets'),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compSurv==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###聚类结果导入
        if(input$validation_method_compSurv==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      surv.info <- integrated_data$clin.info #获取生存和临床数据
      if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
        colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
      } #数据预处理(列名替换)
      if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
        colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
      } #数据预处理(列名替换)
      
      progress$set(value = 2)
      ##Format convertion of the survival time
      if(input$convt_time_compSurv==1){
        convt.time <- 'y'
      }else if(input$convt_time_compSurv==2){
        convt.time <- 'm'
      }else{
        convt.time <- 'd'
      }
      
      ##The x-axis cutoff for showing the maximal survival time
      if(input$surv_cut_compSurv==1){
        surv.cut <- NULL
      }else{
        surv.cut <- input$surv_cut_value_compSurv
      }
      
      ##Estimate probability of surviving beyond a certain number (x) of years (Estimating x-year survival)
      if(input$xyrs_est_compSurv==1){
        xyrs.est <- data.frame(rbind(hot_to_r(input$table.xyrs_est_compSurv)))[1,]
        xyrs.est <- trimws(xyrs.est,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        xyrs.est <- as.numeric(xyrs.est)
      }else{
        xyrs.est <- NULL
      }
      
      ##Colors for each subtype
      if(input$clustcolor_compSurv==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.compSurv_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      ##Method for adjusting p values
      p.adjust.method <- switch(input$p_adjust_method_compSurv,
                                '1' = "holm",
                                '2' = "hochberg",
                                '3' = "hommel",
                                '4' = "bonferroni",
                                '5' = "BH",
                                '6' = "BY",
                                '7' = "fdr",
                                '8' = "none")
      
      ##The way for drawing a horizontal/vertical line at median survival
      if(input$surv_median_line_compSurv==1){
        surv.median.line <- "h"
      }else if(input$surv_median_line_compSurv==2){
        surv.median.line <- "v"
      }else if(input$surv_median_line_compSurv==3){
        surv.median.line <- "hv"
      }else{
        surv.median.line <- "none"
      }
      
      ##The output path for storing the kaplan-meier curve
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      ##The name of the kaplan-meier curve
      if(input$tcga_vali_compSurv==1){
        fig.name <- input$compSurv_FigureName_tcga
      }else{
        fig.name <- input$compSurv_FigureName_validation
      }
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        if(input$tcga_vali_compSurv==1){
          fig.name <- paste0("KAPLAN-MEIER_CURVE_OF_",moic.res$mo.method,"(TCGA)")
        }else{
          fig.name <- paste0("KAPLAN-MEIER_CURVE_OF_",moic.res$mo.method,"(Validation)")
        }
      }
      
      progress$set(value = 3)
      compSurvResult <- compSurvs(moic.res = moic.res,
                                 surv.info = surv.info,
                                 convt.time = convt.time,
                                 surv.cut = surv.cut,
                                 xyrs.est = xyrs.est,
                                 clust.col = clust.col,
                                 p.adjust.method = p.adjust.method,
                                 surv.median.line = surv.median.line,
                                 fig.name = fig.name,
                                 fig.path = fig.path)
      
      #保存compSurv结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compSurv==1){
        save(compSurvResult, file = file.path(RData.path,paste0(tolower(project),".tcga.compSurv.RData")))
      }else{
        save(compSurvResult, file = file.path(RData.path,paste0(tolower(project),".validation.compSurv.RData")))
      }
      
      progress$set(value = 4)
      #绘图展示
      output$compSurv_figure_unit<-renderUI({
        list(column(12,downloadButton("compSurv_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("compSurv_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$compSurv_figure_download<-downloadHandler(
        filename = ifelse(is.null(fig.name),ifelse(input$tcga_vali_compSurv==1,paste0("km_curve_",moic.res$mo.method,"(TCGA).pdf"),paste0("km_curve_",moic.res$mo.method,"(Validation).pdf")),paste0(fig.name,".pdf")),
        content = function(theFile) {
          pdf(file = theFile, width=6, height=7)
          p <- compSurvResult[[length(compSurvResult)]]
          print(p)
          dev.off()
        }
      )
      
      ##页面展示事件
      output$compSurv_figure<-renderPlot({
        
        p <- compSurvResult[[length(compSurvResult)]]
        p
      })
      
      ##RData结果下载按钮响应事件
      output$compSurv_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_compSurv==1,paste0(tolower(project),".tcga.compSurv.RData"),paste0(tolower(project),".validation.compSurv.RData")),
        content = function(theFile) {
          save(compSurvResult,file=theFile)
        }
      )
      
      #compSurv后给予用户的反馈提示
      output$compSurv_finish<-renderUI({
        
        if(input$tcga_vali_compSurv==1){
          list(br(),column(12,h3(strong("The process of comparing survival outcome on tcga datasets has been finished, you can download and check the figure as well as the '.RData' file. Now let's turn to the next step--'Compare clinical features'.")),style="color:darkblack"),br(),
          column(12,downloadButton("compSurv_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of comparing survival outcome on validation datasets has been finished, you can download and check the figure as well as the '.RData' file. Now let's turn to the next step--'Compare clinical features'.")),style="color:darkblack"),br(),
          column(12,downloadButton("compSurv_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
  
    #101.Specify variables in survival and clinical information for summary and statistical tests填写表格
    output$table.variables_compClinvar<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$variables_number_compClinvar,nrow=1))
      colnames(DF)=1:input$variables_number_compClinvar
      rownames(DF)="Variable"
      
      default <- c()
      for(d in 1:input$variables_number_compClinvar){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=60)
    })
    
    #102.Indicate the categorical variables填写表格
    output$table.categorical_compClinvar<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$categorical_number_compClinvar,nrow=1))
      colnames(DF)=1:input$categorical_number_compClinvar
      rownames(DF)="Categorical variable"
      
      default <- c()
      for(d in 1:input$categorical_number_compClinvar){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=130)
    })
    
    #103.Specify the variables for which the p-values should be those of nonparametric tests填写表格
    output$table.nonparametric_compClinvar<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$nonparametric_number_compClinvar,nrow=1))
      colnames(DF)=1:input$nonparametric_number_compClinvar
      rownames(DF)="Nonparametric test variable"
      
      default <- c()
      for(d in 1:input$nonparametric_number_compClinvar){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=180)
    })
    
    #104.Specify the variables for which the p-values should be those of exact tests填写表格
    output$table.exact_compClinvar<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$exact_number_compClinvar,nrow=1))
      colnames(DF)=1:input$exact_number_compClinvar
      rownames(DF)="Exact test variable"
      
      default <- c()
      for(d in 1:input$exact_number_compClinvar){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=180)
    })
    
    #105.compClinvar_process事件响应按钮
    observeEvent(input$compClinvar_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$tcga_vali_compClinvar==1,'Compare the specified clinical variables of different subtypes on tcga datasets','Compare the specified clinical variables of different subtypes on validation datasets'),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compClinvar==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###聚类结果导入
        if(input$validation_method_compClinvar==1){
          #NTP
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          #PAM
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      surv.info <- integrated_data$clin.info #获取生存和临床数据
      # if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
      #   colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
      # } #数据预处理(列名替换)
      # if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
      #   colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
      # } #数据预处理(列名替换)
      # if('age_at_initial_pathologic_diagnosis' %in% colnames(surv.info)){
      #   colnames(surv.info)[which(colnames(surv.info)=='age_at_initial_pathologic_diagnosis')] <- "age"
      # } #数据预处理(列名替换)
      # if('ajcc_pathologic_tumor_stage' %in% colnames(surv.info)){
      #   colnames(surv.info)[which(colnames(surv.info)=='ajcc_pathologic_tumor_stage')] <- "pStage"
      # } #数据预处理(列名替换)
      # if("pstage" %in% tolower(names(surv.info))){
      #   for(i in 1:nrow(surv.info)){
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="[Discrepancy]"){
      #       surv.info$pStage[i] <- names(which(table(surv.info$pStage)==max(table(surv.info$pStage))))
      #     }
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="[Not Available]"){
      #       surv.info$pStage[i] <- names(which(table(surv.info$pStage)==max(table(surv.info$pStage))))
      #     } 
      #   } #缺失值填补(众数填补多分类变量)
      #   for(i in 1:nrow(surv.info)){
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage I" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IB"){
      #       surv.info$pStage[i] <- "Stage I"
      #     }
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage II" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIB"){
      #       surv.info$pStage[i] <- "Stage II"
      #     }
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage III" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIB"){
      #       surv.info$pStage[i] <- "Stage III"
      #     }
      #   } #类别合并
      # }
      # if("age" %in% tolower(names(surv.info))){
      #   for(i in 1:nrow(surv.info)){
      #     if(is.na(surv.info$age[i])==TRUE){
      #       surv.info$age[i] <- median(surv.info$age,na.rm = TRUE)
      #     }
      #   } #缺失值填补(中位数填补连续型变量)
      # }
      
      progress$set(value = 2)
      ##Specify variables in survival and clinical information for summary and statistical tests
      variables_compClinvar <- data.frame(rbind(hot_to_r(input$table.variables_compClinvar)))[1,]
      variables_compClinvar <- trimws(variables_compClinvar,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(length(variables_compClinvar)>1){
        var2comp <- surv.info[,variables_compClinvar]
      }else{
        var2comp_rownames <- rownames(surv.info)
        var2comp <- data.frame(surv.info[,variables_compClinvar])
        colnames(var2comp) <- variables_compClinvar
        rownames(var2comp) <- var2comp_rownames
      }
      
      # ###数据预处理
      # for (i in 1:nrow(var2comp)){
      #   for(j in 1:ncol(var2comp)){
      #     if(is.na(var2comp[i,j])){
      #       next
      #     }
      #     if(var2comp[i,j]=="" | var2comp[i,j]=="[Discrepancy]" | var2comp[i,j]=="[Not Available]"){
      #       var2comp[i,j] <- NA
      #     }
      #   }
      # }
      
      ##Indicate stratifying variable
      if(input$strata_compClinvar==1){
        strata <- input$strata_variable_compClinvar
        strata <- trimws(strata,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }else{
        strata <- NULL
      }
       
      ##Indicate the categorical variables
      if(input$categorical_number_compClinvar>0){
        factorVars <- data.frame(rbind(hot_to_r(input$table.categorical_compClinvar)))[1,]
        factorVars <- trimws(factorVars,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }else{
        factorVars <- NULL
      }
      
      ##Process the obtained data
      if(!('fustat' %in% colnames(var2comp)) & ('OS' %in% colnames(var2comp))){
        colnames(var2comp)[which(colnames(var2comp)=='OS')] <- "fustat"
      } #数据预处理(列名替换)
      if(!('futime' %in% colnames(var2comp)) & ('OS.time' %in% colnames(var2comp))){
        colnames(var2comp)[which(colnames(var2comp)=='OS.time')] <- "futime"
      } #数据预处理(列名替换)
      if('age_at_initial_pathologic_diagnosis' %in% colnames(var2comp)){
        colnames(var2comp)[which(colnames(var2comp)=='age_at_initial_pathologic_diagnosis')] <- "age"
      } #数据预处理(列名替换)
      if('ajcc_pathologic_tumor_stage' %in% colnames(var2comp)){
        colnames(var2comp)[which(colnames(var2comp)=='ajcc_pathologic_tumor_stage')] <- "pStage"
      } #数据预处理(列名替换)
      for (j in 1:ncol(var2comp)){
        for(i in 1:nrow(var2comp)){
          ##NA判定##
          if(is.na(trimws(var2comp[i,j],whitespace = "[ \t\r\n]"))){
            next
          }
          if(trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="[Discrepancy]" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="[Not Available]"){
            var2comp[i,j] <- NA
          }
          ##类别合并##
          if(tolower(colnames(var2comp)[j])=="pstage"){
            if(trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage I" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IB"){
              var2comp[i,j] <- "Stage I"
            }
            if(trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage II" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIB"){
              var2comp[i,j] <- "Stage II"
            }
            if(trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage III" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(var2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIIB"){
              var2comp[i,j] <- "Stage III"
            }
          }
          ##缺失值填补##
          if(colnames(var2comp)[j] %in% factorVars){
            ##众数填补多分类变量
            if(is.na(trimws(var2comp[i,j],whitespace = "[ \t\r\n]"))==TRUE){
              var2comp[i,j] <- names(which(table(var2comp[,j])==max(table(var2comp[,j]))))
            }
          }else{
            ##中位数填补连续型变量
            if(is.na(trimws(var2comp[i,j],whitespace = "[ \t\r\n]"))==TRUE){
              var2comp[i,j] <- median(var2comp[,j],na.rm = TRUE)
            }
          }
        }
      }
      
      ##Specify the variables for which the p-values should be those of nonparametric tests
      if(input$nonparametric_number_compClinvar>0){
        nonnormalVars <- data.frame(rbind(hot_to_r(input$table.nonparametric_compClinvar)))[1,]
        nonnormalVars <- trimws(nonnormalVars,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }else{
        nonnormalVars <- NULL
      }
      
      ##Specify the variables for which the p-values should be those of exact tests
      if(input$exact_number_compClinvar>0){
        exactVars <- data.frame(rbind(hot_to_r(input$table.exact_compClinvar)))[1,]
        exactVars <- trimws(exactVars,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }else{
        exactVars <- NULL
      }
      
      ##Whether NA should be handled as a regular factor level rather than missing value
      if(input$includeNA_compClinvar==1){
        includeNA <- TRUE
      }else{
        includeNA <- FALSE
      }
      
      ##Transform the '.txt' output file to a '.docx' WORD file or not ('.txt' file will be also kept)
      if(input$doWord_compClinvar==1){
        doWord <- TRUE
      }else{
        doWord <- FALSE
      }
      
      ##The output path for storing the output table
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      res.path <- file.path(user_workdir,"tables")
      if(!file.exists(res.path)) {dir.create(res.path,recursive = TRUE)}
      
      ##Indicate the name of the output table
      if(input$tcga_vali_compClinvar==1){
        tab.name <- input$compClinvar_TableName_tcga
      }else{
        tab.name <- input$compClinvar_TableName_validation
      }
      tab.name <- trimws(tab.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(tab.name==""){
        if(input$tcga_vali_compClinvar==1){
          if(input$strata_compClinvar==1){
            tab.name <- paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(TCGA)")
          }else{
            tab.name <- "Summarization_of_clinical_variables_stratified_by_current_subtypes(TCGA)"
          }
        }else{
          if(input$strata_compClinvar==1){
            tab.name <- paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(Validation)")
          }else{
            tab.name <- "Summarization_of_clinical_variables_stratified_by_current_subtypes(Validation)"
          }
        }
      }
      
      progress$set(value = 3)
      compClinvarResult <- compClinvars(moic.res = moic.res,
                                        var2comp = var2comp,
                                        strata = strata,
                                        factorVars = factorVars,
                                        nonnormalVars = nonnormalVars,
                                        exactVars = exactVars,
                                        includeNA = includeNA,
                                        doWord = doWord,
                                        tab.name = tab.name,
                                        res.path = res.path)
      
      #保存compClinvar结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compClinvar==1){
        save(compClinvarResult, file = file.path(RData.path,paste0(tolower(project),".tcga.compClinvar.RData")))
      }else{
        save(compClinvarResult, file = file.path(RData.path,paste0(tolower(project),".validation.compClinvar.RData")))
      }
      
      progress$set(value = 4)
      ##表格标题
      output$table.compClinvarTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$strata_compClinvar==1){
          if(input$tcga_vali_compClinvar==1){
            tags$strong(paste0("Table: Summarization of clinical variables stratified by ",strata," (TCGA)"))
          }else{
            tags$strong(paste0("Table: Summarization of clinical variables stratified by ",strata," (Validation)"))
          }
        }else{
          if(input$tcga_vali_compClinvar==1){
            tags$strong("Table: Summarization of clinical variables stratified by current subtypes (TCGA)")
          }else{
            tags$strong("Table: Summarization of clinical variables stratified by current subtypes (Validation)")
          }
        }
      })
      
      ##表格
      output$table.compClinvar<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(compClinvarResult$compTab),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(compClinvarResult$compTab)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Bfrtip',
                        searching= FALSE,
                        bInfo=FALSE, #not to shown 1/N message
                        display=NULL,
                        bPaginate= FALSE, #suppress page
                        pageLength =100,
                          
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compClinvar==1,ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(TCGA)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(TCGA)"),ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(Validation)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(Validation)")),tab.name),
                               title = ifelse(input$strata_compClinvar==1,ifelse(input$tcga_vali_compClinvar==1,paste0("Table: Summarization of clinical variables stratified by ",strata," (TCGA)"),paste0("Table: Summarization of clinical variables stratified by ",strata," (Validation)")),ifelse(input$tcga_vali_compClinvar==1,"Table: Summarization of clinical variables stratified by current subtypes (TCGA)","Table: Summarization of clinical variables stratified by current subtypes (Validation)"))
                          ),
                              
                          list(extend='csv',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compClinvar==1,ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(TCGA)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(TCGA)"),ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(Validation)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(Validation)")),tab.name),
                               title = ifelse(input$strata_compClinvar==1,ifelse(input$tcga_vali_compClinvar==1,paste0("Table: Summarization of clinical variables stratified by ",strata," (TCGA)"),paste0("Table: Summarization of clinical variables stratified by ",strata," (Validation)")),ifelse(input$tcga_vali_compClinvar==1,"Table: Summarization of clinical variables stratified by current subtypes (TCGA)","Table: Summarization of clinical variables stratified by current subtypes (Validation)"))
                          ),
                          list(extend='excel',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compClinvar==1,ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(TCGA)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(TCGA)"),ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(Validation)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(Validation)")),tab.name),
                               title = ifelse(input$strata_compClinvar==1,ifelse(input$tcga_vali_compClinvar==1,paste0("Table: Summarization of clinical variables stratified by ",strata," (TCGA)"),paste0("Table: Summarization of clinical variables stratified by ",strata," (Validation)")),ifelse(input$tcga_vali_compClinvar==1,"Table: Summarization of clinical variables stratified by current subtypes (TCGA)","Table: Summarization of clinical variables stratified by current subtypes (Validation)"))
                          ),
                              
                          list(extend='print',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compClinvar==1,ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(TCGA)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(TCGA)"),ifelse(input$strata_compClinvar==1,paste0("Summarization_of_clinical_variables_stratified_by_",strata,"(Validation)"),"Summarization_of_clinical_variables_stratified_by_current_subtype(Validation)")),tab.name),
                               title = ifelse(input$strata_compClinvar==1,ifelse(input$tcga_vali_compClinvar==1,paste0("Table: Summarization of clinical variables stratified by ",strata," (TCGA)"),paste0("Table: Summarization of clinical variables stratified by ",strata," (Validation)")),ifelse(input$tcga_vali_compClinvar==1,"Table: Summarization of clinical variables stratified by current subtypes (TCGA)","Table: Summarization of clinical variables stratified by current subtypes (Validation)"))
                          )
                        )
                      )
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.compClinvarNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      ##RData结果下载按钮响应事件
      output$compClinvars_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_compClinvar==1,paste0(tolower(project),".tcga.compClinvar.RData"),paste0(tolower(project),".validation.compClinvar.RData")),
        content = function(theFile) {
          save(compClinvarResult,file=theFile)
        }
      )
      
      ##compClinvar后给予用户的反馈提示
      output$compClinvar_finish<-renderUI({
        
        if(input$tcga_vali_compClinvar==1){
          list(br(),column(12,h3(strong("The process of comparing clinical features on the tcga datasets has been finished, you can download and check the table as well as the '.RData' file. Now let's turn to the next step--'Compare mutational frequency'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compClinvars_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of comparing clinical features on the validation datasets has been finished, you can download and check the table as well as the '.RData' file. Now let's turn to the next step--'Compare mutational frequency'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compClinvars_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #106.Colors for each subtype填写表格
    output$table.compMut_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ##获取聚类数
      if(input$tcga_vali_compMut==1){
        #TCGA
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        if(input$Consensus_N_clust==1){
          N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
        }else{
          if(length(input$multiClusteringAlgorithms)==1){
            #用户指定聚类数
            N.clust <- switch(input$multiClusteringAlgorithms,
                              'iClusterBayes' = input$iClusterBayes_N_clusts,
                              'SNF' = input$SNF_N_clusts,
                              'PINSPlus' = input$PINSPlus_N_clusts,
                              'NEMO' = input$NEMO_N_clusts,
                              'COCA' = input$COCA_N_clusts,
                              'LRAcluster' = input$LRAcluster_N_clusts,
                              'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                              'IntNMF' = input$IntNMF_N_clusts,
                              'CIMLR' = input$CIMLR_N_clusts,
                              'MoCluster' = input$MoCluster_N_clusts)
          }
          if(length(input$multiClusteringAlgorithms)>1){
            N.clust <- input$Consensus_N_clusts #用户指定聚类数
          }
        }
      }else{
        #Validation
        if(input$validation_method_compMut==1){
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        N.clust <- length(unique(integrated_data$clust.res$clust))
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #107.Sample annotations from survival information for oncoprint填写表格
    output$table.compMut_annCol<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$compMut_annCol_number,nrow=3))
      colnames(DF)=1:input$compMut_annCol_number
      rownames(DF)=c("Sample annotation","Continuous or Categorical","Color settings")
      
      default <- c()
      for(d in 1:input$compMut_annCol_number){
        default <- c(default,"")
      }
      DF[1,]=default
      DF[2,]=default
      DF[3,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=180)
    })
    
    #108.compMut_process事件响应按钮
    observeEvent(input$compMut_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$tcga_vali_compMut==1,'Compare mutational frequency among different subtypes on tcga datasets','Compare mutational frequency among different subtypes on validation datasets'),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compMut==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###聚类结果导入
        if(input$validation_method_compMut==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      mut.matrix <- integrated_data$mut.status
      
      progress$set(value = 2)
      #The frequency cutoff for mutation data (Only features that mutated in over than such proportion would be included in testing)
      freq.cutoff <- input$freq_cutoff_compMut
      
      #Statistical method for independence testing
      if(input$test_method_compMut==1){
        test.method <- "fisher"
      }else{
        test.method <- "chisq"
      }
       
      #The correction method for multiple comparison
      p.adj.method <- switch(input$p_adj_method_compMut,
                                '1' = "holm",
                                '2' = "hochberg",
                                '3' = "hommel",
                                '4' = "bonferroni",
                                '5' = "BH",
                                '6' = "BY",
                                '7' = "fdr")
      
      #Transform the '.txt' output file to a '.docx' WORD file or not ('.txt' file will be also kept)
      if(input$doWord_compMut==1){
        doWord <- TRUE
      }else{
        doWord <- FALSE
      }
      
      #The output path for storing the output table
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      res.path <- file.path(user_workdir,"tables")
      
      #The name of the output table
      if(input$tcga_vali_compMut==1){
        tab.name <- input$compMut_TableName_tcga
      }else{
        tab.name <- input$compMut_TableName_validation
      }
      tab.name <- trimws(tab.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(tab.name==""){
        if(input$tcga_vali_compMut==1){
          tab.name <- "Independent_test_between_subtype_and_mutation(TCGA)"
        }else{
          tab.name <- "Independent_test_between_subtype_and_mutation(Validation)"
        }
      }
      
      #Oncoprint
      doPlot <- TRUE
      
      #Perform clustering within each subtype for oncoprint or not
      if(input$innerclust_compMut==1){
        innerclust <- TRUE
      }else{
        innerclust <- FALSE
      }
      
      #The nominal p value cutoff for significant mutations shown in oncoprint
      p.cutoff <- input$p_cutoff_compMut
      
      #The adjusted p value cutoff for significant mutations shown in oncoprint
      p.adj.cutoff <- input$p_adj_cutoff_compMut
      
      #User defined mutation color for oncoprint or not
      if(input$mut_col_compMut==1){
        mut.col <- input$mut_col_value_compMut
        mut.col <- trimws(mut.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }else{
        mut.col <- "#21498D"
      }
      
      #User defined background color for oncoprint or not
      if(input$bg_col_compMut==1){
        bg.col <- input$bg_col_value_compMut
        bg.col <- trimws(bg.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }else{
        bg.col <- "#dcddde"
      }
      
      #Colors for each subtype
      if(input$clustcolor_compMut==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.compMut_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Sample annotations from survival information for oncoprint
      if(input$compMut_annCol==1){
        surv_clin.info <- integrated_data$clin.info #获取生存和临床数据
        # if(!('fustat' %in% colnames(surv_clin.info)) & ('OS' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS')] <- "fustat"
        # } #数据预处理(列名替换)
        # if(!('futime' %in% colnames(surv_clin.info)) & ('OS.time' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS.time')] <- "futime"
        # } #数据预处理(列名替换)
        # if('age_at_initial_pathologic_diagnosis' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='age_at_initial_pathologic_diagnosis')] <- "age"
        # } #数据预处理(列名替换)
        # if('ajcc_pathologic_tumor_stage' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        # } #数据预处理(列名替换)
        # if("pstage" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Discrepancy]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Not Available]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     } 
        #   } #缺失值填补(众数填补多分类变量)
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage I" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IB"){
        #       surv_clin.info$pStage[i] <- "Stage I"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage II" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIB"){
        #       surv_clin.info$pStage[i] <- "Stage II"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage III" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIB"){
        #       surv_clin.info$pStage[i] <- "Stage III"
        #     }
        #   } #类别合并
        # }
        # if("age" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(is.na(surv_clin.info$age[i])==TRUE){
        #       surv_clin.info$age[i] <- median(surv_clin.info$age,na.rm = TRUE)
        #     }
        #   } #缺失值填补(中位数填补连续型变量)
        # }
        ###annCol
        sample_annotation <- data.frame(rbind(hot_to_r(input$table.compMut_annCol)))[1,]
        sample_annotation <- trimws(sample_annotation,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        annCol <- surv_clin.info[,sample_annotation, drop = FALSE]
        ###annColors
        continuous_or_categorical <- data.frame(rbind(hot_to_r(input$table.compMut_annCol)))[2,]
        continuous_or_categorical <- trimws(continuous_or_categorical,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        color_settings <- data.frame(rbind(hot_to_r(input$table.compMut_annCol)))[3,]
        color_settings <- trimws(color_settings,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        ###Process the obtained data
        if(!('fustat' %in% colnames(annCol)) & ('OS' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS')] <- "fustat"
        } #数据预处理(列名替换)
        if(!('futime' %in% colnames(annCol)) & ('OS.time' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS.time')] <- "futime"
        } #数据预处理(列名替换)
        if('age_at_initial_pathologic_diagnosis' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='age_at_initial_pathologic_diagnosis')] <- "age"
        } #数据预处理(列名替换)
        if('ajcc_pathologic_tumor_stage' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        } #数据预处理(列名替换)
        for (j in 1:ncol(annCol)){
          for(i in 1:nrow(annCol)){
            ##NA判定##
            if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))){
              next
            }
            if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Discrepancy]" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Not Available]"){
              annCol[i,j] <- NA
            }
            ##类别合并##
            if(tolower(colnames(annCol)[j])=="pstage"){
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage I" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IB"){
                annCol[i,j] <- "Stage I"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage II" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIB"){
                annCol[i,j] <- "Stage II"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage III" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIB"){
                annCol[i,j] <- "Stage III"
              }
            }
            ##缺失值填补##
            if(continuous_or_categorical[j] == "Categorical"){
              ##众数填补多分类变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- names(which(table(annCol[,j])==max(table(annCol[,j]))))
              }
            }else{
              ##中位数填补连续型变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- median(annCol[,j],na.rm = TRUE)
              }
            }
          }
        }
        annColors <- list()
        for(i in 1:length(sample_annotation)){
          if(continuous_or_categorical[i]=="Continuous"){
            breaks <- c(min(annCol[,sample_annotation[i]]),median(annCol[,sample_annotation[i]]),max(annCol[,sample_annotation[i]]))
            colors <- unlist(strsplit(color_settings[i],";"))
            annColors[[i]] <- circlize::colorRamp2(breaks = breaks, colors = colors)
          }else{
            colors <- unlist(strsplit(color_settings[i],";"))
            names(colors) <- unique(annCol[,sample_annotation[i]])
            annColors[[i]] <- colors
          }
        }
        names(annColors) <- sample_annotation
      }else{
        annCol = NULL
        annColors = NULL
      }
      
      #The width of output figure
      width <- input$width_compMut
      
      #The height of output figure
      height <- input$height_compMut
      
      #The output path for storing the oncoprint
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of output figure
      if(input$tcga_vali_compMut==1){
        fig.name <- input$compMut_FigureName_tcga
      }else{
        fig.name <- input$compMut_FigureName_validation
      }
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        if(input$tcga_vali_compMut==1){
          fig.name <- paste0("oncoprint_for_mutations_with_frequency_over_than_",freq.cutoff*100,"pct(TCGA)")
        }else{
          fig.name <- paste0("oncoprint_for_mutations_with_frequency_over_than_",freq.cutoff*100,"pct(Validation)")
        }
      }
      
      progress$set(value = 3)
      compMutResult <- compMuts(moic.res = moic.res,
                                mut.matrix = mut.matrix,
                                freq.cutoff = freq.cutoff,
                                test.method = test.method,
                                p.adj.method = p.adj.method,
                                doWord = doWord,
                                doPlot = doPlot,
                                innerclust = innerclust,
                                res.path = res.path,
                                tab.name = tab.name,
                                fig.path = fig.path,
                                fig.name = fig.name,
                                annCol = annCol,
                                annColors = annColors,
                                mut.col = mut.col,
                                bg.col = bg.col,
                                p.cutoff = p.cutoff,
                                p.adj.cutoff = p.adj.cutoff,
                                clust.col = clust.col,
                                width = width,
                                height = height)
      
      #保存compMut结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compMut==1){
        save(compMutResult, file = file.path(RData.path,paste0(tolower(project),".tcga.compMut.RData")))
      }else{
        save(compMutResult, file = file.path(RData.path,paste0(tolower(project),".validation.compMut.RData")))
      }
      
      progress$set(value = 4)
      #表格标题
      output$table.compMutTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_compMut==1){
          tags$strong("Table: Independent test between subtype and mutation (TCGA)")
        }else{
          tags$strong("Table: Independent test between subtype and mutation (Validation)")
        }
      })
      
      #表格
      output$table.compMut<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(compMutResult$out),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(compMutResult$out)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Bfrtip',
                        searching= FALSE,
                        bInfo=FALSE, #not to shown 1/N message
                        display=NULL,
                        bPaginate= FALSE, #suppress page
                        pageLength =100,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compMut==1,"Independent_test_between_subtype_and_mutation(TCGA)","Independent_test_between_subtype_and_mutation(Validation)"),tab.name),
                               title = ifelse(input$tcga_vali_compMut==1,"Table: Independent test between subtype and mutation (TCGA)","Table: Independent test between subtype and mutation (Validation)")
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compMut==1,"Independent_test_between_subtype_and_mutation(TCGA)","Independent_test_between_subtype_and_mutation(Validation)"),tab.name),
                               title = ifelse(input$tcga_vali_compMut==1,"Table: Independent test between subtype and mutation (TCGA)","Table: Independent test between subtype and mutation (Validation)")
                          ),
                          list(extend='excel',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compMut==1,"Independent_test_between_subtype_and_mutation(TCGA)","Independent_test_between_subtype_and_mutation(Validation)"),tab.name),
                               title = ifelse(input$tcga_vali_compMut==1,"Table: Independent test between subtype and mutation (TCGA)","Table: Independent test between subtype and mutation (Validation)")
                          ),
                          
                          list(extend='print',
                               filename = ifelse(is.null(tab.name),ifelse(input$tcga_vali_compMut==1,"Independent_test_between_subtype_and_mutation(TCGA)","Independent_test_between_subtype_and_mutation(Validation)"),tab.name),
                               title = ifelse(input$tcga_vali_compMut==1,"Table: Independent test between subtype and mutation (TCGA)","Table: Independent test between subtype and mutation (Validation)")
                          )
                        )
                      ), rownames = FALSE
        )
      }, server=FALSE)
      
      #表格脚注
      output$table.compMutNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$compMut_figure_unit<-renderUI({
        list(column(12,downloadButton("compMut_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("compMut_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$compMut_figure_download<-downloadHandler(
        filename = paste0(ifelse(is.null(fig.name),ifelse(input$tcga_vali_compMut==1,paste0("oncoprint_for_mutations_with_frequency_over_than_",freq.cutoff*100,"pct(TCGA)"),paste0("oncoprint_for_mutations_with_frequency_over_than_",freq.cutoff*100,"pct(Validation)")),fig.name),".pdf"),
        content = function(theFile) {
          pdf(file = theFile, width=width, height=height)
          p <- compMutResult$oncoprint
          draw(p)
          dev.off()
        }
      )
      
      ##页面展示事件
      output$compMut_figure<-renderPlot({

        p <- compMutResult$oncoprint
        draw(p)
      })
      
      ##RData结果下载按钮响应事件
      output$compMuts_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_compMut==1,paste0(tolower(project),".tcga.compMut.RData"),paste0(tolower(project),".validation.compMut.RData")),
        content = function(theFile) {
          save(compMutResult,file=theFile)
        }
      )
      
      #compMut后给予用户的反馈提示
      output$compMut_finish<-renderUI({
        
        if(input$tcga_vali_compMut==1){
          list(br(),column(12,h3(strong("The process of comparing mutational frequency on tcga datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare total mutation burden'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compMuts_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of comparing mutational frequency on validation datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare total mutation burden'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compMuts_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #109.Input variant classification填写表格
    output$table.nonSyn_compTMB<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$nonSyn_compTMB_number,nrow=1))
      colnames(DF)=1:input$nonSyn_compTMB_number
      rownames(DF)="Variant classification"
      
      default <- c()
      for(d in 1:input$nonSyn_compTMB_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=135)
    })
    
    #110.Colors for each subtype填写表格
    output$table.compTMB_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ##获取聚类数
      if(input$tcga_vali_compTMB==1){
        #TCGA
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        
        if(input$Consensus_N_clust==1){
          N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
        }else{
          if(length(input$multiClusteringAlgorithms)==1){
            #用户指定聚类数
            N.clust <- switch(input$multiClusteringAlgorithms,
                              'iClusterBayes' = input$iClusterBayes_N_clusts,
                              'SNF' = input$SNF_N_clusts,
                              'PINSPlus' = input$PINSPlus_N_clusts,
                              'NEMO' = input$NEMO_N_clusts,
                              'COCA' = input$COCA_N_clusts,
                              'LRAcluster' = input$LRAcluster_N_clusts,
                              'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                              'IntNMF' = input$IntNMF_N_clusts,
                              'CIMLR' = input$CIMLR_N_clusts,
                              'MoCluster' = input$MoCluster_N_clusts)
          }
          if(length(input$multiClusteringAlgorithms)>1){
            N.clust <- input$Consensus_N_clusts #用户指定聚类数
          }
        }
      }else{
        #Validation
        if(input$validation_method_compTMB==1){
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        N.clust <- length(unique(integrated_data$clust.res$clust))
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #111.compTMB_process事件响应按钮
    observeEvent(input$compTMB_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$tcga_vali_compTMB==1,'Compare total mutation burden among different subtypes (TCGA)','Compare total mutation burden among different subtypes (Validation)'),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compTMB==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###聚类结果导入
        if(input$validation_method_compTMB==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      maf <- integrated_data$maf
      
      progress$set(value = 2)
      #Whether remove repeated variants in a particuar sample, mapped to multiple transcripts of same gene
      if(input$rmDup_compTMB==1){
        rmDup <- TRUE
      }else{
        rmDup <- FALSE
      }
      
      #Remove possible FLAGS
      if(input$rmFLAGS_compTMB==1){
        rmFLAGS <- TRUE
      }else{
        rmFLAGS <- FALSE
      }
      
      #Whether user defined a list of variant classifications that should be considered as non-synonymous and the rest will be considered synonymous
      if(input$nonSyn_compTMB==1){
        nonSyn <- data.frame(rbind(hot_to_r(input$table.nonSyn_compTMB)))[1,]
        nonSyn <- trimws(nonSyn,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符 
      }else{
        nonSyn <- NULL
      }
      
      #The estimation of exome size
      exome.size <- input$exome_size_compTMB
      
      #The method for statistical testing
      if(input$test_method_compTMB==1){
        test.method <- "parametric"
      }else{
        test.method <- "nonparametric"
      }
      
      #Show the sample size within each subtype at the top of the figure or not
      if(input$show_size_compTMB==1){
        show.size <- TRUE
      }else{
        show.size <- FALSE
      }
      
      #Colors for each subtype
      if(input$clustcolor_compTMB==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.compTMB_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #The width of boxviolin plot
      width <- input$width_compTMB
      
      #The height of boxviolin plot
      height <- input$height_compTMB
      
      #The output path for storing the boxviolin plot
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of output figure
      if(input$tcga_vali_compTMB==1){
        fig.name <- input$compTMB_FigureName_tcga
      }else{
        fig.name <- input$compTMB_FigureName_validation
      }
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        if(input$tcga_vali_compTMB==1){
          fig.name <- "distribution_of_TMB_and_titv(TCGA)"
        }else{
          fig.name <- "distribution_of_TMB_and_titv(Validation)"
        }
      }
      
      progress$set(value = 3)
      compTMBResult <- compTMBs(moic.res = moic.res,
                                maf = maf,
                                rmDup = rmDup,
                                rmFLAGS = rmFLAGS,
                                nonSyn = nonSyn,
                                exome.size = exome.size,
                                clust.col = clust.col,
                                test.method = test.method,
                                show.size = show.size,
                                fig.path = fig.path,
                                fig.name = fig.name,
                                width = width,
                                height = height)
      
      #保存compTMB结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compTMB==1){
        save(compTMBResult, file = file.path(RData.path,paste0(tolower(project),".tcga.compTMB.RData")))
      }else{
        save(compTMBResult, file = file.path(RData.path,paste0(tolower(project),".validation.compTMB.RData")))
      }
      
      progress$set(value = 4)
      ###展示TMB.dat###
      #表格标题
      output$table.compTMBdatTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_compTMB==1){
          tags$strong("Table: Comparison of TMB among identified subtypes (TCGA)")
        }else{
          tags$strong("Table: Comparison of TMB among identified subtypes (Validation)")
        }
      })
      
      #表格
      output$table.compTMBdat<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(compTMBResult$TMB.dat),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(compTMBResult$TMB.dat)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$tcga_vali_compTMB==1,"Comparison_of_TMB_among_identified_subtypes(TCGA)","Comparison_of_TMB_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compTMB==1,"Table: Comparison of TMB among identified subtypes (TCGA)","Table: Comparison of TMB among identified subtypes (Validation)")
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$tcga_vali_compTMB==1,"Comparison_of_TMB_among_identified_subtypes(TCGA)","Comparison_of_TMB_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compTMB==1,"Table: Comparison of TMB among identified subtypes (TCGA)","Table: Comparison of TMB among identified subtypes (Validation)")
                          ),
                          list(extend='excel',
                               filename = ifelse(input$tcga_vali_compTMB==1,"Comparison_of_TMB_among_identified_subtypes(TCGA)","Comparison_of_TMB_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compTMB==1,"Table: Comparison of TMB among identified subtypes (TCGA)","Table: Comparison of TMB among identified subtypes (Validation)")
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$tcga_vali_compTMB==1,"Comparison_of_TMB_among_identified_subtypes(TCGA)","Comparison_of_TMB_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compTMB==1,"Table: Comparison of TMB among identified subtypes (TCGA)","Table: Comparison of TMB among identified subtypes (Validation)")
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      #表格脚注
      output$table.compTMBdatNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$compTMB_figure_unit<-renderUI({
        list(column(12,downloadButton("compTMB_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("compTMB_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$compTMB_figure_download<-downloadHandler(
        
        filename = paste0(ifelse(is.null(fig.name),ifelse(input$tcga_vali_compTMB==1,"distribution_of_TMB_and_titv(TCGA)","distribution_of_TMB_and_titv(Validation)"),fig.name),".pdf"),
        content = function(theFile) {
          
          ##获取项目名称
          project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          
          ##导入数据
          workdir <- getwd() #获取当前工作路径
          user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
          RData.path <- file.path(user_workdir,"RData")
          if(input$tcga_vali_compTMB==1){
            #TCGA
            ###聚类结果导入
            ####聚类方法数为1
            if(length(input$multiClusteringAlgorithms)==1){
              moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
            }
            ####聚类方法数大于1
            if(length(input$multiClusteringAlgorithms)>1){
              cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
              moic.res <- cmoic
            }
            ###多组学数据导入
            integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
          }else{
            #Validation
            ###聚类结果导入
            if(input$validation_method_compTMB==1){
              moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
            }else{
              moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
            }
            ###多组学数据导入
            # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
            integrated_data <- get(load(file.path(user_workdir,"RData.path",paste0(tolower(project),".validation.getElites.RData"))))
          }
          
          maf <- integrated_data$maf
          
          #Whether remove repeated variants in a particuar sample, mapped to multiple transcripts of same gene
          if(input$rmDup_compTMB==1){
            rmDup <- TRUE
          }else{
            rmDup <- FALSE
          }
          
          #Remove possible FLAGS
          if(input$rmFLAGS_compTMB==1){
            rmFLAGS <- TRUE
          }else{
            rmFLAGS <- FALSE
          }
          
          #Whether user defined a list of variant classifications that should be considered as non-synonymous and the rest will be considered synonymous
          if(input$nonSyn_compTMB==1){
            nonSyn <- data.frame(rbind(hot_to_r(input$table.nonSyn_compTMB)))[1,]
            nonSyn <- trimws(nonSyn,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符 
          }else{
            nonSyn <- NULL
          }
          
          #The estimation of exome size
          exome.size <- input$exome_size_compTMB
          
          #The method for statistical testing
          if(input$test_method_compTMB==1){
            test.method <- "parametric"
          }else{
            test.method <- "nonparametric"
          }
          
          #Show the sample size within each subtype at the top of the figure or not
          if(input$show_size_compTMB==1){
            show.size <- TRUE
          }else{
            show.size <- FALSE
          }
          
          #Colors for each subtype
          if(input$clustcolor_compTMB==1){
            clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
          }else{
            clust.col <- data.frame(rbind(hot_to_r(input$table.compTMB_clustcolor)))[1,]
            clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          }
          
          #The width of boxviolin plot
          width <- input$width_compTMB
          
          #The height of boxviolin plot
          height <- input$height_compTMB
          
          #The output path for storing the boxviolin plot
          workdir <- getwd() #获取当前工作路径
          user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
          fig.path <- file.path(user_workdir,"figures")
          
          #The name of output figure
          if(input$tcga_vali_compTMB==1){
            fig.name <- input$compTMB_FigureName_tcga
          }else{
            fig.name <- input$compTMB_FigureName_validation
          }
          fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          if(fig.name==""){
            fig.name <- NULL
          }
            
          label <- c("Tumor_Sample_Barcode",
                     "Hugo_Symbol",
                     "Chromosome",
                     "Start_Position",
                     "End_Position",
                     "Variant_Classification",
                     "Variant_Type",
                     "Reference_Allele",
                     "Tumor_Seq_Allele1",
                     "Tumor_Seq_Allele2")
          
          maf <- as.data.frame(maf)
          
          # check arguments
          if(!all(is.element(label, colnames(maf)))) {
            stop(paste0("maf data must have the following columns: \n ", paste(label, collapse = "\n "),"\n\nmissing required fields from maf: ", paste(setdiff(label, colnames(maf)), collapse = "\n")))
          }
          
          if(!is.element(test.method, c("nonparametric","parametric"))) {
            stop("test.method can be one of nonparametric or parametric.")
          }
          
          # check data
          comsam <- intersect(moic.res$clust.res$samID, unique(maf$Tumor_Sample_Barcode))
          if(length(comsam) == nrow(moic.res$clust.res)) {
            message("--all samples matched.")
          } else {
            message(paste0("--",(nrow(moic.res$clust.res)-length(comsam))," samples mismatched from current subtypes."))
          }
          
          maf <- maf[which(maf$Tumor_Sample_Barcode %in% comsam),]
          clust.res <- moic.res$clust.res[comsam, , drop = FALSE]
          n.moic <- length(unique(clust.res$clust))
          colvec <- clust.col[1:n.moic]
          names(colvec) <- paste0("CS",unique(clust.res$clust))
          col.titv <- c("#E64B35CC", "#4DBBD5CC", "#00A087CC", "#3C5488CC", "#F39B7FCC", "#8491B4CC")
          names(col.titv) <- c("C>T", "T>C", "C>A", "C>G", "T>A", "T>G")
          
          if(rmFLAGS) {
            FLAGS <- c("TTN",
                       "MUC16",
                       "OBSCN",
                       "AHNAK2",
                       "SYNE1",
                       "FLG",
                       "MUC5B",
                       "DNAH17",
                       "PLEC",
                       "DST",
                       "SYNE2",
                       "NEB",
                       "HSPG2",
                       "LAMA5",
                       "AHNAK",
                       "HMCN1",
                       "USH2A",
                       "DNAH11",
                       "MACF1",
                       "MUC17",
                       "DNAH5",
                       "GPR98",
                       "FAT1",
                       "PKD1",
                       "MDN1",
                       "RNF213",
                       "RYR1",
                       "DNAH2",
                       "DNAH3",
                       "DNAH8",
                       "DNAH1",
                       "DNAH9",
                       "ABCA13",
                       "APOB",
                       "SRRM2",
                       "CUBN",
                       "SPTBN5",
                       "PKHD1",
                       "LRP2",
                       "FBN3",
                       "CDH23",
                       "DNAH10",
                       "FAT4",
                       "RYR3",
                       "PKHD1L1",
                       "FAT2",
                       "CSMD1",
                       "PCNT",
                       "COL6A3",
                       "FRAS1",
                       "FCGBP",
                       "DNAH7",
                       "RP1L1",
                       "PCLO",
                       "ZFHX3",
                       "COL7A1",
                       "LRP1B",
                       "FAT3",
                       "EPPK1",
                       "VPS13C",
                       "HRNR",
                       "MKI67",
                       "MYO15A",
                       "STAB1",
                       "ZAN",
                       "UBR4",
                       "VPS13B",
                       "LAMA1",
                       "XIRP2",
                       "BSN",
                       "KMT2C",
                       "ALMS1",
                       "CELSR1",
                       "TG",
                       "LAMA3",
                       "DYNC2H1",
                       "KMT2D",
                       "BRCA2",
                       "CMYA5",
                       "SACS",
                       "STAB2",
                       "AKAP13",
                       "UTRN",
                       "VWF",
                       "VPS13D",
                       "ANK3",
                       "FREM2",
                       "PKD1L1",
                       "LAMA2",
                       "ABCA7",
                       "LRP1",
                       "ASPM",
                       "MYOM2",
                       "PDE4DIP",
                       "TACC2",
                       "MUC2",
                       "TEP1",
                       "HELZ2",
                       "HERC2",
                       "ABCA4")
            
            if(sum(is.element(FLAGS, unique(maf$Hugo_Symbol))) > 0) {
              maf.FLAGS <- maf[which(maf$Hugo_Symbol %in% FLAGS),]
              maf.rmFLAGS <- maf[-which(maf$Hugo_Symbol %in% FLAGS),]
              count.flags <- maf.FLAGS %>% group_by(Hugo_Symbol) %>% summarise(count = dplyr::n())
              message("--remove possible FLAGS as below:")
              head(count.flags)
              
              maf.ob <- read_maf(maf                      = maf.rmFLAGS,
                                 removeDuplicatedVariants = rmDup,
                                 vc_nonSyn                = nonSyn)
            } else {
              maf.ob <- read_maf(maf                      = maf,
                                 removeDuplicatedVariants = rmDup,
                                 vc_nonSyn                = nonSyn)
            }
          }   else {
            maf.ob <- read_maf(maf                      = maf,
                               removeDuplicatedVariants = rmDup,
                               vc_nonSyn                = nonSyn)
          }
          
          # classifies Single Nucleotide Variants into Transitions and Transversions
          titv.dat <- maftools::titv(maf = maf.ob, plot = FALSE, useSyn = FALSE)$fraction.contribution %>%
            as.data.frame() %>%
            mutate(Subtype = paste0("CS",clust.res[as.character(.$Tumor_Sample_Barcode),"clust"]))
          titv.dat.backup <- titv.dat
          titv.dat <- split(titv.dat, f = titv.dat$Subtype)
          
          # extract silent and nonsilent mutations
          maf.silent    <- as.data.frame(maf.ob@maf.silent)
          maf.nonsilent <- as.data.frame(maf.ob@data)
          
          # extract total mutation burden
          TMB.dat <- as.data.frame(maf.ob@variants.per.sample)
          TMB.dat <- data.frame(samID            = as.character(TMB.dat$Tumor_Sample_Barcode),
                                variants         = as.character(TMB.dat$Variants),
                                TMB              = as.numeric(TMB.dat$Variants)/exome.size,
                                log10TMB         = log10(as.numeric(TMB.dat$Variants)/exome.size + 1),
                                Subtype          = paste0("CS", clust.res[as.character(TMB.dat$Tumor_Sample_Barcode), "clust"]),
                                stringsAsFactors = FALSE)
          TMB.dat <- TMB.dat[order(TMB.dat$Subtype), , drop = FALSE]
          TMB.med <- TMB.dat %>% group_by(Subtype) %>% summarize(median = median(TMB)) %>% as.data.frame()
          TMB.dat <- TMB.dat[order(TMB.dat$Subtype, TMB.dat$TMB, decreasing = FALSE), ]
          
          # sample order in bottom panel
          sampleorder <- TMB.dat %>% split(.$Subtype) %>% lapply("[[", 1) %>% lapply(., as.character)
          TMB.plot <- split(TMB.dat, as.factor(TMB.dat$Subtype))
          TMB.plot <- lapply(seq_len(length(TMB.plot)), function(i) {
            x = TMB.plot[[i]]
            x = data.frame(x = seq(i - 1, i, length.out = nrow(x)),
                           TMB = x[, "TMB"],
                           Subtype = x[, "Subtype"])
            return(x)
          })
          names(TMB.plot) <- levels(as.factor(TMB.dat$Subtype))
          
          # prepare titv data
          titv.dat2 <- lapply(TMB.med$Subtype, function(x){
            tmp <- titv.dat[[x]]
            tmp <- tmp[match(sampleorder[[x]], as.character(tmp$Tumor_Sample_Barcode)), ]
            return(tmp)
            if (!all(tmp$Tumor_Sample_Barcode == sampleorder[[x]])){
              stop("inconsistent sample order...")
            }
          })
          
          names(titv.dat2) <- TMB.med$Subtype
          titv.dat2 <- lapply(titv.dat2, function(x){
            x <- as.data.frame(x)
            rownames(x) <- x$Tumor_Sample_Barcode
            x <- x[, setdiff(colnames(x), c("Tumor_Sample_Barcode","Subtype"))]
            x <- t(x)
            #delete samples without mutation
            if (length(which(colSums(x) == 0)) > 0) {
              x = x[, -which(colSums(x) == 0), drop = FALSE]
            }
            return(x)
          })
          
          TMB.med$Median_Mutations_log10 <- log10(TMB.med$median + 1)
          
          # statistical testing
          if(n.moic == 2 & test.method == "nonparametric") {
            statistic <- "wilcox.test"
            TMB.test  <- wilcox.test(TMB.dat$log10TMB ~ TMB.dat$Subtype)$p.value
            cat(paste0("Wilcoxon rank sum test p value = ", formatC(TMB.test, format = "e", digits = 2)))
          }
          
          if(n.moic == 2 & test.method == "parametric") {
            statistic <- "t.test"
            TMB.test  <- t.test(TMB.dat$log10TMB ~ TMB.dat$Subtype)$p.value
            cat(paste0("Student's t test p value = ", formatC(TMB.test, format = "e", digits = 2)))
          }
          
          if(n.moic > 2 & test.method == "nonparametric") {
            statistic <- "kruskal.test"
            TMB.test  <- kruskal.test(TMB.dat$log10TMB ~ TMB.dat$Subtype)$p.value
            pairwise.TMB.test <- pairwise.wilcox.test(TMB.dat$log10TMB,TMB.dat$Subtype,p.adjust.method = "BH")
            # pairwise.TMB.test <- dunnTest(log10TMB ~ as.factor(Subtype),
            #                               data = TMB.dat,
            #                               method = "bh")
            cat(paste0("Kruskal-Wallis rank sum test p value = ", formatC(TMB.test, format = "e", digits = 2),"\npost-hoc pairwise wilcoxon rank sum test with Benjamini-Hochberg adjustment presents below:\n"))
            print(formatC(pairwise.TMB.test$p.value, format = "e", digits = 2))
          }
          
          if(n.moic > 2 & test.method == "parametric") {
            statistic <- "anova"
            TMB.test  <- summary(aov(TMB.dat$log10TMB ~ TMB.dat$Subtype))[[1]][["Pr(>F)"]][1]
            pairwise.TMB.test <- pairwise.t.test(TMB.dat$log10TMB,TMB.dat$Subtype,p.adjust.method = "BH")
            cat(paste0("One-way anova test p value = ", formatC(TMB.test, format = "e", digits = 2),"\npost-hoc pairwise Student's t test with Benjamini-Hochberg adjustment presents below:\n"))
            print(formatC(pairwise.TMB.test$p.value, format = "e", digits = 2))
          }
          
          # start illustration
          if(is.null(fig.name)) {
            outFig <- "distribution of TMB and titv.pdf"
          } else {
            outFig <- paste0(fig.name,".pdf")
          }
          
          pdf(file = theFile, width=width, height=height)
          # base layout
          n1 <- seq(from = 0.105, to = 0.97-(0.97-0.05)*0.04, length.out = n.moic + 1)
          n2 <- n1[2:length(n1)]
          n  <- data.frame(n1 = n1[1:n.moic], n2 = n2, n3 = 0.05, n4 = 0.2) %>%
            rbind(c(0.05, 0.97, 0.25, 1), ., c(0, 0.1, 0.05, 0.25)) %>% as.matrix()
          
          opar <- par(no.readonly = TRUE)
          invisible(suppressWarnings(split.screen(n, erase = TRUE)))
          screen(1, new = TRUE)
          par(xpd = TRUE, mar = c(3, 1, 2, 0), oma = c(0, 0, 0, 0),bty = "o", mgp = c(2, 0.5, 0), tcl=-.25)
          y_lims = range(log10(unlist(lapply(TMB.plot, function(x) max(x[, "TMB"], na.rm = TRUE))) + 1))
          y_lims[1] = 0
          y_lims[2] = ceiling(max(y_lims))
          y_at = y_lims[1]:y_lims[2]
          x_top_label <- as.numeric(unlist(lapply(TMB.plot, nrow)))
          plot(NA, NA, xlim = c(0, length(TMB.plot)), ylim = y_lims,
               xlab = NA, ylab = NA, xaxt="n", yaxt = "n", xaxs = "r", yaxs = "r")
          rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col = "#EAE9E9", border = FALSE)
          grid(col = "white", lty = 1, lwd = 1.5) # add grid
          
          par(new = TRUE, bty="o")
          plot(NA, NA,
               col = "white",
               xlim = c(0, length(TMB.plot)), ylim = y_lims,
               xlab = "", ylab = "",
               xaxt = "n", yaxt = "n")
          
          text(x = length(TMB.plot)/2, y = y_lims[2] - 0.1, cex = 1.2,
               label = paste0(statistic, " p = ", formatC(TMB.test, digits = 1, format = "e")))
          
          # add median TMB
          invisible(lapply(seq_len(nrow(TMB.med)), function(i) {
            segments(x0  = i - 1,
                     x1  = i,
                     y0  = TMB.med[i, "Median_Mutations_log10"],
                     y1  = TMB.med[i, "Median_Mutations_log10"],
                     col = "#2b2d42",
                     lwd = 2)}))
          
          # add scatter TMB
          invisible(lapply(seq_len(length(TMB.plot)), function(i){
            tmp = TMB.plot[[i]]
            points(tmp$x, log10(tmp$TMB + 1), pch = 16, cex = 0.5, col = colvec[i])
          }))
          
          # modify axis
          axis(side = 1, at = seq(0.5, length(TMB.plot) - 0.5, 1), labels = names(TMB.plot),
               las = 1, tick = TRUE, line = 0)
          axis(side = 2, at = y_at, las = 2, line = 0, tick = TRUE, labels = y_at)
          if(show.size) {
            axis(side = 3, at = seq(0.5, length(TMB.plot) - 0.5, 1), labels = paste0("n = ",x_top_label),
                 tick = TRUE, line = 0, cex.axis = 0.9)
          }
          mtext(text = bquote("log"[10]~"(TMB + 1)"), side = 2, line = 1.15, cex = 1.1)
          
          # add titv
          invisible(lapply(seq_len(length(titv.dat2)), function(i){
            screen(i + 1, new = TRUE)
            par(xpd = TRUE, mar = c(0, 0, 0, 0), oma = c(0, 0, 0, 0), bty = "o")
            tmp <- titv.dat2[[i]]
            barplot(tmp, col = col.titv[rownames(tmp)], names.arg = rep("", ncol(tmp)),
                    xaxs = "i", yaxs = "i",
                    axes = FALSE, space = 0, border = NA, lwd = 1.2)
            box()
          }))
          
          # add legend
          screen(n.moic + 2, new = TRUE)
          par(xpd = TRUE, mar = c(0, 0, 0, 0), oma = c(0, 0, 0, 0), bty = "n")
          plot(NA, NA, xlim = c(0, 1), ylim = c(0, 1), axes = FALSE, xlab = NA, ylab = NA)
          legend("center",
                 fill   = col.titv,
                 legend = names(col.titv),
                 border = NA,
                 bty    = "n",
                 cex    = 0.8)
          close.screen(all.screens = TRUE)
          dev.off()
        }
      )
      
      ##页面展示事件
      output$compTMB_figure<-renderPlot({
        
        ##获取项目名称
        project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        
        ##导入数据
        workdir <- getwd() #获取当前工作路径
        user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
        RData.path <- file.path(user_workdir,"RData")
        if(input$tcga_vali_compTMB==1){
          #TCGA
          ###聚类结果导入
          ####聚类方法数为1
          if(length(input$multiClusteringAlgorithms)==1){
            moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
          }
          ####聚类方法数大于1
          if(length(input$multiClusteringAlgorithms)>1){
            cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
            moic.res <- cmoic
          }
          ###多组学数据导入
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        }else{
          #Validation
          ###聚类结果导入
          if(input$validation_method_compTMB==1){
            moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
          }else{
            moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
          }
          ###多组学数据导入
          # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
          integrated_data <- get(load(file.path(user_workdir,"RData.path",paste0(tolower(project),".validation.getElites.RData"))))
        }
        
        maf <- integrated_data$maf
        
        #Whether remove repeated variants in a particuar sample, mapped to multiple transcripts of same gene
        if(input$rmDup_compTMB==1){
          rmDup <- TRUE
        }else{
          rmDup <- FALSE
        }
        
        #Remove possible FLAGS
        if(input$rmFLAGS_compTMB==1){
          rmFLAGS <- TRUE
        }else{
          rmFLAGS <- FALSE
        }
        
        #Whether user defined a list of variant classifications that should be considered as non-synonymous and the rest will be considered synonymous
        if(input$nonSyn_compTMB==1){
          nonSyn <- data.frame(rbind(hot_to_r(input$table.nonSyn_compTMB)))[1,]
          nonSyn <- trimws(nonSyn,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符 
        }else{
          nonSyn <- NULL
        }
        
        #The estimation of exome size
        exome.size <- input$exome_size_compTMB
        
        #The method for statistical testing
        if(input$test_method_compTMB==1){
          test.method <- "parametric"
        }else{
          test.method <- "nonparametric"
        }
        
        #Show the sample size within each subtype at the top of the figure or not
        if(input$show_size_compTMB==1){
          show.size <- TRUE
        }else{
          show.size <- FALSE
        }
        
        #Colors for each subtype
        if(input$clustcolor_compTMB==1){
          clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
        }else{
          clust.col <- data.frame(rbind(hot_to_r(input$table.compTMB_clustcolor)))[1,]
          clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        }
        
        #The width of boxviolin plot
        width <- input$width_compTMB
        
        #The height of boxviolin plot
        height <- input$height_compTMB
        
        #The output path for storing the boxviolin plot
        workdir <- getwd() #获取当前工作路径
        user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
        fig.path <- file.path(user_workdir,"figures")
        
        #The name of output figure
        if(input$tcga_vali_compTMB==1){
          fig.name <- input$compTMB_FigureName_tcga
        }else{
          fig.name <- input$compTMB_FigureName_validation
        }
        fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        if(fig.name==""){
          fig.name <- NULL
        }
        
        label <- c("Tumor_Sample_Barcode",
                   "Hugo_Symbol",
                   "Chromosome",
                   "Start_Position",
                   "End_Position",
                   "Variant_Classification",
                   "Variant_Type",
                   "Reference_Allele",
                   "Tumor_Seq_Allele1",
                   "Tumor_Seq_Allele2")
        
        maf <- as.data.frame(maf)
        
        # check arguments
        if(!all(is.element(label, colnames(maf)))) {
          stop(paste0("maf data must have the following columns: \n ", paste(label, collapse = "\n "),"\n\nmissing required fields from maf: ", paste(setdiff(label, colnames(maf)), collapse = "\n")))
        }
        
        if(!is.element(test.method, c("nonparametric","parametric"))) {
          stop("test.method can be one of nonparametric or parametric.")
        }
        
        # check data
        comsam <- intersect(moic.res$clust.res$samID, unique(maf$Tumor_Sample_Barcode))
        if(length(comsam) == nrow(moic.res$clust.res)) {
          message("--all samples matched.")
        } else {
          message(paste0("--",(nrow(moic.res$clust.res)-length(comsam))," samples mismatched from current subtypes."))
        }
        
        maf <- maf[which(maf$Tumor_Sample_Barcode %in% comsam),]
        clust.res <- moic.res$clust.res[comsam, , drop = FALSE]
        n.moic <- length(unique(clust.res$clust))
        colvec <- clust.col[1:n.moic]
        names(colvec) <- paste0("CS",unique(clust.res$clust))
        col.titv <- c("#E64B35CC", "#4DBBD5CC", "#00A087CC", "#3C5488CC", "#F39B7FCC", "#8491B4CC")
        names(col.titv) <- c("C>T", "T>C", "C>A", "C>G", "T>A", "T>G")
        
        if(rmFLAGS) {
          FLAGS <- c("TTN",
                     "MUC16",
                     "OBSCN",
                     "AHNAK2",
                     "SYNE1",
                     "FLG",
                     "MUC5B",
                     "DNAH17",
                     "PLEC",
                     "DST",
                     "SYNE2",
                     "NEB",
                     "HSPG2",
                     "LAMA5",
                     "AHNAK",
                     "HMCN1",
                     "USH2A",
                     "DNAH11",
                     "MACF1",
                     "MUC17",
                     "DNAH5",
                     "GPR98",
                     "FAT1",
                     "PKD1",
                     "MDN1",
                     "RNF213",
                     "RYR1",
                     "DNAH2",
                     "DNAH3",
                     "DNAH8",
                     "DNAH1",
                     "DNAH9",
                     "ABCA13",
                     "APOB",
                     "SRRM2",
                     "CUBN",
                     "SPTBN5",
                     "PKHD1",
                     "LRP2",
                     "FBN3",
                     "CDH23",
                     "DNAH10",
                     "FAT4",
                     "RYR3",
                     "PKHD1L1",
                     "FAT2",
                     "CSMD1",
                     "PCNT",
                     "COL6A3",
                     "FRAS1",
                     "FCGBP",
                     "DNAH7",
                     "RP1L1",
                     "PCLO",
                     "ZFHX3",
                     "COL7A1",
                     "LRP1B",
                     "FAT3",
                     "EPPK1",
                     "VPS13C",
                     "HRNR",
                     "MKI67",
                     "MYO15A",
                     "STAB1",
                     "ZAN",
                     "UBR4",
                     "VPS13B",
                     "LAMA1",
                     "XIRP2",
                     "BSN",
                     "KMT2C",
                     "ALMS1",
                     "CELSR1",
                     "TG",
                     "LAMA3",
                     "DYNC2H1",
                     "KMT2D",
                     "BRCA2",
                     "CMYA5",
                     "SACS",
                     "STAB2",
                     "AKAP13",
                     "UTRN",
                     "VWF",
                     "VPS13D",
                     "ANK3",
                     "FREM2",
                     "PKD1L1",
                     "LAMA2",
                     "ABCA7",
                     "LRP1",
                     "ASPM",
                     "MYOM2",
                     "PDE4DIP",
                     "TACC2",
                     "MUC2",
                     "TEP1",
                     "HELZ2",
                     "HERC2",
                     "ABCA4")
          
          if(sum(is.element(FLAGS, unique(maf$Hugo_Symbol))) > 0) {
            maf.FLAGS <- maf[which(maf$Hugo_Symbol %in% FLAGS),]
            maf.rmFLAGS <- maf[-which(maf$Hugo_Symbol %in% FLAGS),]
            count.flags <- maf.FLAGS %>% group_by(Hugo_Symbol) %>% summarise(count = dplyr::n())
            message("--remove possible FLAGS as below:")
            head(count.flags)
            
            maf.ob <- read_maf(maf                      = maf.rmFLAGS,
                               removeDuplicatedVariants = rmDup,
                               vc_nonSyn                = nonSyn)
          } else {
            maf.ob <- read_maf(maf                      = maf,
                               removeDuplicatedVariants = rmDup,
                               vc_nonSyn                = nonSyn)
          }
        }   else {
          maf.ob <- read_maf(maf                      = maf,
                             removeDuplicatedVariants = rmDup,
                             vc_nonSyn                = nonSyn)
        }
        
        # classifies Single Nucleotide Variants into Transitions and Transversions
        titv.dat <- maftools::titv(maf = maf.ob, plot = FALSE, useSyn = FALSE)$fraction.contribution %>%
          as.data.frame() %>%
          mutate(Subtype = paste0("CS",clust.res[as.character(.$Tumor_Sample_Barcode),"clust"]))
        titv.dat.backup <- titv.dat
        titv.dat <- split(titv.dat, f = titv.dat$Subtype)
        
        # extract silent and nonsilent mutations
        maf.silent    <- as.data.frame(maf.ob@maf.silent)
        maf.nonsilent <- as.data.frame(maf.ob@data)
        
        # extract total mutation burden
        TMB.dat <- as.data.frame(maf.ob@variants.per.sample)
        TMB.dat <- data.frame(samID            = as.character(TMB.dat$Tumor_Sample_Barcode),
                              variants         = as.character(TMB.dat$Variants),
                              TMB              = as.numeric(TMB.dat$Variants)/exome.size,
                              log10TMB         = log10(as.numeric(TMB.dat$Variants)/exome.size + 1),
                              Subtype          = paste0("CS", clust.res[as.character(TMB.dat$Tumor_Sample_Barcode), "clust"]),
                              stringsAsFactors = FALSE)
        TMB.dat <- TMB.dat[order(TMB.dat$Subtype), , drop = FALSE]
        TMB.med <- TMB.dat %>% group_by(Subtype) %>% summarize(median = median(TMB)) %>% as.data.frame()
        TMB.dat <- TMB.dat[order(TMB.dat$Subtype, TMB.dat$TMB, decreasing = FALSE), ]
        
        # sample order in bottom panel
        sampleorder <- TMB.dat %>% split(.$Subtype) %>% lapply("[[", 1) %>% lapply(., as.character)
        TMB.plot <- split(TMB.dat, as.factor(TMB.dat$Subtype))
        TMB.plot <- lapply(seq_len(length(TMB.plot)), function(i) {
          x = TMB.plot[[i]]
          x = data.frame(x = seq(i - 1, i, length.out = nrow(x)),
                         TMB = x[, "TMB"],
                         Subtype = x[, "Subtype"])
          return(x)
        })
        names(TMB.plot) <- levels(as.factor(TMB.dat$Subtype))
        
        # prepare titv data
        titv.dat2 <- lapply(TMB.med$Subtype, function(x){
          tmp <- titv.dat[[x]]
          tmp <- tmp[match(sampleorder[[x]], as.character(tmp$Tumor_Sample_Barcode)), ]
          return(tmp)
          if (!all(tmp$Tumor_Sample_Barcode == sampleorder[[x]])){
            stop("inconsistent sample order...")
          }
        })
        
        names(titv.dat2) <- TMB.med$Subtype
        titv.dat2 <- lapply(titv.dat2, function(x){
          x <- as.data.frame(x)
          rownames(x) <- x$Tumor_Sample_Barcode
          x <- x[, setdiff(colnames(x), c("Tumor_Sample_Barcode","Subtype"))]
          x <- t(x)
          #delete samples without mutation
          if (length(which(colSums(x) == 0)) > 0) {
            x = x[, -which(colSums(x) == 0), drop = FALSE]
          }
          return(x)
        })
        
        TMB.med$Median_Mutations_log10 <- log10(TMB.med$median + 1)
        
        # statistical testing
        if(n.moic == 2 & test.method == "nonparametric") {
          statistic <- "wilcox.test"
          TMB.test  <- wilcox.test(TMB.dat$log10TMB ~ TMB.dat$Subtype)$p.value
          cat(paste0("Wilcoxon rank sum test p value = ", formatC(TMB.test, format = "e", digits = 2)))
        }
        
        if(n.moic == 2 & test.method == "parametric") {
          statistic <- "t.test"
          TMB.test  <- t.test(TMB.dat$log10TMB ~ TMB.dat$Subtype)$p.value
          cat(paste0("Student's t test p value = ", formatC(TMB.test, format = "e", digits = 2)))
        }
        
        if(n.moic > 2 & test.method == "nonparametric") {
          statistic <- "kruskal.test"
          TMB.test  <- kruskal.test(TMB.dat$log10TMB ~ TMB.dat$Subtype)$p.value
          pairwise.TMB.test <- pairwise.wilcox.test(TMB.dat$log10TMB,TMB.dat$Subtype,p.adjust.method = "BH")
          # pairwise.TMB.test <- dunnTest(log10TMB ~ as.factor(Subtype),
          #                               data = TMB.dat,
          #                               method = "bh")
          cat(paste0("Kruskal-Wallis rank sum test p value = ", formatC(TMB.test, format = "e", digits = 2),"\npost-hoc pairwise wilcoxon rank sum test with Benjamini-Hochberg adjustment presents below:\n"))
          print(formatC(pairwise.TMB.test$p.value, format = "e", digits = 2))
        }
        
        if(n.moic > 2 & test.method == "parametric") {
          statistic <- "anova"
          TMB.test  <- summary(aov(TMB.dat$log10TMB ~ TMB.dat$Subtype))[[1]][["Pr(>F)"]][1]
          pairwise.TMB.test <- pairwise.t.test(TMB.dat$log10TMB,TMB.dat$Subtype,p.adjust.method = "BH")
          cat(paste0("One-way anova test p value = ", formatC(TMB.test, format = "e", digits = 2),"\npost-hoc pairwise Student's t test with Benjamini-Hochberg adjustment presents below:\n"))
          print(formatC(pairwise.TMB.test$p.value, format = "e", digits = 2))
        }
        
        # start illustration
        if(is.null(fig.name)) {
          outFig <- "distribution of TMB and titv.pdf"
        } else {
          outFig <- paste0(fig.name,".pdf")
        }
        
        # base layout
        n1 <- seq(from = 0.105, to = 0.97-(0.97-0.05)*0.04, length.out = n.moic + 1)
        n2 <- n1[2:length(n1)]
        n  <- data.frame(n1 = n1[1:n.moic], n2 = n2, n3 = 0.05, n4 = 0.2) %>%
          rbind(c(0.05, 0.97, 0.25, 1), ., c(0, 0.1, 0.05, 0.25)) %>% as.matrix()
        
        opar <- par(no.readonly = TRUE)
        invisible(suppressWarnings(split.screen(n, erase = TRUE)))
        screen(1, new = TRUE)
        par(xpd = TRUE, mar = c(3, 1, 2, 0), oma = c(0, 0, 0, 0),bty = "o", mgp = c(2, 0.5, 0), tcl=-.25)
        y_lims = range(log10(unlist(lapply(TMB.plot, function(x) max(x[, "TMB"], na.rm = TRUE))) + 1))
        y_lims[1] = 0
        y_lims[2] = ceiling(max(y_lims))
        y_at = y_lims[1]:y_lims[2]
        x_top_label <- as.numeric(unlist(lapply(TMB.plot, nrow)))
        plot(NA, NA, xlim = c(0, length(TMB.plot)), ylim = y_lims,
             xlab = NA, ylab = NA, xaxt="n", yaxt = "n", xaxs = "r", yaxs = "r")
        rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col = "#EAE9E9", border = FALSE)
        grid(col = "white", lty = 1, lwd = 1.5) # add grid
        
        par(new = TRUE, bty="o")
        plot(NA, NA,
             col = "white",
             xlim = c(0, length(TMB.plot)), ylim = y_lims,
             xlab = "", ylab = "",
             xaxt = "n", yaxt = "n")
        
        text(x = length(TMB.plot)/2, y = y_lims[2] - 0.1, cex = 1.2,
             label = paste0(statistic, " p = ", formatC(TMB.test, digits = 1, format = "e")))
        
        # add median TMB
        invisible(lapply(seq_len(nrow(TMB.med)), function(i) {
          segments(x0  = i - 1,
                   x1  = i,
                   y0  = TMB.med[i, "Median_Mutations_log10"],
                   y1  = TMB.med[i, "Median_Mutations_log10"],
                   col = "#2b2d42",
                   lwd = 2)}))
        
        # add scatter TMB
        invisible(lapply(seq_len(length(TMB.plot)), function(i){
          tmp = TMB.plot[[i]]
          points(tmp$x, log10(tmp$TMB + 1), pch = 16, cex = 0.5, col = colvec[i])
        }))
        
        # modify axis
        axis(side = 1, at = seq(0.5, length(TMB.plot) - 0.5, 1), labels = names(TMB.plot),
             las = 1, tick = TRUE, line = 0)
        axis(side = 2, at = y_at, las = 2, line = 0, tick = TRUE, labels = y_at)
        if(show.size) {
          axis(side = 3, at = seq(0.5, length(TMB.plot) - 0.5, 1), labels = paste0("n = ",x_top_label),
               tick = TRUE, line = 0, cex.axis = 0.9)
        }
        mtext(text = bquote("log"[10]~"(TMB + 1)"), side = 2, line = 1.15, cex = 1.1)
        
        # add titv
        invisible(lapply(seq_len(length(titv.dat2)), function(i){
          screen(i + 1, new = TRUE)
          par(xpd = TRUE, mar = c(0, 0, 0, 0), oma = c(0, 0, 0, 0), bty = "o")
          tmp <- titv.dat2[[i]]
          barplot(tmp, col = col.titv[rownames(tmp)], names.arg = rep("", ncol(tmp)),
                  xaxs = "i", yaxs = "i",
                  axes = FALSE, space = 0, border = NA, lwd = 1.2)
          box()
        }))
        
        # add legend
        screen(n.moic + 2, new = TRUE)
        par(xpd = TRUE, mar = c(0, 0, 0, 0), oma = c(0, 0, 0, 0), bty = "n")
        plot(NA, NA, xlim = c(0, 1), ylim = c(0, 1), axes = FALSE, xlab = NA, ylab = NA)
        legend("center",
               fill   = col.titv,
               legend = names(col.titv),
               border = NA,
               bty    = "n",
               cex    = 0.8)
        close.screen(all.screens = TRUE)
      })
      
      ##RData结果下载按钮响应事件
      output$compTMBs_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_compTMB==1,paste0(tolower(project),".tcga.compTMB.RData"),paste0(tolower(project),".validation.compTMB.RData")),
        content = function(theFile) {
          save(compTMBResult,file=theFile)
        }
      )
      
      #compTMB后给予用户的反馈提示
      output$compTMB_finish<-renderUI({
        
        if(input$tcga_vali_compTMB==1){
          list(br(),column(12,h3(strong("The process of comparing total mutation burden on tcga datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare fraction genome altered'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compTMBs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of comparing total mutation burden on validation datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare fraction genome altered'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compTMBs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #112.The mapping color for bars of FGA, FGG and FGL填写表格
    output$table.barcolor_compFGA<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=3,nrow=1))
      colnames(DF)=c("FGA","FGG","FGL")
      rownames(DF)="Color for bars"
      
      default <- c()
      for(d in 1:3){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=150)
    })
    
    #113.Colors for each subtype填写表格
    output$table.compFGA_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ##获取聚类数
      if(input$tcga_vali_compFGA==1){
        #TCGA
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        
        if(input$Consensus_N_clust==1){
          N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
        }else{
          if(length(input$multiClusteringAlgorithms)==1){
            #用户指定聚类数
            N.clust <- switch(input$multiClusteringAlgorithms,
                              'iClusterBayes' = input$iClusterBayes_N_clusts,
                              'SNF' = input$SNF_N_clusts,
                              'PINSPlus' = input$PINSPlus_N_clusts,
                              'NEMO' = input$NEMO_N_clusts,
                              'COCA' = input$COCA_N_clusts,
                              'LRAcluster' = input$LRAcluster_N_clusts,
                              'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                              'IntNMF' = input$IntNMF_N_clusts,
                              'CIMLR' = input$CIMLR_N_clusts,
                              'MoCluster' = input$MoCluster_N_clusts)
          }
          if(length(input$multiClusteringAlgorithms)>1){
            N.clust <- input$Consensus_N_clusts #用户指定聚类数
          }
        }
      }else{
        #Validation
        if(input$validation_method_compFGA==1){
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        N.clust <- length(unique(integrated_data$clust.res$clust))
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #114.compFGA_process事件响应按钮
    observeEvent(input$compFGA_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Calculate and compare FGA, FGG as well as FGL among different subtypes',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compFGA==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###聚类结果导入
        if(input$validation_method_compFGA==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
     
      segment <- integrated_data$segment
      
      ####数据预处理
      if("num.mark" %in% colnames(segment)){
        segment <- segment[,-which(colnames(segment)=="num.mark")]
      }
      if("seg.mean" %in% colnames(segment)){
        colnames(segment)[which(colnames(segment)=="seg.mean")] <- "value"
      }
      if("chromosome" %in% colnames(segment)){
        colnames(segment)[which(colnames(segment)=="chromosome")] <- "chrom"
      }
      segment <- segment[,c("sample","chrom","start","end","value")]
      
      progress$set(value = 2)
      #The type of the 'value' column in segmented copy number dataset
      if(input$iscopynumber_compFGA==1){
        iscopynumber <- TRUE
      }else{
        iscopynumber <- FALSE
      }
      
      #The cutoff for identifying copy-number gain or loss
      cnathreshold <- input$cnathreshold_compFGA
      
      #The method for statistical testing
      if(input$test_method_compFGA==1){
        test.method <- "parametric"
      }else{
        test.method <- "nonparametric"
      }
      
      #The mapping colors for bars of FGA, FGG and FGL
      if(input$barcolor_compFGA==1){
        barcolor <- c("#008B8A", "#F2042C", "#21498D")
      }else{
        barcolor <- data.frame(rbind(hot_to_r(input$table.barcolor_compFGA)))[1,]
        barcolor <- trimws(barcolor,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Colors for each subtype
      if(input$clustcolor_compFGA==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.compFGA_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #The width of barplot
      width <- input$width_compFGA
      
      #The height of barplot
      height <- input$height_compFGA
      
      #The output path for storing the barplot
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the barplot
      if(input$tcga_vali_compFGA==1){
        fig.name <- input$compFGA_FigureName_tcga
      }else{
        fig.name <- input$compFGA_FigureName_validation
      }
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        if(input$tcga_vali_compFGA==1){
          fig.name <- "barplot_of_fraction_genome_altered(TCGA)"
        }else{
          fig.name <- "barplot_of_fraction_genome_altered(Validation)"
        }
      }
      
      progress$set(value = 3)
      compFGAResult <- compFGAs(moic.res = moic.res,
                                segment = segment,
                                iscopynumber = iscopynumber,
                                cnathreshold = cnathreshold,
                                test.method = test.method,
                                barcolor = barcolor,
                                clust.col = clust.col,
                                fig.path = fig.path,
                                fig.name = fig.name,
                                width = width,
                                height = height)
      
      #保存compFGA结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compFGA==1){
        save(compFGAResult, file = file.path(RData.path,paste0(tolower(project),".tcga.compFGA.RData")))
      }else{
        save(compFGAResult, file = file.path(RData.path,paste0(tolower(project),".validation.compFGA.RData")))
      }
      
      progress$set(value = 4)
      ###展示summary###
      #表格标题
      output$table.compFGAsummaryTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_compFGA==1){
          tags$strong("Table: Comparison of fraction genome altered among identified subtypes (TCGA)")
        }else{
          tags$strong("Table: Comparison of fraction genome altered among identified subtypes (Validation)")
        }
      })
      
      #表格
      output$table.compFGAsummary<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(compFGAResult$summary),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(compFGAResult$summary)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$tcga_vali_compFGA==1,"Comparison_of_fraction_genome_altered_among_identified_subtypes(TCGA)","Comparison_of_fraction_genome_altered_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compFGA==1,"Table: Comparison of fraction genome altered among identified subtypes (TCGA)","Table: Comparison of fraction genome altered among identified subtypes (Validation)")
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$tcga_vali_compFGA==1,"Comparison_of_fraction_genome_altered_among_identified_subtypes(TCGA)","Comparison_of_fraction_genome_altered_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compFGA==1,"Table: Comparison of fraction genome altered among identified subtypes (TCGA)","Table: Comparison of fraction genome altered among identified subtypes (Validation)")
                          ),
                          list(extend='excel',
                               filename = ifelse(input$tcga_vali_compFGA==1,"Comparison_of_fraction_genome_altered_among_identified_subtypes(TCGA)","Comparison_of_fraction_genome_altered_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compFGA==1,"Table: Comparison of fraction genome altered among identified subtypes (TCGA)","Table: Comparison of fraction genome altered among identified subtypes (Validation)")
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$tcga_vali_compFGA==1,"Comparison_of_fraction_genome_altered_among_identified_subtypes(TCGA)","Comparison_of_fraction_genome_altered_among_identified_subtypes(Validation)"),
                               title = ifelse(input$tcga_vali_compFGA==1,"Table: Comparison of fraction genome altered among identified subtypes (TCGA)","Table: Comparison of fraction genome altered among identified subtypes (Validation)")
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      #表格脚注
      output$table.compFGAsummaryNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$compFGA_figure_unit<-renderUI({
        list(column(12,downloadButton("compFGA_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("compFGA_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$compFGA_figure_download<-downloadHandler(
        
        filename = paste0(ifelse(is.null(fig.name),ifelse(input$tcga_vali_compFGA==1,"barplot_of_fraction_genome_altered(TCGA)","barplot_of_fraction_genome_altered(Validation)"),fig.name),".pdf"),
        content = function(theFile) {
          
          pal <- compFGAResult$pal
          pal
          ggsave(theFile, width = width, height = height)
        }
      )
      
      ##页面展示事件
      output$compFGA_figure<-renderPlot({
        
        pal <- compFGAResult$pal
        pal
      })
      
      ##RData结果下载按钮响应事件
      output$compFGAs_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_compFGA==1,paste0(tolower(project),".tcga.compFGA.RData"),paste0(tolower(project),".validation.compFGA.RData")),
        content = function(theFile) {
          save(compFGAResult,file=theFile)
        }
      )
      
      #compFGA后给予用户的反馈提示
      output$compFGA_finish<-renderUI({
        
        if(input$tcga_vali_compFGA==1){
          list(br(),column(12,h3(strong("The process of comparing fraction genome altered on tcga datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare drug sensitivity'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compFGAs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of comparing fraction genome altered on validation datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare drug sensitivity'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compFGAs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #115.Colors for each subtype填写表格
    output$table.compDrugsen_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ##获取聚类数
      if(input$tcga_vali_compDrugsen==1){
        #TCGA
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        
        if(input$Consensus_N_clust==1){
          N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
        }else{
          if(length(input$multiClusteringAlgorithms)==1){
            #用户指定聚类数
            N.clust <- switch(input$multiClusteringAlgorithms,
                              'iClusterBayes' = input$iClusterBayes_N_clusts,
                              'SNF' = input$SNF_N_clusts,
                              'PINSPlus' = input$PINSPlus_N_clusts,
                              'NEMO' = input$NEMO_N_clusts,
                              'COCA' = input$COCA_N_clusts,
                              'LRAcluster' = input$LRAcluster_N_clusts,
                              'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                              'IntNMF' = input$IntNMF_N_clusts,
                              'CIMLR' = input$CIMLR_N_clusts,
                              'MoCluster' = input$MoCluster_N_clusts)
          }
          if(length(input$multiClusteringAlgorithms)>1){
            N.clust <- input$Consensus_N_clusts #用户指定聚类数
          }
        }
      }else{
        #Validation
        if(input$validation_method_compDrugsen==1){
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        N.clust <- length(unique(integrated_data$clust.res$clust))
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #116.compDrugsen_process事件响应按钮
    observeEvent(input$compDrugsen_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Estimate and compare the IC50 of specific drug among different subtypes',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compDrugsen==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###聚类结果导入
        if(input$validation_method_compDrugsen==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      if(input$tcga_vali_compDrugsen==1){
        
        if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
          
          ##mRNA
          if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
            tpm <- integrated_data$mRNA.expr
          }
          ##lncRNA
          if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
            tpm <- integrated_data$lncRNA.expr
          }
          ##mRNA & lncRNA
          if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
            tpm_mrna <- integrated_data$mRNA.expr
            tpm_lncrna <- integrated_data$lncRNA.expr
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_tpm <- which(duplicated(c(rownames(tpm_mrna),rownames(tpm_lncrna))))
            tpm <- rbind(tpm_mrna,tpm_lncrna)
            if(length(drop_index_tpm)!=0){
              tpm <- tpm[-drop_index_tpm,]
            }
          }
          norm.expr <- tpm
        }
        
        #如果有fpkm数据则优先使用
        if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
          
          ##mRNA
          if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
            fpkm <- integrated_data$fpkm_mrna
          }
          ##lncRNA
          if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
            fpkm <- integrated_data$fpkm_lncrna
          }
          ##mRNA & lncRNA
          if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
            fpkm_mrna <- integrated_data$fpkm_mrna
            fpkm_lncrna <- integrated_data$fpkm_lncrna
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index <- which(duplicated(c(rownames(fpkm_mrna),rownames(fpkm_lncrna))))
            fpkm <- rbind(fpkm_mrna, fpkm_lncrna)
            if(length(drop_index)!=0){
              fpkm <- fpkm[-drop_index,]
            }
          }
          norm.expr <- fpkm
        }
      }else{
        
        if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
          
          ##mRNA
          if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
            tpm <- integrated_data$mRNA.expr
          }
          ##lncRNA
          if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
            tpm <- integrated_data$lncRNA.expr
          }
          ##mRNA & lncRNA
          if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
            tpm_mrna <- integrated_data$mRNA.expr
            tpm_lncrna <- integrated_data$lncRNA.expr
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_tpm <- which(duplicated(c(rownames(tpm_mrna),rownames(tpm_lncrna))))
            tpm <- rbind(tpm_mrna,tpm_lncrna)
            if(length(drop_index_tpm)!=0){
              tpm <- tpm[-drop_index_tpm,]
            }
          }
          norm.expr <- tpm
        }
        
        #如果有fpkm数据则优先使用
        if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
          
          ##mRNA
          if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
            fpkm <- integrated_data$fpkm_mrna
          }
          ##lncRNA
          if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
            fpkm <- integrated_data$fpkm_lncrna
          }
          ##mRNA & lncRNA
          if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
            fpkm_mrna <- integrated_data$fpkm_mrna
            fpkm_lncrna <- integrated_data$fpkm_lncrna
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index <- which(duplicated(c(rownames(fpkm_mrna),rownames(fpkm_lncrna))))
            fpkm <- rbind(fpkm_mrna, fpkm_lncrna)
            if(length(drop_index)!=0){
              fpkm <- fpkm[-drop_index,]
            }
          }
          # norm.expr <- fpkm[,moic.res$clust.res$samID]
          norm.expr <- fpkm
        }
      }

      progress$set(value = 2)
      #The name of the drug from GDSC for which you would like to predict sensitivity
      drugs <- input$drugs_compDrugsen
      drugs <- trimws(drugs,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The tissue type from which the cell lines originated
      if(input$tissueType_compDrugsen==1){
        tissueType <- switch(input$tissueTypeValue_compDrugsen,
                             '1' = "aero_digestive_tract",
                             '2' = "blood",
                             '3' = "bone",
                             '4' = "breast",
                             '5' = "digestive_system",
                             '6' = "lung",
                             '7' = "nervous_system",
                             '8' = "skin",
                             '9' = "urogenital_system")
      }else{
        tissueType <- "all"
      }
      
      #The method for statistical testing
      if(input$test_method_compDrugsen==1){
        test.method <- "parametric"
      }else{
        test.method <- "nonparametric"
      }
      
      #Colors for each subtype
      if(input$clustcolor_compDrugsen==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.compDrugsen_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #The seed for reproducing the result of comparing drug sensitivity
      seed <- input$seed_compDrugsen
      
      #The width of boxviolin plot
      width <- input$width_compDrugsen
      
      #The height of boxviolin plot
      height <- input$height_compDrugsen
      
      #The output path for storing the barplot
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The prefix for the name of output boxviolin plot
      if(input$tcga_vali_compDrugsen==1){
        prefix <- input$compDrugsen_prefix_tcga
      }else{
        prefix <- input$compDrugsen_prefix_validation
      }
      prefix <- trimws(prefix,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(prefix==""){
        if(input$tcga_vali_compDrugsen==1){
          prefix <- "boxviolin_of_estimated_IC50(TCGA)"
        }else{
          prefix <- "boxviolin_of_estimated_IC50(Validation)"
        }
      }
      
      progress$set(value = 3)
      compDrugsenResult <- compDrugsens(moic.res = moic.res,
                                        norm.expr = norm.expr,
                                        drugs = drugs,
                                        tissueType = tissueType,
                                        test.method = test.method,
                                        clust.col = clust.col,
                                        prefix = prefix,
                                        seed = seed,
                                        fig.path = fig.path,
                                        width = width,
                                        height = height)
      
      #保存compDrugsen结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compDrugsen==1){
        save(compDrugsenResult, file = file.path(RData.path,paste0(tolower(project),".tcga.compDrugsen.",drugs,".RData")))
      }else{
        save(compDrugsenResult, file = file.path(RData.path,paste0(tolower(project),".validation.compDrugsen.",drugs,".RData")))
      }
      
      progress$set(value = 4)
      ###展示predictedBoxdat###
      #表格标题
      output$table.compDrugsenpredictedBoxdatTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_compDrugsen==1){
          tags$strong(paste0("Table: Comparison of estimated IC50 (TCGA) for ",drugs," among identified subtypes"))
        }else{
          tags$strong(paste0("Table: Comparison of estimated IC50 (Validation) for ",drugs," among identified subtypes"))
        }
      })
      
      #表格
      output$table.compDrugsenpredictedBoxdat<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(compDrugsenResult$predictedBoxdat),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(compDrugsenResult$predictedBoxdat)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$tcga_vali_compDrugsen==1,paste0("Comparison_of_estimated_IC50(TCGA)_for_",drugs,"_among_identified_subtypes"),paste0("Comparison_of_estimated_IC50(Validation)_for_",drugs,"_among_identified_subtypes")),
                               title = ifelse(input$tcga_vali_compDrugsen==1,paste0("Table: Comparison of estimated IC50 (TCGA) for ",drugs," among identified subtypes"),paste0("Table: Comparison of estimated IC50 (Validation) for ",drugs," among identified subtypes"))
                          ),
                          list(extend='csv',
                               filename = ifelse(input$tcga_vali_compDrugsen==1,paste0("Comparison_of_estimated_IC50(TCGA)_for_",drugs,"_among_identified_subtypes"),paste0("Comparison_of_estimated_IC50(Validation)_for_",drugs,"_among_identified_subtypes")),
                               title = ifelse(input$tcga_vali_compDrugsen==1,paste0("Table: Comparison of estimated IC50 (TCGA) for ",drugs," among identified subtypes"),paste0("Table: Comparison of estimated IC50 (Validation) for ",drugs," among identified subtypes"))
                          ),
                          list(extend='excel',
                               filename = ifelse(input$tcga_vali_compDrugsen==1,paste0("Comparison_of_estimated_IC50(TCGA)_for_",drugs,"_among_identified_subtypes"),paste0("Comparison_of_estimated_IC50(Validation)_for_",drugs,"_among_identified_subtypes")),
                               title = ifelse(input$tcga_vali_compDrugsen==1,paste0("Table: Comparison of estimated IC50 (TCGA) for ",drugs," among identified subtypes"),paste0("Table: Comparison of estimated IC50 (Validation) for ",drugs," among identified subtypes"))
                          ),
                          list(extend='print',
                               filename = ifelse(input$tcga_vali_compDrugsen==1,paste0("Comparison_of_estimated_IC50(TCGA)_for_",drugs,"_among_identified_subtypes"),paste0("Comparison_of_estimated_IC50(Validation)_for_",drugs,"_among_identified_subtypes")),
                               title = ifelse(input$tcga_vali_compDrugsen==1,paste0("Table: Comparison of estimated IC50 (TCGA) for ",drugs," among identified subtypes"),paste0("Table: Comparison of estimated IC50 (Validation) for ",drugs," among identified subtypes"))
                          )
                        )
                      )
        )
      }, server=FALSE)
      
      #表格脚注
      output$table.compDrugsenpredictedBoxdatNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$compDrugsen_figure_unit<-renderUI({
        list(column(12,downloadButton("compDrugsen_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("compDrugsen_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$compDrugsen_figure_download<-downloadHandler(
        
        filename = paste0(ifelse(is.null(prefix),ifelse(input$tcga_vali_compDrugsen==1,"boxviolin_of_estimated_IC50(TCGA)","boxviolin_of_estimated_IC50(Validation)"),prefix), "_for_", drugs, ".pdf"),
        content = function(theFile) {
          
          p <- compDrugsenResult$p
          p
          ggsave(theFile, width = width, height = height)
        }
      )
      
      ##页面展示事件
      output$compDrugsen_figure<-renderPlot({
        
        p <- compDrugsenResult$p
        p
      })
      
      ##RData结果下载按钮响应事件
      output$compDrugsens_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_compDrugsen==1,paste0(tolower(project),".tcga.compDrugsen.",drugs,".RData"),paste0(tolower(project),".validation.compDrugsen.",drugs,".RData")),
        content = function(theFile) {
          save(compDrugsenResult,file=theFile)
        }
      )
      
      #compDrugsen后给予用户的反馈提示
      output$compDrugsen_finish<-renderUI({
        
        if(input$tcga_vali_compDrugsen==1){
          list(br(),column(12,h3(strong("The process of comparing drug sensitivity on tcga datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare agreement with other subtypes'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compDrugsens_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of comparing drug sensitivity on validation datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Compare agreement with other subtypes'.")),style="color:darkblack"),br(),
               column(12,downloadButton("compDrugsens_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #117.Variable name of traditional subtypes for comparison填写表格
    output$table.subt2comp_compAgree<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$subt2comp_number_compAgree,nrow=1))
      colnames(DF)=1:input$subt2comp_number_compAgree
      rownames(DF)="Variable name"
      
      default <- c()
      for(d in 1:input$subt2comp_number_compAgree){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #118.Colors for each subtype填写表格
    output$table.compAgree_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ##获取聚类数
      if(input$tcga_vali_compAgree==1){
        #TCGA
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        
        if(input$Consensus_N_clust==1){
          N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
        }else{
          if(length(input$multiClusteringAlgorithms)==1){
            #用户指定聚类数
            N.clust <- switch(input$multiClusteringAlgorithms,
                              'iClusterBayes' = input$iClusterBayes_N_clusts,
                              'SNF' = input$SNF_N_clusts,
                              'PINSPlus' = input$PINSPlus_N_clusts,
                              'NEMO' = input$NEMO_N_clusts,
                              'COCA' = input$COCA_N_clusts,
                              'LRAcluster' = input$LRAcluster_N_clusts,
                              'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                              'IntNMF' = input$IntNMF_N_clusts,
                              'CIMLR' = input$CIMLR_N_clusts,
                              'MoCluster' = input$MoCluster_N_clusts)
          }
          if(length(input$multiClusteringAlgorithms)>1){
            N.clust <- input$Consensus_N_clusts #用户指定聚类数
          }
        }
      }else{
        #Validation
        if(input$validation_method_compAgree==1){
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        N.clust <- length(unique(integrated_data$clust.res$clust))
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #119.compAgree_process事件响应按钮
    observeEvent(input$compAgree_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Compute four evaluation indicators (RI, AMI, JI, FM) for agreement of two partitions',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compAgree==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###聚类结果导入
        if(input$validation_method_compAgree==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
        }else{
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
        }
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      surv.info <- integrated_data$clin.info #获取生存和临床数据
      # if(!('fustat' %in% colnames(surv.info)) & ('OS' %in% colnames(surv.info))){
      #   colnames(surv.info)[which(colnames(surv.info)=='OS')] <- "fustat"
      # } #数据预处理(列名替换)
      # if(!('futime' %in% colnames(surv.info)) & ('OS.time' %in% colnames(surv.info))){
      #   colnames(surv.info)[which(colnames(surv.info)=='OS.time')] <- "futime"
      # } #数据预处理(列名替换)
      # if('age_at_initial_pathologic_diagnosis' %in% colnames(surv.info)){
      #   colnames(surv.info)[which(colnames(surv.info)=='age_at_initial_pathologic_diagnosis')] <- "age"
      # } #数据预处理(列名替换)
      # if('ajcc_pathologic_tumor_stage' %in% colnames(surv.info)){
      #   colnames(surv.info)[which(colnames(surv.info)=='ajcc_pathologic_tumor_stage')] <- "pStage"
      # } #数据预处理(列名替换)
      # if("pstage" %in% tolower(names(surv.info))){
      #   for(i in 1:nrow(surv.info)){
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="[Discrepancy]"){
      #       surv.info$pStage[i] <- names(which(table(surv.info$pStage)==max(table(surv.info$pStage))))
      #     }
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="[Not Available]"){
      #       surv.info$pStage[i] <- names(which(table(surv.info$pStage)==max(table(surv.info$pStage))))
      #     } 
      #   } #缺失值填补(众数填补多分类变量)
      #   for(i in 1:nrow(surv.info)){
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage I" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IB"){
      #       surv.info$pStage[i] <- "Stage I"
      #     }
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage II" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIB"){
      #       surv.info$pStage[i] <- "Stage II"
      #     }
      #     if(trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage III" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(surv.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIB"){
      #       surv.info$pStage[i] <- "Stage III"
      #     }
      #   } #类别合并
      # }
      # if("age" %in% tolower(names(surv.info))){
      #   for(i in 1:nrow(surv.info)){
      #     if(is.na(surv.info$age[i])==TRUE){
      #       surv.info$age[i] <- median(surv.info$age,na.rm = TRUE)
      #     }
      #   } #缺失值填补(中位数填补连续型变量)
      # }
      
      progress$set(value = 2)
      #Input the variable name of traditional subtypes in survival and clinical information for comparison
      subt2comp_variables <- data.frame(rbind(hot_to_r(input$table.subt2comp_compAgree)))[1,]
      subt2comp_variables <- trimws(subt2comp_variables,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(length(subt2comp_variables)==1){
        tmp_rownames <- rownames(surv.info)
        subt2comp <- as.data.frame(surv.info[,subt2comp_variables])
        rownames(subt2comp) <- tmp_rownames
        colnames(subt2comp) <- subt2comp_variables
      }
      if(length(subt2comp_variables)>1){
        subt2comp <- surv.info[,subt2comp_variables]
      }
      
      #Process the obtained data
      if(!('fustat' %in% colnames(subt2comp)) & ('OS' %in% colnames(subt2comp))){
        colnames(subt2comp)[which(colnames(subt2comp)=='OS')] <- "fustat"
      } #数据预处理(列名替换)
      if(!('futime' %in% colnames(subt2comp)) & ('OS.time' %in% colnames(subt2comp))){
        colnames(subt2comp)[which(colnames(subt2comp)=='OS.time')] <- "futime"
      } #数据预处理(列名替换)
      if('age_at_initial_pathologic_diagnosis' %in% colnames(subt2comp)){
        colnames(subt2comp)[which(colnames(subt2comp)=='age_at_initial_pathologic_diagnosis')] <- "age"
      } #数据预处理(列名替换)
      if('ajcc_pathologic_tumor_stage' %in% colnames(subt2comp)){
        colnames(subt2comp)[which(colnames(subt2comp)=='ajcc_pathologic_tumor_stage')] <- "pStage"
      } #数据预处理(列名替换)
      for (j in 1:ncol(subt2comp)){
        for(i in 1:nrow(subt2comp)){
          ##NA判定##
          if(is.na(trimws(subt2comp[i,j],whitespace = "[ \t\r\n]"))){
            next
          }
          if(trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="[Discrepancy]" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="[Not Available]"){
            subt2comp[i,j] <- NA
          }
          ##类别合并##
          if(tolower(colnames(subt2comp)[j])=="pstage"){
            if(trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage I" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IB"){
              subt2comp[i,j] <- "Stage I"
            }
            if(trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage II" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIB"){
              subt2comp[i,j] <- "Stage II"
            }
            if(trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage III" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(subt2comp[i,j],whitespace = "[ \t\r\n]")=="Stage IIIB"){
              subt2comp[i,j] <- "Stage III"
            }
          }
          ##缺失值填补##
          ##众数填补多分类变量
          if(is.na(trimws(subt2comp[i,j],whitespace = "[ \t\r\n]"))==TRUE){
            subt2comp[i,j] <- names(which(table(subt2comp[,j])==max(table(subt2comp[,j]))))
          }
        }
      }
      
      #Generate alluvial diagram to show the agreement of different subtypes
      doPlot <- TRUE
      
      #Colors for each subtype
      if(input$clustcolor_compAgree==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.compAgree_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #The width for box in alluvial diagram
      box.width <- input$box_width_compAgree
      
      #The width of the figure
      width <- input$width_compAgree
      
      #The height of the figure
      height <- input$height_compAgree
      
      #The output path for storing the barplot
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the figure
      if(input$tcga_vali_compAgree==1){
        fig.name <- input$compAgree_FigureName_tcga
      }else{
        fig.name <- input$compAgree_FigureName_validation
      }
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        if(input$tcga_vali_compAgree==1){
          fig.name <- "agreement_between_current_subtype_and_other_classifications(TCGA)"
        }else{
          fig.name <- "agreement_between_current_subtype_and_other_classifications(Validation)"
        }
      }
      
      progress$set(value = 3)
      compAgreeResult <- compAgrees(moic.res = moic.res,
                                    subt2comp = subt2comp,
                                    doPlot = doPlot,
                                    clust.col = clust.col,
                                    box.width = box.width,
                                    fig.name = fig.name,
                                    fig.path = fig.path,
                                    width = width,
                                    height = height)
      
      #保存compAgree结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_compAgree==1){
        save(compAgreeResult, file = file.path(RData.path,paste0(tolower(project),".tcga.compAgree.RData")))
      }else{
        save(compAgreeResult, file = file.path(RData.path,paste0(tolower(project),".validation.compAgree.RData")))
      }
      
      progress$set(value = 4)
      ##表格标题
      output$table.compAgreeTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_compAgree==1){
          tags$strong("Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (TCGA)")
        }else{
          tags$strong("Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (Validation)")
        }
      })
      
      ##表格
      output$table.compAgree<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(compAgreeResult$outTab),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(compAgreeResult$outTab)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Bfrtip',
                        searching= FALSE,
                        bInfo=FALSE, #not to shown 1/N message
                        display=NULL,
                        bPaginate= FALSE, #suppress page
                        pageLength =100,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$tcga_vali_compAgree==1,"Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(TCGA)","Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(Validation)"),
                               title = ifelse(input$tcga_vali_compAgree==1,"Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (TCGA)","Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (Validation)")
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$tcga_vali_compAgree==1,"Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(TCGA)","Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(Validation)"),
                               title = ifelse(input$tcga_vali_compAgree==1,"Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (TCGA)","Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (Validation)")
                          ),
                          list(extend='excel',
                               filename = ifelse(input$tcga_vali_compAgree==1,"Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(TCGA)","Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(Validation)"),
                               title = ifelse(input$tcga_vali_compAgree==1,"Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (TCGA)","Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (Validation)")
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$tcga_vali_compAgree==1,"Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(TCGA)","Agreement_of_current_identified_subtypes_with_other_traditional_subtypes_in_survival_and_clinical_information(Validation)"),
                               title = ifelse(input$tcga_vali_compAgree==1,"Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (TCGA)","Table: Agreement of current identified subtypes with other traditional subtypes in survival and clinical information (Validation)")
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.compAgreeNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$compAgree_figure_unit<-renderUI({
        list(column(12,downloadButton("compAgree_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("compAgree_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$compAgree_figure_download<-downloadHandler(
        
        filename = paste0(ifelse(is.null(fig.name),ifelse(input$tcga_vali_compAgree==1,"agreement_between_current_subtype_and_other_classifications(TCGA)","agreement_between_current_subtype_and_other_classifications(Validation)"),fig.name), ".pdf"),
        content = function(theFile) {
          
          bp <- compAgreeResult$bp
          bp
          ggsave(theFile, width = width, height = height)
        }
      )
      
      ##页面展示事件
      output$compAgree_figure<-renderPlot({
        
        bp <- compAgreeResult$bp
        bp
      })
      
      ##RData结果下载按钮响应事件
      output$compAgrees_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_compAgree==1,paste0(tolower(project),".tcga.compAgree.RData"),paste0(tolower(project),".validation.compAgree.RData")),
        content = function(theFile) {
          save(compAgreeResult,file=theFile)
        }
      )
      
      #compAgree后给予用户的反馈提示
      output$compAgree_finish<-renderUI({
        
        if(input$tcga_vali_compAgree==1){
          list(br(),column(12,h3(strong("The process of comparing agreement with other subtypes on tcga datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now all steps in 'COMP Module' have been finished, let's start the final step of MOVICS--RUN Module!")),style="color:darkblack"),br(),
               column(12,downloadButton("compAgrees_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of comparing agreement with other subtypes on validation datasets has been finished, you can download and check the table and figure, as well as the '.RData' file. Now all steps in 'COMP Module' have been finished, let's start the final step of MOVICS--RUN Module!")),style="color:darkblack"),br(),
               column(12,downloadButton("compAgrees_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #120.runDEA_process事件响应按钮
    observeEvent(input$runDEA_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$dea_method_runDEA==1,"Perform differential expression analysis using 'deseq2' algorithm between two classes",ifelse(input$dea_method_runDEA==2,"Perform differential expression analysis using 'edger' algorithm between two classes","Perform differential expression analysis using 'limma' algorithm between two classes")),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ###聚类结果导入
      ####聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
      }
      ####聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
        moic.res <- cmoic
      }
      ###多组学数据导入
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      count <- integrated_data$count #RNA-Seq raw count data
      #tpm or fpkm
      if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
        
        ##mRNA
        if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$mRNA.expr
        }
        ##lncRNA
        if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$lncRNA.expr
        }
        ##mRNA & lncRNA
        if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm_mrna <- integrated_data$mRNA.expr
          tpm_lncrna <- integrated_data$lncRNA.expr
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index_tpm <- which(duplicated(c(rownames(tpm_mrna),rownames(tpm_lncrna))))
          tpm <- rbind(tpm_mrna,tpm_lncrna)
          if(length(drop_index_tpm)!=0){
            tpm <- tpm[-drop_index_tpm,]
          }
        }
        norm.expr <- tpm
      }
      
      #如果有fpkm数据则优先使用
      if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
        
        ##mRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_mrna
        }
        ##lncRNA
        if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_lncrna
        }
        ##mRNA & lncRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm_mrna <- integrated_data$fpkm_mrna
          fpkm_lncrna <- integrated_data$fpkm_lncrna
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index <- which(duplicated(c(rownames(fpkm_mrna),rownames(fpkm_lncrna))))
          fpkm <- rbind(fpkm_mrna, fpkm_lncrna)
          if(length(drop_index)!=0){
            fpkm <- fpkm[-drop_index,]
          }
        }
        norm.expr <- fpkm
      }
      
      progress$set(value = 2)
      #Choose the algorithm for differential expression analysis
      if(input$dea_method_runDEA==1){
        dea.method <- "deseq2"
      }else if(input$dea_method_runDEA==2){
        dea.method <- "edger"
      }else{
        dea.method <- "limma"
      }
      
      #expression data
      if(dea.method=="deseq2" | dea.method=="edger"){
        expr <- count
      }else{
        expr <- norm.expr
      }
      
      #Indicate the prefix of output file (e.g. if the name of output file is 'consensusMOIC_nsclc_limma_test_result.CS1_vs_Others.txt', the prefix would be 'nsclc'. If you do not want a prefix but using the system default prefix, ignore this parameter and type nothing)
      prefix <- input$prefix_runDEA
      prefix <- trimws(prefix,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(prefix==""){
        prefix <- NULL
      }
      
      #The result already exists, overwrite it or skip this step directly
      if(input$overwt_runDEA==1){
        overwt <- TRUE
      }else{
        overwt <- FALSE
      }
      
      #Whether sort adjusted p value in ascending order for output table
      if(input$sort_p_runDEA==1){
        sort.p <- TRUE
      }else{
        sort.p <- FALSE
      }
      
      #Only select 'id', 'log2fc', 'pvalue' and 'padj' columns for output table or not
      if(input$verbose_runDEA==1){
        verbose <- TRUE
      }else{
        verbose <- FALSE
      }
      
      #The output path for storing the output table
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      res.path <- file.path(user_workdir,"tables")
      
      progress$set(value = 3)
      runDEAResult <- runDEAs(dea.method = dea.method,
                              expr = expr,
                              moic.res = moic.res,
                              prefix = prefix,
                              overwt = overwt,
                              sort.p = sort.p,
                              verbose = verbose,
                              res.path = res.path)
      
      #保存runDEA结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(runDEAResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runDEA.",dea.method,".RData")))
      
      progress$set(value = 4)
      ##表格标题
      output$table.runDEATitle<-renderUI({
        # isolate (validInputTS.caption())
        if(dea.method=="deseq2"){
          tags$strong("Table: Differential expression analysis results using 'deseq2' algorithm between two classes")
        }else if(dea.method=="edger"){
          tags$strong("Table: Differential expression analysis results using 'edger' algorithm between two classes")
        }else{
          tags$strong("Table: Differential expression analysis results using 'limma' algorithm between two classes")
        }
      })
      
      ##表格
      output$table.runDEA<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(runDEAResult$dea_result),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(runDEAResult$dea_result)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$dea_method_runDEA==1,"Differential_expression_analysis_results_using_deseq2_algorithm_between_two_classes",ifelse(input$dea_method_runDEA==2,"Differential_expression_analysis_results_using_edger_algorithm_between_two_classes","Differential_expression_analysis_results_using_limma_algorithm_between_two_classes")),
                               title = ifelse(input$dea_method_runDEA==1,"Table: Differential expression analysis results using 'deseq2' algorithm between two classes",ifelse(input$dea_method_runDEA==2,"Table: Differential expression analysis results using 'edger' algorithm between two classes","Table: Differential expression analysis results using 'limma' algorithm between two classes"))
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$dea_method_runDEA==1,"Differential_expression_analysis_results_using_deseq2_algorithm_between_two_classes",ifelse(input$dea_method_runDEA==2,"Differential_expression_analysis_results_using_edger_algorithm_between_two_classes","Differential_expression_analysis_results_using_limma_algorithm_between_two_classes")),
                               title = ifelse(input$dea_method_runDEA==1,"Table: Differential expression analysis results using 'deseq2' algorithm between two classes",ifelse(input$dea_method_runDEA==2,"Table: Differential expression analysis results using 'edger' algorithm between two classes","Table: Differential expression analysis results using 'limma' algorithm between two classes"))
                          ),
                          list(extend='excel',
                               filename = ifelse(input$dea_method_runDEA==1,"Differential_expression_analysis_results_using_deseq2_algorithm_between_two_classes",ifelse(input$dea_method_runDEA==2,"Differential_expression_analysis_results_using_edger_algorithm_between_two_classes","Differential_expression_analysis_results_using_limma_algorithm_between_two_classes")),
                               title = ifelse(input$dea_method_runDEA==1,"Table: Differential expression analysis results using 'deseq2' algorithm between two classes",ifelse(input$dea_method_runDEA==2,"Table: Differential expression analysis results using 'edger' algorithm between two classes","Table: Differential expression analysis results using 'limma' algorithm between two classes"))
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$dea_method_runDEA==1,"Differential_expression_analysis_results_using_deseq2_algorithm_between_two_classes",ifelse(input$dea_method_runDEA==2,"Differential_expression_analysis_results_using_edger_algorithm_between_two_classes","Differential_expression_analysis_results_using_limma_algorithm_between_two_classes")),
                               title = ifelse(input$dea_method_runDEA==1,"Table: Differential expression analysis results using 'deseq2' algorithm between two classes",ifelse(input$dea_method_runDEA==2,"Table: Differential expression analysis results using 'edger' algorithm between two classes","Table: Differential expression analysis results using 'limma' algorithm between two classes"))
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.runDEANote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      ##RData结果下载按钮响应事件
      output$runDEAs_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.runDEA.",dea.method,".RData"),
        content = function(theFile) {
          save(runDEAResult,file=theFile)
        }
      )
      
      #runDEA后给予用户的反馈提示
      output$runDEA_finish<-renderUI({
        
        list(br(),column(12,h3(strong("The process of running differential expression analysis has been finished, you can download and check the table and the '.RData' file. Now let's turn to the next step--'Run biomarker identification procedure'.")),style="color:darkblack"),br(),
             column(12,downloadButton("runDEAs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #121.Sample annotations from survival information for heatmap填写表格
    output$table.runMarker_annCol<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$runMarker_annCol_number,nrow=3))
      colnames(DF)=1:input$runMarker_annCol_number
      rownames(DF)=c("Sample annotation","Continuous or Categorical","Color settings")
      
      default <- c()
      for(d in 1:input$runMarker_annCol_number){
        default <- c(default,"")
      }
      DF[1,]=default
      DF[2,]=default
      DF[3,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=180)
    })
    
    #122.Colors for annotating each cluster at the top of heatmap填写表格
    output$table.runMarker_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      ##获取聚类数
      if(input$Consensus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        if(length(input$multiClusteringAlgorithms)==1){
          #用户指定聚类数
          N.clust <- switch(input$multiClusteringAlgorithms,
                            'iClusterBayes' = input$iClusterBayes_N_clusts,
                            'SNF' = input$SNF_N_clusts,
                            'PINSPlus' = input$PINSPlus_N_clusts,
                            'NEMO' = input$NEMO_N_clusts,
                            'COCA' = input$COCA_N_clusts,
                            'LRAcluster' = input$LRAcluster_N_clusts,
                            'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                            'IntNMF' = input$IntNMF_N_clusts,
                            'CIMLR' = input$CIMLR_N_clusts,
                            'MoCluster' = input$MoCluster_N_clusts)
        }
        if(length(input$multiClusteringAlgorithms)>1){
          N.clust <- input$Consensus_N_clusts #用户指定聚类数
        }
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #123.Colors for heatmap填写表格
    output$table.runMarker_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$runMarker_heatmap_color_number,nrow=1))
      colnames(DF)=1:input$runMarker_heatmap_color_number
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:input$runMarker_heatmap_color_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #124.runMarker_process事件响应按钮
    observeEvent(input$runMarker_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$dirct_runMarker==1,"Identify uniquely and significantly overexpressed biomarkers for each subtype","Identify uniquely and significantly downexpressed biomarkers for each subtype"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ###聚类结果导入
      ####聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
      }
      ####聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
        moic.res <- cmoic
      }
      ###多组学数据导入
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      #tpm or fpkm
      if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
        
        ##mRNA
        if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$mRNA.expr
        }
        ##lncRNA
        if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$lncRNA.expr
        }
        ##mRNA & lncRNA
        if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm_mrna <- integrated_data$mRNA.expr
          tpm_lncrna <- integrated_data$lncRNA.expr
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index_tpm <- which(duplicated(c(rownames(tpm_mrna),rownames(tpm_lncrna))))
          tpm <- rbind(tpm_mrna,tpm_lncrna)
          if(length(drop_index_tpm)!=0){
            tpm <- tpm[-drop_index_tpm,]
          }
        }
        norm.expr <- tpm
      }
      
      #如果有fpkm数据则优先使用
      if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
        
        ##mRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_mrna
        }
        ##lncRNA
        if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_lncrna
        }
        ##mRNA & lncRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm_mrna <- integrated_data$fpkm_mrna
          fpkm_lncrna <- integrated_data$fpkm_lncrna
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index <- which(duplicated(c(rownames(fpkm_mrna),rownames(fpkm_lncrna))))
          fpkm <- rbind(fpkm_mrna, fpkm_lncrna)
          if(length(drop_index)!=0){
            fpkm <- fpkm[-drop_index,]
          }
        }
        norm.expr <- fpkm
      }
      
      progress$set(value = 2)
      #Indicate the algorithm for completed differential expression analysis
      if(input$dea_method_runMarker==1){
        dea.method <- "deseq2"
      }else if(input$dea_method_runMarker==2){
        dea.method <- "edger"
      }else{
        dea.method <- "limma"
      }
      
      #Indicate the prefix of output file (e.g. if the name of output file is 'consensusMOIC_nsclc_limma_test_result.CS1_vs_Others.txt', the prefix would be 'nsclc'. If you do not want a prefix but using the system default prefix, ignore this parameter and type nothing)
      prefix <- input$prefix_runDEA
      prefix <- trimws(prefix,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(prefix==""){
        prefix <- NULL
      }
      
      #The path for saving the files of differential expression analysis
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      dat.path <- file.path(user_workdir,"tables")
      
      #The path for saving the results for identifying subtype-specific markers
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      res.path <- file.path(user_workdir,"tables")
      
      #Indicate the nominal p value for identifying significant markers
      p.cutoff <- input$p_cutoff_runMarker
      
      #Indicate the adjusted p value for identifying significant markers
      p.adj.cutoff <- input$p_adj_cutoff_runMarker
      
      #Indicate the direction of identifying significant marker
      if(input$dirct_runMarker==1){
        dirct <- "up"
      }else{
        dirct <- "down"
      }
      
      #Indicate the number of top markers sorted by log2fc should be identified for each subtype
      n.marker <- input$n_marker_runMarker
      
      #Generate heatmap by using subtype-specific markers
      doplot <- TRUE
      
      #Sample annotations from survival information for heatmap
      if(input$runMarker_annCol==1){
        surv_clin.info <- integrated_data$clin.info #获取生存和临床数据
        # if(!('fustat' %in% colnames(surv_clin.info)) & ('OS' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS')] <- "fustat"
        # } #数据预处理(列名替换)
        # if(!('futime' %in% colnames(surv_clin.info)) & ('OS.time' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS.time')] <- "futime"
        # } #数据预处理(列名替换)
        # if('age_at_initial_pathologic_diagnosis' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='age_at_initial_pathologic_diagnosis')] <- "age"
        # } #数据预处理(列名替换)
        # if('ajcc_pathologic_tumor_stage' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        # } #数据预处理(列名替换)
        # if("pstage" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Discrepancy]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Not Available]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     } 
        #   } #缺失值填补(众数填补多分类变量)
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage I" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IB"){
        #       surv_clin.info$pStage[i] <- "Stage I"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage II" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIB"){
        #       surv_clin.info$pStage[i] <- "Stage II"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage III" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIB"){
        #       surv_clin.info$pStage[i] <- "Stage III"
        #     }
        #   } #类别合并
        # }
        # if("age" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(is.na(surv_clin.info$age[i])==TRUE){
        #       surv_clin.info$age[i] <- median(surv_clin.info$age,na.rm = TRUE)
        #     }
        #   } #缺失值填补(中位数填补连续型变量)
        # }
        ###annCol
        sample_annotation <- data.frame(rbind(hot_to_r(input$table.runMarker_annCol)))[1,]
        sample_annotation <- trimws(sample_annotation,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        annCol <- surv_clin.info[,sample_annotation, drop = FALSE]
        ###annColors
        continuous_or_categorical <- data.frame(rbind(hot_to_r(input$table.runMarker_annCol)))[2,]
        continuous_or_categorical <- trimws(continuous_or_categorical,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        color_settings <- data.frame(rbind(hot_to_r(input$table.runMarker_annCol)))[3,]
        color_settings <- trimws(color_settings,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        ###Process the obtained data
        if(!('fustat' %in% colnames(annCol)) & ('OS' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS')] <- "fustat"
        } #数据预处理(列名替换)
        if(!('futime' %in% colnames(annCol)) & ('OS.time' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS.time')] <- "futime"
        } #数据预处理(列名替换)
        if('age_at_initial_pathologic_diagnosis' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='age_at_initial_pathologic_diagnosis')] <- "age"
        } #数据预处理(列名替换)
        if('ajcc_pathologic_tumor_stage' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        } #数据预处理(列名替换)
        for (j in 1:ncol(annCol)){
          for(i in 1:nrow(annCol)){
            ##NA判定##
            if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))){
              next
            }
            if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Discrepancy]" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Not Available]"){
              annCol[i,j] <- NA
            }
            ##类别合并##
            if(tolower(colnames(annCol)[j])=="pstage"){
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage I" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IB"){
                annCol[i,j] <- "Stage I"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage II" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIB"){
                annCol[i,j] <- "Stage II"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage III" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIB"){
                annCol[i,j] <- "Stage III"
              }
            }
            ##缺失值填补##
            if(continuous_or_categorical[j] == "Categorical"){
              ##众数填补多分类变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- names(which(table(annCol[,j])==max(table(annCol[,j]))))
              }
            }else{
              ##中位数填补连续型变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- median(annCol[,j],na.rm = TRUE)
              }
            }
          }
        }
        annColors <- list()
        for(i in 1:length(sample_annotation)){
          if(continuous_or_categorical[i]=="Continuous"){
            breaks <- c(min(annCol[,sample_annotation[i]]),median(annCol[,sample_annotation[i]]),max(annCol[,sample_annotation[i]]))
            colors <- unlist(strsplit(color_settings[i],";"))
            annColors[[i]] <- circlize::colorRamp2(breaks = breaks, colors = colors)
          }else{
            colors <- unlist(strsplit(color_settings[i],";"))
            names(colors) <- unique(annCol[,sample_annotation[i]])
            annColors[[i]] <- colors
          }
        }
        names(annColors) <- sample_annotation
      }else{
        annCol = NULL
        annColors = NULL
      }
      
      #Colors for annotating each cluster at the top of heatmap
      if(input$runMarker_clustcolor==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.runMarker_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Colors for heatmap
      if(input$runMarker_heatmap_color==1){
        color <- c("#5bc0eb", "black", "#ECE700")
      }else{
        color <- data.frame(rbind(hot_to_r(input$table.runMarker_heatmap_color)))[1,]
        color <- trimws(color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Indicate if expression data should be centered
      if(input$runMarker_centerFlag==1){
        centerFlag <- TRUE
      }else{
        centerFlag <- FALSE
      }
      
      #Indicate if expression data should be scaled
      if(input$runMarker_scaleFlag==1){
        scaleFlag <- TRUE
      }else{
        scaleFlag <- FALSE
      }
      
      #Assign marginal cutoff for truncating values in data
      if(input$halfwidth_runMarker==1){
        halfwidth <- input$halfwidth_value_runMarker
      }else{
        halfwidth <- NULL
      }
      
      #Show rownames (feature names) in heatmap or not
      if(input$runMarker_show_rownames==1){
        show_rownames <- TRUE
      }else{
        show_rownames <- FALSE
      }
      
      #Show colnames (sample ID) in heatmap or not
      if(input$runMarker_show_colnames==1){
        show_colnames <- TRUE
      }else{
        show_colnames <- FALSE
      }
      
      #The output path for storing the marker heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The width of output figure
      width <- input$width_runMarker
      
      #The height of output figure
      height <- input$height_runMarker
      
      #Indicate the prefix of output figure (e.g. if the name of output figure is 'markerheatmap_using_upregulated_genes.pdf', the prefix would be 'markerheatmap'. If you do not want a prefix, ignore this parameter and type nothing)
      fig.name <- input$runMarker_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- NULL
      }
      
      progress$set(value = 3)
      runMarkerResult <- runMarkers(moic.res = moic.res,
                                    dea.method = dea.method,
                                    prefix = prefix,
                                    dat.path = dat.path,
                                    res.path = res.path,
                                    p.cutoff = p.cutoff,
                                    p.adj.cutoff = p.adj.cutoff,
                                    dirct = dirct,
                                    n.marker = n.marker,
                                    doplot = doplot,
                                    norm.expr = norm.expr,
                                    annCol = annCol,
                                    annColors = annColors,
                                    clust.col = clust.col,
                                    halfwidth = halfwidth,
                                    centerFlag = centerFlag,
                                    scaleFlag = scaleFlag,
                                    show_rownames = show_rownames,
                                    show_colnames = show_colnames,
                                    color = color,
                                    fig.path = fig.path,
                                    fig.name = fig.name,
                                    width = width,
                                    height = height)
      
      #保存runMarker结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(runMarkerResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runMarker.",dea.method,"_",dirct,".RData")))
      
      progress$set(value = 4)
      ##表格标题
      output$table.runMarkerTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(dea.method=="deseq2"){
          if(dirct=="up"){
            tags$strong("Table: Subtype-specific upregulated biomarkers for each identified subtype based on differential expression analysis results using 'deseq2' algorithm" )
          }else{
            tags$strong("Table: Subtype-specific downregulated biomarkers for each identified subtype based on differential expression analysis results using 'deseq2' algorithm" )
          }
        }else if(dea.method=="edger"){
          if(dirct=="up"){
            tags$strong("Table: Subtype-specific upregulated biomarkers for each identified subtype based on differential expression analysis results using 'edger' algorithm" )
          }else{
            tags$strong("Table: Subtype-specific downregulated biomarkers for each identified subtype based on differential expression analysis results using 'edger' algorithm" )
          }
        }else{
          if(dirct=="up"){
            tags$strong("Table: Subtype-specific upregulated biomarkers for each identified subtype based on differential expression analysis results using 'limma' algorithm" )
          }else{
            tags$strong("Table: Subtype-specific downregulated biomarkers for each identified subtype based on differential expression analysis results using 'limma' algorithm" )
          }
        }
      })
      
      ##表格
      output$table.runMarker<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(runMarkerResult$templates),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(runMarkerResult$templates)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$dea_method_runMarker==1,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runMarker==1,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$dea_method_runMarker==1,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runMarker==1,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          list(extend='excel',
                               filename = ifelse(input$dea_method_runMarker==1,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runMarker==1,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$dea_method_runMarker==1,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_",dirct,"regulated_biomarkers_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runMarker==1,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runMarker==2,paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: Subtype-specific ",dirct,"regulated biomarkers for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.runMarkerNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$runMarker_figure_unit<-renderUI({
        list(column(12,downloadButton("runMarker_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runMarker_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runMarker_figure_download<-downloadHandler(
        
        filename = ifelse(is.null(fig.name),paste0("markerheatmap_using_",dirct,"regulated_genes(",dea.method,").pdf"),paste0(fig.name, "_using_", dirct,"regulated_genes(",dea.method,").pdf")),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          hm <- runMarkerResult$heatmap
          draw(hm,annotation_legend_side = "left",heatmap_legend_side = "left")
          dev.off()
        }
      )
      
      ##页面展示事件
      output$runMarker_figure<-renderPlot({
        
        hm <- runMarkerResult$heatmap
        draw(hm,annotation_legend_side = "left",heatmap_legend_side = "left")
      })
      
      ##RData结果下载按钮响应事件
      output$runMarkers_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.runMarker.",dea.method,"_",dirct,".RData"),
        content = function(theFile) {
          save(runMarkerResult,file=theFile)
        }
      )
      
      #runMarker后给予用户的反馈提示
      output$runMarker_finish<-renderUI({
        
        list(br(),column(12,h3(strong("The process of running biomarker identification procedure has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Run gene set enrichment analysis'.")),style="color:darkblack"),br(),
             column(12,downloadButton("runMarkers_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #125.基因集富集分析参考文件链接下载处理按钮
    observeEvent(input$gsea_geneset_background_preparation_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download gene set background file using specified url',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 2)
      #从用户指定路径下载基因集富集分析参考文件
      gsea_geneset_background_preparation_downloadUrl <- input$gsea_geneset_background_preparation_downloadUrl
      gsea_geneset_background_preparation_downloadUrl <- trimws(gsea_geneset_background_preparation_downloadUrl,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      gsea_download_path <- file.path(workdir,"gsea_geneset_background_file(download)")
      #if(!file.exists(gsea_download_path)) {dir.create(gsea_download_path,recursive = TRUE)}
      dir.create(gsea_download_path,recursive = TRUE)
      setwd(gsea_download_path)
      wget(gsea_geneset_background_preparation_downloadUrl)
      #download.file(gsea_geneset_background_preparation_downloadUrl,method="auto",destfile = gsea_download_path)  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
      setwd(workdir)
      
      progress$set(value = 3)
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      progress$set(value = 4)
      #文件下载好后给予用户的反馈提示
      output$gsea_geneset_background_preparation_finish2<-renderUI({
        
        list(br(),column(12,h3(strong("The gene set background file has been downloaded in the corresponding folder. Now let's start running gene set enrichment analysis.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #126.基因集富集分析参考文件上传处理按钮
    observeEvent(input$gsea_geneset_background_preparation_process3,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload gene set background file',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 2)
      #已上传基因集富集分析参考文件转移
      gsea_geneset_background_preparation_uploadPath <- input$gsea_geneset_background_preparation_fileUpload$datapath #基因集富集分析参考文件上传路径
      gsea_upload_path <- file.path(workdir,"gsea_geneset_background_file(upload)")
      #if(!file.exists(gsea_upload_path)) {dir.create(gsea_upload_path,recursive = TRUE)}
      dir.create(gsea_upload_path,recursive = TRUE)
      # file.move(gsea_geneset_background_preparation_uploadPath, gsea_upload_path) #将上传的文件移动到工作目录下
      file.copy(gsea_geneset_background_preparation_uploadPath, gsea_upload_path,overwrite=TRUE) #将上传的文件复制到工作目录下
      
      progress$set(value = 3)
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      progress$set(value = 4)
      #文件上传好后给予用户的反馈提示
      output$gsea_geneset_background_preparation_finish3<-renderUI({
        
        list(br(),column(12,h3(strong("The gene set background file has been uploaded to the corresponding folder. Now let's start running gene set enrichment analysis.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #127.Colors for annotating each cluster at the top of heatmap填写表格
    output$table.runGSEA_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      ##获取聚类数
      if(input$Consensus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        if(length(input$multiClusteringAlgorithms)==1){
          #用户指定聚类数
          N.clust <- switch(input$multiClusteringAlgorithms,
                            'iClusterBayes' = input$iClusterBayes_N_clusts,
                            'SNF' = input$SNF_N_clusts,
                            'PINSPlus' = input$PINSPlus_N_clusts,
                            'NEMO' = input$NEMO_N_clusts,
                            'COCA' = input$COCA_N_clusts,
                            'LRAcluster' = input$LRAcluster_N_clusts,
                            'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                            'IntNMF' = input$IntNMF_N_clusts,
                            'CIMLR' = input$CIMLR_N_clusts,
                            'MoCluster' = input$MoCluster_N_clusts)
        }
        if(length(input$multiClusteringAlgorithms)>1){
          N.clust <- input$Consensus_N_clusts #用户指定聚类数
        }
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #128.Colors for heatmap填写表格
    output$table.runGSEA_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=5,nrow=1))
      colnames(DF)=1:5
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:5){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #129.runGSEA_process事件响应按钮
    observeEvent(input$runGSEA_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$dirct_runGSEA==1,"Perform gene set enrichment analysis to identify subtype-specific overexpressed functional pathways for each subtype","Perform gene set enrichment analysis to identify subtype-specific downexpressed functional pathways for each subtype"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ###聚类结果导入
      ####聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
      }
      ####聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
        moic.res <- cmoic
      }
      ###多组学数据导入
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      #tpm or fpkm
      if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
        
        ##mRNA
        if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$mRNA.expr
        }
        ##lncRNA
        if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$lncRNA.expr
        }
        ##mRNA & lncRNA
        if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm_mrna <- integrated_data$mRNA.expr
          tpm_lncrna <- integrated_data$lncRNA.expr
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index_tpm <- which(duplicated(c(rownames(tpm_mrna),rownames(tpm_lncrna))))
          tpm <- rbind(tpm_mrna,tpm_lncrna)
          if(length(drop_index_tpm)!=0){
            tpm <- tpm[-drop_index_tpm,]
          }
        }
        norm.expr <- tpm
      }
      
      #如果有fpkm数据则优先使用
      if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
        
        ##mRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_mrna
        }
        ##lncRNA
        if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_lncrna
        }
        ##mRNA & lncRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm_mrna <- integrated_data$fpkm_mrna
          fpkm_lncrna <- integrated_data$fpkm_lncrna
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index <- which(duplicated(c(rownames(fpkm_mrna),rownames(fpkm_lncrna))))
          fpkm <- rbind(fpkm_mrna, fpkm_lncrna)
          if(length(drop_index)!=0){
            fpkm <- fpkm[-drop_index,]
          }
        }
        norm.expr <- fpkm
      }
      
      progress$set(value = 2)
      #Indicate the algorithm for completed differential expression analysis
      if(input$dea_method_runGSEA==1){
        dea.method <- "deseq2"
      }else if(input$dea_method_runGSEA==2){
        dea.method <- "edger"
      }else{
        dea.method <- "limma"
      }
      
      #Indicate the prefix of output file (e.g. if the name of output file is 'consensusMOIC_nsclc_limma_test_result.CS1_vs_Others.txt', the prefix would be 'nsclc'. If you do not want a prefix but using the system default prefix, ignore this parameter and type nothing)
      prefix <- input$prefix_runDEA
      prefix <- trimws(prefix,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(prefix==""){
        prefix <- NULL
      }
      
      #The path for saving the files of differential expression analysis
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      dat.path <- file.path(user_workdir,"tables")
      
      #The path for saving the results for identifying subtype-specific functional pathways
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      res.path <- file.path(user_workdir,"tables")
      
      #Indicate the direction of identifying significant pathway
      if(input$dirct_runGSEA==1){
        dirct <- "up"
      }else{
        dirct <- "down"
      }
      
      #Indicate the number of top pathways sorted by NES should be identified for each subtype
      n.path <- input$n_path_runGSEA
      
      #The absolute path of gene set background file
      if(input$gsea_geneset_background_preparation==1){
        msigdb.path <- system.file("extdata", "c5.bp.v7.1.symbols.xls", package = "MOVICS", mustWork = TRUE)
      }else if(input$gsea_geneset_background_preparation==2){
        msigdb.path <- file.path(workdir,"gsea_geneset_background_file(download)",list.files(file.path(workdir,"gsea_geneset_background_file(download)")))
      }else{
        msigdb.path <- file.path(workdir,"gsea_geneset_background_file(upload)",list.files(file.path(workdir,"gsea_geneset_background_file(upload)")))
      }
      
      #Indicate the number of permutations for gene set enrichment analysis (1000 by default and 10000 will be better for reproducibility)
      nPerm <- input$nPerm_runGSEA
      
      #Indicate minimal size of each gene set for analysis
      minGSSize <- input$minGSSize_runGSEA
      
      #Indicate maximal size of each gene set for analysis
      maxGSSize <- input$maxGSSize_runGSEA
      
      #Indicate the nominal p value for identifying significant pathways
      p.cutoff <- input$p_cutoff_runGSEA
      
      #Indicate the adjusted p value for identifying significant pathways
      p.adj.cutoff <- input$p_adj_cutoff_runGSEA
      
      #Indicate the method to employ in the estimation of gene set enrichment scores per sample
      if(input$gsva_method_runGSEA==1){
        gsva.method <- "gsva"
      }else if(input$gsva_method_runGSEA==2){
        gsva.method <- "ssgsea"
      }else if(input$gsva_method_runGSEA==3){
        gsva.method <- "zscore"
      }else{
        gsva.method <- "plage"
      }
      
      #Indicate the method to calculate subtype-specific pathway enrichment scores
      if(input$norm_method_runGSEA==1){
        norm.method <- "mean"
      }else{
        norm.method <- "median"
      }
      
      #Colors for annotating each cluster at the top of heatmap
      if(input$runGSEA_clustcolor==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.runGSEA_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Colors for heatmap
      if(input$runGSEA_heatmap_color==1){
        color <- NULL
      }else{
        color <- data.frame(rbind(hot_to_r(input$table.runGSEA_heatmap_color)))[1,]
        color <- trimws(color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #The output path for storing the pathway heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The width of output figure
      width <- input$width_runGSEA
      
      #The height of output figure
      height <- input$height_runGSEA
      
      ##Indicate the prefix of output figure (e.g. if the name of output figure is 'gseaheatmap_using_upregulated_pathways.pdf', the prefix would be 'gseaheatmap'. If you do not want a prefix but using the system default prefix, ignore this parameter and type nothing)
      fig.name <- input$runGSEA_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- NULL
      }
      
      progress$set(value = 3)
      runGSEAResult <- runGSEAs(moic.res = moic.res,
                                dea.method = dea.method,
                                norm.expr = norm.expr,
                                prefix = prefix,
                                dat.path = dat.path,
                                res.path = res.path,
                                dirct = dirct,
                                n.path = n.path,
                                msigdb.path = msigdb.path,
                                nPerm = nPerm,
                                minGSSize = minGSSize,
                                maxGSSize = maxGSSize,
                                p.cutoff = p.cutoff,
                                p.adj.cutoff = p.adj.cutoff,
                                gsva.method = gsva.method,
                                norm.method = norm.method,
                                clust.col = clust.col,
                                color = color,
                                fig.name = fig.name,
                                fig.path = fig.path,
                                width = width,
                                height = height)
      
      #保存runGSEA结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(runGSEAResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runGSEA.",dea.method,"_",dirct,".RData")))
      
      #整合各个亚型的GSEA结果
      gsea_results <- data.frame()
      gseaidList <- c()
      for(i in 1:length(runGSEAResult$gsea.list)){
        gsea_result <- as.data.frame(runGSEAResult$gsea.list[[i]])
        if(dirct == "up"){
          gseaidList <- c(gseaidList, rownames(gsea_result[which(gsea_result$NES > 0 & gsea_result$pvalue < p.cutoff & gsea_result$p.adjust < p.adj.cutoff),]))
        }
        if(dirct == "down"){
          gseaidList <- c(gseaidList, rownames(gsea_result[which(gsea_result$NES < 0 & gsea_result$pvalue < p.cutoff & gsea_result$p.adjust < p.adj.cutoff),]))
        }
        unqlist <- setdiff(gseaidList,gseaidList[duplicated(gseaidList)])
      }
      
      for(j in 1:length(runGSEAResult$gsea.list)){
        gsea_result <- as.data.frame(runGSEAResult$gsea.list[[j]])
        gsea_result <- gsea_result[,setdiff(colnames(gsea_result),"ID")]
        rownames(gsea_result) <- gsea_result[,1]
        if(dirct == "up"){
          outk <- intersect(unqlist, rownames(gsea_result[which(gsea_result$NES > 0 & gsea_result$pvalue < p.cutoff & gsea_result$p.adjust < p.adj.cutoff),]) )
          gsea_result <- gsea_result[outk,c("Description","setSize","enrichmentScore","NES","pvalue","p.adjust","qvalue")]
          gsea_result <- gsea_result[order(gsea_result$NES, decreasing = TRUE),]
          subtype <- rep(paste0("CS",j),nrow(gsea_result))
          gsea_result$Subtype <- subtype
        }else{
          outk <- intersect(unqlist, rownames(gsea_result[which(gsea_result$NES < 0 & gsea_result$pvalue < p.cutoff & gsea_result$p.adjust < p.adj.cutoff),]) )
          gsea_result <- gsea_result[outk,c("Description","setSize","enrichmentScore","NES","pvalue","p.adjust","qvalue")]
          gsea_result <- gsea_result[order(gsea_result$NES, decreasing = FALSE),]
          subtype <- rep(paste0("CS",j),nrow(gsea_result))
          gsea_result$Subtype <- subtype
        }
        if(j==1){
          gsea_results <- gsea_result
        }else{
          gsea_results <- rbind(gsea_results,gsea_result)
        }
      }  
      
      progress$set(value = 4)
      ##GSEA结果表格标题
      output$table.gseaResultTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(dea.method=="deseq2"){
          if(dirct=="up"){
            tags$strong("Table 1: GSEA results with upregulated pathways for each identified subtype based on differential expression analysis results using 'deseq2' algorithm" )
          }else{
            tags$strong("Table 1: GSEA results with downregulated pathways for each identified subtype based on differential expression analysis results using 'deseq2' algorithm" )
          }
        }else if(dea.method=="edger"){
          if(dirct=="up"){
            tags$strong("Table 1: GSEA results with upregulated pathways for each identified subtype based on differential expression analysis results using 'edger' algorithm" )
          }else{
            tags$strong("Table 1: GSEA results with downregulated pathways for each identified subtype based on differential expression analysis results using 'edger' algorithm" )
          }
        }else{
          if(dirct=="up"){
            tags$strong("Table 1: GSEA results with upregulated pathways for each identified subtype based on differential expression analysis results using 'limma' algorithm" )
          }else{
            tags$strong("Table 1: GSEA results with downregulated pathways for each identified subtype based on differential expression analysis results using 'limma' algorithm" )
          }
        }
      })
      
      ##GSEA结果表格
      output$table.gseaResult<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(gsea_results),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(gsea_results)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          list(extend='excel',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("GSEA_results_with_",dirct,"regulated_pathways_for_each_identified_subtype_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'edger' algorithm"),paste0("Table: GSEA results with ",dirct,"regulated pathways for each identified subtype based on differential expression analysis results using 'limma' algorithm")))
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.gseaResultNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      ##Subtype-specific enrichment scores表格标题
      output$table.enrichmentScoresTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(dea.method=="deseq2"){
          if(dirct=="up"){
            tags$strong("Table 2: subtype-specific enrichment scores on upregulated pathways among identified subtypes based on differential expression analysis results using 'deseq2' algorithm" )
          }else{
            tags$strong("Table 2: subtype-specific enrichment scores on downregulated pathways among identified subtypes based on differential expression analysis results using 'deseq2' algorithm" )
          }
        }else if(dea.method=="edger"){
          if(dirct=="up"){
            tags$strong("Table 2: subtype-specific enrichment scores on upregulated pathways among identified subtypes based on differential expression analysis results using 'edger' algorithm" )
          }else{
            tags$strong("Table 2: subtype-specific enrichment scores on downregulated pathways among identified subtypes based on differential expression analysis results using 'edger' algorithm" )
          }
        }else{
          if(dirct=="up"){
            tags$strong("Table 2: subtype-specific enrichment scores on upregulated pathways among identified subtypes based on differential expression analysis results using 'limma' algorithm" )
          }else{
            tags$strong("Table 2: subtype-specific enrichment scores on downregulated pathways among identified subtypes based on differential expression analysis results using 'limma' algorithm" )
          }
        }
      })
      
      ##Subtype-specific enrichment scores表格
      output$table.enrichmentScores<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(runGSEAResult$grouped.es),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(runGSEAResult$grouped.es)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'edger' algorithm"),paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'edger' algorithm"),paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          list(extend='excel',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'edger' algorithm"),paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'limma' algorithm")))
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$dea_method_runGSEA==1,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_deseq2_algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_edger_algorithm"),paste0("Subtype-specific_enrichment_scores_on_",dirct,"regulated_pathways_among_identified_subtypes_based_on_differential_expression_analysis_results_using_limma_algorithm"))),
                               title = ifelse(input$dea_method_runGSEA==1,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'deseq2' algorithm"),ifelse(input$dea_method_runGSEA==2,paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'edger' algorithm"),paste0("Table: subtype-specific enrichment scores on ",dirct,"regulated pathways among identified subtypes based on differential expression analysis results using 'limma' algorithm")))
                          )
                        )
                      )
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.enrichmentScoresNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$runGSEA_figure_unit<-renderUI({
        list(column(12,downloadButton("runGSEA_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runGSEA_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runGSEA_figure_download<-downloadHandler(
        
        filename = ifelse(is.null(fig.name),paste0("gseaheatmap_using_",dirct,"regulated_pathways(",dea.method,").pdf"),paste0(fig.name, "_using_", dirct,"regulated_pathways(",dea.method,").pdf")),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          hm <- runGSEAResult$heatmap
          draw(hm, annotation_legend_side = "left",heatmap_legend_side = "left")
          dev.off()
        }
      )
      
      ##页面展示事件
      output$runGSEA_figure<-renderPlot({
        
        hm <- runGSEAResult$heatmap
        draw(hm, annotation_legend_side = "left",heatmap_legend_side = "left")
      })
      
      ##RData结果下载按钮响应事件
      output$runGSEAs_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.runGSEA.",dea.method,"_",dirct,".RData"),
        content = function(theFile) {
          save(runGSEAResult,file=theFile)
        }
      )
      
      #runGSEA后给予用户的反馈提示
      output$runGSEA_finish<-renderUI({
        
        list(br(),column(12,h3(strong("The process of running gene set enrichment analysis has been finished, you can download and check the tables and figure, as well as the '.RData' file. Now let's turn to the next step--'Run gene set variation analysis'.")),style="color:darkblack"),br(),
             column(12,downloadButton("runGSEAs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
   })
   
    #130.基因集变异分析的感兴趣基因集列表链接下载处理按钮
    observeEvent(input$gsva_geneset_interest_preparation_process2,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Download the gene set list of interest using specified url',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行数据的下载
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 2)
      #从用户指定路径下载感兴趣基因集列表
      gsva_geneset_interest_preparation_downloadUrl <- input$gsva_geneset_interest_preparation_downloadUrl
      gsva_geneset_interest_preparation_downloadUrl <- trimws(gsva_geneset_interest_preparation_downloadUrl,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      gsva_download_path <- file.path(workdir,"gsva_geneset_interest(download)")
      #if(!file.exists(gsva_download_path)) {dir.create(gsva_download_path,recursive = TRUE)}
      dir.create(gsva_download_path,recursive = TRUE)
      setwd(gsva_download_path)
      wget(gsva_geneset_interest_preparation_downloadUrl)
      #download.file(gsva_geneset_interest_preparation_downloadUrl,method="auto",destfile = gsva_download_path)  #服务器代码自动下载(method="auto"根据服务器类型自动选择下载方法)
      setwd(workdir)
      
      progress$set(value = 3)
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      progress$set(value = 4)
      #文件下载好后给予用户的反馈提示
      output$gsva_geneset_interest_preparation_finish2<-renderUI({
        
        list(br(),column(12,h3(strong("The gene set list of interest has been downloaded in the corresponding folder. Now let's start running gene set variation analysis.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #131.基因集变异分析的感兴趣基因集列表上传处理按钮
    observeEvent(input$gsva_geneset_interest_preparation_process3,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = 'Upload the gene set list of interest',
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      #每点击一次按钮就重新进行已上传数据的处理
      workdir <- getwd()
      workdir_origin <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-2)],collapse = "/") #获取总工作目录的路径
      cancers <- data.frame(rbind(hot_to_r(input$table.cancerType)))[1,]
      cancers <- toupper(trimws(cancers,whitespace = "[ \t\r\n]")) #去除字符串前后两端的空格、tab、回到行首以及换行符并将整个字符串转换成大写格式
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      progress$set(value = 2)
      #已上传感兴趣基因集列表转移
      gsva_geneset_interest_preparation_uploadPath <- input$gsva_geneset_interest_preparation_fileUpload$datapath #感兴趣基因集列表上传路径
      gsva_upload_path <- file.path(workdir,"gsva_geneset_interest(upload)")
      #if(!file.exists(gsva_upload_path)) {dir.create(gsva_upload_path,recursive = TRUE)}
      dir.create(gsva_upload_path,recursive = TRUE)
      # file.move(gsva_geneset_interest_preparation_uploadPath, gsva_upload_path) #将上传的文件移动到工作目录下
      file.copy(gsva_geneset_interest_preparation_uploadPath, gsva_upload_path,overwrite=TRUE) #将上传的文件复制到工作目录下
      
      progress$set(value = 3)
      system(paste0("chmod -R 777 ",workdir_origin)) #级联更新总工作目录下所有文件的权限为“777”
      
      progress$set(value = 4)
      #文件上传好后给予用户的反馈提示
      output$gsva_geneset_interest_preparation_finish3<-renderUI({
        
        list(br(),column(12,h3(strong("The gene set list of interest has been uploaded to the corresponding folder. Now let's start running gene set variation analysis.")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存
    })
    
    #132.Sample annotations from survival information for heatmap填写表格
    output$table.runGSVA_annCol<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$runGSVA_annCol_number,nrow=3))
      colnames(DF)=1:input$runGSVA_annCol_number
      rownames(DF)=c("Sample annotation","Continuous or Categorical","Color settings")
      
      default <- c()
      for(d in 1:input$runGSVA_annCol_number){
        default <- c(default,"")
      }
      DF[1,]=default
      DF[2,]=default
      DF[3,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=180)
    })
    
    #133.Colors for annotating each cluster at the top of heatmap填写表格
    output$table.runGSVA_clustcolor<-renderRHandsontable({
      
      #读取Consensus Clustering步骤整理好的数据
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      ##获取聚类数
      if(input$Consensus_N_clust==1){
        N.clust <- integrated_data$"optClustNum"$N.clust #获取最优聚类数
      }else{
        if(length(input$multiClusteringAlgorithms)==1){
          #用户指定聚类数
          N.clust <- switch(input$multiClusteringAlgorithms,
                            'iClusterBayes' = input$iClusterBayes_N_clusts,
                            'SNF' = input$SNF_N_clusts,
                            'PINSPlus' = input$PINSPlus_N_clusts,
                            'NEMO' = input$NEMO_N_clusts,
                            'COCA' = input$COCA_N_clusts,
                            'LRAcluster' = input$LRAcluster_N_clusts,
                            'ConsensusClustering' = input$ConsensusClustering_N_clusts,
                            'IntNMF' = input$IntNMF_N_clusts,
                            'CIMLR' = input$CIMLR_N_clusts,
                            'MoCluster' = input$MoCluster_N_clusts)
        }
        if(length(input$multiClusteringAlgorithms)>1){
          N.clust <- input$Consensus_N_clusts #用户指定聚类数
        }
      }
      
      DF=data.frame(matrix(NA,ncol=N.clust,nrow=1))
      colnames(DF)=unlist(lapply(1:N.clust,function(i) paste("CS",i,sep="")))
      rownames(DF)="Clustering subtypes colors"
      
      default <- c()
      for(d in 1:N.clust){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=75,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=168)
    })
    
    #134.Colors for heatmap填写表格
    output$table.runGSVA_heatmap_color<-renderRHandsontable({
      
      DF=data.frame(matrix(NA,ncol=input$runGSVA_heatmap_color_number,nrow=1))
      colnames(DF)=1:input$runGSVA_heatmap_color_number
      rownames(DF)="Heatmap colors"
      
      default <- c()
      for(d in 1:input$runGSVA_heatmap_color_number){
        default <- c(default,"")
      }
      DF[1,]=default
      
      #设置table的参数
      rhandsontable(DF) %>%  #width = 500, height =340, no need for those options
        hot_cols(colWidths=80,halign="htMiddle") %>%
        hot_rows(rowHeights=30)%>%
        hot_table(highlightCol = TRUE, highlightRow = FALSE,rowHeaderWidth=120)
    })
    
    #135.runGSVA_process事件响应按钮
    observeEvent(input$runGSVA_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = "Perform gene set variation analysis to calculate enrichment score of each sample in each subtype",
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ###聚类结果导入
      ####聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
      }
      ####聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
        moic.res <- cmoic
      }
      ###多组学数据导入
      integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      #tpm or fpkm
      if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
        
        ##mRNA
        if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$mRNA.expr
        }
        ##lncRNA
        if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm <- integrated_data$lncRNA.expr
        }
        ##mRNA & lncRNA
        if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          tpm_mrna <- integrated_data$mRNA.expr
          tpm_lncrna <- integrated_data$lncRNA.expr
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index_tpm <- which(duplicated(c(rownames(tpm_mrna),rownames(tpm_lncrna))))
          tpm <- rbind(tpm_mrna,tpm_lncrna)
          if(length(drop_index_tpm)!=0){
            tpm <- tpm[-drop_index_tpm,]
          }
        }
        norm.expr <- tpm
      }
      
      #如果有fpkm数据则优先使用
      if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
        
        ##mRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_mrna
        }
        ##lncRNA
        if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm <- integrated_data$fpkm_lncrna
        }
        ##mRNA & lncRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          fpkm_mrna <- integrated_data$fpkm_mrna
          fpkm_lncrna <- integrated_data$fpkm_lncrna
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index <- which(duplicated(c(rownames(fpkm_mrna),rownames(fpkm_lncrna))))
          fpkm <- rbind(fpkm_mrna, fpkm_lncrna)
          if(length(drop_index)!=0){
            fpkm <- fpkm[-drop_index,]
          }
        }
        norm.expr <- fpkm
      }
      
      progress$set(value = 2)
      #The absolute path of the gene set list of interest
      if(input$gsva_geneset_interest_preparation==1){
        gset.gmt.path <- system.file("extdata", "gene sets of interest.gmt", package = "MOVICS", mustWork = TRUE)
      }else if(input$gsva_geneset_interest_preparation==2){
        gset.gmt.path <- file.path(workdir,"gsva_geneset_interest(download)",list.files(file.path(workdir,"gsva_geneset_interest(download)")))
      }else{
        gset.gmt.path <- file.path(workdir,"gsva_geneset_interest(upload)",list.files(file.path(workdir,"gsva_geneset_interest(upload)")))
      }
      
      #Indicate the method to employ in the estimation of gene set enrichment scores per sample
      if(input$gsva_method_runGSVA==1){
        gsva.method <- "gsva"
      }else if(input$gsva_method_runGSVA==2){
        gsva.method <- "ssgsea"
      }else if(input$gsva_method_runGSVA==3){
        gsva.method <- "zscore"
      }else{
        gsva.method <- "plage"
      }
      
      #Indicate if enrichment scores should be centered or not
      if(input$centerFlag_runGSVA==1){
        centerFlag <- TRUE
      }else{
        centerFlag <- FALSE
      }
      
      #Indicate if enrichment scores should be scaled or not
      if(input$scaleFlag_runGSVA==1){
        scaleFlag <- TRUE
      }else{
        scaleFlag <- FALSE
      }
      
      #Assign marginal cutoff for truncating enrichment scores
      if(input$halfwidth_runGSVA==1){
        halfwidth <- input$halfwidth_value_runGSVA
      }else{
        halfwidth <- NULL
      }
      
      #Sample annotations from survival information for heatmap
      if(input$runGSVA_annCol==1){
        surv_clin.info <- integrated_data$clin.info #获取生存和临床数据
        # if(!('fustat' %in% colnames(surv_clin.info)) & ('OS' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS')] <- "fustat"
        # } #数据预处理(列名替换)
        # if(!('futime' %in% colnames(surv_clin.info)) & ('OS.time' %in% colnames(surv_clin.info))){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='OS.time')] <- "futime"
        # } #数据预处理(列名替换)
        # if('age_at_initial_pathologic_diagnosis' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='age_at_initial_pathologic_diagnosis')] <- "age"
        # } #数据预处理(列名替换)
        # if('ajcc_pathologic_tumor_stage' %in% colnames(surv_clin.info)){
        #   colnames(surv_clin.info)[which(colnames(surv_clin.info)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        # } #数据预处理(列名替换)
        # if("pstage" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Discrepancy]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="[Not Available]"){
        #       surv_clin.info$pStage[i] <- names(which(table(surv_clin.info$pStage)==max(table(surv_clin.info$pStage))))
        #     } 
        #   } #缺失值填补(众数填补多分类变量)
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage I" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IB"){
        #       surv_clin.info$pStage[i] <- "Stage I"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage II" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIB"){
        #       surv_clin.info$pStage[i] <- "Stage II"
        #     }
        #     if(trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage III" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(surv_clin.info$pStage[i],whitespace = "[ \t\r\n]")=="Stage IIIB"){
        #       surv_clin.info$pStage[i] <- "Stage III"
        #     }
        #   } #类别合并
        # }
        # if("age" %in% tolower(names(surv_clin.info))){
        #   for(i in 1:nrow(surv_clin.info)){
        #     if(is.na(surv_clin.info$age[i])==TRUE){
        #       surv_clin.info$age[i] <- median(surv_clin.info$age,na.rm = TRUE)
        #     }
        #   } #缺失值填补(中位数填补连续型变量)
        # }
        ###annCol
        sample_annotation <- data.frame(rbind(hot_to_r(input$table.runGSVA_annCol)))[1,]
        sample_annotation <- trimws(sample_annotation,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        annCol <- surv_clin.info[,sample_annotation, drop = FALSE]
        ###annColors
        continuous_or_categorical <- data.frame(rbind(hot_to_r(input$table.runGSVA_annCol)))[2,]
        continuous_or_categorical <- trimws(continuous_or_categorical,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        color_settings <- data.frame(rbind(hot_to_r(input$table.runGSVA_annCol)))[3,]
        color_settings <- trimws(color_settings,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        ###Process the obtained data
        if(!('fustat' %in% colnames(annCol)) & ('OS' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS')] <- "fustat"
        } #数据预处理(列名替换)
        if(!('futime' %in% colnames(annCol)) & ('OS.time' %in% colnames(annCol))){
          colnames(annCol)[which(colnames(annCol)=='OS.time')] <- "futime"
        } #数据预处理(列名替换)
        if('age_at_initial_pathologic_diagnosis' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='age_at_initial_pathologic_diagnosis')] <- "age"
        } #数据预处理(列名替换)
        if('ajcc_pathologic_tumor_stage' %in% colnames(annCol)){
          colnames(annCol)[which(colnames(annCol)=='ajcc_pathologic_tumor_stage')] <- "pStage"
        } #数据预处理(列名替换)
        for (j in 1:ncol(annCol)){
          for(i in 1:nrow(annCol)){
            ##NA判定##
            if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))){
              next
            }
            if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Discrepancy]" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="[Not Available]"){
              annCol[i,j] <- NA
            }
            ##类别合并##
            if(tolower(colnames(annCol)[j])=="pstage"){
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage I" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IB"){
                annCol[i,j] <- "Stage I"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage II" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIB"){
                annCol[i,j] <- "Stage II"
              }
              if(trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage III" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIA" | trimws(annCol[i,j],whitespace = "[ \t\r\n]")=="Stage IIIB"){
                annCol[i,j] <- "Stage III"
              }
            }
            ##缺失值填补##
            if(continuous_or_categorical[j] == "Categorical"){
              ##众数填补多分类变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- names(which(table(annCol[,j])==max(table(annCol[,j]))))
              }
            }else{
              ##中位数填补连续型变量
              if(is.na(trimws(annCol[i,j],whitespace = "[ \t\r\n]"))==TRUE){
                annCol[i,j] <- median(annCol[,j],na.rm = TRUE)
              }
            }
          }
        }
        annColors <- list()
        for(i in 1:length(sample_annotation)){
          if(continuous_or_categorical[i]=="Continuous"){
            breaks <- c(min(annCol[,sample_annotation[i]]),median(annCol[,sample_annotation[i]]),max(annCol[,sample_annotation[i]]))
            colors <- unlist(strsplit(color_settings[i],";"))
            annColors[[i]] <- circlize::colorRamp2(breaks = breaks, colors = colors)
          }else{
            colors <- unlist(strsplit(color_settings[i],";"))
            names(colors) <- unique(annCol[,sample_annotation[i]])
            annColors[[i]] <- colors
          }
        }
        names(annColors) <- sample_annotation
      }else{
        annCol = NULL
        annColors = NULL
      }
      
      #Colors for annotating each cluster at the top of heatmap
      if(input$runGSVA_clustcolor==1){
        clust.col <- c("#2EC4B6","#E71D36","#FF9F1C","#BDD5EA","#FFA5AB","#011627","#023E8A","#9D4EDD")
      }else{
        clust.col <- data.frame(rbind(hot_to_r(input$table.runGSVA_clustcolor)))[1,]
        clust.col <- trimws(clust.col,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Colors for heatmap
      if(input$runGSVA_heatmap_color==1){
        color <- c("#366A9B", "#4E98DE", "#DDDDDD", "#FBCFA7", "#F79C4A")
      }else{
        color <- data.frame(rbind(hot_to_r(input$table.runGSVA_heatmap_color)))[1,]
        color <- trimws(color,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      }
      
      #Distance measurement for hierarchical clustering
      distance <- input$runGSVA_distance
      distance <- trimws(distance,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Clustering method for hierarchical clustering
      linkage <- input$runGSVA_linkage
      linkage <- trimws(linkage,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Show rownames (feature names) in heatmap or not
      if(input$show_rownames_runGSVA==1){
        show_rownames <- TRUE
      }else{
        show_rownames <- FALSE
      }
      
      #Show colnames (sample ID) in heatmap or not
      if(input$show_colnames_runGSVA==1){
        show_colnames <- TRUE
      }else{
        show_colnames <- FALSE
      }
      
      #The output path for storing the enrichment heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The width of output figure
      width <- input$width_runGSVA
      
      #The height of output figure
      height <- input$height_runGSVA
      
      #Indicate the prefix of output figure (e.g. if the name of output figure is 'enrichment_heatmap_using_gsva.pdf', the prefix would be 'enrichment_heatmap_using'. If you do not want a prefix but using the system default prefix, ignore this parameter and type nothing)
      fig.name <- input$runGSVA_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- NULL
      }
      
      progress$set(value = 3)
      runGSVAResult <- runGSVAs(moic.res = moic.res,
                                norm.expr = norm.expr,
                                gset.gmt.path = gset.gmt.path,
                                gsva.method = gsva.method,
                                centerFlag = centerFlag,
                                scaleFlag = scaleFlag,
                                halfwidth = halfwidth,
                                annCol = annCol,
                                annColors = annColors,
                                clust.col = clust.col,
                                distance = distance,
                                linkage = linkage,
                                show_rownames = show_rownames,
                                show_colnames = show_colnames,
                                color = color,
                                fig.path = fig.path,
                                fig.name = fig.name,
                                width = width,
                                height = height)
      
      #保存runGSVA结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      save(runGSVAResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runGSVA.RData")))
      
      progress$set(value = 4)
      ##Raw enrichment score结果表格标题
      output$table.gsvaRawTitle<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong(paste0("Table 1: Raw enrichment score based on the given gene set list of interest by using ",gsva.method," method"))
      })
      
      ##Raw enrichment score结果
      output$table.gsvaRaw<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(runGSVAResult$raw.es),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(runGSVAResult$raw.es)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = paste0("Raw_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: Raw enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          ),
                          
                          list(extend='csv',
                               filename = paste0("Raw_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: Raw enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          ),
                          list(extend='excel',
                               filename = paste0("Raw_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: Raw enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          ),
                          
                          list(extend='print',
                               filename = paste0("Raw_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: Raw enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          )
                        )
                      )
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.gsvaRawNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      ##z-scored enrichment score结果表格标题
      output$table.gsvaZscoredTitle<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong(paste0("Table 2: z-scored enrichment score based on the given gene set list of interest by using ",gsva.method," method"))
      })
      
      ##z-scored enrichment score结果
      output$table.gsvaZscored<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(runGSVAResult$scaled.es),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(runGSVAResult$scaled.es)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = paste0("z-scored_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: z-scored enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          ),
                          
                          list(extend='csv',
                               filename = paste0("z-scored_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: z-scored enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          ),
                          list(extend='excel',
                               filename = paste0("z-scored_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: z-scored enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          ),
                          
                          list(extend='print',
                               filename = paste0("z-scored_enrichment_score_based_on_the_given_gene_set_list_of_interest_by_using_",gsva.method,"_method"),
                               title = paste0("Table: z-scored enrichment score based on the given gene set list of interest by using ",gsva.method," method")
                          )
                        )
                      )
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.gsvaZscoredNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$runGSVA_figure_unit<-renderUI({
        list(column(12,downloadButton("runGSVA_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runGSVA_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runGSVA_figure_download<-downloadHandler(
        
        filename = ifelse(is.null(fig.name),paste0("enrichment_heatmap_using_",gsva.method,".pdf"),paste0(fig.name,"_", gsva.method,".pdf")),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          hm <- runGSVAResult$heatmap
          draw(hm)
          dev.off()
        }
      )
      
      ##页面展示事件
      output$runGSVA_figure<-renderPlot({
        
        hm <- runGSVAResult$heatmap
        draw(hm)
      })
      
      ##RData结果下载按钮响应事件
      output$runGSVAs_download<-downloadHandler(
        filename = paste0(tolower(project),".tcga.runGSVA.RData"),
        content = function(theFile) {
          save(runGSVAResult,file=theFile)
        }
      )
      
      #runGSVA后给予用户的反馈提示
      output$runGSVA_finish<-renderUI({
        
        list(br(),column(12,h3(strong("The process of running gene set variation analysis has been finished, you can download and check the tables and figure, as well as the '.RData' file. Now let's turn to the next step--'Run nearest template prediction'.")),style="color:darkblack"),br(),
             column(12,downloadButton("runGSVAs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    #136.runNTP_process事件响应按钮
    observeEvent(input$runNTP_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$tcga_vali_runNTP==1,"Assign potential subtype labels on tcga cohort using Nearest Template Prediction (NTP)","Assign potential subtype labels on validation cohort using Nearest Template Prediction (NTP)"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_runNTP==1){
        #TCGA
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #Validation
        ###多组学数据导入
        # integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      #expr
      if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
        
        ##mRNA
        if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
          expr <- integrated_data$mRNA.expr
        }
        ##lncRNA
        if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          expr <- integrated_data$lncRNA.expr
        }
        ##mRNA & lncRNA
        if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
          expr1 <- integrated_data$mRNA.expr
          expr2 <- integrated_data$lncRNA.expr
          comsam <- intersect(colnames(expr1),colnames(expr2))
          expr1 <- expr1[,comsam]
          expr2 <- expr2[,comsam]
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index_tpm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
          expr <- rbind(expr1,expr2)
          if(length(drop_index_tpm)!=0){
            expr <- expr[-drop_index_tpm,]
          }
        }
      }
      
      #如果有fpkm数据则优先使用
      if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
        
        ##mRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
          expr <- integrated_data$fpkm_mrna
        }
        ##lncRNA
        if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          expr <- integrated_data$fpkm_lncrna
        }
        ##mRNA & lncRNA
        if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
          expr1 <- integrated_data$fpkm_mrna
          expr2 <- integrated_data$fpkm_lncrna
          comsam <- intersect(colnames(expr1),colnames(expr2))
          expr1 <- expr1[,comsam]
          expr2 <- expr2[,comsam]
          #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
          drop_index_fpkm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
          expr <- rbind(expr1,expr2)
          if(length(drop_index_fpkm)!=0){
            expr <- expr[-drop_index_fpkm,]
          }
        }
      }
      
      #Choose template from 'Run biomarker identification procedure'
      ##Clustering method
      ###聚类结果导入
      ####聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
      }
      ####聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
        moic.res <- cmoic
      }
      mo.method <- moic.res$mo.method
      
      ##Indicate the algorithm
      if(input$dea_method_runNTP==1){
        dea_method <- "deseq2"
      }else if(input$dea_method_runNTP==2){
        dea_method <- "edger"
      }else{
        dea_method <- "limma"
      }
      
      ##Indicate the direction
      if(input$dirct_runNTP==1){
        dirct <- "up"
      }else{
        dirct <- "down"
      }
      
      templates_file <- paste(mo.method,"_",dea_method,"_",dirct,"regulated_marker_templates.txt",sep="")
      templates <- read.table(file.path(user_workdir,"tables",templates_file),sep = "\t",check.names = F,stringsAsFactors = F,header = T)
        
      progress$set(value = 2)
      #Indicate if the expression data should be further scaled
      if(input$scaleFlag_runNTP==1){
        scaleFlag <- TRUE
      }else{
        scaleFlag <- FALSE
      }
      
      #Indicate if the expression data should be further centered
      if(input$centerFlag_runNTP==1){
        centerFlag <- TRUE
      }else{
        centerFlag <- FALSE
      }
      
      #Indicate the permutations for p-value estimation
      nPerm <- input$nPerm_runNTP
      
      #Indicate the distance measurement
      if(input$distance_runNTP==1){
        distance <- "cosine"
      }else if(input$distance_runNTP==2){
        distance <- "pearson"
      }else if(input$distance_runNTP==3){
        distance <- "spearman"
      }else{
        distance <- "kendall"
      }
      
      #Input an integer value for p-value reproducibility
      seed <- input$seed_runNTP
      
      #Indicate whether console messages are to be displayed
      verbose <- TRUE
      
      #Indicate whether to produce prediction heatmap
      doPlot <- TRUE
      
      #The width of output figure
      width <- input$width_runNTP
      
      #The height of output figure
      height <- input$height_runNTP
      
      #The output path for storing the nearest template prediction heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the nearest template prediction heatmap
      if(input$tcga_vali_runNTP==1){
        fig.name <- input$runNTP_FigureName_tcga
      }else{
        fig.name <- input$runNTP_FigureName_validation
      }
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        if(input$tcga_vali_runNTP==1){
          fig.name <- "ntpheatmap(TCGA)"
        }else{
          fig.name <- "ntpheatmap(Validation)"
        }
      }
      
      progress$set(value = 3)
      runNTPResult <- runNTPs(expr = expr,
                              templates = templates,
                              scaleFlag = scaleFlag,
                              centerFlag = centerFlag,
                              nPerm = nPerm,
                              distance = distance,
                              seed = seed,
                              verbose = verbose,
                              doPlot = doPlot,
                              fig.path = fig.path,
                              fig.name = fig.name,
                              width = width,
                              height = height)
      
      #保存runNTP结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_runNTP==1){
        save(runNTPResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runNTP(TCGA).RData")))
      }else{
        save(runNTPResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData")))
      }
      
      progress$set(value = 4)
      ##NTP结果表格标题
      output$table.ntpTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_runNTP==1){
          tags$strong("Table: the results of nearest template prediction (TCGA)")
        }else{
          tags$strong("Table: the results of nearest template prediction (Validation)")
        }
      })
      
      ##NTP结果
      output$table.ntp<-DT::renderDataTable({
        # isolate(validInputTS())
        DT::datatable(isolate(runNTPResult$ntp.res),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(runNTPResult$ntp.res)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$tcga_vali_runNTP==1,"The_results_of_nearest_template_prediction(TCGA)","The_results_of_nearest_template_prediction(Validation)"),
                               title = ifelse(input$tcga_vali_runNTP==1,"Table: the results of nearest template prediction (TCGA)","Table: the results of nearest template prediction (Validation)")
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$tcga_vali_runNTP==1,"The_results_of_nearest_template_prediction(TCGA)","The_results_of_nearest_template_prediction(Validation)"),
                               title = ifelse(input$tcga_vali_runNTP==1,"Table: the results of nearest template prediction (TCGA)","Table: the results of nearest template prediction (Validation)")
                          ),
                          list(extend='excel',
                               filename = ifelse(input$tcga_vali_runNTP==1,"The_results_of_nearest_template_prediction(TCGA)","The_results_of_nearest_template_prediction(Validation)"),
                               title = ifelse(input$tcga_vali_runNTP==1,"Table: the results of nearest template prediction (TCGA)","Table: the results of nearest template prediction (Validation)")
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$tcga_vali_runNTP==1,"The_results_of_nearest_template_prediction(TCGA)","The_results_of_nearest_template_prediction(Validation)"),
                               title = ifelse(input$tcga_vali_runNTP==1,"Table: the results of nearest template prediction (TCGA)","Table: the results of nearest template prediction (Validation)")
                          )
                        )
                      )
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.ntpNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      #绘图展示
      output$runNTP_figure_unit<-renderUI({
        list(column(12,downloadButton("runNTP_figure_download",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runNTP_figure"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runNTP_figure_download<-downloadHandler(
        
        filename = paste0(ifelse(input$tcga_vali_runNTP==1,"ntpheatmap(TCGA)","ntpheatmap(Validation)"),".pdf"),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          parameters_list <- runNTPResult$parameters_list
          subHeatmap(emat = parameters_list$emat, res = runNTPResult$ntp.res, templates = parameters_list$templates)
          dev.off()
        }
      )
      
      ##页面展示事件
      output$runNTP_figure<-renderPlot({
        
        parameters_list <- runNTPResult$parameters_list
        subHeatmap(emat = parameters_list$emat, res = runNTPResult$ntp.res, templates = parameters_list$templates)
      })
      
      ##RData结果下载按钮响应事件
      output$runNTPs_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_runNTP==1,paste0(tolower(project),".tcga.runNTP(TCGA).RData"),paste0(tolower(project),".tcga.runNTP.RData")),
        content = function(theFile) {
          save(runNTPResult,file=theFile)
        }
      )
      
      #runNTP后给予用户的反馈提示
      output$runNTP_finish<-renderUI({
        
        if(input$tcga_vali_runNTP==1){
          list(br(),column(12,h3(strong("The process of running nearest template prediction on tcga cohort has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Run partition around medoids classifier'.")),style="color:darkblack"),br(),
               column(12,downloadButton("runNTPs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of running nearest template prediction on validation cohort has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Run partition around medoids classifier'.")),style="color:darkblack"),br(),
               column(12,downloadButton("runNTPs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    #137.runPAM_process事件响应按钮
    observeEvent(input$runPAM_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = ifelse(input$tcga_vali_runPAM==1,"Use partition around medoids (PAM) classifier to predict potential subtype labels on tcga cohort and calculate in-group proportions (IGP) statistics","Use partition around medoids (PAM) classifier to predict potential subtype labels on validation cohort and calculate in-group proportions (IGP) statistics"),
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_runPAM==1){
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
      }else{
        #TCGA
        ###聚类结果导入
        ####聚类方法数为1
        if(length(input$multiClusteringAlgorithms)==1){
          moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
        }
        ####聚类方法数大于1
        if(length(input$multiClusteringAlgorithms)>1){
          cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
          moic.res <- cmoic
        }
        ###多组学数据导入
        train_integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.getClustNum.RData"))))
        
        #Validation
        ###多组学数据导入
        # validation_integrated_data <- get(load(file.path(user_workdir,"Validation_data_preparing",paste0(tolower(project),".validation.RData"))))
        validation_integrated_data <- get(load(file.path(RData.path,paste0(tolower(project),".validation.getElites.RData"))))
      }
      
      #train.expr & test.expr
      if(input$tcga_vali_runPAM==1){
        
        #TCGA(train.expr==test.expr)
        if(("mRNA.expr" %in% names(integrated_data)) | ("lncRNA.expr" %in% names(integrated_data))){
          
          ##mRNA
          if(("mRNA.expr" %in% names(integrated_data)) & !("lncRNA.expr" %in% names(integrated_data))){
            train.expr <- integrated_data$mRNA.expr
            test.expr <- integrated_data$mRNA.expr
          }
          ##lncRNA
          if(!("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
            train.expr <- integrated_data$lncRNA.expr
            test.expr <- integrated_data$lncRNA.expr
          }
          ##mRNA & lncRNA
          if(("mRNA.expr" %in% names(integrated_data)) & ("lncRNA.expr" %in% names(integrated_data))){
            expr1 <- integrated_data$mRNA.expr
            expr2 <- integrated_data$lncRNA.expr
            comsam <- intersect(colnames(expr1),colnames(expr2))
            expr1 <- expr1[,comsam]
            expr2 <- expr2[,comsam]
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_tpm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
            expr <- rbind(expr1,expr2)
            if(length(drop_index_tpm)!=0){
              expr <- expr[-drop_index_tpm,]
            }
            train.expr <- expr
            test.expr <- expr
          }
        }
        
        #如果有fpkm数据则优先使用
        if(("fpkm_mrna" %in% names(integrated_data)) | ("fpkm_lncrna" %in% names(integrated_data))){
          
          ##mRNA
          if(("fpkm_mrna" %in% names(integrated_data)) & !("fpkm_lncrna" %in% names(integrated_data))){
            train.expr <- integrated_data$fpkm_mrna
            test.expr <- integrated_data$fpkm_mrna
          }
          ##lncRNA
          if(!("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
            train.expr <- integrated_data$fpkm_lncrna
            test.expr <- integrated_data$fpkm_lncrna
          }
          ##mRNA & lncRNA
          if(("fpkm_mrna" %in% names(integrated_data)) & ("fpkm_lncrna" %in% names(integrated_data))){
            expr1 <- integrated_data$fpkm_mrna
            expr2 <- integrated_data$fpkm_lncrna
            comsam <- intersect(colnames(expr1),colnames(expr2))
            expr1 <- expr1[,comsam]
            expr2 <- expr2[,comsam]
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_fpkm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
            expr <- rbind(expr1,expr2)
            if(length(drop_index_fpkm)!=0){
              expr <- expr[-drop_index_fpkm,]
            }
            train.expr <- expr
            test.expr <- expr
          }
        }
      }else{
        
        #Validation(train.expr!=test.expr)
        ##train.expr
        if(("mRNA.expr" %in% names(train_integrated_data)) | ("lncRNA.expr" %in% names(train_integrated_data))){
          
          ##mRNA
          if(("mRNA.expr" %in% names(train_integrated_data)) & !("lncRNA.expr" %in% names(train_integrated_data))){
            train.expr <- train_integrated_data$mRNA.expr
          }
          ##lncRNA
          if(!("mRNA.expr" %in% names(train_integrated_data)) & ("lncRNA.expr" %in% names(train_integrated_data))){
            train.expr <- train_integrated_data$lncRNA.expr
          }
          ##mRNA & lncRNA
          if(("mRNA.expr" %in% names(train_integrated_data)) & ("lncRNA.expr" %in% names(train_integrated_data))){
            expr1 <- train_integrated_data$mRNA.expr
            expr2 <- train_integrated_data$lncRNA.expr
            comsam <- intersect(colnames(expr1),colnames(expr2))
            expr1 <- expr1[,comsam]
            expr2 <- expr2[,comsam]
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_tpm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
            expr <- rbind(expr1,expr2)
            if(length(drop_index_tpm)!=0){
              expr <- expr[-drop_index_tpm,]
            }
            train.expr <- expr
          }
        }
        
        #如果有fpkm数据则优先使用
        if(("fpkm_mrna" %in% names(train_integrated_data)) | ("fpkm_lncrna" %in% names(train_integrated_data))){
          
          ##mRNA
          if(("fpkm_mrna" %in% names(train_integrated_data)) & !("fpkm_lncrna" %in% names(train_integrated_data))){
            train.expr <- train_integrated_data$fpkm_mrna
          }
          ##lncRNA
          if(!("fpkm_mrna" %in% names(train_integrated_data)) & ("fpkm_lncrna" %in% names(train_integrated_data))){
            train.expr <- train_integrated_data$fpkm_lncrna
          }
          ##mRNA & lncRNA
          if(("fpkm_mrna" %in% names(train_integrated_data)) & ("fpkm_lncrna" %in% names(train_integrated_data))){
            expr1 <- train_integrated_data$fpkm_mrna
            expr2 <- train_integrated_data$fpkm_lncrna
            comsam <- intersect(colnames(expr1),colnames(expr2))
            expr1 <- expr1[,comsam]
            expr2 <- expr2[,comsam]
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_fpkm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
            expr <- rbind(expr1,expr2)
            if(length(drop_index_fpkm)!=0){
              expr <- expr[-drop_index_fpkm,]
            }
            train.expr <- expr
          }
        }
        
        ##test.expr
        if(("mRNA.expr" %in% names(validation_integrated_data)) | ("lncRNA.expr" %in% names(validation_integrated_data))){
          
          ##mRNA
          if(("mRNA.expr" %in% names(validation_integrated_data)) & !("lncRNA.expr" %in% names(validation_integrated_data))){
            test.expr <- validation_integrated_data$mRNA.expr
          }
          ##lncRNA
          if(!("mRNA.expr" %in% names(validation_integrated_data)) & ("lncRNA.expr" %in% names(validation_integrated_data))){
            test.expr <- validation_integrated_data$lncRNA.expr
          }
          ##mRNA & lncRNA
          if(("mRNA.expr" %in% names(validation_integrated_data)) & ("lncRNA.expr" %in% names(validation_integrated_data))){
            expr1 <- validation_integrated_data$mRNA.expr
            expr2 <- validation_integrated_data$lncRNA.expr
            comsam <- intersect(colnames(expr1),colnames(expr2))
            expr1 <- expr1[,comsam]
            expr2 <- expr2[,comsam]
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_tpm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
            expr <- rbind(expr1,expr2)
            if(length(drop_index_tpm)!=0){
              expr <- expr[-drop_index_tpm,]
            }
            test.expr <- expr
          }
        }
        
        #如果有fpkm数据则优先使用
        if(("fpkm_mrna" %in% names(validation_integrated_data)) | ("fpkm_lncrna" %in% names(validation_integrated_data))){
          
          ##mRNA
          if(("fpkm_mrna" %in% names(validation_integrated_data)) & !("fpkm_lncrna" %in% names(validation_integrated_data))){
            test.expr <- validation_integrated_data$fpkm_mrna
          }
          ##lncRNA
          if(!("fpkm_mrna" %in% names(train_integrated_data)) & ("fpkm_lncrna" %in% names(train_integrated_data))){
            test.expr <- validation_integrated_data$fpkm_lncrna
          }
          ##mRNA & lncRNA
          if(("fpkm_mrna" %in% names(validation_integrated_data)) & ("fpkm_lncrna" %in% names(validation_integrated_data))){
            expr1 <- validation_integrated_data$fpkm_mrna
            expr2 <- validation_integrated_data$fpkm_lncrna
            comsam <- intersect(colnames(expr1),colnames(expr2))
            expr1 <- expr1[,comsam]
            expr2 <- expr2[,comsam]
            #两个文本框合并前先检查是否有重复的行名(基因名),有的话去除重复的靠后的基因名记录
            drop_index_fpkm <- which(duplicated(c(rownames(expr1),rownames(expr2))))
            expr <- rbind(expr1,expr2)
            if(length(drop_index_fpkm)!=0){
              expr <- expr[-drop_index_fpkm,]
            }
            test.expr <- expr
          }
        }
      }
      
      #Deal with missing value
      ##TCGA
      training_na_action <- ""
      if(input$runPAM_training_na==1){
        train.expr.origin <- train.expr
        train.expr <- as.data.frame(na.omit(train.expr))
        if(nrow(train.expr) != nrow(train.expr.origin)) {message(paste0("--",(nrow(train.expr.origin) - nrow(train.expr)), " features with NA values are removed."))}
        training_na_action <- "rm"
      }else{
        train.expr <- as.data.frame(impute::impute.knn(as.matrix(train.expr))$data)
        training_na_action <- "impute"
      }
      
      ##Validation
      testing_na_action <- ""
      if(input$runPAM_testing_na==1){
        test.expr.origin <- test.expr
        test.expr <- as.data.frame(na.omit(test.expr))
        if(nrow(test.expr) != nrow(test.expr.origin)) {message(paste0("--",(nrow(test.expr.origin) - nrow(test.expr)), " features with NA values are removed."))}
        testing_na_action <- "rm"
      }else{
        test.expr <- as.data.frame(impute::impute.knn(as.matrix(test.expr))$data)
        testing_na_action <- "impute"
      }
      
      #Indicate a subset of genes to be used (Use english commas to separate gene names)
      if(input$gene_subset_runPAM==1){
        gene.subset <- input$gene_subset_list_runPAM
        gene.subset <- trimws(gene.subset,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
        gene.subset <- unlist(strsplit(gene.subset,","))
        drop_index <- c()
        for(i in 1:length(gene.subset)){
          gene.subset[i] <- trimws(gene.subset[i],whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
          if(gene.subset[i]==""){
            drop_index <- c(drop_index,i)
          }
        }
        gene.subset <- gene.subset[-drop_index]
      }else{
        gene.subset <- NULL
      }
      
      progress$set(value = 3)
      runPAMResult <- runPAMs(train.expr = train.expr,
                              moic.res = moic.res,
                              test.expr = test.expr,
                              gene.subset = gene.subset,
                              training_na_action = training_na_action,
                              testing_na_action = testing_na_action)
                             
      #保存runPAM结果(.RData)
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      if(input$tcga_vali_runPAM==1){
        save(runPAMResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runPAM(TCGA).RData")))
      }else{
        save(runPAMResult, file = file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData")))
      }
      
      progress$set(value = 4)
      ##PAM的IGP结果表格标题
      output$table.pamIGPTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_runPAM==1){
          tags$strong("Table 1: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (TCGA)")
        }else{
          tags$strong("Table 1: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (Validation)")
        }
      })
      
      ##PAM的IGP结果
      output$table.pamIGP<-DT::renderDataTable({
        # isolate(validInputTS())
        ##将向量IGP转换为文本框
        IGP <- runPAMResult$IGP
        IGP_names <- names(IGP)
        IGP_data_frame <- as.data.frame(as.list(IGP))
        colnames(IGP_data_frame) <- IGP_names
        DT::datatable(isolate(IGP_data_frame),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(IGP_data_frame)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Bfrtip',
                        searching= FALSE,
                        bInfo=FALSE, #not to shown 1/N message
                        display=NULL,
                        bPaginate= FALSE, #suppress page
                        pageLength =100,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$tcga_vali_runPAM==1,"Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(TCGA)","Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (TCGA)","Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (Validation)")
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$tcga_vali_runPAM==1,"Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(TCGA)","Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (TCGA)","Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (Validation)")
                          ),
                          list(extend='excel',
                               filename = ifelse(input$tcga_vali_runPAM==1,"Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(TCGA)","Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (TCGA)","Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (Validation)")
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$tcga_vali_runPAM==1,"Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(TCGA)","Evaluation_of_similarity_and_reproducibility_of_the_acquired_subtypes_between_discovery_and_validation_cohorts_using_in-group_proportion(IGP)statistic(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (TCGA)","Table: evaluation of similarity and reproducibility of the acquired subtypes between discovery and validation cohorts using in-group proportion (IGP) statistic (Validation)")
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.pamIGPNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      ##PAM的预测结果表格标题
      output$table.pamPredictionTitle<-renderUI({
        # isolate (validInputTS.caption())
        if(input$tcga_vali_runPAM==1){
          tags$strong("Table 2: the results of partition around medoids classifier (TCGA)")
        }else{
          tags$strong("Table 2: the results of partition around medoids classifier (Validation)")
        }
      })
      
      ##PAM的预测结果
      output$table.pamPrediction<-DT::renderDataTable({
        # isolate(validInputTS())
        ##获取并处理
        clust.res <- runPAMResult$clust.res
        clust.res$clust <- paste0("CS",clust.res$clust)
        colnames(clust.res) <- c("sampleID","prediction")
        DT::datatable(isolate(clust.res),
                      extensions = 'Buttons', options =list(
                        columnDefs = list(list(className = 'dt-center', targets = 1:(dim(clust.res)[2]-1))),
                        colReorder = TRUE,
                        scrollX = TRUE,
                        rowReorder = TRUE,
                        dom = 'Blfrtip',
                        # dom = 'Bfrtip',
                        # searching= FALSE,
                        # bInfo=FALSE, #not to shown 1/N message
                        # display=NULL,
                        # bPaginate= FALSE, #suppress page
                        pageLength =10,
                        
                        buttons=list(
                          list(extend='copy',
                               filename = ifelse(input$tcga_vali_runPAM==1,"The_results_of_partition_around_medoids_classifier(TCGA)","The_results_of_partition_around_medoids_classifier(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: the results of partition around medoids classifier (TCGA)","Table: the results of partition around medoids classifier (Validation)")
                          ),
                          
                          list(extend='csv',
                               filename = ifelse(input$tcga_vali_runPAM==1,"The_results_of_partition_around_medoids_classifier(TCGA)","The_results_of_partition_around_medoids_classifier(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: the results of partition around medoids classifier (TCGA)","Table: the results of partition around medoids classifier (Validation)")
                          ),
                          list(extend='excel',
                               filename = ifelse(input$tcga_vali_runPAM==1,"The_results_of_partition_around_medoids_classifier(TCGA)","The_results_of_partition_around_medoids_classifier(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: the results of partition around medoids classifier (TCGA)","Table: the results of partition around medoids classifier (Validation)")
                          ),
                          
                          list(extend='print',
                               filename = ifelse(input$tcga_vali_runPAM==1,"The_results_of_partition_around_medoids_classifier(TCGA)","The_results_of_partition_around_medoids_classifier(Validation)"),
                               title = ifelse(input$tcga_vali_runPAM==1,"Table: the results of partition around medoids classifier (TCGA)","Table: the results of partition around medoids classifier (Validation)")
                          )
                        )
                      ),rownames = FALSE
        )
      }, server=FALSE)
      
      ##表格脚注
      output$table.pamPredictionNote<-renderUI({
        # isolate (validInputTS.caption())
        tags$strong("Note: ...")
      })
      
      ##RData结果下载按钮响应事件
      output$runPAMs_download<-downloadHandler(
        filename = ifelse(input$tcga_vali_runPAM==1,paste0(tolower(project),".tcga.runPAM(TCGA).RData"),paste0(tolower(project),".tcga.runPAM.RData")),
        content = function(theFile) {
          save(runPAMResult,file=theFile)
        }
      )
      
      #runNTP后给予用户的反馈提示
      output$runNTP_finish<-renderUI({
        
        if(input$tcga_vali_runNTP==1){
          list(br(),column(12,h3(strong("The process of running nearest template prediction on tcga cohort has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Run partition around medoids classifier'.")),style="color:darkblack"),br(),
               column(12,downloadButton("runNTPs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of running nearest template prediction on validation cohort has been finished, you can download and check the table and figure, as well as the '.RData' file. Now let's turn to the next step--'Run partition around medoids classifier'.")),style="color:darkblack"),br(),
               column(12,downloadButton("runNTPs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #runPAM后给予用户的反馈提示
      output$runPAM_finish<-renderUI({
        
        if(input$tcga_vali_runPAM==1){
          list(br(),column(12,h3(strong("The process of running partition around medoids classifier on tcga cohort has been finished, you can download and check the tables as well as the '.RData' file. Now let's turn to the next step--'Run consistency evaluation using Kappa statistics'.")),style="color:darkblack"),br(),
               column(12,downloadButton("runPAMs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }else{
          list(br(),column(12,h3(strong("The process of running partition around medoids classifier on validation cohort has been finished, you can download and check the tables as well as the '.RData' file. Now let's turn to the next step--'Run consistency evaluation using Kappa statistics'.")),style="color:darkblack"),br(),
               column(12,downloadButton("runPAMs_download",strong("Download RData file of the results"),style="background-color:#47AEE9"))) ###下载按钮
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    #138.runKappa1_process事件响应按钮
    observeEvent(input$runKappa1_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = "Run consistency evaluation using Kappa statistics between current subtypes derived from multi-omics clustering and NTP-predicted subtypes on tcga cohort",
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      #TCGA
      ###聚类结果导入
      ####聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
      }
      ####聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
        moic.res <- cmoic
      }
      subt1 <- moic.res
      
      ###NTP结果导入
      ntp.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP(TCGA).RData"))))
      subt2 <- ntp.res
      
      progress$set(value = 2)
      #Indicate the label of the first subtype
      subt1.lab <- input$subt1_lab_runKappa1
      subt1.lab <- trimws(subt1.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Indicate the label of the second subtype
      subt2.lab <- input$subt2_lab_runKappa1
      subt2.lab <- trimws(subt2.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The width of output figure
      width <- input$width_runKappa1
      
      #The height of output figure
      height <- input$height_runKappa1
      
      #The output path for storing the consistency heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the consistency heatmap
      fig.name <- input$runKappa1_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "constheatmap(CMOIC_VS_NTP_TCGA)"
      }
      
      progress$set(value = 3)
      #绘图展示
      output$runKappa_figure_unit1<-renderUI({
        list(column(12,downloadButton("runKappa_figure_download1",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runKappa_figure1"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runKappa_figure_download1<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          runKappas(subt1 = subt1$clust.res$clust,
                    subt2 = subt2$clust.res$clust,
                    subt1.lab = subt1.lab,
                    subt2.lab = subt2.lab,
                    fig.path = fig.path,
                    fig.name = fig.name,
                    width = width,
                    height = height)
          dev.off()
        }
      )
      
      progress$set(value = 4)
      ##页面展示事件
      output$runKappa_figure1<-renderPlot({
        
        runKappas(subt1 = subt1$clust.res$clust,
                  subt2 = subt2$clust.res$clust,
                  subt1.lab = subt1.lab,
                  subt2.lab = subt2.lab,
                  fig.path = fig.path,
                  fig.name = fig.name,
                  width = width,
                  height = height)
      })
      
      #runKappa后给予用户的反馈提示
      output$runKappa_finish1<-renderUI({
        
        if(((1 %in% input$ntp_pam_runKappa) & (!(2 %in% input$ntp_pam_runKappa)) & (!(3 %in% input$ntp_pam_runKappa)) & (!(4 %in% input$ntp_pam_runKappa))) | ((((1 %in% input$ntp_pam_runKappa) & (3 %in% input$ntp_pam_runKappa)) | ((1 %in% input$ntp_pam_runKappa) & (4 %in% input$ntp_pam_runKappa))) & !((3 %in% input$ntp_pam_runKappa) & (4 %in% input$ntp_pam_runKappa)))){
          
          list(br(),column(12,h3(strong("The process of running consistency evaluation using Kappa statistics between current subtypes derived from multi-omics clustering and NTP-predicted subtypes on tcga cohort has been finished, you can download and check the figure. Now all steps in 'RUN Module' have been finished!"),),style="color:darkblack"))
        }
        else{
          
          list(br(),column(12,h3(strong("The process of running consistency evaluation using Kappa statistics between current subtypes derived from multi-omics clustering and NTP-predicted subtypes on tcga cohort has been finished, you can download and check the figure. Now keep on the next part of 'Run consistency evaluation using Kappa statistics'."),),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    #139.runKappa2_process事件响应按钮
    observeEvent(input$runKappa2_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = "Run consistency evaluation using Kappa statistics between current subtypes derived from multi-omics clustering and PAM-predicted subtypes on tcga cohort",
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      #TCGA
      ###聚类结果导入
      ####聚类方法数为1
      if(length(input$multiClusteringAlgorithms)==1){
        moic.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.",input$multiClusteringAlgorithms,".RData"))))
      }
      ####聚类方法数大于1
      if(length(input$multiClusteringAlgorithms)>1){
        cmoic <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.cmoic.RData"))))
        moic.res <- cmoic
      }
      subt1 <- moic.res
      
      ###PAM结果导入
      pam.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM(TCGA).RData"))))
      subt2 <- pam.res
      
      progress$set(value = 2)
      #Indicate the label of the first subtype
      subt1.lab <- input$subt1_lab_runKappa2
      subt1.lab <- trimws(subt1.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Indicate the label of the second subtype
      subt2.lab <- input$subt2_lab_runKappa2
      subt2.lab <- trimws(subt2.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The width of output figure
      width <- input$width_runKappa2
      
      #The height of output figure
      height <- input$height_runKappa2
      
      #The output path for storing the consistency heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the consistency heatmap
      fig.name <- input$runKappa2_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "constheatmap(CMOIC_VS_PAM_TCGA)"
      }
      
      progress$set(value = 3)
      #绘图展示
      output$runKappa_figure_unit2<-renderUI({
        list(column(12,downloadButton("runKappa_figure_download2",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runKappa_figure2"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runKappa_figure_download2<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          runKappas(subt1 = subt1$clust.res$clust,
                    subt2 = subt2$clust.res$clust,
                    subt1.lab = subt1.lab,
                    subt2.lab = subt2.lab,
                    fig.path = fig.path,
                    fig.name = fig.name,
                    width = width,
                    height = height)
          dev.off()
        }
      )
      
      progress$set(value = 4)
      ##页面展示事件
      output$runKappa_figure2<-renderPlot({
        
        runKappas(subt1 = subt1$clust.res$clust,
                  subt2 = subt2$clust.res$clust,
                  subt1.lab = subt1.lab,
                  subt2.lab = subt2.lab,
                  fig.path = fig.path,
                  fig.name = fig.name,
                  width = width,
                  height = height)
      })
      
      #runKappa后给予用户的反馈提示
      output$runKappa_finish2<-renderUI({
        
        if(((2 %in% input$ntp_pam_runKappa) & (!(1 %in% input$ntp_pam_runKappa)) & (!(3 %in% input$ntp_pam_runKappa)) & (!(4 %in% input$ntp_pam_runKappa))) | ((((2 %in% input$ntp_pam_runKappa) & (3 %in% input$ntp_pam_runKappa)) | ((2 %in% input$ntp_pam_runKappa) & (4 %in% input$ntp_pam_runKappa))) & !((3 %in% input$ntp_pam_runKappa) & (4 %in% input$ntp_pam_runKappa)))){
          
          list(br(),column(12,h3(strong("The process of running consistency evaluation using Kappa statistics between current subtypes derived from multi-omics clustering and PAM-predicted subtypes on tcga cohort has been finished, you can download and check the figure. Now all steps in 'RUN Module' have been finished!")),style="color:darkblack"))
        }
        else{

          list(br(),column(12,h3(strong("The process of running consistency evaluation using Kappa statistics between current subtypes derived from multi-omics clustering and PAM-predicted subtypes on tcga cohort has been finished, you can download and check the figure. Now keep on the next part of 'Run consistency evaluation using Kappa statistics'.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    #140.runKappa3_process事件响应按钮
    observeEvent(input$runKappa3_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = "Run consistency evaluation using Kappa statistics between NTP-predicted subtypes and PAM-predicted subtypes on tcga cohort",
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ###NTP结果导入
      ntp.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP(TCGA).RData"))))
      subt1 <- ntp.res
      
      ###PAM结果导入
      pam.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM(TCGA).RData"))))
      subt2 <- pam.res
      
      progress$set(value = 2)
      #Indicate the label of the first subtype
      subt1.lab <- input$subt1_lab_runKappa3
      subt1.lab <- trimws(subt1.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Indicate the label of the second subtype
      subt2.lab <- input$subt2_lab_runKappa3
      subt2.lab <- trimws(subt2.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The width of output figure
      width <- input$width_runKappa3
      
      #The height of output figure
      height <- input$height_runKappa3
      
      #The output path for storing the consistency heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the consistency heatmap
      fig.name <- input$runKappa3_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "constheatmap(NTP_VS_PAM_TCGA)"
      }
      
      progress$set(value = 3)
      #绘图展示
      output$runKappa_figure_unit3<-renderUI({
        list(column(12,downloadButton("runKappa_figure_download3",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runKappa_figure3"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runKappa_figure_download3<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          runKappas(subt1 = subt1$clust.res$clust,
                    subt2 = subt2$clust.res$clust,
                    subt1.lab = subt1.lab,
                    subt2.lab = subt2.lab,
                    fig.path = fig.path,
                    fig.name = fig.name,
                    width = width,
                    height = height)
          dev.off()
        }
      )
      
      progress$set(value = 4)
      ##页面展示事件
      output$runKappa_figure3<-renderPlot({
        
        runKappas(subt1 = subt1$clust.res$clust,
                  subt2 = subt2$clust.res$clust,
                  subt1.lab = subt1.lab,
                  subt2.lab = subt2.lab,
                  fig.path = fig.path,
                  fig.name = fig.name,
                  width = width,
                  height = height)
      })
      
      #runKappa后给予用户的反馈提示
      output$runKappa_finish3<-renderUI({
        
        if(!((3 %in% input$ntp_pam_runKappa) & (4 %in% input$ntp_pam_runKappa))){
          
          list(br(),column(12,h3(strong("The process of running consistency evaluation using Kappa statistics between NTP-predicted subtypes and PAM-predicted subtypes on tcga cohort has been finished, you can download and check the figure. Now all steps in 'RUN Module' have been finished!")),style="color:darkblack"))
        }else{
          
          list(br(),column(12,h3(strong("The process of running consistency evaluation using Kappa statistics between NTP-predicted subtypes and PAM-predicted subtypes on tcga cohort has been finished, you can download and check the figure. Now keep on the next part of 'Run consistency evaluation using Kappa statistics'.")),style="color:darkblack"))
        }
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    
    #141.runKappa4_process事件响应按钮
    observeEvent(input$runKappa4_process,{
      
      progress <- Progress$new(session, min=1, max=4) #设置进度条
      on.exit(progress$close())
      progress$set(message = "Run consistency evaluation using Kappa statistics between NTP-predicted subtypes and PAM-predicted subtypes on validation cohort",
                   detail = 'This may take a while...')
      progress$set(value = 1)
      
      ##获取项目名称
      project <- trimws(input$projectName,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      ##导入数据
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      RData.path <- file.path(user_workdir,"RData")
      ###NTP结果导入
      ntp.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runNTP.RData"))))
      subt1 <- ntp.res
      
      ###PAM结果导入
      pam.res <- get(load(file.path(RData.path,paste0(tolower(project),".tcga.runPAM.RData"))))
      subt2 <- pam.res
      
      progress$set(value = 2)
      #Indicate the label of the first subtype
      subt1.lab <- input$subt1_lab_runKappa4
      subt1.lab <- trimws(subt1.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #Indicate the label of the second subtype
      subt2.lab <- input$subt2_lab_runKappa4
      subt2.lab <- trimws(subt2.lab,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      
      #The width of output figure
      width <- input$width_runKappa4
      
      #The height of output figure
      height <- input$height_runKappa4
      
      #The output path for storing the consistency heatmap
      workdir <- getwd() #获取当前工作路径
      user_workdir <- paste(unlist(strsplit(workdir,"/"))[1:(length(unlist(strsplit(workdir,"/")))-1)],collapse = "/") #获取当前用户目录
      fig.path <- file.path(user_workdir,"figures")
      
      #The name of the consistency heatmap
      fig.name <- input$runKappa4_FigureName
      fig.name <- trimws(fig.name,whitespace = "[ \t\r\n]") #去除字符串前后两端的空格、tab、回到行首以及换行符
      if(fig.name==""){
        fig.name <- "constheatmap(NTP_VS_PAM_Validation)"
      }
      
      progress$set(value = 3)
      #绘图展示
      output$runKappa_figure_unit4<-renderUI({
        list(column(12,downloadButton("runKappa_figure_download4",strong("Download pdf of the figure"),style="background-color:#47AEE9")), ##下载按钮
             
             column(12,plotOutput("runKappa_figure4"))) ##页面展示
      })
      
      ##下载按钮响应事件
      output$runKappa_figure_download4<-downloadHandler(
        
        filename = paste0(fig.name,".pdf"),
        content = function(theFile) {
          
          pdf(file = theFile, width=width, height=height)
          runKappas(subt1 = subt1$clust.res$clust,
                    subt2 = subt2$clust.res$clust,
                    subt1.lab = subt1.lab,
                    subt2.lab = subt2.lab,
                    fig.path = fig.path,
                    fig.name = fig.name,
                    width = width,
                    height = height)
          dev.off()
        }
      )
      
      progress$set(value = 4)
      ##页面展示事件
      output$runKappa_figure4<-renderPlot({
        
        runKappas(subt1 = subt1$clust.res$clust,
                  subt2 = subt2$clust.res$clust,
                  subt1.lab = subt1.lab,
                  subt2.lab = subt2.lab,
                  fig.path = fig.path,
                  fig.name = fig.name,
                  width = width,
                  height = height)
      })
      
      #runKappa后给予用户的反馈提示
      output$runKappa_finish4<-renderUI({
        
        list(br(),column(12,h3(strong("The process of running consistency evaluation using Kappa statistics between NTP-predicted subtypes and PAM-predicted subtypes on validation cohort has been finished, you can download and check the figure. Now all steps in 'RUN Module' have been finished!")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    #142.MOVICS_finish_process事件响应按钮
    observeEvent(input$MOVICS_finish_process,{
      
      #完成MOVICS所有步骤后给予用户的反馈提示
      output$MOVICS_finish<-renderUI({
        
        list(br(),column(12,h3(strong("Now all steps in MOVICS have been finished, you can download and check all the tables and figures, as well as the '.RData' files. Then you can further process the obtained results, thanks for using MOVICS RShiny and we are looking forward to your valuable feedback and another visit!")),style="color:darkblack"))
      })
      
      #rm(list=ls()) #清空变量
      gc() #释放内存 
    })
    
    ##143.Users Guide
    output$usersGuide<-renderUI({
      tags$iframe(style="height:600px; width:100%",src="Users Guide.pdf")
    })
    
    ##144.About
    output$about<-renderUI({
      tags$iframe(style="height:600px; width:100%",src="About.pdf")
    })
    
})

